diff --git a/mod/data/field/text/field.class.php b/mod/data/field/text/field.class.php
old mode 100755
new mode 100644
index 248d2b4..36c391a
--- a/mod/data/field/text/field.class.php
+++ b/mod/data/field/text/field.class.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 ///////////////////////////////////////////////////////////////////////////
 //                                                                       //
 // NOTICE OF COPYRIGHT                                                   //
@@ -26,10 +26,6 @@ class data_field_text extends data_field_base {
 
     var $type = 'text';
 
-    function data_field_text($field=0, $data=0) {
-        parent::data_field_base($field, $data);
-    }
-
     function display_search_field($value = '') {
         return '<input type="text" size="16" name="f_'.$this->field->id.'" value="'.$value.'" />';
     }
@@ -39,9 +35,14 @@ class data_field_text extends data_field_base {
     }
 
     function generate_sql($tablealias, $value) {
-        return " ({$tablealias}.fieldid = {$this->field->id} AND {$tablealias}.content LIKE '%{$value}%') ";
+        global $DB;
+
+        static $i=0;
+        $i++;
+        $name = "df_text_$i";
+        return array(" ({$tablealias}.fieldid = {$this->field->id} AND ".$DB->sql_like("{$tablealias}.content", ":$name", false).") ", array($name=>"%$value%"));
     }
 
 }
 
-?>
+
