diff --git a/lib/smarty/plugins/function.fetch.php b/lib/smarty/plugins/function.fetch.php
index 81b1bfc..c66ef3b 100644
--- a/lib/smarty/plugins/function.fetch.php
+++ b/lib/smarty/plugins/function.fetch.php
@@ -181,7 +181,7 @@ function smarty_function_fetch($params, &$smarty)
                         $content .= fgets($fp,4096);
                     }
                     fclose($fp);
-                    $csplit = split("\r\n\r\n",$content,2);
+                    $csplit = explode("\r\n\r\n",$content,2);
 
                     $content = $csplit[1];
 
