diff --git a/filter/multilang/filtersettings.php b/filter/multilang/filtersettings.php
index 895aa0e..a407911 100644
--- a/filter/multilang/filtersettings.php
+++ b/filter/multilang/filtersettings.php
@@ -1,6 +1,9 @@
-<?php  //$Id$
+<?php
 
-$settings->add(new admin_setting_configcheckbox('filter_multilang_force_old', 'filter_multilang_force_old',
-                   get_string('multilangforceold', 'admin'), 0));
+defined('MOODLE_INTERNAL') || die;
+
+if ($ADMIN->fulltree) {
 
-?>
+    $settings->add(new admin_setting_configcheckbox('filter_multilang_force_old', 'filter_multilang_force_old',
+                       get_string('multilangforceold', 'admin'), 0));
+}
