diff --git a/admin/settings/plugins.php b/admin/settings/plugins.php
index 0e86f80..3bd5b15 100644
--- a/admin/settings/plugins.php
+++ b/admin/settings/plugins.php
@@ -474,6 +474,15 @@ foreach (get_plugin_list('report') as $plugin => $plugindir) {
     $ADMIN->add('reports', new admin_externalpage('report'.$plugin, $reportname, $www_path, 'moodle/site:viewreports'));
 }
 
+$ADMIN->add('root', new admin_category('elis', get_string('elis', 'elis_core')));
+// Now add ELIS plugins
+
+foreach (get_plugin_list('elis') as $plugin => $plugindir) {
+    $settings_path = "$plugindir/settings.php";
+    if (file_exists($settings_path)) {
+        include($settings_path);
+    }
+}
 
 /// Add all local plugins - must be always last!
 if ($hassiteconfig) {
