diff --git a/admin/settings/server.php b/admin/settings/server.php
index f56484b..049663d 100644
--- a/admin/settings/server.php
+++ b/admin/settings/server.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 
 // This file defines settingpages and externalpages under the "server" category
 
@@ -10,54 +10,15 @@ $temp = new admin_settingpage('systempaths', get_string('systempaths','admin'));
 $temp->add(new admin_setting_configselect('gdversion', get_string('gdversion','admin'), get_string('configgdversion', 'admin'), check_gd_version(), array('0' => get_string('gdnot'),
                                                                                                                                                           '1' => get_string('gd1'),
                                                                                                                                                           '2' => get_string('gd2'))));
-$temp->add(new admin_setting_configexecutable('zip', get_string('pathtozip','admin'), get_string('configzip', 'admin'), ''));
-$temp->add(new admin_setting_configexecutable('unzip', get_string('pathtounzip','admin'), get_string('configunzip', 'admin'), ''));
 $temp->add(new admin_setting_configexecutable('pathtodu', get_string('pathtodu', 'admin'), get_string('configpathtodu', 'admin'), ''));
 $temp->add(new admin_setting_configexecutable('aspellpath', get_string('aspellpath', 'admin'), get_string('edhelpaspellpath'), ''));
-$ADMIN->add('server', $temp, 0);
-
-
-
-// "email" settingpage
-$temp = new admin_settingpage('mail', get_string('mail','admin'));
-$temp->add(new admin_setting_configtext('smtphosts', get_string('smtphosts', 'admin'), get_string('configsmtphosts', 'admin'), '', PARAM_RAW));
-$temp->add(new admin_setting_configtext('smtpuser', get_string('smtpuser', 'admin'), get_string('configsmtpuser', 'admin'), '', PARAM_NOTAGS));
-$temp->add(new admin_setting_configpasswordunmask('smtppass', get_string('smtppass', 'admin'), get_string('configsmtpuser', 'admin'), ''));
-$temp->add(new admin_setting_configtext('smtpmaxbulk', get_string('smtpmaxbulk', 'admin'), get_string('configsmtpmaxbulk', 'admin'), 1, PARAM_INT));
-$temp->add(new admin_setting_configtext('noreplyaddress', get_string('noreplyaddress', 'admin'), get_string('confignoreplyaddress', 'admin'), 'noreply@' . $_SERVER['HTTP_HOST'], PARAM_NOTAGS));
-$temp->add(new admin_setting_configselect('digestmailtime', get_string('digestmailtime', 'admin'), get_string('configdigestmailtime', 'admin'), 17, array('00' => '00',
-                                                                                                                                                          '01' => '01',
-                                                                                                                                                          '02' => '02',
-                                                                                                                                                          '03' => '03',
-                                                                                                                                                          '04' => '04',
-                                                                                                                                                          '05' => '05',
-                                                                                                                                                          '06' => '06',
-                                                                                                                                                          '07' => '07',
-                                                                                                                                                          '08' => '08',
-                                                                                                                                                          '09' => '09',
-                                                                                                                                                          '10' => '10',
-                                                                                                                                                          '11' => '11',
-                                                                                                                                                          '12' => '12',
-                                                                                                                                                          '13' => '13',
-                                                                                                                                                          '14' => '14',
-                                                                                                                                                          '15' => '15',
-                                                                                                                                                          '16' => '16',
-                                                                                                                                                          '17' => '17',
-                                                                                                                                                          '18' => '18',
-                                                                                                                                                          '19' => '19',
-                                                                                                                                                          '20' => '20',
-                                                                                                                                                          '21' => '21',
-                                                                                                                                                          '22' => '22',
-                                                                                                                                                          '23' => '23')));
-$charsets = get_list_of_charsets();
-unset($charsets['UTF-8']); // not needed here
-$options = array();
-$options['0'] = 'UTF-8';
-$options = array_merge($options, $charsets);
-$temp->add(new admin_setting_configselect('sitemailcharset', get_string('sitemailcharset', 'admin'), get_string('configsitemailcharset','admin'), '0', $options));
-$temp->add(new admin_setting_configcheckbox('allowusermailcharset', get_string('allowusermailcharset', 'admin'), get_string('configallowusermailcharset', 'admin'), 0));
-$options = array('LF'=>'LF', 'CRLF'=>'CRLF');
-$temp->add(new admin_setting_configselect('mailnewline', get_string('mailnewline', 'admin'), get_string('configmailnewline','admin'), 'LF', $options));
+$temp->add(new admin_setting_configexecutable('pathtodot', get_string('pathtodot', 'admin'), get_string('pathtodot_help', 'admin'), ''));
+$ADMIN->add('server', $temp);
+
+
+
+// "supportcontact" settingpage
+$temp = new admin_settingpage('supportcontact', get_string('supportcontact','admin'));
 if (isloggedin()) {
     global $USER;
     $primaryadminemail = $USER->email;
@@ -71,13 +32,12 @@ if (isloggedin()) {
 $temp->add(new admin_setting_configtext('supportname', get_string('supportname', 'admin'), get_string('configsupportname', 'admin'), $primaryadminname, PARAM_NOTAGS));
 $temp->add(new admin_setting_configtext('supportemail', get_string('supportemail', 'admin'), get_string('configsupportemail', 'admin'), $primaryadminemail, PARAM_NOTAGS));
 $temp->add(new admin_setting_configtext('supportpage', get_string('supportpage', 'admin'), get_string('configsupportpage', 'admin'), '', PARAM_URL));
-$ADMIN->add('server', $temp, 100);
-
+$ADMIN->add('server', $temp);
 
 
 // "sessionhandling" settingpage
 $temp = new admin_settingpage('sessionhandling', get_string('sessionhandling', 'admin'));
-$temp->add(new admin_setting_configcheckbox('dbsessions', get_string('dbsessions', 'admin'), get_string('configdbsessions', 'admin'), 0));
+$temp->add(new admin_setting_configcheckbox('dbsessions', get_string('dbsessions', 'admin'), get_string('configdbsessions', 'admin'), 1));
 $temp->add(new admin_setting_configselect('sessiontimeout', get_string('sessiontimeout', 'admin'), get_string('configsessiontimeout', 'admin'), 7200, array(14400 => get_string('numhours', '', 4),
                                                                                                                                                       10800 => get_string('numhours', '', 3),
                                                                                                                                                       7200 => get_string('numhours', '', 2),
@@ -90,29 +50,11 @@ $temp->add(new admin_setting_configselect('sessiontimeout', get_string('sessiont
 $temp->add(new admin_setting_configtext('sessioncookie', get_string('sessioncookie', 'admin'), get_string('configsessioncookie', 'admin'), '', PARAM_ALPHANUM));
 $temp->add(new admin_setting_configtext('sessioncookiepath', get_string('sessioncookiepath', 'admin'), get_string('configsessioncookiepath', 'admin'), '/', PARAM_LOCALURL));
 $temp->add(new admin_setting_configtext('sessioncookiedomain', get_string('sessioncookiedomain', 'admin'), get_string('configsessioncookiedomain', 'admin'), '', PARAM_TEXT, 50));
-$ADMIN->add('server', $temp, 50);
-
-
-
-// "rss" settingpage
-$temp = new admin_settingpage('rss', get_string('rss'));
-$temp->add(new admin_setting_configcheckbox('enablerssfeeds', get_string('enablerssfeeds', 'admin'), get_string('configenablerssfeeds', 'admin'), 0));
-$ADMIN->add('server', $temp);
-
-
-// "debugging" settingpage
-$temp = new admin_settingpage('debugging', get_string('debugging', 'admin'));
-$temp->add(new admin_setting_special_debug());
-$temp->add(new admin_setting_configcheckbox('debugdisplay', get_string('debugdisplay', 'admin'), get_string('configdebugdisplay', 'admin'), ini_get_bool('display_errors')));
-$temp->add(new admin_setting_configcheckbox('xmlstrictheaders', get_string('xmlstrictheaders', 'admin'), get_string('configxmlstrictheaders', 'admin'), 0));
-$temp->add(new admin_setting_configcheckbox('debugsmtp', get_string('debugsmtp', 'admin'), get_string('configdebugsmtp', 'admin'), 0));
-$temp->add(new admin_setting_configcheckbox('perfdebug', get_string('perfdebug', 'admin'), get_string('configperfdebug', 'admin'), '7', '15', '7'));
 $ADMIN->add('server', $temp);
 
 
 // "stats" settingpage
-$temp = new admin_settingpage('stats', get_string('stats'));
-$temp->add(new admin_setting_configcheckbox('enablestats', get_string('enablestats', 'admin'), get_string('configenablestats', 'admin'), 0));
+$temp = new admin_settingpage('stats', get_string('stats'), 'moodle/site:config', empty($CFG->enablestats));
 $temp->add(new admin_setting_configselect('statsfirstrun', get_string('statsfirstrun', 'admin'), get_string('configstatsfirstrun', 'admin'), 'none', array('none' => get_string('none'),
                                                                                                                                                            60*60*24*7 => get_string('numweeks','moodle',1),
                                                                                                                                                            60*60*24*14 => get_string('numweeks','moodle',2),
@@ -138,9 +80,6 @@ $temp->add(new admin_setting_configselect('statsmaxruntime', get_string('statsma
 $temp->add(new admin_setting_configtext('statsruntimedays', get_string('statsruntimedays', 'admin'), get_string('configstatsruntimedays', 'admin'), 31, PARAM_INT));
 $temp->add(new admin_setting_configtime('statsruntimestarthour', 'statsruntimestartminute', get_string('statsruntimestart', 'admin'), get_string('configstatsruntimestart', 'admin'), array('h' => 0, 'm' => 0)));
 $temp->add(new admin_setting_configtext('statsuserthreshold', get_string('statsuserthreshold', 'admin'), get_string('configstatsuserthreshold', 'admin'), 0, PARAM_INT));
-
-$options = array(0=>0, 1=>1, 2=>2, 3=>3, 4=>4, 5=>5, 6=>6);
-$temp->add(new admin_setting_configselect('statscatdepth', get_string('statscatdepth', 'admin'), get_string('configstatscatdepth', 'admin'), 1, $options));
 $ADMIN->add('server', $temp);
 
 
@@ -165,24 +104,18 @@ if (defined('CURLPROXY_SOCKS5')) {
 $temp->add(new admin_setting_configselect('proxytype', get_string('proxytype', 'admin'), get_string('configproxytype','admin'), 'HTTP', $options));
 $temp->add(new admin_setting_configtext('proxyuser', get_string('proxyuser', 'admin'), get_string('configproxyuser', 'admin'), ''));
 $temp->add(new admin_setting_configpasswordunmask('proxypassword', get_string('proxypassword', 'admin'), get_string('configproxypassword', 'admin'), ''));
+$temp->add(new admin_setting_configtext('proxybypass', get_string('proxybypass', 'admin'), get_string('configproxybypass', 'admin'), 'localhost, 127.0.0.1'));
 $ADMIN->add('server', $temp);
 
-$ADMIN->add('server', new admin_externalpage('maintenancemode', get_string('sitemaintenancemode', 'admin'), "$CFG->wwwroot/$CFG->admin/maintenance.php"));
-
+$temp = new admin_settingpage('maintenancemode', get_string('sitemaintenancemode', 'admin'));
+$options = array(0=>get_string('disable'), 1=>get_string('enable'));
+$temp->add(new admin_setting_configselect('maintenance_enabled', get_string('sitemaintenancemode', 'admin'),
+                                          get_string('helpsitemaintenance', 'admin'), 0, $options));
+$temp->add(new admin_setting_confightmleditor('maintenance_message', get_string('optionalmaintenancemessage', 'admin'),
+                                              '', ''));
+$ADMIN->add('server', $temp);
 
 $temp = new admin_settingpage('cleanup', get_string('cleanup', 'admin'));
-$temp->add(new admin_setting_configselect('longtimenosee', get_string('longtimenosee', 'admin'), get_string('configlongtimenosee', 'admin'), 120, array(0 => get_string('never'),
-                                                                                                                                                        1000 => get_string('numdays', '', 1000),
-                                                                                                                                                        365 => get_string('numdays', '', 365),
-                                                                                                                                                        180 => get_string('numdays', '', 180),
-                                                                                                                                                        150 => get_string('numdays', '', 150),
-                                                                                                                                                        120 => get_string('numdays', '', 120),
-                                                                                                                                                        90 => get_string('numdays', '', 90),
-                                                                                                                                                        60 => get_string('numdays', '', 60),
-                                                                                                                                                        30 => get_string('numdays', '', 30),
-                                                                                                                                                        21 => get_string('numdays', '', 21),
-                                                                                                                                                        14 => get_string('numdays', '', 14),
-                                                                                                                                                        7 => get_string('numdays', '', 7) )));
 $temp->add(new admin_setting_configselect('deleteunconfirmed', get_string('deleteunconfirmed', 'admin'), get_string('configdeleteunconfirmed', 'admin'), 168, array(0 => get_string('never'),
                                                                                                                                                                     168 => get_string('numdays', '', 7),
                                                                                                                                                                     144 => get_string('numdays', '', 6),
@@ -204,6 +137,8 @@ $temp->add(new admin_setting_configselect('deleteincompleteusers', get_string('d
                                                                                                                                                                     48 => get_string('numdays', '', 2),
                                                                                                                                                                     24 => get_string('numdays', '', 1))));
 
+$temp->add(new admin_setting_configcheckbox('logguests', get_string('logguests', 'admin'),
+                                            get_string('logguests_help', 'admin'), 1));
 $temp->add(new admin_setting_configselect('loglifetime', get_string('loglifetime', 'admin'), get_string('configloglifetime', 'admin'), 0, array(0 => get_string('neverdeletelogs'),
                                                                                                                                                 1000 => get_string('numdays', '', 1000),
                                                                                                                                                 365 => get_string('numdays', '', 365),
@@ -219,10 +154,10 @@ $temp->add(new admin_setting_configselect('loglifetime', get_string('loglifetime
 
 
 $temp->add(new admin_setting_configcheckbox('disablegradehistory', get_string('disablegradehistory', 'grades'),
-                                            get_string('configdisablegradehistory', 'grades'), 0, PARAM_INT));
+                                            get_string('disablegradehistory_help', 'grades'), 0));
 
 $temp->add(new admin_setting_configselect('gradehistorylifetime', get_string('gradehistorylifetime', 'grades'),
-                                          get_string('configgradehistorylifetime', 'grades'), 0, array(0 => get_string('neverdeletehistory', 'grades'),
+                                          get_string('gradehistorylifetime_help', 'grades'), 0, array(0 => get_string('neverdeletehistory', 'grades'),
                                                                                                    1000 => get_string('numdays', '', 1000),
                                                                                                     365 => get_string('numdays', '', 365),
                                                                                                     180 => get_string('numdays', '', 180),
@@ -243,8 +178,10 @@ $ADMIN->add('server', new admin_externalpage('phpinfo', get_string('phpinfo'), "
 // "performance" settingpage
 $temp = new admin_settingpage('performance', get_string('performance', 'admin'));
 
+$temp->add(new admin_setting_configtext('numcoursesincombo', get_string('numcoursesincombo', 'admin'), get_string('numcoursesincombo_help', 'admin'), 500));
+
 $temp->add(new admin_setting_configselect('extramemorylimit', get_string('extramemorylimit', 'admin'),
-                                          get_string('configextramemorylimit', 'admin'), '128M',
+                                          get_string('configextramemorylimit', 'admin'), '512M',
                                           // if this option is set to 0, default 128M will be used
                                           array( '64M' => '64M',
                                                  '128M' => '128M',
@@ -252,6 +189,12 @@ $temp->add(new admin_setting_configselect('extramemorylimit', get_string('extram
                                                  '512M' => '512M',
                                                  '1024M' => '1024M'
                                              )));
+$temp->add(new admin_setting_configtext('curlcache', get_string('curlcache', 'admin'),
+                                        get_string('configcurlcache', 'admin'), 120, PARAM_INT));
+
+$temp->add(new admin_setting_configtext('curltimeoutkbitrate', get_string('curltimeoutkbitrate', 'admin'),
+                                        get_string('curltimeoutkbitrate_help', 'admin'), 56, PARAM_INT));
+/* //TODO: we need to fix code instead of relying on slow rcache, enable this once we have some code that is actually using it
 $temp->add(new admin_setting_special_selectsetup('cachetype', get_string('cachetype', 'admin'),
                                           get_string('configcachetype', 'admin'), '',
                                           array( '' => get_string('none'),
@@ -273,18 +216,11 @@ $temp->add(new admin_setting_configselect('memcachedpconn', get_string('memcache
                                           get_string('configmemcachedpconn', 'admin'), 0,
                                           array( '0' => get_string('no'),
                                                  '1' => get_string('yes'))));
+*/
 $ADMIN->add('server', $temp);
 
-if ($CFG->dbfamily === 'mysql') {
-    if (file_exists("$CFG->dirroot/$CFG->admin/mysql/frame.php")) {
-        $ADMIN->add('server', new admin_externalpage('database', get_string('managedatabase'), "$CFG->wwwroot/$CFG->admin/mysql/frame.php"));
-    }
-} else if ($CFG->dbfamily === 'postgres') {
-    if (file_exists("$CFG->dirroot/$CFG->admin/pgsql/frame.php")) {
-        $ADMIN->add('server', new admin_externalpage('database', get_string('managedatabase'), "$CFG->wwwroot/$CFG->admin/pgsql/frame.php"));
-    }
-}
 
-} // end of speedup
+$ADMIN->add('server', new admin_externalpage('adminregistration', get_string('registration','admin'), "$CFG->wwwroot/$CFG->admin/registration/index.php"));
+$ADMIN->add('root', new admin_externalpage('bloglevelupgrade', get_string('bloglevelupgrade', 'admin'), $CFG->wwwroot.'/'.$CFG->admin.'/blocklevelupgrade.php', 'moodle/site:config', true));
 
-?>
+} // end of speedup
