diff --git a/login/confirm.php b/login/confirm.php
index cc86086..5fad838 100644
--- a/login/confirm.php
+++ b/login/confirm.php
@@ -1,25 +1,52 @@
-<?php // $Id$
-
-    require_once("../config.php");
-
-    $data = optional_param('data', '', PARAM_CLEAN);  // Formatted as:  secret/username
-
-    $p = optional_param('p', '', PARAM_ALPHANUM);     // Old parameter:  secret
-    $s = optional_param('s', '', PARAM_CLEAN);        // Old parameter:  username
-
-    if (empty($CFG->registerauth)) {
-        error("Sorry, you may not use this page.");
-    }
-    $authplugin = get_auth_plugin($CFG->registerauth);
-
-    if (!$authplugin->can_confirm()) {
-        error("Sorry, you may not use this page.");
-    }
-
-    if (!empty($data) || (!empty($p) && !empty($s))) {
+<?php
+
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Confirm self registered user.
+ *
+ * @package    core
+ * @subpackage auth
+ * @copyright  1999 Martin Dougiamas  http://dougiamas.com
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+require('../config.php');
+
+$data = optional_param('data', '', PARAM_RAW);  // Formatted as:  secret/username
+
+$p = optional_param('p', '', PARAM_ALPHANUM);   // Old parameter:  secret
+$s = optional_param('s', '', PARAM_RAW);        // Old parameter:  username
+
+$PAGE->set_url('/login/confirm.php');
+$PAGE->set_context(get_context_instance(CONTEXT_SYSTEM));
+
+if (empty($CFG->registerauth)) {
+    print_error('cannotusepage2');
+}
+$authplugin = get_auth_plugin($CFG->registerauth);
+
+if (!$authplugin->can_confirm()) {
+    print_error('cannotusepage2');
+}
+
+if (!empty($data) || (!empty($p) && !empty($s))) {
 
         if (!empty($data)) {
-            $dataelements = explode('/',$data, 2); // Stop after 1st slash. Rest is username. MDL-7647
+        $dataelements = explode('/', $data, 2); // Stop after 1st slash. Rest is username. MDL-7647
             $usersecret = $dataelements[0];
             $username   = $dataelements[1];
         } else {
@@ -31,24 +58,27 @@
 
         if ($confirmed == AUTH_CONFIRM_ALREADY) {
             $user = get_complete_user_data('username', $username);
-            print_header(get_string("alreadyconfirmed"), get_string("alreadyconfirmed"), array(), "");
-            print_box_start('generalbox centerpara boxwidthnormal boxaligncenter');
+        $PAGE->navbar->add(get_string("alreadyconfirmed"));
+        $PAGE->set_title(get_string("alreadyconfirmed"));
+        $PAGE->set_heading($COURSE->fullname);
+        echo $OUTPUT->header();
+        echo $OUTPUT->box_start('generalbox centerpara boxwidthnormal boxaligncenter');
             echo "<h3>".get_string("thanks").", ". fullname($user) . "</h3>\n";
             echo "<p>".get_string("alreadyconfirmed")."</p>\n";
-            print_single_button("$CFG->wwwroot/course/", null, get_string('courses'));
-            print_box_end();
-            print_footer();
+        echo $OUTPUT->single_button("$CFG->wwwroot/course/", get_string('courses'));
+        echo $OUTPUT->box_end();
+        echo $OUTPUT->footer();
             exit;
 
         } else if ($confirmed == AUTH_CONFIRM_OK) {
 
             // The user has confirmed successfully, let's log them in
 
-            if (!$USER = get_complete_user_data('username', $username)) {
-                error("Something serious is wrong with the database");
+        if (!$user = get_complete_user_data('username', $username)) {
+            print_error('cannotfinduser', '', '', s($username));
             }
 
-            set_moodle_cookie($USER->username);
+        complete_user_login($user);
 
             if ( ! empty($SESSION->wantsurl) ) {   // Send them where they were going
                 $goto = $SESSION->wantsurl;
@@ -56,21 +86,22 @@
                 redirect($goto);
             }
 
-            print_header(get_string("confirmed"), get_string("confirmed"), array(), "");
-            print_box_start('generalbox centerpara boxwidthnormal boxaligncenter');
+        $PAGE->navbar->add(get_string("confirmed"));
+        $PAGE->set_title(get_string("confirmed"));
+        $PAGE->set_heading($COURSE->fullname);
+        echo $OUTPUT->header();
+        echo $OUTPUT->box_start('generalbox centerpara boxwidthnormal boxaligncenter');
             echo "<h3>".get_string("thanks").", ". fullname($USER) . "</h3>\n";
             echo "<p>".get_string("confirmed")."</p>\n";
-            print_single_button("$CFG->wwwroot/course/", null, get_string('courses'));
-            print_box_end();
-            print_footer();
+        echo $OUTPUT->single_button("$CFG->wwwroot/course/", get_string('courses'));
+        echo $OUTPUT->box_end();
+        echo $OUTPUT->footer();
             exit;
         } else {
-            error("Invalid confirmation data");
+        print_error('invalidconfirmdata');
         }
-    } else {
+} else {
         print_error("errorwhenconfirming");
-    }
-
-    redirect("$CFG->wwwroot/");
+}
 
-?>
+redirect("$CFG->wwwroot/");
