diff --git a/login/forgot_password_form.php b/login/forgot_password_form.php
index 684eabc..1a6e441 100644
--- a/login/forgot_password_form.php
+++ b/login/forgot_password_form.php
@@ -1,26 +1,57 @@
-<?php //$Id$
+<?php
+
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Reset forgotten password form definition.
+ *
+ * @package    core
+ * @subpackage auth
+ * @copyright  2006 Petr Skoda {@link http://skodak.org}
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+defined('MOODLE_INTERNAL') || die();
 
 require_once $CFG->libdir.'/formslib.php';
 
 class login_forgot_password_form extends moodleform {
 
     function definition() {
-        $mform    =& $this->_form;
-        $renderer =& $mform->defaultRenderer();
+        $mform    = $this->_form;
 
-        $mform->addElement('header', '', get_string('passwordforgotten'), '');
+        $mform->addElement('header', '', get_string('searchbyusername'), '');
 
         $mform->addElement('text', 'username', get_string('username'));
         $mform->setType('username', PARAM_RAW);
 
+        $submitlabel = get_string('search');
+        $mform->addElement('submit', 'submitbutton', $submitlabel);
+
+        $mform->addElement('header', '', get_string('searchbyemail'), '');
+
         $mform->addElement('text', 'email', get_string('email'));
         $mform->setType('email', PARAM_RAW);
 
-        $this->add_action_buttons(true, get_string('ok'));
+        $submitlabel = get_string('search');
+        $mform->addElement('submit', 'submitbutton', $submitlabel);
     }
 
     function validation($data, $files) {
-        global $CFG;
+        global $CFG, $DB;
 
         $errors = parent::validation($data, $files);
 
@@ -32,7 +63,7 @@ class login_forgot_password_form extends moodleform {
             if (!validate_email($data['email'])) {
                 $errors['email'] = get_string('invalidemail');
 
-            } else if (count_records('user', 'email', $data['email']) > 1) {
+            } else if ($DB->count_records('user', array('email'=>$data['email'])) > 1) {
                 $errors['email'] = get_string('forgottenduplicate');
 
             } else {
@@ -61,5 +92,3 @@ class login_forgot_password_form extends moodleform {
     }
 
 }
-
-?>
\ No newline at end of file
