diff --git a/lib/yui/formautosubmit/formautosubmit.js b/lib/yui/formautosubmit/formautosubmit.js
index 01d6668..39525db 100644
--- a/lib/yui/formautosubmit/formautosubmit.js
+++ b/lib/yui/formautosubmit/formautosubmit.js
@@ -40,9 +40,17 @@ YUI.add('moodle-core-formautosubmit',
                 }
 
                 // Assign this select items 'nothing' value and lastindex (current value)
+                if (this.get('selectid')) {
                 var thisselect = Y.one('select#' + this.get('selectid'));
+                    if (thisselect) {
+                        if (this.get('nothing')) {
                 thisselect.setData('nothing', this.get('nothing'));
+                        }
                 thisselect.setData('startindex', thisselect.get('selectedIndex'));
+                    } else {
+                        Y.log("Warning: A single_select element was renderered, but the output is not displayed on the page.");
+                    }
+                }
             },
 
             /**
