diff --git a/user/profile/field/textarea/field.class.php b/user/profile/field/textarea/field.class.php
index 956c092..eeffebc 100644
--- a/user/profile/field/textarea/field.class.php
+++ b/user/profile/field/textarea/field.class.php
@@ -1,4 +1,4 @@
-<?php //$Id$
+<?php
 
 class profile_field_textarea extends profile_field_base {
 
@@ -7,8 +7,8 @@ class profile_field_textarea extends profile_field_base {
         $rows = $this->field->param2;
 
         /// Create the form field
-        $mform->addElement('htmleditor', $this->inputname, format_string($this->field->name), array('cols'=>$cols, 'rows'=>$rows));
-        $mform->setType($this->inputname, PARAM_CLEAN);
+        $mform->addElement('editor', $this->inputname, format_string($this->field->name), null, null);
+        $mform->setType($this->inputname, PARAM_RAW); // we MUST clean this before display!
     }
 
     /// Overwrite base class method, data in this field type is potentially too large to be
@@ -17,6 +17,28 @@ class profile_field_textarea extends profile_field_base {
         return false;
     }
 
+    function edit_save_data_preprocess($data, &$datarecord) {
+        if (is_array($data)) {
+            $datarecord->dataformat = $data['format'];
+            $data = $data['text'];
+        }
+        return $data;
+    }
+
+    function edit_load_user_data(&$user) {
+        if ($this->data !== NULL) {
+            $this->data = clean_text($this->data, $this->dataformat);
+            $user->{$this->inputname} = array('text'=>$this->data, 'format'=>$this->dataformat);
+        }
+    }
+
+    /**
+     * Display the data for this field
+     */
+    function display_data() {
+        return format_text($this->data, $this->dataformat, array('overflowdiv'=>true));
+    }
+
 }
 
-?>
+
