diff --git a/grade/grading/form/guide/lib.php b/grade/grading/form/guide/lib.php
index 942b19b..c5292ba 100644
--- a/grade/grading/form/guide/lib.php
+++ b/grade/grading/form/guide/lib.php
@@ -820,9 +820,10 @@ class gradingform_guide_instance extends gradingform_instance {
     /**
      * Calculates the grade to be pushed to the gradebook
      *
-     * @return float|int the valid grade from $this->get_controller()->get_grade_range()
+     * @return int the valid grade from $this->get_controller()->get_grade_range()
      */
     public function get_grade() {
+        global $DB, $USER;
         $grade = $this->get_guide_filling();
 
         if (!($scores = $this->get_controller()->get_min_max_score()) || $scores['maxscore'] <= $scores['minscore']) {
@@ -841,12 +842,8 @@ class gradingform_guide_instance extends gradingform_instance {
         foreach ($grade['criteria'] as $record) {
             $curscore += $record['score'];
         }
-        $gradeoffset = ($curscore-$scores['minscore'])/($scores['maxscore']-$scores['minscore'])*
-            ($maxgrade-$mingrade);
-        if ($this->get_controller()->get_allow_grade_decimals()) {
-            return $gradeoffset + $mingrade;
-        }
-        return round($gradeoffset, 0) + $mingrade;
+        return round(($curscore-$scores['minscore'])/($scores['maxscore']-$scores['minscore'])*
+            ($maxgrade-$mingrade), 0) + $mingrade;
     }
 
     /**
