diff --git a/grade/grading/form/guide/version.php b/grade/grading/form/guide/version.php
index 48ba85c..6d92a4f 100644
--- a/grade/grading/form/guide/version.php
+++ b/grade/grading/form/guide/version.php
@@ -25,6 +25,6 @@
 defined('MOODLE_INTERNAL') || die();
 
 $plugin->component  = 'gradingform_guide';
-$plugin->version    = 2013051401.05;
-$plugin->requires   = 2013051401.05;
+$plugin->version    = 2013050100;
+$plugin->requires   = 2013050100;
 $plugin->maturity   = MATURITY_STABLE;
\ No newline at end of file
