diff --git a/mod/survey/db/upgrade.php b/mod/survey/db/upgrade.php
index 7c240e9..b9d06a2 100644
--- a/mod/survey/db/upgrade.php
+++ b/mod/survey/db/upgrade.php
@@ -1,4 +1,4 @@
-<?php  //$Id$
+<?php
 
 // This file keeps track of upgrades to 
 // the survey module
@@ -9,32 +9,55 @@
 //
 // The upgrade function in this file will attempt
 // to perform all the necessary actions to upgrade
-// your older installtion to the current version.
+// your older installation to the current version.
 //
 // If there's something it cannot do itself, it
 // will tell you what you need to do.
 //
 // The commands in here will all be database-neutral,
-// using the functions defined in lib/ddllib.php
-
-function xmldb_survey_upgrade($oldversion=0) {
-
-    global $CFG, $THEME, $db;
-
-    $result = true;
+// using the methods of database_manager class
+//
+// Please do not forget to use upgrade_set_timeout()
+// before any action that may take longer time to finish.
 
-/// And upgrade begins here. For each one, you'll need one 
-/// block of code similar to the next one. Please, delete 
-/// this comment lines once this file start handling proper
-/// upgrade code.
+function xmldb_survey_upgrade($oldversion) {
+    global $CFG, $DB;
 
-/// if ($result && $oldversion < YYYYMMDD00) { //New version in version.php
-///     $result = result of "/lib/ddllib.php" function calls
-/// }
+    $dbman = $DB->get_manager();
 
 //===== 1.9.0 upgrade line ======//
 
-    return $result;
+    if ($oldversion < 2009042002) {
+
+    /// Define field introformat to be added to survey
+        $table = new xmldb_table('survey');
+        $field = new xmldb_field('introformat', XMLDB_TYPE_INTEGER, '4', XMLDB_UNSIGNED, XMLDB_NOTNULL, null, '0', 'intro');
+
+    /// Conditionally launch add field introformat
+        if (!$dbman->field_exists($table, $field)) {
+            $dbman->add_field($table, $field);
+        }
+
+        // conditionally migrate to html format in intro
+        if ($CFG->texteditors !== 'textarea') {
+            $rs = $DB->get_recordset('survey', array('introformat'=>FORMAT_MOODLE), '', 'id,intro,introformat');
+            foreach ($rs as $s) {
+                $s->intro       = text_to_html($s->intro, false, false, true);
+                $s->introformat = FORMAT_HTML;
+                $DB->update_record('survey', $s);
+                upgrade_set_timeout();
+            }
+            $rs->close();
+        }
+
+    /// survey savepoint reached
+        upgrade_mod_savepoint(true, 2009042002, 'survey');
+    }
+
+    // Moodle v2.1.0 release upgrade line
+    // Put any upgrade step following this
+
+    return true;
 }
 
-?>
+
