diff --git a/auth/pam/auth.php b/auth/pam/auth.php
index 48c9aa0..c0afbe0 100644
--- a/auth/pam/auth.php
+++ b/auth/pam/auth.php
@@ -55,8 +55,8 @@ class auth_plugin_pam extends auth_plugin_base {
      * Returns true if the username and password work and false if they are
      * wrong or don't exist.
      *
-     * @param string $username The username (with system magic quotes)
-     * @param string $password The password (with system magic quotes)
+     * @param string $username The username
+     * @param string $password The password
      * @return bool Authentication success or failure.
      */
     function user_login ($username, $password) {
@@ -68,7 +68,7 @@ class auth_plugin_pam extends auth_plugin_base {
 
         // call_time_pass_reference of errormessage is deprecated - throws warnings in multiauth
         //if (pam_auth($username, $password, &$errormessage)) {
-        if (pam_auth(stripslashes($username), stripslashes($password))) {
+        if (pam_auth($username, $password)) {
             return true;
         }
         else {
@@ -121,4 +121,4 @@ class auth_plugin_pam extends auth_plugin_base {
 
 }
 
-?>
+
