diff --git a/mnet/publickey.php b/mnet/publickey.php
index 7ed80c7..ac018ae 100644
--- a/mnet/publickey.php
+++ b/mnet/publickey.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 /**
  * Print this server's public key and exit
  *
@@ -10,7 +10,11 @@
 
 require_once(dirname(dirname(__FILE__)) . '/config.php');
 require_once $CFG->dirroot.'/mnet/lib.php';
-header("Content-type: text/plain");
+
+if ($CFG->mnet_dispatcher_mode === 'off') {
+    print_error('mnetdisabled', 'mnet');
+}
+
+header("Content-type: text/plain; charset=utf-8");
 $keypair = mnet_get_keypair();
 echo $keypair['certificate'];
-?>
