diff --git a/search/Zend/Search/Lucene/Index/SegmentInfo.php b/search/Zend/Search/Lucene/Index/SegmentInfo.php
index 4f84832..5270ee6 100644
--- a/search/Zend/Search/Lucene/Index/SegmentInfo.php
+++ b/search/Zend/Search/Lucene/Index/SegmentInfo.php
@@ -27,11 +27,7 @@ require_once 'Zend/Search/Lucene/Index/DictionaryLoader.php';
 require_once 'Zend/Search/Lucene/Exception.php';
 
 /** Zend_Search_Lucene_LockManager */
-if (!empty($CFG->block_search_softlock)){
-    require_once 'Zend/Search/Lucene/SoftLockManager.php';
-} else {
-    require_once 'Zend/Search/Lucene/LockManager.php';
-}
+require_once 'Zend/Search/Lucene/LockManager.php';
 
 
 /**
