diff --git a/search/Zend/Search/Lucene/Index/Writer.php b/search/Zend/Search/Lucene/Index/Writer.php
index 8442989..fb68151 100644
--- a/search/Zend/Search/Lucene/Index/Writer.php
+++ b/search/Zend/Search/Lucene/Index/Writer.php
@@ -30,12 +30,7 @@ require_once 'Zend/Search/Lucene/Index/SegmentInfo.php';
 require_once 'Zend/Search/Lucene/Index/SegmentMerger.php';
 
 /** Zend_Search_Lucene_LockManager */
-if (!empty($CFG->block_search_softlock)){
-    require_once 'Zend/Search/Lucene/SoftLockManager.php';
-} else {
-    require_once 'Zend/Search/Lucene/LockManager.php';
-}
-
+require_once 'Zend/Search/Lucene/LockManager.php';
 
 
 
