diff --git a/blocks/messages/block_messages.php b/blocks/messages/block_messages.php
index 7552b59..a9979c6 100644
--- a/blocks/messages/block_messages.php
+++ b/blocks/messages/block_messages.php
@@ -1,16 +1,19 @@
-<?PHP //$Id$
+<?php
 
 class block_messages extends block_base {
     function init() {
-        $this->title = get_string('messages','message');
-        $this->version = 2007101509;
+        $this->title = get_string('pluginname', 'block_messages');
     }
 
     function get_content() {
-        global $USER, $CFG;
+        global $USER, $CFG, $DB, $OUTPUT;
 
         if (!$CFG->messaging) {
-            return ''; 
+            $this->content->text = '';
+            if ($this->page->user_is_editing()) {
+                $this->content->text = get_string('disabled', 'message');
+            }
+            return $this->content;
         }
 
         if ($this->content !== NULL) {
@@ -21,19 +24,19 @@ class block_messages extends block_base {
         $this->content->text = '';
         $this->content->footer = '';
         
-        if (empty($this->instance) or empty($USER->id) or isguest() or empty($CFG->messaging)) {
+        if (empty($this->instance) or !isloggedin() or isguestuser() or empty($CFG->messaging)) {
             return $this->content;
         }
 
-        $this->content->footer = '<a href="'.$CFG->wwwroot.'/message/index.php" onclick="this.target=\'message\'; return openpopup(\'/message/index.php\', \'message\', \'menubar=0,location=0,scrollbars,status,resizable,width=400,height=500\', 0);">'.get_string('messages', 'message').'</a>...';
+        $link = '/message/index.php';
+        $action = null; //this was using popup_action() but popping up a fullsize window seems wrong
+        $this->content->footer = $OUTPUT->action_link($link, get_string('messages', 'message'), $action);
 
-        $users = get_records_sql("SELECT m.useridfrom as id, COUNT(m.useridfrom) as count,
-                                         u.firstname, u.lastname, u.picture, u.imagealt, u.lastaccess
-                                       FROM {$CFG->prefix}user u, 
-                                            {$CFG->prefix}message m 
-                                       WHERE m.useridto = '$USER->id' 
-                                         AND u.id = m.useridfrom
-                                    GROUP BY m.useridfrom, u.firstname,u.lastname,u.picture,u.lastaccess,u.imagealt");
+        $ufields = user_picture::fields('u', array('lastaccess'));
+        $users = $DB->get_records_sql("SELECT $ufields, COUNT(m.useridfrom) AS count
+                                         FROM {user} u, {message} m
+                                        WHERE m.useridto = ? AND u.id = m.useridfrom AND m.notification = 0
+                                     GROUP BY $ufields", array($USER->id));
 
 
         //Now, we have in users, the list of users to show
@@ -43,10 +46,16 @@ class block_messages extends block_base {
             foreach ($users as $user) {
                 $timeago = format_time(time() - $user->lastaccess);
                 $this->content->text .= '<li class="listentry"><div class="user"><a href="'.$CFG->wwwroot.'/user/view.php?id='.$user->id.'&amp;course='.SITEID.'" title="'.$timeago.'">';
-                $this->content->text .= print_user_picture($user, SITEID, $user->picture, 0, true, false, '', false);
+                $this->content->text .= $OUTPUT->user_picture($user, array('courseid'=>SITEID)); //TODO: user might not have capability to view frontpage profile :-(
                 $this->content->text .= fullname($user).'</a></div>';
-                $this->content->text .= '<div class="message"><a href="'.$CFG->wwwroot.'/message/discussion.php?id='.$user->id.'" onclick="this.target=\'message_'.$user->id.'\'; return openpopup(\'/message/discussion.php?id='.$user->id.'\', \'message_'.$user->id.'\', \'menubar=0,location=0,scrollbars,status,resizable,width=400,height=500\', 0);"><img class="iconsmall" src="'.$CFG->pixpath.'/t/message.gif" alt="" />&nbsp;'.$user->count.'</a>';
-                $this->content->text .= '</div></li>';
+
+                $link = '/message/index.php?usergroup=unread&id='.$user->id;
+                $anchortagcontents = '<img class="iconsmall" src="'.$OUTPUT->pix_url('t/message') . '" alt="" />&nbsp;'.$user->count;
+
+                $action = null; // popup is gone now
+                $anchortag = $OUTPUT->action_link($link, $anchortagcontents, $action);
+
+                $this->content->text .= '<div class="message">'.$anchortag.'</div></li>';
             }
             $this->content->text .= '</ul>'; 
         } else {
@@ -59,4 +68,4 @@ class block_messages extends block_base {
     }
 }
 
-?>
+
