diff --git a/mod/glossary/formats/entrylist/entrylist_format.php b/mod/glossary/formats/entrylist/entrylist_format.php
index 47b2e54..ae1b66b 100644
--- a/mod/glossary/formats/entrylist/entrylist_format.php
+++ b/mod/glossary/formats/entrylist/entrylist_format.php
@@ -1,7 +1,7 @@
-<?php  // $Id$
+<?php
 
-function glossary_show_entry_entrylist($course, $cm, $glossary, $entry, $mode='', $hook='', $printicons=1, $ratings=NULL, $aliases=true) {
-    global $USER;
+function glossary_show_entry_entrylist($course, $cm, $glossary, $entry, $mode='', $hook='', $printicons=1, $aliases=true) {
+    global $USER, $OUTPUT;
 
     $return = false;
 
@@ -11,17 +11,23 @@ function glossary_show_entry_entrylist($course, $cm, $glossary, $entry, $mode=''
     echo '<td class="entry">';
     if ($entry) {
         glossary_print_entry_approval($cm, $entry, $mode);
-        echo "<div class=\"concept\"><a href=\"showentry.php?courseid=$course->id&amp;eid=$entry->id&amp;displayformat=dictionary\" target=\"_blank\" onclick=\"return openpopup('/mod/glossary/showentry.php?courseid=$course->id&amp;eid=$entry->id&amp;displayformat=dictionary', 'entry', 'menubar=0,location=0,scrollbars,resizable,width=600,height=450', 0);\">";
-        glossary_print_entry_concept($entry);
-        echo '</a></div> ';
+
+        $anchortagcontents = glossary_print_entry_concept($entry, true);
+
+        $link = "/mod/glossary/showentry.php?courseid={$course->id}&eid={$entry->id}&displayformat=dictionary";
+        $action = new popup_action('click', $link.'&popup=1', 'entry',array('title'=>'entry','width'=>600,'height'=>450));
+
+        $anchor = $OUTPUT->action_link($link, $anchortagcontents, $action);
+
+        echo "<div class=\"concept\">$anchor</div> ";
         echo '</td><td align="right" class="entrylowersection">';
         if ($printicons) {
             glossary_print_entry_icons($course, $cm, $glossary, $entry, $mode, $hook,'print');
         }
-        if ($ratings) {
+        if (!empty($entry->rating)) {
             echo '<br />';
             echo '<span class="ratings">';
-            $return = glossary_print_entry_ratings($course, $entry, $ratings);
+            $return = glossary_print_entry_ratings($course, $entry);
             echo '</span>';
         }
         echo '<br />';
@@ -36,7 +42,7 @@ function glossary_show_entry_entrylist($course, $cm, $glossary, $entry, $mode=''
     return $return;
 }
 
-function glossary_print_entry_entrylist($course, $cm, $glossary, $entry, $mode='', $hook='', $printicons=1, $ratings=NULL) {
+function glossary_print_entry_entrylist($course, $cm, $glossary, $entry, $mode='', $hook='', $printicons=1) {
 
     //The print view for this format is different from the normal view, so we implement it here completely
     global $CFG, $USER;
@@ -51,13 +57,11 @@ function glossary_print_entry_entrylist($course, $cm, $glossary, $entry, $mode='
     echo '<b>';
     glossary_print_entry_concept($entry);
     echo ':</b> ';
-    glossary_print_entry_definition($entry);
-    $return = glossary_print_entry_lower_section($course, $cm, $glossary, $entry, $mode, $hook, false, false, false);
+    glossary_print_entry_definition($entry, $glossary, $cm);
+    glossary_print_entry_lower_section($course, $cm, $glossary, $entry, $mode, $hook, false, false);
     echo '</td>';
     echo '</tr>';
     echo "</table>\n";
-
-    return $return;
 }
 
-?>
+
