diff --git a/grade/grading/form/rubric/version.php b/grade/grading/form/rubric/version.php
index 020b4ae..d695639 100644
--- a/grade/grading/form/rubric/version.php
+++ b/grade/grading/form/rubric/version.php
@@ -25,8 +25,8 @@
 defined('MOODLE_INTERNAL') || die();
 
 $plugin->component  = 'gradingform_rubric';
-$plugin->version    = 2013051401.05;
+$plugin->version    = 2013050100;
 
-$plugin->requires   = 2013051401.05;
+$plugin->requires   = 2013050100;
 
 $plugin->maturity   = MATURITY_STABLE;
