diff --git a/lib/tests/moodlelib_test.php b/lib/tests/moodlelib_test.php
index 4f8d6dc..b1e4a1c 100644
--- a/lib/tests/moodlelib_test.php
+++ b/lib/tests/moodlelib_test.php
@@ -2358,6 +2358,7 @@ class moodlelib_testcase extends advanced_testcase {
     }
 
     public function test_email_to_user() {
+        global $CFG;
         $this->resetAfterTest();
 
         $user1 = $this->getDataGenerator()->create_user();
@@ -2387,5 +2388,21 @@ class moodlelib_testcase extends advanced_testcase {
         $this->assertSame($messagetext2, trim($result[1]->body));
         $this->assertSame($user2->email, $result[1]->to);
         $this->assertSame($user1->email, $result[1]->from);
+
+        email_to_user($user1, $user2, $subject, $messagetext);
+        $this->assertDebuggingCalled('Unit tests must not send real emails! Use $this->start_phpmailer_redirection()');
+
+        // Test $CFG->emailonlyfromnoreplyaddress.
+        set_config('emailonlyfromnoreplyaddress', 1);
+        $this->assertNotEmpty($CFG->emailonlyfromnoreplyaddress);
+        $sink = $this->redirectEmails();
+        email_to_user($user1, $user2, $subject, $messagetext);
+        unset_config('emailonlyfromnoreplyaddress');
+        email_to_user($user1, $user2, $subject, $messagetext);
+        $result = $sink->get_messages();
+        $this->assertEquals($CFG->noreplyaddress, $result[0]->from);
+        $this->assertNotEquals($CFG->noreplyaddress, $result[1]->from);
+        $sink->close();
     }
+
 }
