diff --git a/lib/form/button.php b/lib/form/button.php
index 31aa667..eb16bad 100644
--- a/lib/form/button.php
+++ b/lib/form/button.php
@@ -16,8 +16,6 @@
 // | Authors: Adam Daniel <adaniel1@eesus.jnj.com>                        |
 // |          Bertrand Mansion <bmansion@mamasam.com>                     |
 // +----------------------------------------------------------------------+
-//
-// $Id$
 
 require_once("HTML/QuickForm/button.php");
 
@@ -49,18 +47,7 @@ class MoodleQuickForm_button extends HTML_QuickForm_button
      * @param string $function function name to call to get html
      */
     function setHelpButton($helpbuttonargs, $function='helpbutton'){
-        if (!is_array($helpbuttonargs)){
-            $helpbuttonargs=array($helpbuttonargs);
-        }else{
-            $helpbuttonargs=$helpbuttonargs;
-        }
-        //we do this to to return html instead of printing it
-        //without having to specify it in every call to make a button.
-        if ('helpbutton' == $function){
-            $defaultargs=array('', '', 'moodle', true, false, '', true);
-            $helpbuttonargs=$helpbuttonargs + $defaultargs ;
-        }
-        $this->_helpbutton=call_user_func_array($function, $helpbuttonargs);
+        debugging('component setHelpButton() is not used any more, please use $mform->setHelpButton() instead');
     }
     /**
      * get html for help button
@@ -84,4 +71,3 @@ class MoodleQuickForm_button extends HTML_QuickForm_button
         }
     }
 } //end class MoodleQuickForm_button
-?>
