diff --git a/lib/form/checkbox.php b/lib/form/checkbox.php
index d3c0262..237c796 100644
--- a/lib/form/checkbox.php
+++ b/lib/form/checkbox.php
@@ -25,18 +25,7 @@ class MoodleQuickForm_checkbox extends HTML_QuickForm_checkbox{
      * @param string $function function name to call to get html
      */
     function setHelpButton($helpbuttonargs, $function='helpbutton'){
-        if (!is_array($helpbuttonargs)){
-            $helpbuttonargs=array($helpbuttonargs);
-        }else{
-            $helpbuttonargs=$helpbuttonargs;
-        }
-        //we do this to to return html instead of printing it
-        //without having to specify it in every call to make a button.
-        if ('helpbutton' == $function){
-            $defaultargs=array('', '', 'moodle', true, false, '', true);
-            $helpbuttonargs=$helpbuttonargs + $defaultargs ;
-        }
-        $this->_helpbutton=call_user_func_array($function, $helpbuttonargs);
+        debugging('component setHelpButton() is not used any more, please use $mform->setHelpButton() instead');
     }
     /**
      * get html for help button
@@ -124,4 +113,3 @@ class MoodleQuickForm_checkbox extends HTML_QuickForm_checkbox{
         return $output;
     } //end func getFrozenHtml
 }
-?>
\ No newline at end of file
