diff --git a/lib/form/dateselector.php b/lib/form/dateselector.php
index 38d035d..49814a1 100644
--- a/lib/form/dateselector.php
+++ b/lib/form/dateselector.php
@@ -149,7 +149,7 @@ class MoodleQuickForm_date_selector extends MoodleQuickForm_group
                     $value = time();
                 }
                 if (!is_array($value)) {
-                    $currentdate = usergetdate($value);
+                    $currentdate = usergetdate($value, $this->_options['timezone']);
                     $value = array(
                         'day' => $currentdate['mday'],
                         'month' => $currentdate['mon'],
@@ -192,9 +192,30 @@ class MoodleQuickForm_date_selector extends MoodleQuickForm_group
     // }}}
     // {{{ accept()
 
+    /**
+    * RL edit
+    * Accepts a renderer
+    *
+    * @param object     An HTML_QuickForm_Renderer object
+    * @param bool       Whether a group is required
+    * @param string     An error message associated with a group
+    * @access public
+    * @return void
+    */
     function accept(&$renderer, $required = false, $error = null)
     {
+        //only display a disable static element or don't display the optional check box if this element is frozen
+         // 3 is the index where the checkbox is located
+        if($this->isFrozen() && $this->_options['optional'] && !$this->_elements[3]->getChecked()){
+            $this->_elements = array();
+            $this->_elements[] =& MoodleQuickForm::createElement('static', 'disabled', '', get_string('disabled', 'filters'));
+        } else if($this->isFrozen() && isset($this->_elements[3])) {
+            // Remove the frozen checkbox when frozen
+            unset($this->_elements[3]);
+        }
+
         $renderer->renderElement($this, $required, $error);
+        //RL edit end
     }
 
     // }}}
