diff --git a/lib/form/dateselector.php b/lib/form/dateselector.php
index cb7ab39..e160907 100644
--- a/lib/form/dateselector.php
+++ b/lib/form/dateselector.php
@@ -238,6 +238,17 @@ class MoodleQuickForm_date_selector extends MoodleQuickForm_group
      */
     function accept(&$renderer, $required = false, $error = null)
     {
+        // RL Edit - only display a disable static element or don't display the optional check box if this element is frozen
+        // 4 is the index where the checkbox is located
+        if ($this->isFrozen() && $this->_options['optional'] && !$this->_elements[4]->getChecked()){
+            $this->_elements = array();
+            $this->_elements[] =& MoodleQuickForm::createElement('static', 'disabled', '', get_string('disabled', 'filters'));
+        } else if ($this->isFrozen() && isset($this->_elements[3])) {
+            // Remove the frozen checkbox when frozen
+            unset($this->_elements[3]);
+        }
+        // End: RL Edit
+
         $renderer->renderElement($this, $required, $error);
     }
 
