diff --git a/lib/form/dateselector.php b/lib/form/dateselector.php
index 3687e61..49814a1 100644
--- a/lib/form/dateselector.php
+++ b/lib/form/dateselector.php
@@ -1,15 +1,38 @@
 <?php
+
+///////////////////////////////////////////////////////////////////////////
+//                                                                       //
+// NOTICE OF COPYRIGHT                                                   //
+//                                                                       //
+// Moodle - Modular Object-Oriented Dynamic Learning Environment         //
+//          http://moodle.org                                            //
+//                                                                       //
+// Copyright (C) 1999 onwards Martin Dougiamas  http://dougiamas.com     //
+//                                                                       //
+// This program is free software; you can redistribute it and/or modify  //
+// it under the terms of the GNU General Public License as published by  //
+// the Free Software Foundation; either version 2 of the License, or     //
+// (at your option) any later version.                                   //
+//                                                                       //
+// This program is distributed in the hope that it will be useful,       //
+// but WITHOUT ANY WARRANTY; without even the implied warranty of        //
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         //
+// GNU General Public License for more details:                          //
+//                                                                       //
+//          http://www.gnu.org/copyleft/gpl.html                         //
+//                                                                       //
+///////////////////////////////////////////////////////////////////////////
+
 global $CFG;
-require_once "$CFG->libdir/form/group.php";
-require_once "$CFG->libdir/formslib.php";
+require_once($CFG->libdir . '/form/group.php');
+require_once($CFG->libdir . '/formslib.php');
 
 /**
  * Class for a group of elements used to input a date.
  *
  * Emulates moodle print_date_selector function
  *
- * @author Jamie Pratt <me@jamiep.org>
- * @access public
+ * @package formslib
  */
 class MoodleQuickForm_date_selector extends MoodleQuickForm_group
 {
@@ -22,15 +45,15 @@ class MoodleQuickForm_date_selector extends MoodleQuickForm_group
     * applydst => apply users daylight savings adjustment?
     * optional => if true, show a checkbox beside the date to turn it on (or off)
     */
-    var $_options = array('startyear'=>1970, 'stopyear'=>2020,
-                    'timezone'=>99, 'applydst'=>true, 'optional'=>false);
+    protected $_options = array('startyear' => 1970, 'stopyear' => 2020,
+            'timezone' => 99, 'applydst' => true, 'optional' => false);
 
    /**
     * These complement separators, they are appended to the resultant HTML
     * @access   private
     * @var      array
     */
-    var $_wrap = array('', '');
+    protected $_wrap = array('', '');
 
    /**
     * Class constructor
@@ -59,6 +82,7 @@ class MoodleQuickForm_date_selector extends MoodleQuickForm_group
                 }
             }
         }
+        form_init_date_js();
     }
 
     // }}}
@@ -81,7 +105,7 @@ class MoodleQuickForm_date_selector extends MoodleQuickForm_group
         $this->_elements[] =& MoodleQuickForm::createElement('select', 'year', get_string('year', 'form'), $years, $this->getAttributes(), true);
         // If optional we add a checkbox which the user can use to turn if on
         if($this->_options['optional']) {
-            $this->_elements[] =& MoodleQuickForm::createElement('checkbox', 'off', null, get_string('disable'), $this->getAttributes(), true);
+            $this->_elements[] =& MoodleQuickForm::createElement('checkbox', 'enabled', null, get_string('enable'), $this->getAttributes(), true);
         }
         foreach ($this->_elements as $element){
             if (method_exists($element, 'setHiddenLabel')){
@@ -125,25 +149,27 @@ class MoodleQuickForm_date_selector extends MoodleQuickForm_group
                     $value = time();
                 }
                 if (!is_array($value)) {
-                    $currentdate = usergetdate($value);
+                    $currentdate = usergetdate($value, $this->_options['timezone']);
                     $value = array(
                         'day' => $currentdate['mday'],
                         'month' => $currentdate['mon'],
                         'year' => $currentdate['year']);
                     // If optional, default to off, unless a date was provided
                      if($this->_options['optional']) {
-                        $value['off'] = ($requestvalue == 0) ? true : false;
+                        $value['enabled'] = $requestvalue != 0;
                     }
                 } else {
-                    $value['off'] = (isset($value['off'])) ? true : false;
+                    $value['enabled'] = isset($value['enabled']);
                 }
                 if (null !== $value){
                     $this->setValue($value);
                 }
                 break;
             case 'createElement':
-                if($arg[2]['optional']) {
-                    $caller->disabledIf($arg[0], $arg[0].'[off]', 'checked');
+                // Optional is an optional param, if its set we need to add a disabledIf rule.
+                // If its empty or not specified then its not an optional dateselector.
+                if (!empty($arg[2]['optional']) && !empty($arg[0])) {
+                    $caller->disabledIf($arg[0], $arg[0].'[enabled]');
                 }
                 return parent::onQuickFormEvent($event, $arg, $caller);
                 break;
@@ -157,7 +183,7 @@ class MoodleQuickForm_date_selector extends MoodleQuickForm_group
     function toHtml()
     {
         include_once('HTML/QuickForm/Renderer/Default.php');
-        $renderer =& new HTML_QuickForm_Renderer_Default();
+        $renderer = new HTML_QuickForm_Renderer_Default();
         $renderer->setElementTemplate('{element}');
         parent::accept($renderer);
         return $this->_wrap[0] . $renderer->toHtml() . $this->_wrap[1];
@@ -166,9 +192,30 @@ class MoodleQuickForm_date_selector extends MoodleQuickForm_group
     // }}}
     // {{{ accept()
 
+    /**
+    * RL edit
+    * Accepts a renderer
+    *
+    * @param object     An HTML_QuickForm_Renderer object
+    * @param bool       Whether a group is required
+    * @param string     An error message associated with a group
+    * @access public
+    * @return void
+    */
     function accept(&$renderer, $required = false, $error = null)
     {
+        //only display a disable static element or don't display the optional check box if this element is frozen
+         // 3 is the index where the checkbox is located
+        if($this->isFrozen() && $this->_options['optional'] && !$this->_elements[3]->getChecked()){
+            $this->_elements = array();
+            $this->_elements[] =& MoodleQuickForm::createElement('static', 'disabled', '', get_string('disabled', 'filters'));
+        } else if($this->isFrozen() && isset($this->_elements[3])) {
+            // Remove the frozen checkbox when frozen
+            unset($this->_elements[3]);
+        }
+
         $renderer->renderElement($this, $required, $error);
+        //RL edit end
     }
 
     // }}}
@@ -193,16 +240,16 @@ class MoodleQuickForm_date_selector extends MoodleQuickForm_group
         if (count($valuearray)){
             if($this->_options['optional']) {
                 // If checkbox is on, the value is zero, so go no further
-                if(!empty($valuearray['off'])) {
-                    $value[$this->getName()]=0;
+                if(empty($valuearray['enabled'])) {
+                    $value[$this->getName()] = 0;
                     return $value;
                 }
             }
 
-            $value[$this->getName()]=make_timestamp($valuearray['year'],
+            $value[$this->getName()] = make_timestamp($valuearray['year'],
                                    $valuearray['month'],
                                    $valuearray['day'],
-                                   0,0,0,
+                                   0, 0, 0,
                                    $this->_options['timezone'],
                                    $this->_options['applydst']);
 
@@ -214,4 +261,3 @@ class MoodleQuickForm_date_selector extends MoodleQuickForm_group
 
     // }}}
 }
-?>
