diff --git a/lib/form/dateselector.php b/lib/form/dateselector.php
index 3687e61..650b294 100644
--- a/lib/form/dateselector.php
+++ b/lib/form/dateselector.php
@@ -125,11 +125,11 @@ class MoodleQuickForm_date_selector extends MoodleQuickForm_group
                     $value = time();
                 }
                 if (!is_array($value)) {
-                    $currentdate = usergetdate($value);
+                    $currentdate = usergetdate($value, $this->_options['timezone']);
                     $value = array(
-                        'day' => $currentdate['mday'],
-                        'month' => $currentdate['mon'],
-                        'year' => $currentdate['year']);
+                        'day' => (int)$currentdate['mday'],
+                        'month' => (int)$currentdate['mon'],
+                        'year' => (int)$currentdate['year']);
                     // If optional, default to off, unless a date was provided
                      if($this->_options['optional']) {
                         $value['off'] = ($requestvalue == 0) ? true : false;
