diff --git a/lib/form/file.php b/lib/form/file.php
index 6d7726f..2f69f60 100644
--- a/lib/form/file.php
+++ b/lib/form/file.php
@@ -15,6 +15,7 @@ class MoodleQuickForm_file extends HTML_QuickForm_file{
      */
     var $_helpbutton='';
     function MoodleQuickForm_file($elementName=null, $elementLabel=null, $attributes=null) {
+        debugging('file forms element is deprecated, please use new filepicker instead');
         parent::HTML_QuickForm_file($elementName, $elementLabel, $attributes);
     }
     /**
@@ -25,18 +26,7 @@ class MoodleQuickForm_file extends HTML_QuickForm_file{
      * @param string $function function name to call to get html
      */
     function setHelpButton($helpbuttonargs, $function='helpbutton'){
-        if (!is_array($helpbuttonargs)){
-            $helpbuttonargs=array($helpbuttonargs);
-        }else{
-            $helpbuttonargs=$helpbuttonargs;
-        }
-        //we do this to to return html instead of printing it
-        //without having to specify it in every call to make a button.
-        if ('helpbutton' == $function){
-            $defaultargs=array('', '', 'moodle', true, false, '', true);
-            $helpbuttonargs=$helpbuttonargs + $defaultargs ;
-        }
-        $this->_helpbutton=call_user_func_array($function, $helpbuttonargs);
+        debugging('component setHelpButton() is not used any more, please use $mform->setHelpButton() instead');
     }
     /**
      * set html for help button
@@ -75,4 +65,3 @@ class MoodleQuickForm_file extends HTML_QuickForm_file{
     }
 
 }
-?>
\ No newline at end of file
