diff --git a/lib/form/filemanager.js b/lib/form/filemanager.js
index cbd528a..1e879f2 100644
--- a/lib/form/filemanager.js
+++ b/lib/form/filemanager.js
@@ -77,6 +77,7 @@ M.form_filemanager.init = function(Y, options) {
             this.filepicker_options.maxbytes = this.maxbytes;
             this.filepicker_options.env = 'filemanager';
             this.filepicker_options.itemid = options.itemid;
+            this.filepicker_options.accepted_types = options.accepted_types;
 
             if (options.filecount) {
                 this.filecount = options.filecount;
