diff --git a/lib/form/form.js b/lib/form/form.js
index db5def3..c747a0f 100644
--- a/lib/form/form.js
+++ b/lib/form/form.js
@@ -29,6 +29,7 @@ M.form.initShowAdvanced = function(Y, config) {
             this._advButtons = Y.all('.showadvancedbtn');
             if (this._advButtons.size() > 0) {
                 this._stateInput = new Y.NodeList(document.getElementsByName('mform_showadvanced_last'));
+                this._advButtons.set('type', 'button');
                 this._advButtons.on('click', this.switchState, this);
                 this._advButtons.set('type', 'button');
             }
