diff --git a/lib/form/header.php b/lib/form/header.php
index 7270402..9f22503 100644
--- a/lib/form/header.php
+++ b/lib/form/header.php
@@ -1,5 +1,4 @@
 <?php
-// $Id$
 
 require_once 'HTML/QuickForm/header.php';
 
@@ -44,18 +43,7 @@ class MoodleQuickForm_header extends HTML_QuickForm_header
      * @param string $function function name to call to get html
      */
     function setHelpButton($helpbuttonargs, $function='helpbutton'){
-        if (!is_array($helpbuttonargs)){
-            $helpbuttonargs=array($helpbuttonargs);
-        }else{
-            $helpbuttonargs=$helpbuttonargs;
-        }
-        //we do this to to return html instead of printing it
-        //without having to specify it in every call to make a button.
-        if ('helpbutton' == $function){
-            $defaultargs=array('', '', 'moodle', true, false, '', true);
-            $helpbuttonargs=$helpbuttonargs + $defaultargs ;
-        }
-        $this->_helpbutton=call_user_func_array($function, $helpbuttonargs);
+        debugging('component setHelpButton() is not used any more, please use $mform->setHelpButton() instead');
     }
     /**
      * get html for help button
@@ -67,4 +55,3 @@ class MoodleQuickForm_header extends HTML_QuickForm_header
         return $this->_helpbutton;
     }
 } //end class MoodleQuickForm_header
-?>
