diff --git a/lib/form/recaptcha.php b/lib/form/recaptcha.php
index 9b0fc2a..0282efd 100644
--- a/lib/form/recaptcha.php
+++ b/lib/form/recaptcha.php
@@ -1,4 +1,6 @@
 <?php
+require_once('HTML/QuickForm/input.php');
+
 /* vim: set expandtab tabstop=4 shiftwidth=4 softtabstop=4: */
 
 /**
@@ -7,7 +9,6 @@
  * @category  Admin
  * @package   admin
  * @author    Nicolas Connault <nicolasconnault@gmail.com>
- * @version   $Id$
  */
 
 
@@ -51,16 +52,11 @@ class MoodleQuickForm_recaptcha extends HTML_QuickForm_input {
      * @return    string
      */
     function toHtml() {
-        global $CFG;
+        global $CFG, $PAGE;
         require_once $CFG->libdir . '/recaptchalib.php';
 
-        //Accessibility: don't specify a tabindex  MDL-20144
-        $html = '<script type="text/javascript">
-            var RecaptchaOptions = {
-                theme : \'custom\',
-                custom_theme_widget : \'recaptcha_widget\'
-            };
-              </script>' . "\n";
+        $recaptureoptions = Array('theme'=>'custom', 'custom_theme_widget'=>'recaptcha_widget');
+        $html = html_writer::script(js_writer::set_variable('RecaptchaOptions', $recaptureoptions));
 
         $attributes = $this->getAttributes();
         if (empty($attributes['error_message'])) {
@@ -104,18 +100,7 @@ class MoodleQuickForm_recaptcha extends HTML_QuickForm_input {
      * @param string $function function name to call to get html
      */
     function setHelpButton($helpbuttonargs, $function='helpbutton'){
-        if (!is_array($helpbuttonargs)){
-            $helpbuttonargs=array($helpbuttonargs);
-        }else{
-            $helpbuttonargs=$helpbuttonargs;
-        }
-        //we do this to to return html instead of printing it
-        //without having to specify it in every call to make a button.
-        if ('helpbutton' == $function){
-            $defaultargs=array('', '', 'moodle', true, false, '', true);
-            $helpbuttonargs=$helpbuttonargs + $defaultargs ;
-        }
-        $this->_helpbutton=call_user_func_array($function, $helpbuttonargs);
+        debugging('component setHelpButton() is not used any more, please use $mform->setHelpButton() instead');
     }
     
     /**
@@ -145,5 +130,3 @@ class MoodleQuickForm_recaptcha extends HTML_QuickForm_input {
         return true;
     }
 }
-
-?>
