diff --git a/lib/form/selectgroups.php b/lib/form/selectgroups.php
index 3c85b1a..01ee9ed 100644
--- a/lib/form/selectgroups.php
+++ b/lib/form/selectgroups.php
@@ -16,8 +16,6 @@
 // | Authors: Adam Daniel <adaniel1@eesus.jnj.com>                        |
 // |          Bertrand Mansion <bmansion@mamasam.com>                     |
 // +----------------------------------------------------------------------+
-//
-// $Id$
 
 require_once('HTML/QuickForm/element.php');
 
@@ -103,7 +101,7 @@ class MoodleQuickForm_selectgroups extends HTML_QuickForm_element {
     function setSelected($values)
     {
         if (is_string($values) && $this->getMultiple()) {
-            $values = split("[ ]?,[ ]?", $values);
+            $values = preg_split("/[ ]?,[ ]?/", $values);
         }
         if (is_array($values)) {
             $this->_values = array_values($values);
@@ -576,18 +574,7 @@ class MoodleQuickForm_selectgroups extends HTML_QuickForm_element {
      * @param string $function function name to call to get html
      */
     function setHelpButton($helpbuttonargs, $function='helpbutton'){
-        if (!is_array($helpbuttonargs)){
-            $helpbuttonargs=array($helpbuttonargs);
-        }else{
-            $helpbuttonargs=$helpbuttonargs;
-        }
-        //we do this to to return html instead of printing it
-        //without having to specify it in every call to make a button.
-        if ('helpbutton' == $function){
-            $defaultargs=array('', '', 'moodle', true, false, '', true);
-            $helpbuttonargs=$helpbuttonargs + $defaultargs ;
-        }
-        $this->_helpbutton=call_user_func_array($function, $helpbuttonargs);
+        debugging('component setHelpButton() is not used any more, please use $mform->setHelpButton() instead');
     }
     /**
      * get html for help button
@@ -614,4 +601,3 @@ class MoodleQuickForm_selectgroups extends HTML_QuickForm_element {
         }
     } 
 }
-?>
