diff --git a/lib/form/select.php b/lib/form/select.php
index 3054b5c..f6bcfbd 100644
--- a/lib/form/select.php
+++ b/lib/form/select.php
@@ -57,18 +57,7 @@ class MoodleQuickForm_select extends HTML_QuickForm_select{
      * @param string $function function name to call to get html
      */
     function setHelpButton($helpbuttonargs, $function='helpbutton'){
-        if (!is_array($helpbuttonargs)){
-            $helpbuttonargs=array($helpbuttonargs);
-        }else{
-            $helpbuttonargs=$helpbuttonargs;
-        }
-        //we do this to to return html instead of printing it
-        //without having to specify it in every call to make a button.
-        if ('helpbutton' == $function){
-            $defaultargs=array('', '', 'moodle', true, false, '', true);
-            $helpbuttonargs=$helpbuttonargs + $defaultargs ;
-        }
-        $this->_helpbutton=call_user_func_array($function, $helpbuttonargs);
+        debugging('component setHelpButton() is not used any more, please use $mform->setHelpButton() instead');
     }
     /**
      * get html for help button
@@ -128,7 +117,6 @@ class MoodleQuickForm_select extends HTML_QuickForm_select{
             return 'default';
         }
     }
-
    /**
     * We check the options and return only the values that _could_ have been
     * selected. We also return a scalar value if select is not "multiple"
@@ -165,5 +153,3 @@ class MoodleQuickForm_select extends HTML_QuickForm_select{
         }
     }
 }
-
-?>
\ No newline at end of file
