diff --git a/lib/form/static.php b/lib/form/static.php
index 56274c0..8258439 100644
--- a/lib/form/static.php
+++ b/lib/form/static.php
@@ -26,18 +26,7 @@ class MoodleQuickForm_static extends HTML_QuickForm_static{
      * @param string $function function name to call to get html
      */
     function setHelpButton($helpbuttonargs, $function='helpbutton'){
-        if (!is_array($helpbuttonargs)){
-            $helpbuttonargs=array($helpbuttonargs);
-        }else{
-            $helpbuttonargs=$helpbuttonargs;
-        }
-        //we do this to to return html instead of printing it
-        //without having to specify it in every call to make a button.
-        if ('helpbutton' == $function){
-            $defaultargs=array('', '', 'moodle', true, false, '', true);
-            $helpbuttonargs=$helpbuttonargs + $defaultargs ;
-        }
-        $this->_helpbutton=call_user_func_array($function, $helpbuttonargs);
+        debugging('component setHelpButton() is not used any more, please use $mform->setHelpButton() instead');
     }
     /**
      * get html for help button
@@ -53,4 +42,3 @@ class MoodleQuickForm_static extends HTML_QuickForm_static{
         return $this->_elementTemplateType;
     }
 }
-?>
\ No newline at end of file
