diff --git a/mod/data/field/menu/field.class.php b/mod/data/field/menu/field.class.php
old mode 100755
new mode 100644
index 7b374e8..6a948a1
--- a/mod/data/field/menu/field.class.php
+++ b/mod/data/field/menu/field.class.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 ///////////////////////////////////////////////////////////////////////////
 //                                                                       //
 // NOTICE OF COPYRIGHT                                                   //
@@ -26,14 +26,11 @@ class data_field_menu extends data_field_base {
 
     var $type = 'menu';
 
-    function data_field_menu($field=0, $data=0) {
-        parent::data_field_base($field, $data);
-    }
-
     function display_add_field($recordid=0) {
+        global $DB, $OUTPUT;
 
         if ($recordid){
-            $content = get_field('data_content', 'content', 'fieldid', $this->field->id, 'recordid', $recordid);
+            $content = $DB->get_field('data_content', 'content', array('fieldid'=>$this->field->id, 'recordid'=>$recordid));
             $content = trim($content);
         } else {
             $content = '';
@@ -49,8 +46,8 @@ class data_field_menu extends data_field_base {
             }
         }
 
-        $str .= choose_from_menu($options, 'field_'.$this->field->id, $content,
-                                 get_string('menuchoose', 'data'), '', '', true, false, 0, 'field_'.$this->field->id);
+
+        $str .= html_writer::select($options, 'field_'.$this->field->id, $content, array(''=>get_string('menuchoose', 'data')), array('id'=>'field_'.$this->field->id));
 
         $str .= '</div>';
 
@@ -58,15 +55,15 @@ class data_field_menu extends data_field_base {
     }
 
     function display_search_field($content = '') {
-        global $CFG;
+        global $CFG, $DB;
 
-        $varcharcontent = sql_compare_text('content', 255);
+        $varcharcontent =  $DB->sql_compare_text('content', 255);
         $sql = "SELECT DISTINCT $varcharcontent AS content
-                  FROM {$CFG->prefix}data_content
-                 WHERE fieldid={$this->field->id} AND content IS NOT NULL";
+                  FROM {data_content}
+                 WHERE fieldid=? AND content IS NOT NULL";
 
         $usedoptions = array();
-        if ($used = get_records_sql($sql)) {
+        if ($used = $DB->get_records_sql($sql, array($this->field->id))) {
             foreach ($used as $data) {
                 $value = $data->content;
                 if ($value === '') {
@@ -89,7 +86,7 @@ class data_field_menu extends data_field_base {
             return '';
         }
 
-        return choose_from_menu($options, 'f_'.$this->field->id, stripslashes($content), 'choose', '', 0, true);
+        return html_writer::select($options, 'f_'.$this->field->id, $content);
     }
 
      function parse_search_field() {
@@ -97,10 +94,16 @@ class data_field_menu extends data_field_base {
      }
 
     function generate_sql($tablealias, $value) {
-        $varcharcontent = sql_compare_text("{$tablealias}.content", 255);
-        return " ({$tablealias}.fieldid = {$this->field->id} AND $varcharcontent = '$value') ";
+        global $DB;
+
+        static $i=0;
+        $i++;
+        $name = "df_menu_$i";
+        $varcharcontent = $DB->sql_compare_text("{$tablealias}.content", 255);
+
+        return array(" ({$tablealias}.fieldid = {$this->field->id} AND $varcharcontent = :$name) ", array($name=>$value));
     }
 
 }
 
-?>
+
