diff --git a/course/report/participation/mod.php b/course/report/participation/mod.php
index 3842814..5c1fbff 100644
--- a/course/report/participation/mod.php
+++ b/course/report/participation/mod.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 
     if (!defined('MOODLE_INTERNAL')) {
         die('Direct access to this script is forbidden.');    ///  It must be included from a Moodle page
@@ -11,4 +11,4 @@
         echo "$participationreport</a>\n";
         echo '</p>';
     }
-?>
+
