diff --git a/grade/report/grader/db/access.php b/grade/report/grader/db/access.php
index 5906e90..41715d8 100644
--- a/grade/report/grader/db/access.php
+++ b/grade/report/grader/db/access.php
@@ -15,18 +15,18 @@
 // You should have received a copy of the GNU General Public License
 // along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
 
-$gradereport_grader_capabilities = array(
+$capabilities = array(
 
     'gradereport/grader:view' => array(
         'riskbitmask' => RISK_PERSONAL,
         'captype' => 'read',
         'contextlevel' => CONTEXT_COURSE,
-        'legacy' => array(
+        'archetypes' => array(
             'teacher' => CAP_ALLOW,
             'editingteacher' => CAP_ALLOW,
-            'admin' => CAP_ALLOW
+            'manager' => CAP_ALLOW
         )
     )
 );
 
-?>
+
