diff --git a/admin/roles/allow.php b/admin/roles/allow.php
index 52078e7..75496ff 100644
--- a/admin/roles/allow.php
+++ b/admin/roles/allow.php
@@ -55,6 +55,9 @@ if (optional_param('submit', false, PARAM_BOOL) && data_submitted() && confirm_s
 
 $controller->load_current_settings();
 
+// ELIS-3687: this fix may change if/when MDL-30036 is addressed!
+$PAGE->requires->css('/admin/styles.css'); // ELIS-3687
+
 // Display the editing form.
 echo $OUTPUT->header();
 
@@ -65,10 +68,14 @@ $table = $controller->get_table();
 
 echo $OUTPUT->box($controller->get_intro_text());
 
+echo '<div class="role_tables">'; // ELIS-3687
+
 echo '<form action="' . $baseurl . '" method="post">';
 echo '<input type="hidden" name="sesskey" value="' . sesskey() . '" />';
 echo html_writer::table($table);
 echo '<div class="buttons"><input type="submit" name="submit" value="'.get_string('savechanges').'"/>';
 echo '</div></form>';
 
+echo '</div>'; // ELIS-3687
+
 echo $OUTPUT->footer();
