diff --git a/version.php b/version.php
index f89d604..4312eeb 100644
--- a/version.php
+++ b/version.php
@@ -30,7 +30,7 @@
 defined('MOODLE_INTERNAL') || die();
 
 
-$version  = 2012062504.00;              // YYYYMMDD      = weekly release date of this DEV branch
+$version  = 2012062504.01;              // YYYYMMDD      = weekly release date of this DEV branch
                                         //         RR    = release increments - 00 in DEV branches
                                         //           .XX = incremental changes
 
