diff --git a/grade/export/grade_export_form.php b/grade/export/grade_export_form.php
old mode 100755
new mode 100644
index 2b83815..540694c
--- a/grade/export/grade_export_form.php
+++ b/grade/export/grade_export_form.php
@@ -23,7 +23,7 @@ require_once $CFG->libdir.'/formslib.php';
 
 class grade_export_form extends moodleform {
     function definition() {
-        global $CFG, $COURSE, $USER;
+        global $CFG, $COURSE, $USER, $DB;
 
         $mform =& $this->_form;
         if (isset($this->_customdata)) {  // hardcoding plugin names here is hacky
@@ -82,22 +82,24 @@ class grade_export_form extends moodleform {
         if (!empty($CFG->gradepublishing) and !empty($features['publishing'])) {
             $mform->addElement('header', 'publishing', get_string('publishing', 'grades'));
             $options = array(get_string('nopublish', 'grades'), get_string('createnewkey', 'userkey'));
-            if ($keys = get_records_select('user_private_key', "script='grade/export' AND instance={$COURSE->id} AND userid={$USER->id}")) {
+            $keys = $DB->get_records_select('user_private_key', "script='grade/export' AND instance=? AND userid=?",
+                            array($COURSE->id, $USER->id));
+            if ($keys) {
                 foreach ($keys as $key) {
                     $options[$key->value] = $key->value; // TODO: add more details - ip restriction, valid until ??
                 }
             }
             $mform->addElement('select', 'key', get_string('userkey', 'userkey'), $options);
-            $mform->setHelpButton('key', array('userkey', get_string('userkey', 'userkey'), 'grade'));
+            $mform->addHelpButton('key', 'userkey', 'userkey');
             $mform->addElement('static', 'keymanagerlink', get_string('keymanager', 'userkey'),
                     '<a href="'.$CFG->wwwroot.'/grade/export/keymanager.php?id='.$COURSE->id.'">'.get_string('keymanager', 'userkey').'</a>');
 
             $mform->addElement('text', 'iprestriction', get_string('keyiprestriction', 'userkey'), array('size'=>80));
-            $mform->setHelpButton('iprestriction', array('keyiprestriction', get_string('keyiprestriction', 'userkey'), 'userkey'));
+            $mform->addHelpButton('iprestriction', 'keyiprestriction', 'userkey');
             $mform->setDefault('iprestriction', getremoteaddr()); // own IP - just in case somebody does not know what user key is
 
             $mform->addElement('date_time_selector', 'validuntil', get_string('keyvaliduntil', 'userkey'), array('optional'=>true));
-            $mform->setHelpButton('validuntil', array('keyvaliduntil', get_string('keyvaliduntil', 'userkey'), 'userkey'));
+            $mform->addHelpButton('validuntil', 'keyvaliduntil', 'userkey');
             $mform->setDefault('validuntil', time()+3600*24*7); // only 1 week default duration - just in case somebody does not know what user key is
 
             $mform->disabledIf('iprestriction', 'key', 'noteq', 1);
@@ -135,4 +137,4 @@ class grade_export_form extends moodleform {
 
     }
 }
-?>
+
