diff --git a/question/type/multianswer/db/upgrade.php b/question/type/multianswer/db/upgrade.php
index eb3abaf..cc832b9 100644
--- a/question/type/multianswer/db/upgrade.php
+++ b/question/type/multianswer/db/upgrade.php
@@ -1,67 +1,63 @@
-<?php  // $Id$
-
-// This file keeps track of upgrades to 
-// the multianswer qtype plugin
-//
-// Sometimes, changes between versions involve
-// alterations to database structures and other
-// major things that may break installations.
+<?php
+// This file is part of Moodle - http://moodle.org/
 //
-// The upgrade function in this file will attempt
-// to perform all the necessary actions to upgrade
-// your older installtion to the current version.
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
 //
-// If there's something it cannot do itself, it
-// will tell you what you need to do.
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
 //
-// The commands in here will all be database-neutral,
-// using the functions defined in lib/ddllib.php
-
-function xmldb_qtype_multianswer_upgrade($oldversion=0) {
-
-    global $CFG, $THEME, $db;
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
 
-    $result = true;
-
-/// And upgrade begins here. For each one, you'll need one 
-/// block of code similar to the next one. Please, delete 
-/// this comment lines once this file start handling proper
-/// upgrade code.
+/**
+ * Multi-answer question type upgrade code.
+ *
+ * @package    qtype
+ * @subpackage multianswer
+ * @copyright  1999 onwards Martin Dougiamas {@link http://moodle.com}
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
 
-/// if ($result && $oldversion < YYYYMMDD00) { //New version in version.php
-///     $result = result of "/lib/ddllib.php" function calls
-/// }
 
-    if ($result && $oldversion < 2008050800) {
-        question_multianswer_fix_subquestion_parents_and_categories();
-    }
+defined('MOODLE_INTERNAL') || die();
 
-    return $result;
-}
 
 /**
- * Due to MDL-14750, subquestions of multianswer questions restored from backup will
- * have the wrong parent, and due to MDL-10899 subquestions of multianswer questions
- * that have been moved between categories will be in the wrong category, This code fixes these up.
+ * Upgrade code for the multi-answer question type.
+ * @param int $oldversion the version we are upgrading from.
  */
-function question_multianswer_fix_subquestion_parents_and_categories() {
-    global $CFG;
+function xmldb_qtype_multianswer_upgrade($oldversion) {
+    global $CFG, $DB;
+
+    $dbman = $DB->get_manager();
+
+    if ($oldversion < 2008050800) {
+        //hey - no functions here in this file !!!!!!!
 
-    $result = true;
-    $rs = get_recordset_sql('SELECT q.id, q.category, qma.sequence FROM ' . $CFG->prefix .
-            'question q JOIN ' . $CFG->prefix . 'question_multianswer qma ON q.id = qma.question');
-    if ($rs) {
-        while ($q = rs_fetch_next_record($rs)) {
+        $rs = $DB->get_recordset_sql("SELECT q.id, q.category, qma.sequence
+                                        FROM {question} q
+                                        JOIN {question_multianswer} qma ON q.id = qma.question");
+        foreach ($rs as $q) {
             if (!empty($q->sequence)) {
-                $result = $result && execute_sql('UPDATE ' . $CFG->prefix . 'question' .
-                        ' SET parent = ' . $q->id . ', category = ' . $q->category .
-                        ' WHERE id IN (' . $q->sequence . ') AND parent <> 0');
+                $DB->execute("UPDATE {question}
+                                 SET parent = ?, category = ?
+                               WHERE id IN ($q->sequence) AND parent <> 0",
+                             array($q->id, $q->category));
             }
         }
-        rs_close($rs);
-    } else {
-        $result = false;
+        $rs->close();
+
+        /// multianswer savepoint reached
+        upgrade_plugin_savepoint(true, 2008050800, 'qtype', 'multianswer');
     }
-    return $result;
+
+    // Moodle v2.1.0 release upgrade line
+    // Put any upgrade step following this
+
+    return true;
 }
-?>
