diff --git a/mod/quiz/accessrule/securewindow/rule.php b/mod/quiz/accessrule/securewindow/rule.php
index ede3f93..bbeda54 100644
--- a/mod/quiz/accessrule/securewindow/rule.php
+++ b/mod/quiz/accessrule/securewindow/rule.php
@@ -73,7 +73,7 @@ class quizaccess_securewindow extends quiz_access_rule_base {
         $page->set_popup_notification_allowed(false); // Prevent message notifications.
         $page->set_title($this->quizobj->get_course()->shortname . ': ' . $page->title);
         $page->set_cacheable(false);
-        $page->set_pagelayout('popup');
+        $page->set_pagelayout('secure');
 
         if ($this->quizobj->is_preview_user()) {
             return;
