diff --git a/user/profile/field/text/field.class.php b/user/profile/field/text/field.class.php
index 2c0ae1a..0667653 100644
--- a/user/profile/field/text/field.class.php
+++ b/user/profile/field/text/field.class.php
@@ -1,7 +1,31 @@
-<?php //$Id$
+<?php
 
 class profile_field_text extends profile_field_base {
 
+    /**
+     * Overwrite the base class to display the data for this field
+     */
+    function display_data() {
+        /// Default formatting
+        $data = parent::display_data();
+
+        /// Are we creating a link?
+        if (!empty($this->field->param4) and !empty($data)) {
+
+            /// Define the target
+            if (! empty($this->field->param5)) {
+                $target = 'target="'.$this->field->param5.'"';
+            } else {
+                $target = '';
+            }
+
+            /// Create the link
+            $data = '<a href="'.str_replace('$$', urlencode($data), $this->field->param4).'" '.$target.'>'.htmlspecialchars($data).'</a>';
+        }
+
+        return $data;
+    }
+
     function edit_field_add(&$mform) {
         $size = $this->field->param1;
         $maxlength = $this->field->param2;
@@ -14,4 +38,4 @@ class profile_field_text extends profile_field_base {
 
 }
 
-?>
+
