diff --git a/auth/cas/languages.php b/auth/cas/languages.php
index 9edb8e7..d2fe22b 100644
--- a/auth/cas/languages.php
+++ b/auth/cas/languages.php
@@ -1,17 +1,15 @@
-<?PHP
-
-// version $Id$
-
-// List of CAS langages.
-
-// You can add langages in /CAS/langage.
-
-// Please send them to http://esup-phpcas.sourceforge.net
-
-$CASLANGUAGES = array (
-
-"english" => "English",
-
-"french" => "French");
-
-?>
+<?php
+
+$caslangprefix = 'PHPCAS_LANG_';
+$CASLANGUAGES = array ();
+
+$consts = get_defined_constants(true);
+foreach ($consts['user'] as $key => $value) {
+    if (substr($key, 0, strlen($caslangprefix)) == $caslangprefix) {
+        $CASLANGUAGES[$value] = $value;
+    }
+}
+if (empty($CASLANGUAGES)) {
+    $CASLANGUAGES = array ('english' => 'english',
+                           'french'  => 'french');
+}
