diff --git a/grade/export/xml/index.php b/grade/export/xml/index.php
old mode 100755
new mode 100644
index 29c0101..bc4d925
--- a/grade/export/xml/index.php
+++ b/grade/export/xml/index.php
@@ -21,7 +21,9 @@ require_once 'grade_export_xml.php';
 
 $id = required_param('id', PARAM_INT); // course id
 
-if (!$course = get_record('course', 'id', $id)) {
+$PAGE->set_url('/grade/export/xml/index.php', array('id'=>$id));
+
+if (!$course = $DB->get_record('course', array('id'=>$id))) {
     print_error('nocourseid');
 }
 
@@ -31,19 +33,20 @@ $context = get_context_instance(CONTEXT_COURSE, $id);
 require_capability('moodle/grade:export', $context);
 require_capability('gradeexport/xml:view', $context);
 
-print_grade_page_head($COURSE->id, 'export', 'xml', get_string('exportto', 'grades') . ' ' . get_string('modulename', 'gradeexport_xml'));
+print_grade_page_head($COURSE->id, 'export', 'xml', get_string('exportto', 'grades') . ' ' . get_string('pluginname', 'gradeexport_xml'));
 
 if (!empty($CFG->gradepublishing)) {
     $CFG->gradepublishing = has_capability('gradeexport/xml:publish', $context);
 }
 
+//'idnumberrequired'=>true excludes grade items that dont have an ID to use during import
 $mform = new grade_export_form(null, array('idnumberrequired'=>true, 'publishing'=>true, 'updategradesonly'=>true));
 
 $groupmode    = groups_get_course_groupmode($course);   // Groups are being used
 $currentgroup = groups_get_course_group($course, true);
 if ($groupmode == SEPARATEGROUPS and !$currentgroup and !has_capability('moodle/site:accessallgroups', $context)) {
-    print_heading(get_string("notingroup"));
-    print_footer($course);
+    echo $OUTPUT->heading(get_string("notingroup"));
+    echo $OUTPUT->footer();
     die;    
 }
 
@@ -54,8 +57,11 @@ if ($data = $mform->get_data()) {
     // print the grades on screen for feedbacks
     $export->process_form($data);
     $export->print_continue();
-    $export->display_preview(true);
-    print_footer($course);
+
+    $export->display_preview(true); //true == skip users without idnumber as they cannot be identified when importing
+    echo $OUTPUT->container(get_string('useridnumberwarning','gradeexport_xml'), 'useridnumberwarning mdl-align');
+
+    echo $OUTPUT->footer();
     exit;
 }
 
@@ -64,5 +70,5 @@ echo '<div class="clearer"></div>';
 
 $mform->display();
 
-print_footer();
-?>
+echo $OUTPUT->footer();
+
