diff --git a/grade/export/xml/version.php b/grade/export/xml/version.php
index f8e4303..dca0503 100644
--- a/grade/export/xml/version.php
+++ b/grade/export/xml/version.php
@@ -15,7 +15,7 @@
 // You should have received a copy of the GNU General Public License
 // along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
 
-$plugin->version  = 2007092700;
-$plugin->requires = 2007101000;
+$plugin->version  = 2010090501;
+$plugin->requires = 2010090501;
+
 
-?>
