diff --git a/mod/glossary/formats/fullwithauthor/fullwithauthor_format.php b/mod/glossary/formats/fullwithauthor/fullwithauthor_format.php
index d4f4863..0d34805 100644
--- a/mod/glossary/formats/fullwithauthor/fullwithauthor_format.php
+++ b/mod/glossary/formats/fullwithauthor/fullwithauthor_format.php
@@ -1,19 +1,18 @@
-<?php  // $Id$
+<?php
 
-function glossary_show_entry_fullwithauthor($course, $cm, $glossary, $entry, $mode="", $hook="", $printicons=1, $ratings=NULL, $aliases=true) {
-    global $CFG, $USER;
+function glossary_show_entry_fullwithauthor($course, $cm, $glossary, $entry, $mode="", $hook="", $printicons=1, $aliases=true) {
+    global $CFG, $USER, $DB, $OUTPUT;
 
 
-    $user = get_record('user', 'id', $entry->userid);
+    $user = $DB->get_record('user', array('id'=>$entry->userid));
     $strby = get_string('writtenby', 'glossary');
 
-    $return = false;
     if ($entry) {
         echo '<table class="glossarypost fullwithauthor" cellspacing="0">';
         echo '<tr valign="top">';
         
         echo '<td class="picture">';
-        print_user_picture($user, $course->id, $user->picture);
+        echo $OUTPUT->user_picture($user, array('courseid'=>$course->id));
         echo '</td>';
         
         echo '<th class="entryheader">';
@@ -23,7 +22,7 @@ function glossary_show_entry_fullwithauthor($course, $cm, $glossary, $entry, $mo
         echo '</div>';
 
         $fullname = fullname($user);
-        $by = new object();
+        $by = new stdClass();
         $by->name = '<a href="'.$CFG->wwwroot.'/user/view.php?id='.$user->id.'&amp;course='.$course->id.'">'.$fullname.'</a>';
         $by->date = userdate($entry->timemodified);
         echo '<span class="author">'.get_string('bynameondate', 'forum', $by).'</span>';
@@ -32,7 +31,7 @@ function glossary_show_entry_fullwithauthor($course, $cm, $glossary, $entry, $mo
         echo '<td class="entryattachment">';
 
         glossary_print_entry_approval($cm, $entry, $mode);
-        glossary_print_entry_attachment($entry,'html','right');
+        glossary_print_entry_attachment($entry, $cm, 'html', 'right');
         echo '</td>';
 
         echo '</tr>';
@@ -41,14 +40,14 @@ function glossary_show_entry_fullwithauthor($course, $cm, $glossary, $entry, $mo
         echo '<td class="left">&nbsp;</td>';
         echo '<td colspan="2" class="entry">';
 
-        glossary_print_entry_definition($entry);
+        glossary_print_entry_definition($entry, $glossary, $cm);
 
         echo '</td></tr>';
         echo '<tr valign="top">';
         echo '<td class="left">&nbsp;</td>';
         echo '<td colspan="2" class="entrylowersection">';
         
-        $return = glossary_print_entry_lower_section($course, $cm, $glossary, $entry, $mode, $hook, $printicons, $ratings, $aliases);
+        glossary_print_entry_lower_section($course, $cm, $glossary, $entry, $mode, $hook, $printicons, $aliases);
         echo ' ';
         echo '</td></tr>';
         echo "</table>\n";
@@ -57,10 +56,9 @@ function glossary_show_entry_fullwithauthor($course, $cm, $glossary, $entry, $mo
         print_string('noentry', 'glossary');
         echo '</div>';
     }
-    return $return;
 }
 
-function glossary_print_entry_fullwithauthor($course, $cm, $glossary, $entry, $mode="", $hook="", $printicons=1, $ratings=NULL) {
+function glossary_print_entry_fullwithauthor($course, $cm, $glossary, $entry, $mode="", $hook="", $printicons=1) {
 
     //The print view for this format is exactly the normal view, so we use it
 
@@ -72,4 +70,4 @@ function glossary_print_entry_fullwithauthor($course, $cm, $glossary, $entry, $m
 
 }
 
-?>
+
