diff --git a/mod/glossary/formats.php b/mod/glossary/formats.php
index 19ee665..5d65009 100644
--- a/mod/glossary/formats.php
+++ b/mod/glossary/formats.php
@@ -1,32 +1,39 @@
-<?php    // $Id$
-    /// This file allows to manage the default behaviour of the display formats
+<?php
 
-    require_once("../../config.php");
-    require_once($CFG->libdir.'/adminlib.php');
-    require_once("lib.php");
+/// This file allows to manage the default behaviour of the display formats
 
-    $id   = required_param('id', PARAM_INT);
-    $mode = optional_param('mode');
+require_once("../../config.php");
+require_once($CFG->libdir.'/adminlib.php');
+require_once("lib.php");
 
-    admin_externalpage_setup('managemodules'); // this is hacky, tehre should be a special hidden page for it
+$id   = required_param('id', PARAM_INT);
+$mode = optional_param('mode', '', PARAM_ACTION);
 
-    if ( !$displayformat = get_record("glossary_formats","id",$id) ) {
-        error ("Invalid Glossary Format");
-    }
+$url = new moodle_url('/mod/glossary/formats.php', array('id'=>$id));
+if ($mode !== '') {
+    $url->param('mode', $mode);
+}
+$PAGE->set_url($url);
+
+admin_externalpage_setup('managemodules'); // this is hacky, tehre should be a special hidden page for it
+
+if ( !$displayformat = $DB->get_record("glossary_formats", array("id"=>$id))) {
+    print_error('invalidglossaryformat', 'glossary');
+}
 
-    $form = data_submitted();
-    if ( $mode == 'visible' and confirm_sesskey()) {
+$form = data_submitted();
+if ( $mode == 'visible' and confirm_sesskey()) {
         if ( $displayformat ) {
             if ( $displayformat->visible ) {
                 $displayformat->visible = 0;
             } else {
                 $displayformat->visible = 1;
             }
-            update_record("glossary_formats",$displayformat);
+        $DB->update_record("glossary_formats",$displayformat);
         }
         redirect("$CFG->wwwroot/$CFG->admin/settings.php?section=modsettingglossary#glossary_formats_header");
         die;
-    } elseif ( $mode == 'edit' and $form and confirm_sesskey()) {
+} elseif ( $mode == 'edit' and $form and confirm_sesskey()) {
 
         $displayformat->popupformatname = $form->popupformatname;
         $displayformat->showgroup   = $form->showgroup;
@@ -35,33 +42,33 @@
         $displayformat->sortkey     = $form->sortkey;
         $displayformat->sortorder   = $form->sortorder;
 
-        update_record("glossary_formats",$displayformat);
+    $DB->update_record("glossary_formats",$displayformat);
         redirect("$CFG->wwwroot/$CFG->admin/settings.php?section=modsettingglossary#glossary_formats_header");
         die;
-    }
+}
 
-    $strmodulename = get_string("modulename", "glossary");
-    $strdisplayformats = get_string("displayformats","glossary");
+$strmodulename = get_string("modulename", "glossary");
+$strdisplayformats = get_string("displayformats","glossary");
 
-    admin_externalpage_print_header();
+echo $OUTPUT->header();
 
-    print_heading($strmodulename . ': ' . get_string("displayformats","glossary"));
+echo $OUTPUT->heading($strmodulename . ': ' . get_string("displayformats","glossary"));
 
-    print_simple_box("<center>".get_string("configwarning", 'admin')."</center>", "center", "60%");
-    echo "<br />";
+echo $OUTPUT->box(get_string("configwarning", 'admin'), "generalbox boxaligncenter boxwidthnormal");
+echo "<br />";
 
-    $yes = get_string("yes");
-    $no  = get_string("no");
+$yes = get_string("yes");
+$no  = get_string("no");
 
-    echo '<form method="post" action="formats.php" id="form">';
-    echo '<table width="90%" align="center" class="generalbox">';
-    ?>
-    <tr>
+echo '<form method="post" action="formats.php" id="form">';
+echo '<table width="90%" align="center" class="generalbox">';
+?>
+<tr>
         <td colspan="3" align="center"><strong>
         <?php echo get_string('displayformat'.$displayformat->name,'glossary'); ?>
         </strong></td>
-    </tr>
-    <tr valign="top">
+</tr>
+<tr valign="top">
         <td align="right" width="20%"><?php print_string('popupformat','glossary'); ?></td>
         <td>
      <?php
@@ -77,18 +84,18 @@
         //Sort it
         asort($formats);
 
-        choose_from_menu($formats,'popupformatname',$displayformat->popupformatname);
+    echo html_writer::select($formats, 'popupformatname', $displayformat->popupformatname, false);
      ?>
         </td>
         <td width="60%">
         <?php print_string("cnfrelatedview", "glossary") ?><br /><br />
         </td>
-    </tr>
-    <tr valign="top">
+</tr>
+<tr valign="top">
         <td align="right" width="20%"><?php print_string('defaultmode','glossary'); ?></td>
         <td>
         <select size="1" name="defaultmode">
-    <?php
+<?php
         $sletter = '';
         $scat = '';
         $sauthor = '';
@@ -110,7 +117,7 @@
             $sauthor = ' selected="selected" ';
         break;
         }
-    ?>
+?>
         <option value="letter" <?php p($sletter)?>><?php print_string("letter", "glossary"); ?></option>
         <option value="cat" <?php p($scat)?>><?php print_string("cat", "glossary"); ?></option>
         <option value="date" <?php p($sdate)?>><?php print_string("date", "glossary"); ?></option>
@@ -120,12 +127,12 @@
         <td width="60%">
         <?php print_string("cnfdefaultmode", "glossary") ?><br /><br />
         </td>
-    </tr>
-    <tr valign="top">
+</tr>
+<tr valign="top">
         <td align="right" width="20%"><?php print_string('defaulthook','glossary'); ?></td>
         <td>
         <select size="1" name="defaulthook">
-    <?php
+<?php
         $sall = '';
         $sspecial = '';
         $sallcategories = '';
@@ -147,7 +154,7 @@
             $snocategorised = ' selected="selected" ';
         break;
         }
-    ?>
+?>
         <option value="ALL" <?php p($sall)?>><?php p(get_string("allentries","glossary"))?></option>
         <option value="SPECIAL" <?php p($sspecial)?>><?php p(get_string("special","glossary"))?></option>
         <option value="0" <?php p($sallcategories)?>><?php p(get_string("allcategories","glossary"))?></option>
@@ -157,12 +164,12 @@
         <td width="60%">
         <?php print_string("cnfdefaulthook", "glossary") ?><br /><br />
         </td>
-    </tr>
-    <tr valign="top">
+</tr>
+<tr valign="top">
         <td align="right" width="20%"><?php print_string('defaultsortkey','glossary'); ?></td>
         <td>
         <select size="1" name="sortkey">
-    <?php
+<?php
         $sfname = '';
         $slname = '';
         $supdate = '';
@@ -184,7 +191,7 @@
             $supdate = ' selected="selected" ';
         break;
         }
-    ?>
+?>
         <option value="CREATION" <?php p($screation)?>><?php p(get_string("sortbycreation","glossary"))?></option>
         <option value="UPDATE" <?php p($supdate)?>><?php p(get_string("sortbylastupdate","glossary"))?></option>
         <option value="FIRSTNAME" <?php p($sfname)?>><?php p(get_string("firstname"))?></option>
@@ -194,12 +201,12 @@
         <td width="60%">
         <?php print_string("cnfsortkey", "glossary") ?><br /><br />
         </td>
-    </tr>
-    <tr valign="top">
+</tr>
+<tr valign="top">
         <td align="right" width="20%"><?php print_string('defaultsortorder','glossary'); ?></td>
         <td>
         <select size="1" name="sortorder">
-    <?php
+<?php
         $sasc = '';
         $sdesc = '';
         switch ( strtolower($displayformat->sortorder) ) {
@@ -211,7 +218,7 @@
             $sdesc = ' selected="selected" ';
         break;
         }
-    ?>
+?>
         <option value="asc" <?php p($sasc)?>><?php p(get_string("ascending","glossary"))?></option>
         <option value="desc" <?php p($sdesc)?>><?php p(get_string("descending","glossary"))?></option>
         </select>
@@ -219,12 +226,12 @@
         <td width="60%">
         <?php print_string("cnfsortorder", "glossary") ?><br /><br />
         </td>
-    </tr>
-    <tr valign="top">
+</tr>
+<tr valign="top">
         <td align="right" width="20%"><?php print_string("includegroupbreaks", "glossary"); ?>:</td>
         <td>
         <select size="1" name="showgroup">
-    <?php
+<?php
         $yselected = "";
         $nselected = "";
         if ($displayformat->showgroup) {
@@ -232,7 +239,7 @@
         } else {
             $nselected = " selected=\"selected\" ";
         }
-    ?>
+?>
         <option value="1" <?php echo $yselected ?>><?php p($yes)?></option>
         <option value="0" <?php echo $nselected ?>><?php p($no)?></option>
         </select>
@@ -240,17 +247,17 @@
         <td width="60%">
         <?php print_string("cnfshowgroup", "glossary") ?><br /><br />
         </td>
-    </tr>
-    <tr>
+</tr>
+<tr>
         <td colspan="3" align="center">
         <input type="submit" value="<?php print_string("savechanges") ?>" /></td>
-    </tr>
-    <input type="hidden" name="id"    value="<?php p($id) ?>" />
-    <input type="hidden" name="sesskey" value="<?php echo sesskey() ?>" />
-    <input type="hidden" name="mode"    value="edit" />
-    <?php
+</tr>
+<input type="hidden" name="id"    value="<?php p($id) ?>" />
+<input type="hidden" name="sesskey" value="<?php echo sesskey() ?>" />
+<input type="hidden" name="mode"    value="edit" />
+<?php
 
-    echo '</table></form>';
+echo '</table></form>';
 
-    print_footer();
+echo $OUTPUT->footer();
 ?>
