diff --git a/filter/algebra/filter.php b/filter/algebra/filter.php
index bdd6d1d..54ffaff 100644
--- a/filter/algebra/filter.php
+++ b/filter/algebra/filter.php
@@ -1,26 +1,33 @@
-<?PHP
-/////////////////////////////////////////////////////////////////////////////
-//                                                                         //
-// NOTICE OF COPYRIGHT                                                     //
-//                                                                         //
-// Moodle - Filter for converting simple calculator-type algebraic        //
-// expressions to cached gif images                                        //
-//                                                                         //
-// Copyright (C) 2004 Zbigniew Fiedorowicz fiedorow@math.ohio-state.edu    //
-// Originally based on code provided by Bruno Vernier bruno@vsbeducation.ca//
-// This program is free software; you can redistribute it and/or modify    //
-// it under the terms of the GNU General Public License as published by    //
-// the Free Software Foundation; either version 2 of the License, or       //
-// (at your option) any later version.                                     //
-//                                                                         //
-// This program is distributed in the hope that it will be useful,         //
-// but WITHOUT ANY WARRANTY; without even the implied warranty of          //
-// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the           //
-// GNU General Public License for more details:                            //
-//                                                                         //
-//          http://www.gnu.org/copyleft/gpl.html                           //
-//                                                                         //
-/////////////////////////////////////////////////////////////////////////////
+<?php
+
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Moodle - Filter for converting simple calculator-type algebraic
+ * expressions to cached gif images
+ *
+ * @package    filter
+ * @subpackage algebra
+ * @copyright  2004 Zbigniew Fiedorowicz fiedorow@math.ohio-state.edu
+ *             Originally based on code provided by Bruno Vernier bruno@vsbeducation.ca
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+defined('MOODLE_INTERNAL') || die();
+
 //-------------------------------------------------------------------------
 // NOTE: This Moodle text filter converts algebraic expressions delimited
 // by either @@...@@ or by <algebra...>...</algebra> tags
@@ -37,21 +44,17 @@
 // You will then need to edit your moodle/config.php to invoke mathml_filter.php
 //-------------------------------------------------------------------------
 
-function string_file_picture_algebra($imagefile, $tex= "", $height="", $width="", $align="middle") {
+function filter_algebra_image($imagefile, $tex= "", $height="", $width="", $align="middle") {
   // Given the path to a picture file in a course, or a URL,
   // this function includes the picture in the page.
-  global $CFG;
+  global $CFG, $OUTPUT;
 
   $output = "";
-  $origtex = $tex;
   $style = 'style="border:0px; vertical-align:'.$align.';';
+  $title = '';
   if ($tex) {
-    $tex = str_replace('&','&amp;',$tex);
-    $tex = str_replace('<','&lt;',$tex);
-    $tex = str_replace('>','&gt;',$tex);
-    $tex = str_replace('"','&quot;',$tex);
-    $tex = str_replace("\'",'&#39;',$tex);
-    $title = "title=\"$tex\"";
+    $tex = html_entity_decode($tex, ENT_QUOTES, 'UTF-8');
+    $title = 'title="'.s($tex).'"';
   }
   if ($height) {
     $style .= " height:{$height}px;";
@@ -60,50 +63,50 @@ function string_file_picture_algebra($imagefile, $tex= "", $height="", $width=""
     $style .= " width:{$width}px;";
   }
   $style .= '"';
+  $anchorcontents = '';
   if ($imagefile) {
-    if (!file_exists("$CFG->dataroot/filter/algebra/$imagefile") && has_capability('moodle/site:config', get_context_instance(CONTEXT_SYSTEM))) {
-      $output .= "<a href=\"$CFG->wwwroot/filter/algebra/algebradebug.php\">";
+    $anchorcontents .= "<img $title alt=\"".s($tex)."\" src=\"";
+    if ($CFG->slasharguments) {        // Use this method if possible for better caching
+      $anchorcontents .= "$CFG->wwwroot/filter/algebra/pix.php/$imagefile";
     } else {
-      $output .= "<a target=\"popup\" title=\"TeX\" href=";
-      $output .= "\"$CFG->wwwroot/filter/algebra/displaytex.php?";
-      $output .= urlencode($tex) . "\" onclick=\"return openpopup('/filter/algebra/displaytex.php?";
-      $output .= urlencode($tex) . "', 'popup', 'menubar=0,location=0,scrollbars,";
-      $output .= "resizable,width=300,height=240', 0);\">";
+      $anchorcontents .= "$CFG->wwwroot/filter/algebra/pix.php?file=$imagefile";
     }
-    $output .= "<img $title alt=\"".s($origtex)."\" src=\"";
-    if ($CFG->slasharguments) {        // Use this method if possible for better caching
-      $output .= "$CFG->wwwroot/filter/algebra/pix.php/$imagefile";
+    $anchorcontents .= "\" $style />";
+
+    if (!file_exists("$CFG->dataroot/filter/algebra/$imagefile") && has_capability('moodle/site:config', get_context_instance(CONTEXT_SYSTEM))) {
+        $link = '/filter/algebra/algebradebug.php';
+        $action = null;
     } else {
-      $output .= "$CFG->wwwroot/filter/algebra/pix.php?file=$imagefile";
+        $link = new moodle_url('/filter/tex/displaytex.php', array('texexp'=>$tex));
+        $action = new popup_action('click', $link, 'popup', array('width'=>320,'height'=>240)); //TODO: the popups do not work when text caching is enabled!!
     }
-    $output .= "\" $style />";
-    $output .= "</a>";
+    $output .= $OUTPUT->action_link($link, $anchorcontents, $action, array('title'=>'TeX'));
+
   } else {
     $output .= "Error: must pass URL or course";
   }
   return $output;
 }
 
-
-function algebra_filter ($courseid, $text) {
-
-    global $CFG;
+class filter_algebra extends moodle_text_filter {
+    function filter($text, array $options = array()){
+        global $CFG, $DB;
 
     /// Do a quick check using stripos to avoid unnecessary wor
     if (!preg_match('/<algebra/i',$text) && !strstr($text,'@@')) {
         return $text;
     }
 
-#    //restrict filtering to forum 130 (Maths Tools on moodle.org)
+//restrict filtering to forum 130 (Maths Tools on moodle.org)
 #    $scriptname = $_SERVER['SCRIPT_NAME'];
 #    if (!strstr($scriptname,'/forum/')) {
 #        return $text;
 #    }
 #    if (strstr($scriptname,'post.php')) {
 #        $parent = forum_get_post_full($_GET['reply']);
-#        $discussion = get_record("forum_discussions","id",$parent->discussion);
+#        $discussion = $DB->get_record("forum_discussions",array("id"=>$parent->discussion));
 #    } else if (strstr($scriptname,'discuss.php')) {
-#        $discussion = get_record("forum_discussions","id",$_GET['d'] );
+#        $discussion = $DB->get_record("forum_discussions",array("id"=>$_GET['d']));
 #    } else {
 #        return $text;
 #    }
@@ -140,7 +143,7 @@ function algebra_filter ($courseid, $text) {
         }
         $md5 =  md5($algebra);
         $filename =  $md5  . ".gif";
-        if (! $texcache = get_record("cache_filters","filter","algebra", "md5key", $md5)) {
+            if (! $texcache = $DB->get_record("cache_filters",array("filter"=>"algebra", "md5key"=>$md5))) {
            $algebra = str_replace('&lt;','<',$algebra);
            $algebra = str_replace('&gt;','>',$algebra);
            $algebra = str_replace('<>','#',$algebra);
@@ -216,21 +219,23 @@ function algebra_filter ($courseid, $text) {
               $texexp = preg_replace('/\\\int\\\left\((.+?),(.+?),(.+?)\\\right\)/s','\int_'. "{\$2}^{\$3}\$1 ",$texexp);
               $texexp = preg_replace('/\\\int\\\left\((.+?d[a-z])\\\right\)/s','\int '. "\$1 ",$texexp);
               $texexp = preg_replace('/\\\lim\\\left\((.+?),(.+?),(.+?)\\\right\)/s','\lim_'. "{\$2\\to \$3}\$1 ",$texexp);
+                  $texexp = str_replace('\mbox', '', $texexp); // now blacklisted in tex, sorry
               $texcache->filter = 'algebra';
               $texcache->version = 1;
               $texcache->md5key = $md5;
-              $texcache->rawtext = addslashes($texexp);
+                  $texcache->rawtext = $texexp;
               $texcache->timemodified = time();
-              insert_record("cache_filters",$texcache, false);
-              $text = str_replace( $matches[0][$i], string_file_picture_algebra($filename, $texexp, '', '', $align), $text);
+                  $DB->insert_record("cache_filters", $texcache, false);
+                  $text = str_replace( $matches[0][$i], filter_algebra_image($filename, $texexp, '', '', $align), $text);
            } else {
               $text = str_replace( $matches[0][$i],"<b>Undetermined error:</b> ",$text);
            }
         } else {
-           $text = str_replace( $matches[0][$i], string_file_picture_algebra($filename, $texcache->rawtext), $text);
+               $text = str_replace( $matches[0][$i], filter_algebra_image($filename, $texcache->rawtext), $text);
         }
     }
     return $text;
+    }
 }
 
-?>
+
