diff --git a/user/filters/courserole.php b/user/filters/courserole.php
index f200322..49cb54c 100644
--- a/user/filters/courserole.php
+++ b/user/filters/courserole.php
@@ -1,4 +1,4 @@
-<?php //$Id$
+<?php
 
 require_once($CFG->dirroot .'/user/filters/lib.php');
 
@@ -22,7 +22,7 @@ class user_filter_courserole extends user_filter_type {
      */
     function get_roles() {
         $context = get_context_instance(CONTEXT_SYSTEM);
-        $roles = array(0=> get_string('anyrole','filters')) + get_assignable_roles($context);
+        $roles = array(0=> get_string('anyrole','filters')) + get_default_enrol_roles($context);
         return $roles;
     }
 
@@ -31,6 +31,8 @@ class user_filter_courserole extends user_filter_type {
      * @return array of course categories
      */
     function get_course_categories() {
+        global $CFG;
+        require_once($CFG->dirroot.'/course/lib.php');
         $displaylist = array();
         $parentlist = array();
         make_categories_list($displaylist, $parentlist);
@@ -47,7 +49,6 @@ class user_filter_courserole extends user_filter_type {
         $objs[] =& $mform->createElement('select', $this->_name .'_ct', null, $this->get_course_categories());
         $objs[] =& $mform->createElement('text', $this->_name, null);
         $grp =& $mform->addElement('group', $this->_name.'_grp', $this->_label, $objs, '', false);
-        $grp->setHelpButton(array('courserole', $this->_label, 'filters'));
         if ($this->_advanced) {
             $mform->setAdvanced($this->_name.'_grp');
         }
@@ -78,34 +79,42 @@ class user_filter_courserole extends user_filter_type {
     /**
      * Returns the condition to be used with SQL where
      * @param array $data filter settings
-     * @return string the filtering condition or null if the filter is disabled
+     * @return array sql string and $params
      */
     function get_sql_filter($data) {
-        global $CFG;
-        $value      = addslashes($data['value']);
+        global $CFG, $DB;
+        static $counter = 0;
+        $name = 'ex_courserole'.$counter++;
+
+        $value      = $data['value'];
         $roleid     = $data['roleid'];
         $categoryid = $data['categoryid'];
 
+        $params = array();
+
         if (empty($value) and empty($roleid) and empty($categoryid)) {
-            return '';
+            return array('', $params);
         }
 
-        $timenow = round(time(), 100); // rounding - enable sql caching
-        $where = "b.contextlevel=50 AND a.timestart<$timenow AND (a.timeend=0 OR a.timeend>$timenow)";
+        $where = "b.contextlevel=50";
         if ($roleid) {
-            $where .= " AND a.roleid=$roleid";
+            $where .= " AND a.roleid = :roleid";
+            $params['roleid'] = $roleid;
+
         }
         if ($categoryid) {
-            $where .= " AND c.category=$categoryid";
+            $where .= " AND c.category = :categoryid";
+            $params['categoryid'] = $categoryid;
         }
         if ($value) {
-            $where .= " AND c.shortname ".sql_ilike()." '$value'";
+            $where .= " AND c.shortname = :$name";
+            $params[$name] = $value;
         }
-        return "id IN (SELECT userid
-                         FROM {$CFG->prefix}role_assignments a
-                   INNER JOIN {$CFG->prefix}context b ON a.contextid=b.id
-                   INNER JOIN {$CFG->prefix}course c ON b.instanceid=c.id
-                        WHERE $where)";
+        return array("id IN (SELECT userid
+                               FROM {role_assignments} a
+                         INNER JOIN {context} b ON a.contextid=b.id
+                         INNER JOIN {course} c ON b.instanceid=c.id
+                              WHERE $where)", $params);
     }
 
     /**
@@ -114,22 +123,24 @@ class user_filter_courserole extends user_filter_type {
      * @return string active filter label
      */
     function get_label($data) {
+        global $DB;
+
         $value      = $data['value'];
         $roleid     = $data['roleid'];
         $categoryid = $data['categoryid'];
 
-        $a = new object();
+        $a = new stdClass();
         $a->label = $this->_label;
 
         if ($roleid) {
-            $rolename = get_field('role', 'name', 'id', $roleid);
+            $rolename = $DB->get_field('role', 'name', array('id'=>$roleid));
             $a->rolename = '"'.format_string($rolename).'"';
         } else {
             $a->rolename = get_string('anyrole', 'filters');
         }
 
         if ($categoryid) {
-            $catname = get_field('course_categories', 'name', 'id', $categoryid);
+            $catname = $DB->get_field('course_categories', 'name', array('id'=>$categoryid));
             $a->categoryname = '"'.format_string($catname).'"';
         } else {
             $a->categoryname = get_string('anycategory', 'filters');
@@ -137,7 +148,7 @@ class user_filter_courserole extends user_filter_type {
 
         if ($value) {
             $a->coursename = '"'.s($value).'"';
-            if (!record_exists('course', 'shortname', addslashes($value))) {
+            if (!$DB->record_exists('course', array('shortname'=>$value))) {
                 return '<span class="notifyproblem">'.get_string('courserolelabelerror', 'filters', $a).'</span>';
             }
         } else {
