diff --git a/user/filters/date.php b/user/filters/date.php
old mode 100755
new mode 100644
index 9ff2c14..a39c199
--- a/user/filters/date.php
+++ b/user/filters/date.php
@@ -1,4 +1,4 @@
-<?php //$Id$
+<?php
 
 require_once($CFG->dirroot.'/user/filters/lib.php');
 
@@ -35,10 +35,8 @@ class user_filter_date extends user_filter_type {
         $objs[] =& $mform->createElement('static', $this->_name.'_break', null, '<br/>');
         $objs[] =& $mform->createElement('checkbox', $this->_name.'_eck', null, get_string('isbefore', 'filters'));
         $objs[] =& $mform->createElement('date_selector', $this->_name.'_edt', null);
-        $objs[] = & $mform->createElement('checkbox', $this->_name.'_never', null, get_string('includenever', 'filters'));
 
         $grp =& $mform->addElement('group', $this->_name.'_grp', $this->_label, $objs, '', false);
-        $grp->setHelpButton(array('date',$this->_label,'filters'));
 
         if ($this->_advanced) {
             $mform->setAdvanced($this->_name.'_grp');
@@ -50,8 +48,6 @@ class user_filter_date extends user_filter_type {
         $mform->disabledIf($this->_name.'_edt[day]', $this->_name.'_eck', 'notchecked');
         $mform->disabledIf($this->_name.'_edt[month]', $this->_name.'_eck', 'notchecked');
         $mform->disabledIf($this->_name.'_edt[year]', $this->_name.'_eck', 'notchecked');
-
-        $mform->disabledIf($this->_name.'_never', $this->_name.'_eck', 'notchecked');
     }
 
     /**
@@ -64,7 +60,6 @@ class user_filter_date extends user_filter_type {
         $sdt = $this->_name.'_sdt';
         $eck = $this->_name.'_eck';
         $edt = $this->_name.'_edt';
-        $never = $this->_name.'_never';
 
         if (!array_key_exists($sck, $formdata) and !array_key_exists($eck, $formdata)) {
             return false;
@@ -81,11 +76,6 @@ class user_filter_date extends user_filter_type {
         } else {
             $data['before'] = 0;
         }
-        if (array_key_exists($never, $formdata)) {
-            $data['never'] = $formdata->$never;
-        } else {
-            $data['never'] = 0;
-        }
 
         return $data;
     }
@@ -93,25 +83,19 @@ class user_filter_date extends user_filter_type {
     /**
      * Returns the condition to be used with SQL where
      * @param array $data filter settings
-     * @return string the filtering condition or null if the filter is disabled
+     * @return array sql string and $params
      */
     function get_sql_filter($data) {
-        $after  = $data['after'];
-        $before = $data['before'];
-        $never = $data['never'];
+        $after  = (int)$data['after'];
+        $before = (int)$data['before'];
+
         $field  = $this->_field;
 
         if (empty($after) and empty($before)) {
-            return '';
+            return array('', array());
         }
 
-        $res = '';
-
-        if (!empty($never)) {
-            $res .= " $field >= 0 " ;
-        } else {
-            $res .= " $field > 0 " ;
-        }
+        $res = " $field >= 0 " ;
 
         if ($after) {
             $res .= " AND $field >= $after";
@@ -120,8 +104,7 @@ class user_filter_date extends user_filter_type {
         if ($before) {
             $res .= " AND $field <= $before";
         }
-
-        return $res;
+        return array($res, array());
     }
 
     /**
@@ -132,30 +115,20 @@ class user_filter_date extends user_filter_type {
     function get_label($data) {
         $after  = $data['after'];
         $before = $data['before'];
-        $never = $data['never'];
         $field  = $this->_field;
 
-        $a = new object();
+        $a = new stdClass();
         $a->label  = $this->_label;
         $a->after  = userdate($after);
         $a->before = userdate($before);
 
-        if ($never) {
-            $strnever = ' ('.get_string('includenever', 'filters').')';
-        } else {
-            $strnever = '';
-        }
-
         if ($after and $before) {
-            return get_string('datelabelisbetween', 'filters', $a).$strnever;
-
+            return get_string('datelabelisbetween', 'filters', $a);
         } else if ($after) {
-            return get_string('datelabelisafter', 'filters', $a).$strnever;;
-
+            return get_string('datelabelisafter', 'filters', $a);
         } else if ($before) {
-            return get_string('datelabelisbefore', 'filters', $a).$strnever;;
+            return get_string('datelabelisbefore', 'filters', $a);
         }
         return '';
     }
 }
-?>
