diff --git a/user/filters/profilefield.php b/user/filters/profilefield.php
index 454a4e2..3e5be1f 100644
--- a/user/filters/profilefield.php
+++ b/user/filters/profilefield.php
@@ -1,4 +1,4 @@
-<?php //$Id$
+<?php
 
 require_once($CFG->dirroot.'/user/filters/lib.php');
 
@@ -37,7 +37,8 @@ class user_filter_profilefield extends user_filter_type {
      * @return array of profile fields
      */
     function get_profile_fields() {
-        if (!$fields = get_records_select('user_info_field', '', 'shortname', 'id,shortname')) {
+        global $DB;
+        if (!$fields = $DB->get_records('user_info_field', null, 'shortname', 'id,shortname')) {
             return null;
         }
         $res = array(0 => get_string('anyfield', 'filters'));
@@ -61,7 +62,6 @@ class user_filter_profilefield extends user_filter_type {
         $objs[] =& $mform->createElement('select', $this->_name.'_op', null, $this->get_operators());
         $objs[] =& $mform->createElement('text', $this->_name, null);
         $grp =& $mform->addElement('group', $this->_name.'_grp', $this->_label, $objs, '', false);
-        $grp->setHelpButton(array('profilefield',$this->_label,'filters'));
         if ($this->_advanced) {
             $mform->setAdvanced($this->_name.'_grp');
         }
@@ -97,10 +97,12 @@ class user_filter_profilefield extends user_filter_type {
     /**
      * Returns the condition to be used with SQL where
      * @param array $data filter settings
-     * @return string the filtering condition or null if the filter is disabled
+     * @return array sql string and $params
      */
     function get_sql_filter($data) {
-        global $CFG;
+        global $CFG, $DB;
+        static $counter = 0;
+        $name = 'ex_profilefield'.$counter++;
 
         $profile_fields = $this->get_profile_fields();
         if (empty($profile_fields)) {
@@ -109,15 +111,15 @@ class user_filter_profilefield extends user_filter_type {
 
         $profile  = $data['profile'];
         $operator = $data['operator'];
-        $value    = addslashes($data['value']);
+        $value    = $data['value'];
 
+        $params = array();
         if (!array_key_exists($profile, $profile_fields)) {
-            return '';
+            return array('', array());
         } 
 
         $where = "";
         $op = " IN ";
-        $ilike = sql_ilike();
 
         if ($operator < 5 and $value === '') {
             return '';
@@ -125,17 +127,29 @@ class user_filter_profilefield extends user_filter_type {
 
         switch($operator) {
             case 0: // contains
-                $where = "data $ilike '%$value%'"; break;
+                $where = $DB->sql_like('data', ":$name", false, false);
+                $params[$name] = "%$value%";
+                break;
             case 1: // does not contain
-                $where = "data NOT $ilike '%$value%'"; break;
+                $where = $DB->sql_like('data', ":$name", false, false, true);
+                $params[$name] = "%$value%";
+                break;
             case 2: // equal to
-                $where = "data $ilike '$value'"; break;
+                $where = $DB->sql_like('data', ":$name", false, false);
+                $params[$name] = "$value";
+                break;
             case 3: // starts with
-                $where = "data $ilike '$value%'"; break;
+                $where = $DB->sql_like('data', ":$name", false, false);
+                $params[$name] = "$value%";
+                break;
             case 4: // ends with
-                $where = "data $ilike '%$value'"; break;
+                $where = $DB->sql_like('data', ":$name", false, false);
+                $params[$name] = "%$value";
+                break;
             case 5: // empty
-                $where = "data=''"; break;
+                $where = "data = :$name";
+                $params[$name] = "";
+                break;
             case 6: // is not defined
                 $op = " NOT IN "; break;
             case 7: // is defined
@@ -150,7 +164,7 @@ class user_filter_profilefield extends user_filter_type {
         if ($where !== '') {
             $where = "WHERE $where";
         }
-        return "id $op (SELECT userid FROM {$CFG->prefix}user_info_data $where)";
+        return array("id $op (SELECT userid FROM {user_info_data} $where)", $params);
     }
 
     /**
@@ -174,7 +188,7 @@ class user_filter_profilefield extends user_filter_type {
             return '';
         } 
 
-        $a = new object();
+        $a = new stdClass();
         $a->label    = $this->_label;
         $a->value    = $value;
         $a->profile  = $profile_fields[$profile];
