diff --git a/user/filters/user_filter_forms.php b/user/filters/user_filter_forms.php
index 3186ebc..865aafe 100644
--- a/user/filters/user_filter_forms.php
+++ b/user/filters/user_filter_forms.php
@@ -29,6 +29,18 @@ class user_add_filter_form extends moodleform {
         // Don't use last advanced state
         $mform->setShowAdvanced(false);
     }
+
+    // RL EDIT: BJB130214 - from rlmoodle.reporting.git
+    /**
+     * Obtain the list of elements that that one or more elements depend on
+     *
+     * @return  array  List of the names of the appropriate elements
+     */
+    function get_dependencies() {
+        // fetch from QuickForm
+        return $this->_form->_dependencies;
+    }
+    // End RL EDIT
 }
 
 class user_active_filter_form extends moodleform {
