diff --git a/notes/edit_form.php b/notes/edit_form.php
index 5116789..979b432 100644
--- a/notes/edit_form.php
+++ b/notes/edit_form.php
@@ -1,4 +1,4 @@
-<?php  // $Id$
+<?php
 
 if (!defined('MOODLE_INTERNAL')) {
     die('Direct access to this script is forbidden.');    ///  It must be included from a Moodle page
@@ -10,18 +10,16 @@ class note_edit_form extends moodleform {
 
     function definition() {
         $mform    =& $this->_form;
-        $strcontent = get_string('content', 'notes');
-        $strpublishstate = get_string('publishstate', 'notes');
         $mform->addElement('header', 'general', get_string('note', 'notes'));
 
-        $mform->addElement('textarea', 'content', $strcontent, array('rows'=>15, 'cols'=>40));
+        $mform->addElement('textarea', 'content', get_string('content', 'notes'), array('rows'=>15, 'cols'=>40));
         $mform->setType('content', PARAM_RAW);
         $mform->addRule('content', get_string('nocontent', 'notes'), 'required', null, 'client');
 
-        $mform->addElement('select', 'publishstate', $strpublishstate, note_get_state_names());
+        $mform->addElement('select', 'publishstate', get_string('publishstate', 'notes'), note_get_state_names());
         $mform->setDefault('publishstate', NOTES_STATE_PUBLIC);
         $mform->setType('publishstate', PARAM_ALPHA);
-        $mform->setHelpButton('publishstate', array('status', $strpublishstate, 'notes'));
+        $mform->addHelpButton('publishstate', 'publishstate', 'notes');
 
         $this->add_action_buttons();
 
@@ -35,4 +33,3 @@ class note_edit_form extends moodleform {
         $mform->setType('id', PARAM_INT);
     }
 }
-?>
