diff --git a/mod/chat/version.php b/mod/chat/version.php
index f9dc159..55337a9 100644
--- a/mod/chat/version.php
+++ b/mod/chat/version.php
@@ -1,12 +1,10 @@
-<?php // $Id$
+<?php
 
 /////////////////////////////////////////////////////////////////////////////////
 ///  Code fragment to define the version of chat
 ///  This fragment is called by moodle_needs_upgrading() and /admin/index.php
 /////////////////////////////////////////////////////////////////////////////////
 
-$module->version  = 2009031100;   // The (date) version of this module
-$module->requires = 2007101509;  // Requires this Moodle version
+$module->version  = 2010080302;   // The (date) version of this module
+$module->requires = 2010080300;  // Requires this Moodle version
 $module->cron     = 300;          // How often should cron check this module (seconds)?
-
-?>
