diff --git a/admin/xmldb/actions/XMLDBAction.class.php b/admin/xmldb/actions/XMLDBAction.class.php
index 81a2a58..f65e3a9 100644
--- a/admin/xmldb/actions/XMLDBAction.class.php
+++ b/admin/xmldb/actions/XMLDBAction.class.php
@@ -1,32 +1,36 @@
-<?php // $Id$
-
-///////////////////////////////////////////////////////////////////////////
-//                                                                       //
-// NOTICE OF COPYRIGHT                                                   //
-//                                                                       //
-// Moodle - Modular Object-Oriented Dynamic Learning Environment         //
-//          http://moodle.com                                            //
-//                                                                       //
-// Copyright (C) 1999 onwards Martin Dougiamas        http://dougiamas.com  //
-//           (C) 2001-3001 Eloy Lafuente (stronk7) http://contiento.com  //
-//                                                                       //
-// This program is free software; you can redistribute it and/or modify  //
-// it under the terms of the GNU General Public License as published by  //
-// the Free Software Foundation; either version 2 of the License, or     //
-// (at your option) any later version.                                   //
-//                                                                       //
-// This program is distributed in the hope that it will be useful,       //
-// but WITHOUT ANY WARRANTY; without even the implied warranty of        //
-// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         //
-// GNU General Public License for more details:                          //
-//                                                                       //
-//          http://www.gnu.org/copyleft/gpl.html                         //
-//                                                                       //
-///////////////////////////////////////////////////////////////////////////
-
-/// This is the main action class. It implements all the basic
-/// functionalities to be shared by each action.
+<?php
 
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * @package   xmldb-editor
+ * @copyright 2003 onwards Eloy Lafuente (stronk7) {@link http://stronk7.com}
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+/**
+ * Main xmldb action clasee
+ *
+ * Main xmldb action class. It implements all the basic
+ * functionalities to be shared by each action.
+ *
+ * @package   xmldb-editor
+ * @copyright 2003 onwards Eloy Lafuente (stronk7) {@link http://stronk7.com}
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
 class XMLDBAction {
 
     var $does_generate;  //Type of value returned by the invoke method
@@ -47,7 +51,7 @@ class XMLDBAction {
 
     var $postaction;     //Action to execute at the end of the invoke script
 
-    var $sesskey_protected; // Actions must be protected by sesskey mechanishm
+    var $sesskey_protected; // Actions must be protected by sesskey mechanism
 
     /**
      * Constructor
@@ -101,7 +105,7 @@ class XMLDBAction {
     }
 
     /**
-     * getPostAtion method, returns the action to launch after executing
+     * getPostAction method, returns the action to launch after executing
      * another one
      */
     function getPostAction() {
@@ -122,7 +126,11 @@ class XMLDBAction {
      */
     function loadStrings($strings) {
     /// Load some commonly used strings
+        if (get_string_manager()->string_exists($this->title, 'xmldb')) {
         $this->str['title'] = get_string($this->title, 'xmldb');
+        } else {
+            $this->str['title'] = $this->title;
+        }
 
     /// Now process the $strings array loading it in the $str atribute
         if ($strings) {
@@ -148,7 +156,7 @@ class XMLDBAction {
     /// If we are used any dir, save it in the lastused session object
     /// Some actions can use it to perform positioning
         if ($lastused = optional_param ('dir', NULL, PARAM_PATH)) {
-            $SESSION->lastused = stripslashes_safe($lastused);
+            $SESSION->lastused = $lastused;
         }
 
         $this->postaction = optional_param ('postaction', NULL, PARAM_ALPHAEXT);
@@ -195,5 +203,59 @@ class XMLDBAction {
         }
         return $result;
     }
+
+    /**
+     * This function will generate the PHP code needed to
+     * implement the upgrade_xxxx_savepoint() php calls in
+     * upgrade code generated from the editor. It's used by
+     * the view_structure_php and view_table_php actions
+     *
+     * @param xmldb_structure structure object containing all the info
+     * @return string PHP code to be used to mark a reached savepoint
+     */
+    function upgrade_savepoint_php($structure) {
+
+        $path = $structure->getPath();
+
+    /// Trim "db" from path
+        $path = dirname($path);
+
+    /// Get pluginname, plugindir and plugintype
+        $pluginname = basename($path);
+        if ($path == 'lib') { /// exception for lib (not proper plugin)
+            $plugindir = 'lib';
+            $plugintype = 'lib';
+        } else { /// rest of plugins
+            //TODO: this is not nice and may fail, plugintype should be passed around somehow instead
+            $plugintypes = get_plugin_types(false);
+            $plugindir = dirname($path);
+            $plugindir = str_replace('\\', '/', $plugindir);
+            $plugintype = array_search($plugindir, $plugintypes);
+        }
+
+        $result = '';
+
+        switch ($plugintype ) {
+            case 'lib': /// has own savepoint function
+                $result = XMLDB_LINEFEED .
+                          '        // Main savepoint reached' . XMLDB_LINEFEED .
+                          '        upgrade_main_savepoint(true, XXXXXXXXXX);' . XMLDB_LINEFEED;
+                break;
+            case 'mod': /// has own savepoint function
+                $result = XMLDB_LINEFEED .
+                          '        // ' . $pluginname . ' savepoint reached' . XMLDB_LINEFEED .
+                          '        upgrade_mod_savepoint(true, XXXXXXXXXX, ' . "'$pluginname'" . ');' . XMLDB_LINEFEED;
+                break;
+            case 'block': /// has own savepoint function
+                $result = XMLDB_LINEFEED .
+                          '        // ' . $pluginname . ' savepoint reached' . XMLDB_LINEFEED .
+                          '        upgrade_block_savepoint(true, XXXXXXXXXX, ' . "'$pluginname'" . ');' . XMLDB_LINEFEED;
+                break;
+            default: /// rest of plugins
+                $result = XMLDB_LINEFEED .
+                          '        // ' . $pluginname . ' savepoint reached' . XMLDB_LINEFEED .
+                          '        upgrade_plugin_savepoint(true, XXXXXXXXXX, ' . "'$plugintype'" . ', ' . "'$pluginname'" . ');' . XMLDB_LINEFEED;
+        }
+        return $result;
+    }
 }
-?>
