diff --git a/grade/edit/letter/edit_form.php b/grade/edit/letter/edit_form.php
index 8228142..9931484 100644
--- a/grade/edit/letter/edit_form.php
+++ b/grade/edit/letter/edit_form.php
@@ -15,7 +15,6 @@
 // You should have received a copy of the GNU General Public License
 // along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
 
-
 if (!defined('MOODLE_INTERNAL')) {
     die('Direct access to this script is forbidden.');    ///  It must be included from a Moodle page
 }
@@ -24,7 +23,7 @@ require_once $CFG->libdir.'/formslib.php';
 
 class edit_letter_form extends moodleform {
 
-    function definition() {
+    public function definition() {
         $mform =& $this->_form;
         $num   = $this->_customdata['num'];
         $admin = $this->_customdata['admin'];
@@ -34,11 +33,9 @@ class edit_letter_form extends moodleform {
         // Only show "override site defaults" checkbox if editing the course grade letters
         if (!$admin) {
             $mform->addElement('checkbox', 'override', get_string('overridesitedefaultgradedisplaytype', 'grades'));
-            $mform->setHelpButton('override', array('overridesitedefaultgradedisplaytype', get_string('overridesitedefaultgradedisplaytype', 'grades'), 'grade'));
+            $mform->addHelpButton('override', 'overridesitedefaultgradedisplaytype', 'grades');
         }
 
-        $gradeletterhelp   = get_string('configgradeletter', 'grades');
-        $gradeboundaryhelp = get_string('configgradeboundary', 'grades');
         $gradeletter       = get_string('gradeletter', 'grades');
         $gradeboundary     = get_string('gradeboundary', 'grades');
 
@@ -53,7 +50,7 @@ class edit_letter_form extends moodleform {
 
             $mform->addElement('text', $gradelettername, $gradeletter." $i");
             if ($i == 1) {
-                $mform->setHelpButton($gradelettername, array('gradeletter', get_string('gradeletter', 'grades'), 'grade'));
+                $mform->addHelpButton($gradelettername, 'gradeletter', 'grades');
             }
             $mform->setType($gradelettername, PARAM_TEXT);
             
@@ -64,7 +61,7 @@ class edit_letter_form extends moodleform {
 
             $mform->addElement('select', $gradeboundaryname, $gradeboundary." $i", $percentages);
             if ($i == 1) {
-                $mform->setHelpButton($gradeboundaryname, array('gradeboundary', get_string('gradeboundary', 'grades'), 'grade'));
+                $mform->addHelpButton($gradeboundaryname, 'gradeboundary', 'grades');
             }
             $mform->setDefault($gradeboundaryname, -1);
             $mform->setType($gradeboundaryname, PARAM_INT);
@@ -85,4 +82,4 @@ class edit_letter_form extends moodleform {
 
 }
 
-?>
+
