diff --git a/search/indexersplash.php b/search/indexersplash.php
index 5227e8b..78ac128 100644
--- a/search/indexersplash.php
+++ b/search/indexersplash.php
@@ -1,59 +1,60 @@
 <?php
-/**
-* Global Search Engine for Moodle
-*
-* @package search
-* @category core
-* @subpackage search_engine
-* @author Michael Champanis (mchampan) [cynnical@gmail.com], Valery Fremaux [valery.fremaux@club-internet.fr] > 1.8
-* @date 2008/03/31
-* @license http://www.gnu.org/copyleft/gpl.html GNU Public License
-*
-* This file serves as a splash-screen (entry page) to the indexer script -
-* it is in place to prevent accidental reindexing which can lead to a loss
-* of time, amongst other things.
-*/
+    /**
+    * Global Search Engine for Moodle
+    *
+    * @package search
+    * @category core
+    * @subpackage search_engine
+    * @author Michael Champanis (mchampan) [cynnical@gmail.com], Valery Fremaux [valery.fremaux@club-internet.fr] > 1.8
+    * @date 2008/03/31
+    * @version prepared for 2.0
+    * @license http://www.gnu.org/copyleft/gpl.html GNU Public License
+    *
+    * This file serves as a splash-screen (entry page) to the indexer script -
+    * it is in place to prevent accidental reindexing which can lead to a loss
+    * of time, amongst other things.
+    */
 
-/**
-* includes and requires
-*/
-require_once('../config.php');
-require_once($CFG->dirroot.'/search/lib.php');
-
-/// makes inclusions of the Zend Engine more reliable
+    /**
+    * includes and requires
+    */
+    require_once('../config.php');
 
+    /// makes inclusions of the Zend Engine more reliable
     ini_set('include_path', $CFG->dirroot.DIRECTORY_SEPARATOR.'search'.PATH_SEPARATOR.ini_get('include_path'));
 
-/// check global search is enabled 
+    require_once($CFG->dirroot.'/search/lib.php');
+    
+    /// check global search is enabled 
 
     require_login();
     
     if (empty($CFG->enableglobalsearch)) {
-        error(get_string('globalsearchdisabled', 'search'));
+        print_error('globalsearchdisabled', 'search');
     }
     
-    if (!has_capability('moodle/site:doanything', get_context_instance(CONTEXT_SYSTEM))) {
-        error(get_string('beadmin', 'search'), $CFG->wwwroot.'/login/index.php');
+    if (!has_capability('moodle/site:config', get_context_instance(CONTEXT_SYSTEM))) {
+        print_error('beadmin', 'search', get_login_url());
     } 
     
-    require_once($CFG->dirroot.'/search/indexlib.php');
+    require_once("$CFG->dirroot/search/indexlib.php");
     $indexinfo = new IndexInfo();
     
     if ($indexinfo->valid()) {
         $strsearch = get_string('search', 'search');
         $strquery  = get_string('stats');
         
-        $navlinks[] = array('name' => $strsearch, 'link' => "index.php", 'type' => 'misc');
-        $navlinks[] = array('name' => $strquery, 'link' => "stats.php", 'type' => 'misc');
-        $navlinks[] = array('name' => get_string('runindexer','search'), 'link' => null, 'type' => 'misc');
-        // if ($CFG->version <= 2007021541){ // 1.8 branch stable timestamp NOT RELIABLE
-        if (!function_exists('build_navigation')){ // 1.8 branch stable timestamp
-            $navigation = '';
-        } else {
-            $navigation = build_navigation($navlinks);
-        }
+        // print page header
         $site = get_site();
-        print_header("$strsearch", "$site->fullname" , $navigation, "", "", true, "&nbsp;", navmenu($site));
+
+        $PAGE->set_url('/search/indexersplash.php');
+        $PAGE->set_context(get_context_instance(CONTEXT_SYSTEM));
+        $PAGE->navbar->add($strsearch, new moodle_url('/search/index.php'));
+        $PAGE->navbar->add($strquery, new moodle_url('/search/stats.php'));
+        $PAGE->navbar->add(get_string('runindexer','search'));
+        $PAGE->set_title($strsearch);
+        $PAGE->set_heading($site->fullname);
+        echo $OUTPUT->header();
      
         mtrace("<pre>The data directory ($indexinfo->path) contains $indexinfo->filecount files, and\n"
               ."there are ".$indexinfo->dbcount." records in the <em>block_search_documents</em> table.\n"
@@ -69,9 +70,8 @@ require_once($CFG->dirroot.'/search/lib.php');
               ."<a href='tests/index.php'>Test indexing</a> or "
               ."<a href='indexer.php?areyousure=yes'>Continue indexing</a> or <a href='index.php'>Back to query page</a>."
               ."</pre>");
-        print_footer();
-    } 
-    else {
+        echo $OUTPUT->footer();
+    } else {
         header('Location: indexer.php?areyousure=yes');
     }
 ?>
\ No newline at end of file
