diff --git a/search/indexlib.php b/search/indexlib.php
index 75a89f9..3c2075e 100644
--- a/search/indexlib.php
+++ b/search/indexlib.php
@@ -7,6 +7,7 @@
 * @subpackage search_engine
 * @author Michael Champanis (mchampan) [cynnical@gmail.com], Valery Fremaux [valery.fremaux@club-internet.fr] > 1.8
 * @date 2008/03/31
+* @version prepared for 2.0
 * @license http://www.gnu.org/copyleft/gpl.html GNU Public License
 * 
 * Index info class
@@ -37,7 +38,7 @@ class IndexInfo {
             $time;        //date index was generated
     
     public function __construct($path = SEARCH_INDEX_PATH) {
-        global $CFG, $db;
+        global $CFG, $DB;
         
         $this->path = $path;
         
@@ -65,27 +66,25 @@ class IndexInfo {
         $db_exists = false; //for now
         
         //get all the current tables in moodle
-        $admin_tables = $db->MetaTables();
-        
-        //TODO: use new IndexDBControl class for database checks?
+        $admin_tables = $DB->get_tables();
         
         //check if our search table exists
-        if (in_array($CFG->prefix.SEARCH_DATABASE_TABLE, $admin_tables)) {
+        if (in_array(SEARCH_DATABASE_TABLE, $admin_tables)) {
             //retrieve database information if it does
             $db_exists = true;
             
             //total documents
-            $this->dbcount = count_records(SEARCH_DATABASE_TABLE);
+            $this->dbcount = $DB->count_records(SEARCH_DATABASE_TABLE);
             
             //individual document types
-            // $types = search_get_document_types();
-            $types = search_collect_searchables(true, false);
-            sort($types);
+            $types = search_collect_searchables(false, false);
+            asort($types);
             
-            foreach($types as $type) {
-                $c = count_records(SEARCH_DATABASE_TABLE, 'doctype', $type);
-                $this->types[$type] = (int)$c;
+            foreach(array_keys($types) as $type) {
+                $c = $DB->count_records(SEARCH_DATABASE_TABLE, array('doctype' => $type));
+                $types[$type]->records = (int)$c;
             }
+            $this->types = $types;
         } else {
             $this->dbcount = 0;
             $this->types = array();
@@ -178,14 +177,13 @@ class IndexDBControl {
     /**
     * does the table exist?
     * @deprecated
-    * @uses CFG, db
+    * @uses $CFG, $DB
     */
     public function checkTableExists() {
-        global $CFG, $db;
+        global $CFG, $DB;
         
-        $table = SEARCH_DATABASE_TABLE;
-        $tables = $db->MetaTables();
-        if (in_array($CFG->prefix.$table, $tables)) {
+        $tables = $DB->get_tables();
+        if (in_array(SEARCH_DATABASE_TABLE, $tables)) {
             return true;
         } 
         else {
@@ -194,17 +192,19 @@ class IndexDBControl {
     } //checkTableExists
 
     /**
+    * NEVER USED
+    *
     * is our database setup valid?
     * @uses db, CFG
     * @deprecated Database is installed at install and should not be dropped out
-    */
+    *
     public function checkDB() {
         global $CFG, $db;
         
         $sqlfile = "{$CFG->dirroot}/search/db/$CFG->dbtype.sql";
         $ret = false;
         if ($this->checkTableExists()) {
-            execute_sql('drop table '.$CFG->prefix.SEARCH_DATABASE_TABLE, false);
+            execute_sql('drop table '.SEARCH_DATABASE_TABLE, false);
         }
 
         //turn output buffering on - to hide modify_database() output
@@ -214,15 +214,15 @@ class IndexDBControl {
         //chuck the buffer and resume normal operation
         ob_end_clean(); 
         return $ret;
-    } //checkDB
+    } //checkDB */
 
     /**
     * add a document record to the table
     * @param document must be a Lucene SearchDocument instance
-    * @uses db, CFG
+    * @uses $CFG, $DB
     */
     public function addDocument($document=null) {
-        global $db, $CFG;
+        global $DB, $CFG;
         
         if ($document == null) {
              return false;
@@ -232,17 +232,16 @@ class IndexDBControl {
         $doc->doctype   = $document->doctype;
         $doc->docid     = $document->docid;
         $doc->itemtype  = $document->itemtype;
-        $doc->title     = search_escape_string($document->title);
-        $doc->url       = search_escape_string($document->url);
+        $doc->title     = $document->title;
+        $doc->url       = $document->url;
         $doc->updated   = time();
         $doc->docdate   = $document->date;
         $doc->courseid  = $document->course_id;
         $doc->groupid   = $document->group_id;
         
-        if ($doc->groupid < 0) $doc->groupid = 0;
-        
         //insert summary into db
-        $id = insert_record(SEARCH_DATABASE_TABLE, $doc);
+        $table = SEARCH_DATABASE_TABLE;
+        $id = $DB->insert_record($table, $doc);
         
         return $id;
     } 
@@ -250,12 +249,13 @@ class IndexDBControl {
     /**
     * remove a document record from the index
     * @param document must be a Lucene document instance, or at least a dbid enveloppe
-    * @uses db
+    * @uses $DB
     */
     public function delDocument($document) {
-        global $db;
+        global $DB;
         
-        delete_records(SEARCH_DATABASE_TABLE, 'id', $document->dbid);
+        $table = SEARCH_DATABASE_TABLE;
+        $DB->delete_records($table, array('id' => $document->dbid));
     }
 } 
 
