diff --git a/search/querylib.php b/search/querylib.php
index b7b9714..6f91f92 100644
--- a/search/querylib.php
+++ b/search/querylib.php
@@ -13,7 +13,7 @@
 /**
 * includes and requires
 */
-require_once("{$CFG->dirroot}/search/Zend/Search/Lucene.php");
+require_once($CFG->dirroot.'/search/Zend/Search/Lucene.php');
 
 define('DEFAULT_POPUP_SETTINGS', "\"menubar=0,location=0,scrollbars,resizable,width=600,height=450\"");
 
@@ -220,6 +220,7 @@ class SearchQuery {
     private function process_results($all=false) {
         global $USER;
 
+        // unneeded since changing the default Zend Lexer
         // $term = mb_convert_case($this->term, MB_CASE_LOWER, 'UTF-8');
         $term = $this->term;
         $page = optional_param('page', 1, PARAM_INT);
@@ -325,8 +326,6 @@ class SearchQuery {
     */
     public function page_numbers() {
       $pages  = $this->total_pages();
-      // $query  = htmlentities($this->term);
-      // http://moodle.org/mod/forum/discuss.php?d=115788
       $query = htmlentities($this->term,ENT_NOQUOTES,'utf-8');
       $page   = $this->pagenumber;
       $next   = get_string('next', 'search');
@@ -388,29 +387,29 @@ class SearchQuery {
     * // TODO reorder parameters more consistently
     */
     private function can_display(&$user, $this_id, $doctype, $course_id, $group_id, $path, $item_type, $context_id, &$searchables) {
-        global $CFG;
+        global $CFG, $DB;
        
       /**
       * course related checks
       */
       // admins can see everything, anyway.
-      if (has_capability('moodle/site:doanything', get_context_instance(CONTEXT_SYSTEM))){
+      if (has_capability('moodle/site:config', get_context_instance(CONTEXT_SYSTEM))){
         return true;
       }
             
         // first check course compatibility against user : enrolled users to that course can see. 
-        $myCourses = get_my_courses($user->id);
+        $myCourses = enrol_get_users_courses($user->id, true);
         $unenroled = !in_array($course_id, array_keys($myCourses));
         
         // if guests are allowed, logged guest can see
-        $isallowedguest = (isguest()) ? get_field('course', 'guest', 'id', $course_id) : false ;
+        $isallowedguest = false; //TODO: this will be harder to do now because we do not have guest field in course table any more
         
         if ($unenroled && !$isallowedguest){
             return false;
         }
         
         // if user is enrolled or is allowed user and course is hidden, can he see it ?
-        $visibility = get_field('course', 'visible', 'id', $course_id);
+        $visibility = $DB->get_field('course', 'visible', array('id' => $course_id));
         if ($visibility <= 0){
             if (!has_capability('moodle/course:viewhiddencourses', get_context_instance(CONTEXT_COURSE, $course_id))){
                 return false;
@@ -431,7 +430,7 @@ class SearchQuery {
         if ($searchable_instance->location == 'internal'){
             include_once "{$CFG->dirroot}/search/documents/{$doctype}_document.php";
         } else {
-            include_once "{$CFG->dirroot}/{$searchable_instance->location}/$doctype/search_document.php";
+            include_once "{$CFG->dirroot}/{$searchable_instance->location}/{$doctype}/search_document.php";
         }
         $access_check_function = "{$doctype}_check_text_access";
         
@@ -492,13 +491,5 @@ class SearchQuery {
     public function get_results_per_page() {
       return $this->results_per_page;
     }
-
-    /**
-    *
-    */    
-    public function __destruct(){
-        unset($this->index);
-    }
-
 }
 ?>
\ No newline at end of file
