diff --git a/search/searchtypes.php b/search/searchtypes.php
index 308c89d..0291262 100644
--- a/search/searchtypes.php
+++ b/search/searchtypes.php
@@ -8,6 +8,7 @@
 * @subpackage search_engine
 * @author Michael Champanis (mchampan) [cynnical@gmail.com], Valery Fremaux [valery.fremaux@club-internet.fr] > 1.8
 * @date 2008/03/31
+* @version prepared for 2.0
 * @license http://www.gnu.org/copyleft/gpl.html GNU Public License
 *
 * Searcheable types
@@ -30,5 +31,6 @@ define('SEARCH_TYPE_LABEL', 'label');
 define('SEARCH_EXTRAS', 'user');
 define('SEARCH_TYPE_USER', 'user');
 define('PATH_FOR_SEARCH_TYPE_USER', 'user');
+define('TYPE_FOR_SEARCH_TYPE_USER', 'core');
 
 ?>
\ No newline at end of file
