diff --git a/grade/export/xls/grade_export_xls.php b/grade/export/xls/grade_export_xls.php
old mode 100755
new mode 100644
index 964e820..f1efc6e
--- a/grade/export/xls/grade_export_xls.php
+++ b/grade/export/xls/grade_export_xls.php
@@ -19,12 +19,12 @@ require_once($CFG->dirroot.'/grade/export/lib.php');
 
 class grade_export_xls extends grade_export {
 
-    var $plugin = 'xls';
+    public $plugin = 'xls';
 
     /**
      * To be implemented by child classes
      */
-    function print_grades() {
+    public function print_grades() {
         global $CFG;
         require_once($CFG->dirroot.'/lib/excellib.class.php');
 
@@ -33,7 +33,8 @@ class grade_export_xls extends grade_export {
         $strgrades = get_string('grades');
 
     /// Calculate file name
-        $downloadfilename = clean_filename("{$this->course->shortname} $strgrades.xls");
+        $shortname = format_string($this->course->shortname, true, array('context' => get_context_instance(CONTEXT_COURSE, $this->course->id)));
+        $downloadfilename = clean_filename("$shortname $strgrades.xls");
     /// Creating a workbook
         $workbook = new MoodleExcelWorkbook("-");
     /// Sending HTTP headers
@@ -103,4 +104,4 @@ class grade_export_xls extends grade_export {
     }
 }
 
-?>
+
