diff --git a/admin/report/spamcleaner/settings.php b/admin/report/spamcleaner/settings.php
index 025f91a..7efc70a 100644
--- a/admin/report/spamcleaner/settings.php
+++ b/admin/report/spamcleaner/settings.php
@@ -1,5 +1,7 @@
 <?php
 
+defined('MOODLE_INTERNAL') || die;
+
 // spam cleaner
-$ADMIN->add('reports', new admin_externalpage('reportspamcleaner', get_string('spamcleaner','report_spamcleaner'), "$CFG->wwwroot/$CFG->admin/report/spamcleaner/index.php", 'moodle/site:config'));
+$ADMIN->add('reports', new admin_externalpage('reportspamcleaner', get_string('pluginname', 'report_spamcleaner'), "$CFG->wwwroot/$CFG->admin/report/spamcleaner/index.php", 'moodle/site:config'));
 
