diff --git a/course/simpletest/testcourselib.php b/course/simpletest/testcourselib.php
old mode 100755
new mode 100644
index 367baa3..2104d84
--- a/course/simpletest/testcourselib.php
+++ b/course/simpletest/testcourselib.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 
 ///////////////////////////////////////////////////////////////////////////
 //                                                                       //
@@ -38,13 +38,46 @@ if (!defined('MOODLE_INTERNAL')) {
 require_once($CFG->dirroot . '/course/lib.php');
 
 class courselib_test extends UnitTestCase {
+    var $realDB;
+    public  static $includecoverage = array('course/lib.php');
 
     function setUp() {
+        global $DB;
+        Mock::generate(get_class($DB), 'mockDB');
+        $this->realDB = $DB;
+        $DB = new mockDB();
     }
 
     function tearDown() {
+        global $DB;
+        $DB = $this->realDB;
     }
 
+    function testMoveSection() {
+        global $DB;
+        $course = new stdClass();
+        $course->numsections = 10;
+        $course->id = 1;
+        $sections = array(20 => 0, 21 => 1, 22 => 2, 23 => 3, 24 => 4, 25 => 5);
+
+        $DB->setReturnValueAt(0, 'get_records_menu', $sections);
+        $DB->expectAt(0, 'set_field', array('course_sections', 'section', 0, array('id' => 20)));
+        $DB->expectAt(1, 'set_field', array('course_sections', 'section', 1, array('id' => 21)));
+        $DB->expectAt(2, 'set_field', array('course_sections', 'section', 2, array('id' => 23)));
+        $DB->expectAt(3, 'set_field', array('course_sections', 'section', 3, array('id' => 24)));
+        $DB->expectAt(4, 'set_field', array('course_sections', 'section', 4, array('id' => 22)));
+        $DB->expectAt(5, 'set_field', array('course_sections', 'section', 5, array('id' => 25)));
+        move_section_to($course, 2, 4);
+
+        $DB->setReturnValueAt(1, 'get_records_menu', $sections);
+        $DB->expectAt(6, 'set_field', array('course_sections', 'section', 0, array('id' => 20)));
+        $DB->expectAt(7, 'set_field', array('course_sections', 'section', 1, array('id' => 24)));
+        $DB->expectAt(8, 'set_field', array('course_sections', 'section', 2, array('id' => 21)));
+        $DB->expectAt(9, 'set_field', array('course_sections', 'section', 3, array('id' => 22)));
+        $DB->expectAt(10, 'set_field', array('course_sections', 'section', 4, array('id' => 23)));
+        $DB->expectAt(11, 'set_field', array('course_sections', 'section', 5, array('id' => 25)));
+        move_section_to($course, 4, 0);
+    }
 
     function testReorderSections() {
         $sections = array(20 => 0, 21 => 1, 22 => 2, 23 => 3, 24 => 4, 25 => 5);
