diff --git a/filter/censor/filtersettings.php b/filter/censor/filtersettings.php
index 6effb9c..323d1ef 100644
--- a/filter/censor/filtersettings.php
+++ b/filter/censor/filtersettings.php
@@ -1,6 +1,9 @@
-<?php  //$Id$
+<?php
 
-$settings->add(new admin_setting_configtextarea('filter_censor_badwords', get_string('badwordslist','admin'),
-               get_string('badwordsconfig', 'admin').'<br />'.get_string('badwordsdefault', 'admin'), ''));
+defined('MOODLE_INTERNAL') || die;
+
+if ($ADMIN->fulltree) {
 
-?>
+    $settings->add(new admin_setting_configtextarea('filter_censor_badwords', get_string('badwordslist','admin'),
+                   get_string('badwordsconfig', 'admin').'<br />'.get_string('badwordsdefault', 'admin'), ''));
+}
