diff --git a/admin/tool/unsuproles/index.php b/admin/tool/unsuproles/index.php
index e1c6cf0..3135376 100644
--- a/admin/tool/unsuproles/index.php
+++ b/admin/tool/unsuproles/index.php
@@ -61,7 +61,17 @@ if ($action === 'delete') {
     //show confirmation
     echo $OUTPUT->header();
     $yesurl = new moodle_url($PAGE->url, array('roleid'=>$roleid, 'contextlevel'=>$contextlevel, 'action'=>'delete', 'confirm'=>1, 'sesskey'=>sesskey()));
+    try {
         $levelname = get_contextlevel_name($contextlevel);
+    } catch (Exception $e) {
+        if (file_exists($CFG->dirroot.'/elis/program/accesslib.php')) {
+            require_once($CFG->dirroot.'/elis/program/accesslib.php');
+            $contextclass = context_elis_helper::get_class_for_level($contextlevel);
+            if (!empty($contextclass)) {
+                $levelname = $contextclass::get_level_name();
+            }
+        }
+    }
     $rolename = format_string($role->name);
     $message = get_string('confirmdelete', 'tool_unsuproles', array('level'=>$levelname, 'role'=>$rolename));
     echo $OUTPUT->confirm($message, $yesurl, $PAGE->url);
@@ -96,7 +106,17 @@ if (!$problems) {
     $roles = get_all_roles();
     $data = array();
     foreach ($problems as $problem) {
+        try {
             $levelname = get_contextlevel_name($problem->contextlevel);
+        } catch (Exception $e) {
+            if (file_exists($CFG->dirroot.'/elis/program/accesslib.php')) {
+                require_once($CFG->dirroot.'/elis/program/accesslib.php');
+                $contextclass = context_elis_helper::get_class_for_level($problem->contextlevel);
+                if (!empty($contextclass)) {
+                    $levelname = $contextclass::get_level_name();
+                }
+            }
+        }
         $rolename = format_string($roles[$problem->roleid]->name);
         //TODO: show list of users if count low
         $count = $problem->racount;
