diff --git a/mod/glossary/formats/faq/faq_format.php b/mod/glossary/formats/faq/faq_format.php
index 0780a58..99ec759 100644
--- a/mod/glossary/formats/faq/faq_format.php
+++ b/mod/glossary/formats/faq/faq_format.php
@@ -1,8 +1,7 @@
-<?php  // $Id$
+<?php
 
-function glossary_show_entry_faq($course, $cm, $glossary, $entry, $mode="", $hook="", $printicons=1, $ratings=NULL, $aliases=true) {
+function glossary_show_entry_faq($course, $cm, $glossary, $entry, $mode="", $hook="", $printicons=1, $aliases=true) {
     global $USER;
-    $return = false;
     if ( $entry ) {
 
         echo '<table class="glossarypost faq" cellspacing="0">';
@@ -21,7 +20,7 @@ function glossary_show_entry_faq($course, $cm, $glossary, $entry, $mode="", $hoo
         echo '<td class="entryattachment">';
 
         glossary_print_entry_approval($cm, $entry, $mode);
-        glossary_print_entry_attachment($entry,'html','right');
+        glossary_print_entry_attachment($entry, $cm, 'html', 'right');
         echo '</td>';
 
         echo '</tr>';
@@ -30,11 +29,11 @@ function glossary_show_entry_faq($course, $cm, $glossary, $entry, $mode="", $hoo
         echo '<td colspan="2" class="entry">';
         echo '<b>'.get_string('answer','glossary').':</b> ';
 
-        glossary_print_entry_definition($entry);
+        glossary_print_entry_definition($entry, $glossary, $cm);
 
         echo '</td></tr>';
         echo '<tr valign="top"><td colspan="3" class="entrylowersection">';
-        $return = glossary_print_entry_lower_section($course, $cm, $glossary, $entry, $mode, $hook, $printicons, $ratings, $aliases);
+        glossary_print_entry_lower_section($course, $cm, $glossary, $entry, $mode, $hook, $printicons, $aliases);
         echo '</td></tr></table>';
 
     } else {
@@ -42,10 +41,9 @@ function glossary_show_entry_faq($course, $cm, $glossary, $entry, $mode="", $hoo
         print_string('noentry', 'glossary');
         echo '</div>';
     }
-    return $return;
 }
 
-function glossary_print_entry_faq($course, $cm, $glossary, $entry, $mode='', $hook='', $printicons=1, $ratings=NULL) {
+function glossary_print_entry_faq($course, $cm, $glossary, $entry, $mode='', $hook='', $printicons=1) {
 
     //The print view for this format is exactly the normal view, so we use it
 
@@ -57,4 +55,4 @@ function glossary_print_entry_faq($course, $cm, $glossary, $entry, $mode='', $ho
 
 }
 
-?>
+
