diff --git a/question/simpletest/testimportexport.php b/question/simpletest/testimportexport.php
index e2f2812..87aaa91 100644
--- a/question/simpletest/testimportexport.php
+++ b/question/simpletest/testimportexport.php
@@ -1,5 +1,4 @@
 <?php
-
 // This file is part of Moodle - http://moodle.org/
 //
 // Moodle is free software: you can redistribute it and/or modify
@@ -15,16 +14,18 @@
 // You should have received a copy of the GNU General Public License
 // along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
 
-
 /**
  * Unit tests for the question import and export system.
  *
- * @package core
+ * @package    moodlecore
  * @subpackage questionbank
- * @copyright 2010 The Open University
+ * @copyright  2009 The Open University
  * @license http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
 
+
+defined('MOODLE_INTERNAL') || die();
+
 require_once($CFG->libdir . '/questionlib.php');
 require_once($CFG->dirroot . '/question/format.php');
 
@@ -72,11 +73,11 @@ class qformat_default_test extends UnitTestCase {
         $path = '$course$/Tim\'s questions/Tricky things like // //// and so on/Category name ending in // / // and one that starts with one/<span lang="en" class="multilang">Matematically<//span> <span lang="sv" class="multilang">Matematiskt (svenska)<//span>';
         $this->assertEqual(array(
                     '$course$',
-                    "Tim\\'s questions",
+                    "Tim's questions",
                     "Tricky things like / // and so on",
                     'Category name ending in /',
                     '/ and one that starts with one',
-                    '<span lang=\"en\" class=\"multilang\">Matematically</span> <span lang=\"sv\" class=\"multilang\">Matematiskt (svenska)</span>'
+                    '<span lang="en" class="multilang">Matematically</span> <span lang="sv" class="multilang">Matematiskt (svenska)</span>'
                 ), $format->split_category_path($path));
     }
 
