diff --git a/mod/chat/gui_sockets/chatinput.php b/mod/chat/gui_sockets/chatinput.php
index 74b2cd7..fe369b5 100644
--- a/mod/chat/gui_sockets/chatinput.php
+++ b/mod/chat/gui_sockets/chatinput.php
@@ -1,52 +1,35 @@
-<?php  // $Id$
-
-    $nomoodlecookie = true;     // Session not needed!
-
-    require('../../../config.php');
-    require('../lib.php');
-
-    $chat_sid = required_param('chat_sid', PARAM_ALPHANUM);
-
-    if (!$chatuser = get_record('chat_users', 'sid', $chat_sid)) {
-        error('Not logged in!');
-    }
-
-    //Get the user theme
-    $USER = get_record('user', 'id', $chatuser->userid);
-
-    //Setup course, lang and theme
-    course_setup($chatuser->course);
-
-    ob_start();
-    ?>
-<script type="text/javascript">
-scroll_active = true;
-function empty_field_and_submit() {
-    var cf   = document.getElementById('sendform');
-    var inpf = document.getElementById('inputform');
-    cf.chat_msgidnr.value = parseInt(cf.chat_msgidnr.value) + 1;
-    cf.chat_message.value = inpf.chat_message.value;
-    inpf.chat_message.value='';
-    cf.submit();
-    inpf.chat_message.focus();
-    return false;
-}
-function setfocus() {
-    document.getElementsByName("chat_message")[0].focus(); 
+<?php
+
+define('NO_MOODLE_COOKIES', true); // session not used here
+
+require('../../../config.php');
+require('../lib.php');
+
+$chat_sid = required_param('chat_sid', PARAM_ALPHANUM);
+
+$PAGE->set_url('/mod/chat/gui_sockets/chatinput.php', array('chat_sid'=>$chat_sid));
+
+if (!$chatuser = $DB->get_record('chat_users', array('sid'=>$chat_sid))) {
+    print_error('notlogged', 'chat');
 }
-</script>
-    <?php
 
-    $meta = ob_get_clean();
-    // TODO: there will be two onload in body tag, does it matter?
-    print_header('', '', '', 'inputform.chat_message', $meta, false, '&nbsp;', '', false, 'onload="setfocus();"');
+//Get the user theme
+$USER = $DB->get_record('user', array('id'=>$chatuser->userid));
+
+//Setup course, lang and theme
+$PAGE->set_course($DB->get_record('course', array('id' => $chatuser->course)));
+$PAGE->requires->js('/mod/chat/gui_sockets/chat_gui_sockets.js', true);
+$PAGE->requires->js_function_call('setfocus');
+$PAGE->set_focuscontrol('chat_message');
+$PAGE->set_cacheable(false);
+echo $OUTPUT->header();
 
 ?>
 
     <form action="../empty.php" method="get" target="empty" id="inputform"
           onsubmit="return empty_field_and_submit();">
-        <input type="text" name="chat_message" size="60" value="" />
-        <?php helpbutton("chatting", get_string("helpchatting", "chat"), "chat", true, false); ?>
+        <input type="text" name="chat_message" id="chat_message" size="60" value="" />
+        <?php echo $OUTPUT->help_icon('usingchat', 'chat'); ?>
     </form>
     
     <form action="<?php echo "http://$CFG->chat_serverhost:$CFG->chat_serverport/"; ?>" method="get" target="empty" id="sendform">
@@ -56,5 +39,5 @@ function setfocus() {
         <input type="hidden" name="chat_sid" value="<?php echo $chat_sid ?>" />
     </form>
 <?php
-    print_footer('empty');
+    echo $OUTPUT->footer();
 ?>
