diff --git a/mod/chat/gui_sockets/index.php b/mod/chat/gui_sockets/index.php
index 0b68d28..4a5de8b 100644
--- a/mod/chat/gui_sockets/index.php
+++ b/mod/chat/gui_sockets/index.php
@@ -1,65 +1,66 @@
-<?php  // $Id$
+<?php
 
-    require_once('../../../config.php');
-    require_once('../lib.php');
+require_once('../../../config.php');
+require_once('../lib.php');
 
-    $id      = required_param('id', PARAM_INT);
-    $groupid = optional_param('groupid', 0, PARAM_INT); //only for teachers
+$id      = required_param('id', PARAM_INT);
+$groupid = optional_param('groupid', 0, PARAM_INT); //only for teachers
 
-    if (!$chat = get_record('chat', 'id', $id)) {
-        error('Could not find that chat room!');
-    }
+$url = new moodle_url('/mod/chat/gui_sockets/index.php', array('id'=>$id));
+if ($groupid !== 0) {
+    $url->param('groupid', $groupid);
+}
+$PAGE->set_url($url);
 
-    if (!$course = get_record('course', 'id', $chat->course)) {
-        error('Could not find the course this belongs to!');
-    }
+if (!$chat = $DB->get_record('chat', array('id'=>$id))) {
+    print_error('invalidid', 'chat');
+}
 
-    if (!$cm = get_coursemodule_from_instance('chat', $chat->id, $course->id)) {
-        error('Course Module ID was incorrect');
-    }
+if (!$course = $DB->get_record('course', array('id'=>$chat->course))) {
+    print_error('invalidcourseid');
+}
 
-    require_login($course->id, false, $cm);
+if (!$cm = get_coursemodule_from_instance('chat', $chat->id, $course->id)) {
+    print_error('invalidcoursemodule');
+}
 
-    if (isguest()) {
-        error('Guest does not have access to chat rooms');
-    }
+require_login($course, false, $cm);
+$context = get_context_instance(CONTEXT_MODULE, $cm->id);
 
-    if (!$cm->visible and !has_capability('moodle/course:viewhiddenactivities', get_context_instance(CONTEXT_MODULE, $cm->id))) {
-        print_header();
-        notice(get_string("activityiscurrentlyhidden"));
-    }
+if (isguestuser()) {
+    print_error('noguests', 'chat');
+}
 
 /// Check to see if groups are being used here
      if ($groupmode = groups_get_activity_groupmode($cm)) {   // Groups are being used
         if ($groupid = groups_get_activity_group($cm)) {
-            if (!$group = groups_get_group($groupid, false)) {
-                error("That group (id $groupid) doesn't exist!");
+        if (!$group = groups_get_group($groupid)) {
+            print_error('invalidgroupid');
             }
             $groupname = ': '.$group->name;
         } else {
             $groupname = ': '.get_string('allparticipants');
         }
-    } else {
+} else {
         $groupid = 0;
         $groupname = '';
-    }
+}
 
-    $strchat = get_string('modulename', 'chat'); // must be before current_language() in chat_login_user() to force course language!!!
-
-    if (!$chat_sid = chat_login_user($chat->id, 'sockets', $groupid, $course)) {
-        error('Could not log in to chat room!!');
-    }
+$strchat = get_string('modulename', 'chat'); // must be before current_language() in chat_login_user() to force course language!!!
 
-    $params = "chat_sid=$chat_sid";
+if (!$chat_sid = chat_login_user($chat->id, 'sockets', $groupid, $course)) {
+    print_error('cantlogin');
+}
 
-?>
+$params = "chat_sid=$chat_sid";
+$courseshortname = format_string($course->shortname, true, array('context' => get_context_instance(CONTEXT_COURSE, $course->id)));
 
-<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Frameset//EN" "http://www.w3.org/TR/html4/frameset.dtd">
+?><!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Frameset//EN" "http://www.w3.org/TR/html4/frameset.dtd">
 <html>
  <head>
   <meta http-equiv="content-type" content="text/html; charset=utf-8" />
   <title>
-   <?php echo "$strchat: " . format_string($course->shortname) . ": ".format_string($chat->name,true)."$groupname" ?>
+   <?php echo "$strchat: " . $courseshortname . ": " . format_string($chat->name, true, array('context' => $context)) . "$groupname" ?>
   </title>
  </head>
  <frameset cols="*,200" border="5" framespacing="no" frameborder="yes" marginwidth="2" marginheight="1">
