diff --git a/mod/chat/gui_header_js/insert.php b/mod/chat/gui_header_js/insert.php
index 62f0dc8..6b842d2 100644
--- a/mod/chat/gui_header_js/insert.php
+++ b/mod/chat/gui_header_js/insert.php
@@ -1,83 +1,76 @@
-<?php  // $Id$
+<?php
 
-    include('../../../config.php');
-    include('../lib.php');
+include('../../../config.php');
+include('../lib.php');
 
-    $chat_sid     = required_param('chat_sid', PARAM_ALPHANUM);
-    $chat_message = required_param('chat_message', PARAM_RAW);
+$chat_sid     = required_param('chat_sid', PARAM_ALPHANUM);
+$chat_message = required_param('chat_message', PARAM_RAW);
 
-    if (!$chatuser = get_record('chat_users', 'sid', $chat_sid)) {
-        error('Not logged in!');
-    }
+$PAGE->set_url('/mod/chat/gui_header_js/insert.php', array('chat_sid'=>$chat_sid,'chat_message'=>$chat_message));
 
-    if (!$chat = get_record('chat', 'id', $chatuser->chatid)) {
-        error('No chat found');
-    }
+if (!$chatuser = $DB->get_record('chat_users', array('sid'=>$chat_sid))) {
+    print_error('notlogged', 'chat');
+}
 
-    if (!$course = get_record('course', 'id', $chat->course)) {
-        error('Could not find the course this belongs to!');
-    }
+if (!$chat = $DB->get_record('chat', array('id'=>$chatuser->chatid))) {
+    print_error('nochat', 'chat');
+}
 
-    if (!$cm = get_coursemodule_from_instance('chat', $chat->id, $course->id)) {
-        error('Course Module ID was incorrect');
-    }
+if (!$course = $DB->get_record('course', array('id'=>$chat->course))) {
+    print_error('invalidcourseid');
+}
     
-    require_login($course->id, false, $cm);
+if (!$cm = get_coursemodule_from_instance('chat', $chat->id, $course->id)) {
+    print_error('invalidcoursemodule');
+}
 
-    if (isguest()) {
-        error('Guest does not have access to chat rooms');
-    }
+require_login($course, false, $cm);
+
+if (isguestuser()) {
+    print_error('noguests');
+}
 
-    session_write_close();
+session_get_instance()->write_close();
 
 /// Delete old users now
 
-    chat_delete_old_users();
+chat_delete_old_users();
 
 /// Clean up the message
 
-    $chat_message = addslashes(clean_text(stripslashes($chat_message), FORMAT_MOODLE));  // Strip bad tags
+$chat_message = clean_text($chat_message, FORMAT_MOODLE);  // Strip bad tags
 
 /// Add the message to the database
 
-    if (!empty($chat_message)) {
+if (!empty($chat_message)) {
 
-        $message = new object();
+    $message = new stdClass();
         $message->chatid = $chatuser->chatid;
         $message->userid = $chatuser->userid;
         $message->groupid = $chatuser->groupid;
         $message->message = $chat_message;
         $message->timestamp = time();
 
-        if (!insert_record('chat_messages', $message)) {
-            error('Could not insert a chat message!');
-        }
+    $DB->insert_record('chat_messages', $message);
+    $DB->insert_record('chat_messages_current', $message);
 
         $chatuser->lastmessageping = time() - 2;
-        update_record('chat_users', $chatuser);
+    $DB->update_record('chat_users', $chatuser);
 
         if ($cm = get_coursemodule_from_instance('chat', $chat->id, $course->id)) {
             add_to_log($course->id, 'chat', 'talk', "view.php?id=$cm->id", $chat->id, $cm->id);
         }
-    }
+}
 
-    if ($chatuser->version == 'header_js') {
-        /// force msg referesh ASAP
-        if ($CFG->chat_normal_updatemode == 'jsupdated') {  // See bug MDL-6791
-            echo '<script type="text/javascript">'.
-                 "//<![CDATA[ \n".
-                 '  parent.input.enableForm();'.
-                 "//]]> \n".
-                 '</script>';
-        } else {
-            echo '<script type="text/javascript">'.
-                 "//<![CDATA[ \n".
-                 '  parent.jsupdate.location.href = parent.jsupdate.document.anchors[0].href;'.
-                 '  parent.input.enableForm();'.
-                 "//]]> \n".
-                 '</script>';
-        }
-    }
+if ($chatuser->version == 'header_js') {
+
+    $forcerefreshasap = ($CFG->chat_normal_updatemode != 'jsupdated'); // See bug MDL-6791
+
+    $module = array(
+        'name'      => 'mod_chat_header',
+        'fullpath'  => '/mod/chat/gui_header_js/module.js'
+    );
+    $PAGE->requires->js_init_call('M.mod_chat_header.init_insert_nojsupdated', array($forcerefreshasap), true, $module);
+}
 
-    redirect('../empty.php');
-?>
+redirect('../empty.php');
\ No newline at end of file
