diff --git a/mod/chat/gui_header_js/jsupdate.php b/mod/chat/gui_header_js/jsupdate.php
index 1fe83cb..cbff840 100644
--- a/mod/chat/gui_header_js/jsupdate.php
+++ b/mod/chat/gui_header_js/jsupdate.php
@@ -1,84 +1,97 @@
-<?php  // $Id$
+<?php
 
-    $nomoodlecookie = true;     // Session not needed!
+define('NO_MOODLE_COOKIES', true); // session not used here
 
-    require('../../../config.php');
-    require('../lib.php');
+require('../../../config.php');
+require('../lib.php');
 
-    $chat_sid      = required_param('chat_sid', PARAM_ALPHANUM);
-    $chat_lasttime = optional_param('chat_lasttime', 0, PARAM_INT);
-    $chat_lastrow  = optional_param('chat_lastrow', 1, PARAM_INT);
+$chat_sid      = required_param('chat_sid', PARAM_ALPHANUM);
+$chat_lasttime = optional_param('chat_lasttime', 0, PARAM_INT);
+$chat_lastrow  = optional_param('chat_lastrow', 1, PARAM_INT);
 
-    if (!$chatuser = get_record('chat_users', 'sid', $chat_sid)) {
-        error('Not logged in!');
-    }
+$url = new moodle_url('/mod/chat/gui_header_js/jsupdate.php', array('chat_sid'=>$chat_sid));
+if ($chat_lasttime !== 0) {
+    $url->param('chat_lasttime', $chat_lasttime);
+}
+if ($chat_lastrow !== 1) {
+    $url->param('chat_lastrow', $chat_lastrow);
+}
+$PAGE->set_url($url);
 
-    //Get the minimal course
-    if (!$course = get_record('course','id',$chatuser->course,'','','','','id,theme,lang')) {
-        error('incorrect course id');
-    }
 
-    //Get the user theme and enough info to be used in chat_format_message() which passes it along to
-    if (!$USER = get_record('user','id',$chatuser->userid)) { // no optimisation here, it would break again in future!
-        error('User does not exist!');
-    }
-    $USER->description = '';
+if (!$chatuser = $DB->get_record('chat_users', array('sid'=>$chat_sid))) {
+    print_error('notlogged', 'chat');
+}
 
-    //Setup course, lang and theme
-    course_setup($course);
+//Get the minimal course
+if (!$course = $DB->get_record('course', array('id'=>$chatuser->course))) {
+    print_error('invalidcourseid');
+}
 
-    // force deleting of timed out users if there is a silence in room or just entering
-    if ((time() - $chat_lasttime) > $CFG->chat_old_ping) {
+//Get the user theme and enough info to be used in chat_format_message() which passes it along to
+if (!$USER = $DB->get_record('user', array('id'=>$chatuser->userid))) { // no optimisation here, it would break again in future!
+    print_error('invaliduser');
+}
+$USER->description = '';
+
+//Setup course, lang and theme
+$PAGE->set_course($course);
+
+// force deleting of timed out users if there is a silence in room or just entering
+if ((time() - $chat_lasttime) > $CFG->chat_old_ping) {
         // must be done before chat_get_latest_message!!!
         chat_delete_old_users();
-    }
+}
 
-    if ($message = chat_get_latest_message($chatuser->chatid, $chatuser->groupid)) {
+if ($message = chat_get_latest_message($chatuser->chatid, $chatuser->groupid)) {
         $chat_newlasttime = $message->timestamp;
-    } else {
+} else {
         $chat_newlasttime = 0;
-    }
+}
 
-    if ($chat_lasttime == 0) { //display some previous messages
+if ($chat_lasttime == 0) { //display some previous messages
         $chat_lasttime = time() - $CFG->chat_old_ping; //TO DO - any better value??
-    }
+}
+
+$timenow    = time();
 
-    $timenow    = time();
+$params = array('groupid'=>$chatuser->groupid, 'chatid'=>$chatuser->chatid, 'lasttime'=>$chat_lasttime);
 
-    $groupselect = $chatuser->groupid ? " AND (groupid='".$chatuser->groupid."' OR groupid='0') " : "";
+$groupselect = $chatuser->groupid ? " AND (groupid=:groupid OR groupid=0) " : "";
 
-    $messages = get_records_select("chat_messages",
-                        "chatid = '$chatuser->chatid' AND timestamp > '$chat_lasttime' $groupselect",
+$messages = $DB->get_records_select("chat_messages_current",
+                    "chatid = :chatid AND timestamp > :lasttime $groupselect", $params,
                         "timestamp ASC");
 
-    if ($messages) {
+if ($messages) {
         $num = count($messages);
-    } else {
+} else {
         $num = 0;
-    }
+}
 
-    $chat_newrow = ($chat_lastrow + $num) % 2;
+$chat_newrow = ($chat_lastrow + $num) % 2;
 
-    // no &amp; in url, does not work in header!
-    $refreshurl = "{$CFG->wwwroot}/mod/chat/gui_header_js/jsupdate.php?chat_sid=$chat_sid&chat_lasttime=$chat_newlasttime&chat_lastrow=$chat_newrow";
-    $refreshurlamp = "{$CFG->wwwroot}/mod/chat/gui_header_js/jsupdate.php?chat_sid=$chat_sid&amp;chat_lasttime=$chat_newlasttime&amp;chat_lastrow=$chat_newrow";
+// no &amp; in url, does not work in header!
+$refreshurl = "{$CFG->wwwroot}/mod/chat/gui_header_js/jsupdate.php?chat_sid=$chat_sid&chat_lasttime=$chat_newlasttime&chat_lastrow=$chat_newrow";
+$refreshurlamp = "{$CFG->wwwroot}/mod/chat/gui_header_js/jsupdate.php?chat_sid=$chat_sid&amp;chat_lasttime=$chat_newlasttime&amp;chat_lastrow=$chat_newrow";
 
-    header('Expires: Sun, 28 Dec 1997 09:32:45 GMT');
-    header('Last-Modified: '.gmdate('D, d M Y H:i:s').' GMT');
-    header('Cache-Control: no-cache, must-revalidate');
-    header('Pragma: no-cache');
-    header('Content-Type: text/html; charset=utf-8');
-    header("Refresh: $CFG->chat_refresh_room; url=$refreshurl");
+header('Expires: Sun, 28 Dec 1997 09:32:45 GMT');
+header('Last-Modified: '.gmdate('D, d M Y H:i:s').' GMT');
+header('Cache-Control: no-cache, must-revalidate');
+header('Pragma: no-cache');
+header('Content-Type: text/html; charset=utf-8');
+header("Refresh: $CFG->chat_refresh_room; url=$refreshurl");
 
-    /// required stylesheets
-    $stylesheetshtml = '';
-    foreach ($CFG->stylesheets as $stylesheet) {
+/// required stylesheets
+$stylesheetshtml = '';
+/*foreach ($CFG->stylesheets as $stylesheet) {
+    //TODO: MDL-21120
         $stylesheetshtml .= '<link rel="stylesheet" type="text/css" href="'.$stylesheet.'" />';
-    }
+}*/
 
-    // use ob to be able to send Content-Length headers
-    // needed for Keep-Alive to work
-    ob_start();
+// use ob to be able to send Content-Length headers
+// needed for Keep-Alive to work
+ob_start();
 
 ?>
 <!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
@@ -127,7 +140,7 @@
         }
 
         $chatuser->lastping = time();
-        set_field('chat_users', 'lastping', $chatuser->lastping, 'id', $chatuser->id  );
+        $DB->set_field('chat_users', 'lastping', $chatuser->lastping, array('id'=>$chatuser->id));
 
         if ($refreshusers) {
         ?>
