diff --git a/grade/report/overview/db/access.php b/grade/report/overview/db/access.php
index 523c9e5..3e29f06 100644
--- a/grade/report/overview/db/access.php
+++ b/grade/report/overview/db/access.php
@@ -15,18 +15,18 @@
 // You should have received a copy of the GNU General Public License
 // along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
 
-$gradereport_overview_capabilities = array(
+$capabilities = array(
 
     'gradereport/overview:view' => array(
         'riskbitmask' => RISK_PERSONAL,
         'captype' => 'read',
         'contextlevel' => CONTEXT_COURSE,
-        'legacy' => array(
+        'archetypes' => array(
             'student' => CAP_ALLOW,
-            'admin' => CAP_ALLOW
+            'manager' => CAP_ALLOW
         )
     )
 
 );
 
-?>
+
