diff --git a/lib/simpletestlib/extensions/phpunit_test_case.php b/lib/simpletestlib/extensions/phpunit_test_case.php
index 1628ed7..2dd3cd0 100644
--- a/lib/simpletestlib/extensions/phpunit_test_case.php
+++ b/lib/simpletestlib/extensions/phpunit_test_case.php
@@ -26,7 +26,7 @@
          *    @param $label        Test name to display.
          *    @public
          */
-        function TestCase($label) {
+        function TestCase($label = false) {
             $this->SimpleTestCase($label);
         }
         
