diff --git a/auth/nntp/auth.php b/auth/nntp/auth.php
index 8af71ce..495b505 100644
--- a/auth/nntp/auth.php
+++ b/auth/nntp/auth.php
@@ -35,25 +35,25 @@ class auth_plugin_nntp extends auth_plugin_base {
      * Returns true if the username and password work and false if they are
      * wrong or don't exist.
      *
-     * @param string $username The username (with system magic quotes)
-     * @param string $password The password (with system magic quotes)
+     * @param string $username The username
+     * @param string $password The password
      * @return bool Authentication success or failure.
      */
     function user_login ($username, $password) {
         if (! function_exists('imap_open')) {
-            print_error('auth_nntpnotinstalled','auth');
+            print_error('auth_nntpnotinstalled','auth_nntp');
             exit;
         }
 
         global $CFG;
 
         // try each multiple host
-        $hosts = split(';', $this->config->host);
+        $hosts = explode(';', $this->config->host);
         foreach ($hosts as $host) {
             $host = '{' . trim($host) . ':' . $this->config->port . '/nntp}';
 
             error_reporting(0);
-            $connection = imap_open($host, stripslashes($username), stripslashes($password), OP_HALFOPEN);
+            $connection = imap_open($host, $username, $password, OP_HALFOPEN);
             error_reporting($CFG->debug);
 
             if ($connection) {
@@ -124,4 +124,4 @@ class auth_plugin_nntp extends auth_plugin_base {
 
 }
 
-?>
+
