diff --git a/lib/simpletest/testgradelib.php b/lib/simpletest/testgradelib.php
index 7d731bd..60c5a8e 100644
--- a/lib/simpletest/testgradelib.php
+++ b/lib/simpletest/testgradelib.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 
 ///////////////////////////////////////////////////////////////////////////
 //                                                                       //
@@ -43,4 +43,4 @@ class gradelib_test extends grade_test {
     }
 }
 
-?>
+
