diff --git a/backup/log.php b/backup/log.php
index 7ef38b1..3f83de6 100644
--- a/backup/log.php
+++ b/backup/log.php
@@ -1,4 +1,4 @@
-<?php  // $Id$
+<?php
        // log.php - old scheduled backups report. Now redirecting
        // to the new admin one
 
@@ -6,8 +6,6 @@
 
     require_login();
 
-    require_capability('moodle/site:backup', get_context_instance(CONTEXT_SYSTEM));
+    require_capability('moodle/backup:backupcourse', get_context_instance(CONTEXT_SYSTEM));
 
     redirect("$CFG->wwwroot/$CFG->admin/report/backups/index.php", '', 'admin', 1);
-
-?>
