diff --git a/grade/report/grader/tabs.php b/grade/report/grader/tabs.php
index dd83ddb..c1e4d40 100644
--- a/grade/report/grader/tabs.php
+++ b/grade/report/grader/tabs.php
@@ -19,7 +19,7 @@
     $tabcontext = get_context_instance(CONTEXT_COURSE, $COURSE->id);
     $row[] = new tabobject('graderreport',
                            $CFG->wwwroot.'/grade/report/grader/index.php?id='.$courseid,
-                           get_string('modulename', 'gradereport_grader'));
+                           get_string('pluginname', 'gradereport_grader'));
     if (has_capability('moodle/grade:manage',$tabcontext ) ||
         has_capability('moodle/grade:edit', $tabcontext) ||
         has_capability('gradereport/grader:view', $tabcontext)) {
@@ -32,4 +32,4 @@
     echo '<div class="gradedisplay">';
     print_tabs($tabs, $currenttab);
     echo '</div>';
-?>
+
