diff --git a/mod/data/simpletest/testpreset.php b/mod/data/simpletest/testpreset.php
index e7160a0..b543693 100644
--- a/mod/data/simpletest/testpreset.php
+++ b/mod/data/simpletest/testpreset.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 
 ///////////////////////////////////////////////////////////////////////////
 //                                                                       //
@@ -35,9 +35,10 @@ if (!defined('MOODLE_INTERNAL')) {
     die('Direct access to this script is forbidden.');    ///  It must be included from a Moodle page
 }
 
-require_once($CFG->dirroot . '/mod/data/preset_class.php');
+require_once($CFG->dirroot . '/mod/data/lib.php');
 
 class data_preset_test extends UnitTestCase {
+    public static $includecoverage = array('mod/data/lib.php');
     
     function setUp() {
     }
@@ -61,4 +62,4 @@ class data_preset_test extends UnitTestCase {
     }
 }
 
-?>
+
