diff --git a/user/profile/field/menu/define.class.php b/user/profile/field/menu/define.class.php
index 78e9d6e..330472c 100644
--- a/user/profile/field/menu/define.class.php
+++ b/user/profile/field/menu/define.class.php
@@ -1,4 +1,4 @@
-<?php  //$Id$
+<?php
 
 class profile_define_menu extends profile_define_base {
 
@@ -38,4 +38,4 @@ class profile_define_menu extends profile_define_base {
 
 }
 
-?>
+
