diff --git a/blocks/search/block_search.php b/blocks/search/block_search.php
index 7fdddbf..1aeb08c 100644
--- a/blocks/search/block_search.php
+++ b/blocks/search/block_search.php
@@ -1,25 +1,20 @@
 <?php
   
-  /* This is the global search shortcut block - a single query can be entered, and
+/** This is the global search shortcut block - a single query can be entered, and
   * the user will be redirected to the query page where they can enter more
   *  advanced queries, and view the results of their search. When searching from
   *  this block, the broadest possible selection of documents is searched.
   *  
-  *
-  *  Todo: make strings -> get_string()  
-  * 
   * @package search
   * @subpackage search block
   * @author: Michael Champanis (mchampan), reengineered by Valery Fremaux 
   * @date: 2006 06 25
   */
      
-  class block_search extends block_base {
+class block_search extends block_base {
     
     function init() {
-      $this->title = get_string('blockname', 'block_search');
-      $this->cron = 1;
-      $this->version = 2008031500;
+      $this->title = get_string('pluginname', 'block_search');
     } //init  
     
     // only one instance of this block is required
@@ -36,7 +31,7 @@
       global $CFG;
 
       if (empty($CFG->enableglobalsearch)) {
-        return get_string('disabledsearch', 'search');
+        return '';
       }
 
       //cache block contents
@@ -46,22 +41,14 @@
       
       $this->content = new stdClass;
       
-      //lazy check for the moment
-      if (check_php_version("5.0.0")) {        
-        //fetch values if defined in admin, otherwise use defaults
-        $label  = (!empty($CFG->block_search_text)) ? $CFG->block_search_text : get_string('searchmoodle', 'block_search');
-        $button = (!empty($CFG->block_search_button)) ? $CFG->block_search_button : get_string('go', 'block_search');
-        
         //basic search form
         $this->content->text =
             '<form id="searchquery" method="get" action="'. $CFG->wwwroot .'/search/query.php"><div>'
-          . '<label for="block_search_q">'. $label .'</label>'
+          . '<label for="block_search_q">' . get_string('searchmoodle', 'block_search') . '</label>'
           . '<input id="block_search_q" type="text" name="query_string" />'
-          . '<input type="submit" value="'.$button.'" />'
+          . '<input id="block_instance_id" type="hidden" name="block_instanceid" value="' . $this->instance->id . '"/>'
+          . '<input type="submit" value="' . s(get_string('go', 'block_search')) . '" />'
           . '</div></form>';
-      } else {
-        $this->content->text = "Sorry folks, PHP 5 is needed for the new search module.";
-      } //else
         
       //no footer, thanks
       $this->content->footer = '';
@@ -75,7 +62,6 @@
     
     /**
     * wraps up to search engine cron
-    *
     */
     function cron(){
         global $CFG;
@@ -85,4 +71,3 @@
       
   } //block_search
 
-?>
