diff --git a/blocks/admin_bookmarks/delete.php b/blocks/admin_bookmarks/delete.php
index cc06f06..7abe599 100644
--- a/blocks/admin_bookmarks/delete.php
+++ b/blocks/admin_bookmarks/delete.php
@@ -1,11 +1,13 @@
-<?php // $Id$
+<?php
 
 require('../../config.php');
 
 require_once($CFG->libdir.'/adminlib.php');
 
 require_login();
-$adminroot =& admin_get_root(false, false); // settings not required - only pages
+$context = get_context_instance(CONTEXT_SYSTEM);
+$PAGE->set_context($context);
+$adminroot = admin_get_root(false, false); // settings not required - only pages
 
 if ($section = optional_param('section', '', PARAM_SAFEDIR) and confirm_sesskey()) {
 
@@ -26,9 +28,9 @@ if ($section = optional_param('section', '', PARAM_SAFEDIR) and confirm_sesskey(
 
         $temp = $adminroot->locate($section);
 
-        if (is_a($temp, 'admin_externalpage')) {
+        if ($temp instanceof admin_externalpage) {
             redirect($temp->url, get_string('bookmarkdeleted','admin'));
-        } elseif (is_a($temp, 'admin_settingpage')) {
+        } elseif ($temp instanceof admin_settingpage) {
             redirect($CFG->wwwroot . '/' . $CFG->admin . '/settings.php?section=' . $section);
         } else {
             redirect($CFG->wwwroot);
@@ -46,4 +48,4 @@ if ($section = optional_param('section', '', PARAM_SAFEDIR) and confirm_sesskey(
     die;
 }
 
-?>
+
