diff --git a/backup/bb/restore_bb.php b/backup/bb/restore_bb.php
index 5b29054..3e01adf 100644
--- a/backup/bb/restore_bb.php
+++ b/backup/bb/restore_bb.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 // This file facilitates the conversion of a Blackboard course export
 // into a Moodle course export.  It assumes an unzipped directory and makes in-place alterations.
 
@@ -32,14 +32,15 @@ function choose_bb_xsl($manifest){
 
 
 function blackboard_convert($dir){
-    global $CFG;
+    global $CFG, $OUTPUT;
 
+    throw new coding_exception('bb_convert was not converted to new file api yet, sorry');
 
     // Check for a Blackboard manifest file
     if (is_readable($dir.'/imsmanifest.xml') && !is_readable($dir.'/moodle.xml')){
 
         if (!function_exists('xslt_create')) {  // XSLT MUST be installed for this to work
-            notify('You need the XSLT library installed in PHP to open this Blackboard file');
+            echo $OUTPUT->notification('You need the XSLT library installed in PHP to open this Blackboard file');
             return false;
         }
 
@@ -52,7 +53,7 @@ function blackboard_convert($dir){
 
         // The XSL file must be in the same directory as the Blackboard files when it is processed
         if (!copy($CFG->dirroot."/backup/bb/$xslt_file", "$dir/$xslt_file")) {
-            notify('Could not copy the XSLT file to '."$dir/$xslt_file");
+            echo $OUTPUT->notification('Could not copy the XSLT file to '."$dir/$xslt_file");
             return false;
         }
 
@@ -66,7 +67,7 @@ function blackboard_convert($dir){
         // The XSL processor will open them as needed.
         $xsltproc = xslt_create();
         if (!xslt_process($xsltproc, 'imsmanifest.xml', "$dir/$xslt_file", "$dir/moodle.xml")) {
-            notify('Failed writing xml file');
+            echo $OUTPUT->notification('Failed writing xml file');
             chdir($startdir);
             return false;
         }
@@ -120,4 +121,3 @@ function rename_hexfiles($subdir) {
         closedir($handle);
     }
 }
-?>
\ No newline at end of file
