diff --git a/admin/report/courseoverview/settings.php b/admin/report/courseoverview/settings.php
index 16c390c..c6a0589 100644
--- a/admin/report/courseoverview/settings.php
+++ b/admin/report/courseoverview/settings.php
@@ -1,3 +1,5 @@
-<?php  // $Id$
-$ADMIN->add('reports', new admin_externalpage('reportcourseoverview', get_string('courseoverview', 'admin'), "$CFG->wwwroot/$CFG->admin/report/courseoverview/index.php",'report/courseoverview:view'));
-?>
\ No newline at end of file
+<?php
+
+defined('MOODLE_INTERNAL') || die;
+
+$ADMIN->add('reports', new admin_externalpage('reportcourseoverview', get_string('pluginname', 'report_courseoverview'), "$CFG->wwwroot/$CFG->admin/report/courseoverview/index.php",'report/courseoverview:view'));
