diff --git a/admin/report/courseoverview/version.php b/admin/report/courseoverview/version.php
index 4473706..a2be4bb 100644
--- a/admin/report/courseoverview/version.php
+++ b/admin/report/courseoverview/version.php
@@ -1,4 +1,4 @@
-<?PHP // $Id$
+<?PHP
 
 ///////////////////////////////////////////////////////////////////////////
 //                                                                       //
@@ -23,7 +23,5 @@
 //                                                                       //
 ///////////////////////////////////////////////////////////////////////////
 
-$plugin->version  = 2007101503;
-$plugin->requires = 2007101532;
-
-?>
+$plugin->version  = 2010090501;
+$plugin->requires = 2010090501;
