diff --git a/admin/xmldb/actions/move_updown_key/move_updown_key.class.php b/admin/xmldb/actions/move_updown_key/move_updown_key.class.php
index bf5bb3d..9b697ed 100644
--- a/admin/xmldb/actions/move_updown_key/move_updown_key.class.php
+++ b/admin/xmldb/actions/move_updown_key/move_updown_key.class.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 
 ///////////////////////////////////////////////////////////////////////////
 //                                                                       //
@@ -62,7 +62,7 @@ class move_updown_key extends XMLDBAction {
     /// Do the job, setting result as needed
     /// Get the dir containing the file
         $dirpath = required_param('dir', PARAM_PATH);
-        $dirpath = $CFG->dirroot . stripslashes_safe($dirpath);
+        $dirpath = $CFG->dirroot . $dirpath;
 
     /// Get the correct dirs
         if (!empty($XMLDB->dbdirs)) {
@@ -143,4 +143,4 @@ class move_updown_key extends XMLDBAction {
         return $result;
     }
 }
-?>
+
