diff --git a/admin/report/stats/settings.php b/admin/report/stats/settings.php
index ce1d3e3..0283339 100644
--- a/admin/report/stats/settings.php
+++ b/admin/report/stats/settings.php
@@ -1,4 +1,6 @@
-<?php  // $Id$
+<?php
+
+defined('MOODLE_INTERNAL') || die;
+
 // just a link to course report
 $ADMIN->add('reports', new admin_externalpage('reportstats', get_string('stats', 'admin'), "$CFG->wwwroot/course/report/stats/index.php", 'coursereport/stats:view'));
-?>
\ No newline at end of file
