diff --git a/grade/edit/outcome/edit_form.php b/grade/edit/outcome/edit_form.php
index 5bccdb4..6c1893e 100644
--- a/grade/edit/outcome/edit_form.php
+++ b/grade/edit/outcome/edit_form.php
@@ -15,7 +15,6 @@
 // You should have received a copy of the GNU General Public License
 // along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
 
-
 if (!defined('MOODLE_INTERNAL')) {
     die('Direct access to this script is forbidden.');    ///  It must be included from a Moodle page
 }
@@ -23,32 +22,32 @@ if (!defined('MOODLE_INTERNAL')) {
 require_once $CFG->libdir.'/formslib.php';
 
 class edit_outcome_form extends moodleform {
-    function definition() {
+    public function definition() {
         global $CFG, $COURSE;
         $mform =& $this->_form;
 
         // visible elements
         $mform->addElement('header', 'general', get_string('outcomes', 'grades'));
 
-        $mform->addElement('text', 'fullname', get_string('fullname'), 'size="40"');
+        $mform->addElement('text', 'fullname', get_string('outcomefullname', 'grades'), 'size="40"');
         $mform->addRule('fullname', get_string('required'), 'required');
         $mform->setType('fullname', PARAM_TEXT);
 
-        $mform->addElement('text', 'shortname', get_string('shortname'), 'size="20"');
+        $mform->addElement('text', 'shortname', get_string('outcomeshortname', 'grades'), 'size="20"');
         $mform->addRule('shortname', get_string('required'), 'required');
         $mform->setType('shortname', PARAM_NOTAGS);
 
         $mform->addElement('advcheckbox', 'standard', get_string('outcomestandard', 'grades'));
-        $mform->setHelpButton('standard', array('outcomestandard', get_string('outcomestandard'), 'grade'));
+        $mform->addHelpButton('standard', 'outcomestandard', 'grades');
 
         $options = array();
 
         $mform->addElement('selectwithlink', 'scaleid', get_string('scale'), $options, null,
             array('link' => $CFG->wwwroot.'/grade/edit/scale/edit.php?courseid='.$COURSE->id, 'label' => get_string('scalescustomcreate')));
-        $mform->setHelpButton('scaleid', array('scaleid', get_string('scale'), 'grade'));
+        $mform->addHelpButton('scaleid', 'typescale', 'grades');
         $mform->addRule('scaleid', get_string('required'), 'required');
 
-        $mform->addElement('htmleditor', 'description', get_string('description'), array('cols'=>80, 'rows'=>20));
+        $mform->addElement('editor', 'description_editor', get_string('description'), null, $this->_customdata['editoroptions']);
 
 
         // hidden params
@@ -151,4 +150,4 @@ class edit_outcome_form extends moodleform {
 
 }
 
-?>
+
