diff --git a/grade/edit/outcome/export.php b/grade/edit/outcome/export.php
index 30633d5..dbb510f 100644
--- a/grade/edit/outcome/export.php
+++ b/grade/edit/outcome/export.php
@@ -26,7 +26,7 @@ $action   = optional_param('action', '', PARAM_ALPHA);
 
 /// Make sure they can even access this course
 if ($courseid) {
-    if (!$course = get_record('course', 'id', $courseid)) {
+    if (!$course = $DB->get_record('course', array('id' => $courseid))) {
         print_error('nocourseid');
     }
     require_login($course);
@@ -42,9 +42,8 @@ if ($courseid) {
     admin_externalpage_setup('outcomes');
 }
 
-if (!confirm_sesskey()) {
-    break;
-}
+require_sesskey();
+
 // $outcome = grade_outcome::fetch(array('id'=>$outcomeid));
 
 $systemcontext = get_context_instance(CONTEXT_SYSTEM);
@@ -70,12 +69,12 @@ foreach($outcomes as $outcome) {
 
     $line[] = $outcome->get_name();
     $line[] = $outcome->get_shortname();
-    $line[] = $outcome->description;
+    $line[] = $outcome->get_description();
     
     $scale = $outcome->load_scale();
     $line[] = $scale->get_name();
     $line[] = $scale->compact_items();
-    $line[] = $scale->description;
+    $line[] = $scale->get_description();
     
     echo format_csv($line, ';', '"');
 }
