diff --git a/admin/report/unittest/db/access.php b/admin/report/unittest/db/access.php
index dd9915f..da0247e 100644
--- a/admin/report/unittest/db/access.php
+++ b/admin/report/unittest/db/access.php
@@ -1,4 +1,4 @@
-<?php  // $Id$
+<?php
 
 ///////////////////////////////////////////////////////////////////////////
 //                                                                       //
@@ -23,18 +23,16 @@
 //                                                                       //
 ///////////////////////////////////////////////////////////////////////////
 
-$report_unittest_capabilities = array(
+$capabilities = array(
 
     'report/unittest:view' => array(
         'riskbitmask' => RISK_DATALOSS,
         'captype' => 'read',
         'contextlevel' => CONTEXT_SYSTEM,
-        'legacy' => array(
-            'admin' => CAP_ALLOW
+        'archetypes' => array(
+            'manager' => CAP_ALLOW
         ),
 
         'clonepermissionsfrom' => 'moodle/site:config',
     )
 );
-
-?>
