diff --git a/admin/user/user_bulk_forms.php b/admin/user/user_bulk_forms.php
index c5bbeb7..ea210c2 100644
--- a/admin/user/user_bulk_forms.php
+++ b/admin/user/user_bulk_forms.php
@@ -1,4 +1,4 @@
-<?php //$Id$
+<?php
 
 require_once($CFG->libdir.'/formslib.php');
 require_once($CFG->libdir.'/datalib.php');
@@ -24,10 +24,13 @@ class user_bulk_action_form extends moodleform {
         if (has_capability('moodle/user:update', $syscontext)) {
             $actions[5] = get_string('download', 'admin');
         }
+        if (has_capability('moodle/role:assign', $syscontext)){
+             //TODO: MDL-24064
+            //$actions[6] = get_string('enrolmultipleusers', 'admin');
+        }
         if (has_capability('moodle/user:update', $syscontext)) {
-            $actions[6] = get_string('forcepasswordchange');
+            $actions[7] = get_string('forcepasswordchange');
         }
-
         $objs = array();
         $objs[] =& $mform->createElement('select', 'action', null, $actions);
         $objs[] =& $mform->createElement('submit', 'doaction', get_string('go'));
@@ -52,7 +55,7 @@ class user_bulk_form extends moodleform {
             if ($total == $acount) {
                 $achoices[0] = get_string('allusers', 'bulkusers', $total);
             } else {
-                $a = new object();
+                $a = new stdClass();
                 $a->total  = $total;
                 $a->count = $acount;
                 $achoices[0] = get_string('allfilteredusers', 'bulkusers', $a);
@@ -68,7 +71,7 @@ class user_bulk_form extends moodleform {
         }
 
         if (is_array($susers)) {
-            $a = new object();
+            $a = new stdClass();
             $a->total  = $total;
             $a->count = $scount;
             $schoices[0] = get_string('allselectedusers', 'bulkusers', $a);
@@ -91,8 +94,8 @@ class user_bulk_form extends moodleform {
         $objs[1]->setMultiple(true);
 
 
-        $grp =& $mform->addElement('group', 'usersgrp', get_string('users'), $objs, ' ', false);
-        $grp->setHelpButton(array('lists', get_string('users'), 'bulkusers'));
+        $grp =& $mform->addElement('group', 'usersgrp', get_string('users', 'bulkusers'), $objs, ' ', false);
+        $mform->addHelpButton('usersgrp', 'users', 'bulkusers');
 
         $mform->addElement('static', 'comment');
 
@@ -102,11 +105,10 @@ class user_bulk_form extends moodleform {
         $objs[] =& $mform->createElement('submit', 'addall', get_string('addall', 'bulkusers'));
         $objs[] =& $mform->createElement('submit', 'removeall', get_string('removeall', 'bulkusers'));
         $grp =& $mform->addElement('group', 'buttonsgrp', get_string('selectedlist', 'bulkusers'), $objs, array(' ', '<br />'), false);
-        $grp->setHelpButton(array('selectedlist', get_string('selectedlist', 'bulkusers'), 'bulkusers'));
+        $mform->addHelpButton('buttonsgrp', 'selectedlist', 'bulkusers');
 
         $renderer =& $mform->defaultRenderer();
         $template = '<label class="qflabel" style="vertical-align:top">{label}</label> {element}';
         $renderer->setGroupElementTemplate($template, 'usersgrp');
     }
 }
-?>
