diff --git a/admin/user/user_bulk_download.php b/admin/user/user_bulk_download.php
old mode 100755
new mode 100644
index d71bbea..6b5881c
--- a/admin/user/user_bulk_download.php
+++ b/admin/user/user_bulk_download.php
@@ -1,4 +1,4 @@
-<?php //$Id$
+<?php
 /**
 * script for downloading of user lists
 */
@@ -8,6 +8,7 @@ require_once($CFG->libdir.'/adminlib.php');
 
 $format = optional_param('format', '', PARAM_ALPHA);
 
+require_login();
 admin_externalpage_setup('userbulk');
 require_capability('moodle/user:update', get_context_instance(CONTEXT_SYSTEM));
 
@@ -37,7 +38,7 @@ if ($format) {
                     'msn'       => 'msn',
                     'country'   => 'country');
 
-    if ($extrafields = get_records_select('user_info_field')) {
+    if ($extrafields = $DB->get_records('user_info_field')) {
         foreach ($extrafields as $n=>$v){
             $fields['profile_field_'.$v->shortname] = 'profile_field_'.$v->shortname;
         }
@@ -52,23 +53,23 @@ if ($format) {
     die;
 }
 
-admin_externalpage_print_header();
-print_heading(get_string('download', 'admin'));
+echo $OUTPUT->header();
+echo $OUTPUT->heading(get_string('download', 'admin'));
 
-print_box_start();
+echo $OUTPUT->box_start();
 echo '<ul>';
 echo '<li><a href="user_bulk_download.php?format=csv">'.get_string('downloadtext').'</a></li>';
 echo '<li><a href="user_bulk_download.php?format=ods">'.get_string('downloadods').'</a></li>';
 echo '<li><a href="user_bulk_download.php?format=xls">'.get_string('downloadexcel').'</a></li>';
 echo '</ul>';
-print_box_end();
+echo $OUTPUT->box_end();
 
-print_continue($return);
+echo $OUTPUT->continue_button($return);
 
-print_footer();
+echo $OUTPUT->footer();
 
 function user_download_ods($fields) {
-    global $CFG, $SESSION;
+    global $CFG, $SESSION, $DB;
 
     require_once("$CFG->libdir/odslib.class.php");
     require_once($CFG->dirroot.'/user/profile/lib.php');
@@ -89,7 +90,7 @@ function user_download_ods($fields) {
 
     $row = 1;
     foreach ($SESSION->bulk_users as $userid) {
-        if (!$user = get_record('user', 'id', $userid)) {
+        if (!$user = $DB->get_record('user', array('id'=>$userid))) {
             continue;
         }
         $col = 0;
@@ -106,7 +107,7 @@ function user_download_ods($fields) {
 }
 
 function user_download_xls($fields) {
-    global $CFG, $SESSION;
+    global $CFG, $SESSION, $DB;
 
     require_once("$CFG->libdir/excellib.class.php");
     require_once($CFG->dirroot.'/user/profile/lib.php');
@@ -127,7 +128,7 @@ function user_download_xls($fields) {
 
     $row = 1;
     foreach ($SESSION->bulk_users as $userid) {
-        if (!$user = get_record('user', 'id', $userid)) {
+        if (!$user = $DB->get_record('user', array('id'=>$userid))) {
             continue;
         }
         $col = 0;
@@ -144,7 +145,7 @@ function user_download_xls($fields) {
 }
 
 function user_download_csv($fields) {
-    global $CFG, $SESSION;
+    global $CFG, $SESSION, $DB;
     
     require_once($CFG->dirroot.'/user/profile/lib.php');
 
@@ -156,7 +157,7 @@ function user_download_csv($fields) {
     header("Cache-Control: must-revalidate,post-check=0,pre-check=0");
     header("Pragma: public");
 
-    $delimiter = get_string('listsep');
+    $delimiter = get_string('listsep', 'langconfig');
     $encdelim  = '&#'.ord($delimiter);
 
     $row = array(); 
@@ -167,7 +168,7 @@ function user_download_csv($fields) {
 
     foreach ($SESSION->bulk_users as $userid) {
         $row = array();
-        if (!$user = get_record('user', 'id', $userid)) {
+        if (!$user = $DB->get_record('user', array('id'=>$userid))) {
             continue;
         }
         profile_load_data($user);
@@ -178,5 +179,3 @@ function user_download_csv($fields) {
     }
     die;
 }
-
-?>
