diff --git a/admin/user/user_bulk_delete.php b/admin/user/user_bulk_delete.php
old mode 100755
new mode 100644
index 413c833..a8b6705
--- a/admin/user/user_bulk_delete.php
+++ b/admin/user/user_bulk_delete.php
@@ -1,4 +1,4 @@
-<?php //$Id$
+<?php
 /**
 * script for bulk user delete operations
 */
@@ -8,6 +8,7 @@ require_once($CFG->libdir.'/adminlib.php');
 
 $confirm = optional_param('confirm', 0, PARAM_BOOL);
 
+require_login();
 admin_externalpage_setup('userbulk');
 require_capability('moodle/user:delete', get_context_instance(CONTEXT_SYSTEM));
 
@@ -17,36 +18,33 @@ if (empty($SESSION->bulk_users)) {
     redirect($return);
 }
 
-admin_externalpage_print_header();
+echo $OUTPUT->header();
 
 //TODO: add support for large number of users
 
 if ($confirm and confirm_sesskey()) {
-    $primaryadmin = get_admin();
 
-    $in = implode(',', $SESSION->bulk_users);
-    if ($rs = get_recordset_select('user', "id IN ($in)")) {
-        while ($user = rs_fetch_next_record($rs)) {
-            if ($primaryadmin->id != $user->id and $USER->id != $user->id and delete_user($user)) {
+    list($in, $params) = $DB->get_in_or_equal($SESSION->bulk_users);
+    $rs = $DB->get_recordset_select('user', "id $in", $params);
+    foreach ($rs as $user) {
+        if (!is_siteadmin($user) and $USER->id != $user->id and delete_user($user)) {
                 unset($SESSION->bulk_users[$user->id]);
             } else {
-                notify(get_string('deletednot', '', fullname($user, true)));
+            echo $OUTPUT->notification(get_string('deletednot', '', fullname($user, true)));
             }
         }
-        rs_close($rs);
-    }
+    $rs->close();
+    session_gc(); // remove stale sessions
     redirect($return, get_string('changessaved'));
 
 } else {
-    $in = implode(',', $SESSION->bulk_users);
-    $userlist = get_records_select_menu('user', "id IN ($in)", 'fullname', 'id,'.sql_fullname().' AS fullname');
+    list($in, $params) = $DB->get_in_or_equal($SESSION->bulk_users);
+    $userlist = $DB->get_records_select_menu('user', "id $in", $params, 'fullname', 'id,'.$DB->sql_fullname().' AS fullname');
     $usernames = implode(', ', $userlist);
-    $optionsyes = array();
-    $optionsyes['confirm'] = 1;
-    $optionsyes['sesskey'] = sesskey();
-    print_heading(get_string('confirmation', 'admin'));
-    notice_yesno(get_string('deletecheckfull', '', $usernames), 'user_bulk_delete.php', 'user_bulk.php', $optionsyes, NULL, 'post', 'get');
+    echo $OUTPUT->heading(get_string('confirmation', 'admin'));
+    $formcontinue = new single_button(new moodle_url('user_bulk_delete.php', array('confirm' => 1)), get_string('yes'));
+    $formcancel = new single_button(new moodle_url('user_bulk.php'), get_string('no'), 'get');
+    echo $OUTPUT->confirm(get_string('deletecheckfull', '', $usernames), $formcontinue, $formcancel);
 }
 
-admin_externalpage_print_footer();
-?>
+echo $OUTPUT->footer();
