diff --git a/grade/report/user/db/access.php b/grade/report/user/db/access.php
index fbf024f..d19d468 100644
--- a/grade/report/user/db/access.php
+++ b/grade/report/user/db/access.php
@@ -15,19 +15,19 @@
 // You should have received a copy of the GNU General Public License
 // along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
 
-$gradereport_user_capabilities = array(
+$capabilities = array(
 
     'gradereport/user:view' => array(
         'riskbitmask' => RISK_PERSONAL,
         'captype' => 'read',
         'contextlevel' => CONTEXT_COURSE,
-        'legacy' => array(
+        'archetypes' => array(
             'student' => CAP_ALLOW,
             'teacher' => CAP_ALLOW,
             'editingteacher' => CAP_ALLOW,
-            'admin' => CAP_ALLOW
+            'manager' => CAP_ALLOW
         )
     ),
 );
 
-?>
+
