diff --git a/admin/xmldb/actions/view_structure_sql/view_structure_sql.class.php b/admin/xmldb/actions/view_structure_sql/view_structure_sql.class.php
index 0e078d1..136e0cb 100644
--- a/admin/xmldb/actions/view_structure_sql/view_structure_sql.class.php
+++ b/admin/xmldb/actions/view_structure_sql/view_structure_sql.class.php
@@ -1,32 +1,34 @@
-<?php // $Id$
-
-///////////////////////////////////////////////////////////////////////////
-//                                                                       //
-// NOTICE OF COPYRIGHT                                                   //
-//                                                                       //
-// Moodle - Modular Object-Oriented Dynamic Learning Environment         //
-//          http://moodle.com                                            //
-//                                                                       //
-// Copyright (C) 1999 onwards Martin Dougiamas        http://dougiamas.com  //
-//           (C) 2001-3001 Eloy Lafuente (stronk7) http://contiento.com  //
-//                                                                       //
-// This program is free software; you can redistribute it and/or modify  //
-// it under the terms of the GNU General Public License as published by  //
-// the Free Software Foundation; either version 2 of the License, or     //
-// (at your option) any later version.                                   //
-//                                                                       //
-// This program is distributed in the hope that it will be useful,       //
-// but WITHOUT ANY WARRANTY; without even the implied warranty of        //
-// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         //
-// GNU General Public License for more details:                          //
-//                                                                       //
-//          http://www.gnu.org/copyleft/gpl.html                         //
-//                                                                       //
-///////////////////////////////////////////////////////////////////////////
-
-/// This class will show the SQL generated for the selected RDBMS for
-/// the entire XMLDB file
+<?php
+
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * @package   xmldb-editor
+ * @copyright 2003 onwards Eloy Lafuente (stronk7) {@link http://stronk7.com}
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
 
+/**
+ * This class will show the SQL generated for the selected RDBMS for
+ * the entire XMLDB file
+ *
+ * @package   xmldb-editor
+ * @copyright 2003 onwards Eloy Lafuente (stronk7) {@link http://stronk7.com}
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
 class view_structure_sql extends XMLDBAction {
 
     /**
@@ -59,12 +61,13 @@ class view_structure_sql extends XMLDBAction {
         $this->does_generate = ACTION_GENERATE_HTML;
 
     /// These are always here
-        global $CFG, $XMLDB;
+        global $CFG, $XMLDB, $DB;
+        $dbman = $DB->get_manager();
 
     /// Do the job, setting result as needed
     /// Get the dir containing the file
         $dirpath = required_param('dir', PARAM_PATH);
-        $dirpath = $CFG->dirroot . stripslashes_safe($dirpath);
+        $dirpath = $CFG->dirroot . $dirpath;
 
     /// Get the correct dirs
         if (!empty($XMLDB->dbdirs)) {
@@ -78,23 +81,6 @@ class view_structure_sql extends XMLDBAction {
         }
     /// ADD YOUR CODE HERE
 
-    /// Get parameters
-        $generatorparam = optional_param('generator', null, PARAM_ALPHANUM);
-        if (empty($generatorparam)) {
-            $generatorparam = $CFG->dbtype;
-        }
-
-    /// Calculate list of available SQL generators
-        $plugins = get_list_of_plugins('lib/xmldb/classes/generators');
-        $generators = array();
-        foreach($plugins as $plugin) {
-            $generators[$plugin] = $plugin;
-        }
-    /// Check we have the selected generator
-        if (!in_array($generatorparam, $generators)) {
-            $generatorparam = reset($generators);
-        }
-
         /// The back to edit table button
         $b = ' <p class="centerpara buttons">';
         $b .= '<a href="index.php?action=edit_xml_file&amp;dir=' . urlencode(str_replace($CFG->dirroot, '', $dirpath)) . '">[' . $this->str['back'] . ']</a>';
@@ -102,15 +88,10 @@ class view_structure_sql extends XMLDBAction {
         $o = $b;
 
         $o.= '    <table id="formelements" class="boxaligncenter" cellpadding="5">';
-        $o.= '      <tr><td align="center">' . $this->str['selectdb'];
-
-    /// Show the popup of generators
-        $url = 'index.php?action=view_structure_sql&amp;dir=' . urlencode(str_replace($CFG->dirroot, '', $dirpath)) . '&amp;generator=';
-        $o.= popup_form($url, $generators, 'selectgenerator', $generatorparam, '', '', '' , true);
-        $o.= '      </td></tr>';
         $o.= '      <tr><td><textarea cols="80" rows="32">';
     /// Get an array of statements
-        if ($starr = $structure->getCreateStructureSQL($generatorparam, $CFG->prefix)) {
+        if ($starr = $DB->get_manager()->generator->getCreateStructureSQL($structure)) {
+            $starr = $dbman->generator->getEndedStatements($starr);
             $sqltext = '';
             foreach ($starr as $st) {
                 $sqltext .= s($st) . "\n\n";
@@ -132,4 +113,4 @@ class view_structure_sql extends XMLDBAction {
         return $result;
     }
 }
-?>
+
