diff --git a/lib/adodb/adodb-csvlib.inc.php b/lib/adodb/adodb-csvlib.inc.php
index c82c965..1c0d081 100644
--- a/lib/adodb/adodb-csvlib.inc.php
+++ b/lib/adodb/adodb-csvlib.inc.php
@@ -8,7 +8,7 @@ $ADODB_INCLUDED_CSV = 1;
 
 /* 
 
-  V4.98 13 Feb 2008  (c) 2000-2008 John Lim (jlim#natsoft.com.my). All rights reserved.
+  V5.11 5 May 2010   (c) 2000-2010 John Lim (jlim#natsoft.com). All rights reserved.
   Released under both BSD license and Lesser GPL library license. 
   Whenever there is any discrepancy between the two licenses, 
   the BSD license will take precedence. See License.txt. 
@@ -54,7 +54,7 @@ $ADODB_INCLUDED_CSV = 1;
 		$line = "====1,$tt,$sql\n";
 		
 		if ($rs->databaseType == 'array') {
-			$rows =& $rs->_array;
+			$rows = $rs->_array;
 		} else {
 			$rows = array();
 			while (!$rs->EOF) {	
@@ -64,13 +64,14 @@ $ADODB_INCLUDED_CSV = 1;
 		}
 		
 		for($i=0; $i < $max; $i++) {
-			$o =& $rs->FetchField($i);
+			$o = $rs->FetchField($i);
 			$flds[] = $o;
 		}
 	
 		$savefetch = isset($rs->adodbFetchMode) ? $rs->adodbFetchMode : $rs->fetchMode;
 		$class = $rs->connection->arrayClass;
 		$rs2 = new $class();
+		$rs2->timeCreated = $rs->timeCreated; # memcache fix
 		$rs2->sql = $rs->sql;
 		$rs2->oldProvider = $rs->dataProvider; 
 		$rs2->InitArrayFields($rows,$flds);
@@ -90,7 +91,7 @@ $ADODB_INCLUDED_CSV = 1;
 *			error occurred in sql INSERT/UPDATE/DELETE, 
 *			empty recordset is returned
 */
-	function &csv2rs($url,&$err,$timeout=0, $rsclass='ADORecordSet_array')
+	function csv2rs($url,&$err,$timeout=0, $rsclass='ADORecordSet_array')
 	{
 		$false = false;
 		$err = false;
@@ -287,7 +288,7 @@ $ADODB_INCLUDED_CSV = 1;
 			fclose($fd);
 			
 			if ($ok) {
-				chmod($tmpname,0644);
+				@chmod($tmpname,0644);
 				// the tricky moment
 				@unlink($filename);
 				if (!@rename($tmpname,$filename)) {
@@ -305,7 +306,7 @@ $ADODB_INCLUDED_CSV = 1;
 			if (fwrite( $fd, $contents )) $ok = true;
 			else $ok = false;
 			fclose($fd);
-			chmod($filename,0644);
+			@chmod($filename,0644);
 		}else {
 			fclose($fd);
 			if ($debug)ADOConnection::outp( " Failed acquiring lock for $filename<br>\n");
