diff --git a/lib/adodb/adodb-time.inc.php b/lib/adodb/adodb-time.inc.php
index 1ae5076..d62f678 100644
--- a/lib/adodb/adodb-time.inc.php
+++ b/lib/adodb/adodb-time.inc.php
@@ -1006,7 +1006,6 @@ function adodb_tz_offset($gmt,$isphp5)
 		return sprintf('%s%02d%02d',($gmt<=0)?'+':'-',floor($zhrs),($zhrs-$hrs)*60); 
 	else
 		return sprintf('%s%02d%02d',($gmt<0)?'+':'-',floor($zhrs),($zhrs-$hrs)*60); 
-	break;
 }
 
 
@@ -1075,6 +1074,9 @@ global $ADODB_DATETIME_CLASS;
 	*/
 	for ($i=0; $i < $max; $i++) {
 		switch($fmt[$i]) {
+		case 'e':
+			$dates .= date('e');
+			break;
 		case 'T': 
 			if ($ADODB_DATETIME_CLASS) {
 				$dt = new DateTime();
@@ -1215,7 +1217,7 @@ function adodb_mktime($hr,$min,$sec,$mon=false,$day=false,$year=false,$is_dst=fa
 		
 		// for windows, we don't check 1970 because with timezone differences, 
 		// 1 Jan 1970 could generate negative timestamp, which is illegal
-		$usephpfns = (1971 < $year && $year < 2038
+		$usephpfns = (1970 < $year && $year < 2038
 			|| !defined('ADODB_NO_NEGATIVE_TS') && (1901 < $year && $year < 2038)
 			); 
 			
@@ -1423,4 +1425,5 @@ global $ADODB_DATE_LOCALE;
 	return $ret;
 }
 
+
 ?>
\ No newline at end of file
