diff --git a/admin/xmldb/actions/edit_field/edit_field.class.php b/admin/xmldb/actions/edit_field/edit_field.class.php
index d06ba02..d812f97 100644
--- a/admin/xmldb/actions/edit_field/edit_field.class.php
+++ b/admin/xmldb/actions/edit_field/edit_field.class.php
@@ -1,31 +1,33 @@
-<?php // $Id$
-
-///////////////////////////////////////////////////////////////////////////
-//                                                                       //
-// NOTICE OF COPYRIGHT                                                   //
-//                                                                       //
-// Moodle - Modular Object-Oriented Dynamic Learning Environment         //
-//          http://moodle.com                                            //
-//                                                                       //
-// Copyright (C) 1999 onwards Martin Dougiamas        http://dougiamas.com  //
-//           (C) 2001-3001 Eloy Lafuente (stronk7) http://contiento.com  //
-//                                                                       //
-// This program is free software; you can redistribute it and/or modify  //
-// it under the terms of the GNU General Public License as published by  //
-// the Free Software Foundation; either version 2 of the License, or     //
-// (at your option) any later version.                                   //
-//                                                                       //
-// This program is distributed in the hope that it will be useful,       //
-// but WITHOUT ANY WARRANTY; without even the implied warranty of        //
-// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         //
-// GNU General Public License for more details:                          //
-//                                                                       //
-//          http://www.gnu.org/copyleft/gpl.html                         //
-//                                                                       //
-///////////////////////////////////////////////////////////////////////////
-
-/// This class will provide the interface for all the edit field actions
+<?php
+
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * @package   xmldb-editor
+ * @copyright 2003 onwards Eloy Lafuente (stronk7) {@link http://stronk7.com}
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
 
+/**
+ * This class will provide the interface for all the edit field actions
+ *
+ * @package   xmldb-editor
+ * @copyright 2003 onwards Eloy Lafuente (stronk7) {@link http://stronk7.com}
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
 class edit_field extends XMLDBAction {
 
     /**
@@ -40,6 +42,7 @@ class edit_field extends XMLDBAction {
     /// Get needed strings
         $this->loadStrings(array(
             'change' => 'xmldb',
+            'float2numbernote' => 'xmldb',
             'vieworiginal' => 'xmldb',
             'viewedited' => 'xmldb',
             'yes' => '',
@@ -62,12 +65,12 @@ class edit_field extends XMLDBAction {
         $this->does_generate = ACTION_GENERATE_HTML;
 
     /// These are always here
-        global $CFG, $XMLDB;
+        global $CFG, $XMLDB, $OUTPUT;
 
     /// Do the job, setting result as needed
     /// Get the dir containing the file
         $dirpath = required_param('dir', PARAM_PATH);
-        $dirpath = $CFG->dirroot . stripslashes_safe($dirpath);
+        $dirpath = $CFG->dirroot . $dirpath;
 
     /// Get the correct dirs
         if (!empty($XMLDB->dbdirs)) {
@@ -98,8 +101,15 @@ class edit_field extends XMLDBAction {
         $dbdir =& $XMLDB->dbdirs[$dirpath];
         $origstructure =& $dbdir->xml_file->getStructure();
 
+        $o = ''; /// Output starts
+
+    /// If field is XMLDB_TYPE_FLOAT, comment about to migrate it to XMLDB_TYPE_NUMBER
+        if ($field->getType() == XMLDB_TYPE_FLOAT) {
+            $o .= '<p>' . $this->str['float2numbernote'] . '</p>';
+        }
+
     /// Add the main form
-        $o = '<form id="form" action="index.php" method="post">';
+        $o.= '<form id="form" action="index.php" method="post">';
         $o.= '    <div>';
         $o.= '    <input type="hidden" name ="dir" value="' . str_replace($CFG->dirroot, '', $dirpath) . '" />';
         $o.= '    <input type="hidden" name ="table" value="' . $tableparam .'" />';
@@ -119,7 +129,7 @@ class edit_field extends XMLDBAction {
         }
     /// XMLDB field comment
         $o.= '      <tr valign="top"><td><label for="comment" accesskey="c">Comment:</label></td><td colspan="2"><textarea name="comment" rows="3" cols="80" id="comment">' . s($field->getComment()) . '</textarea></td></tr>';
-    /// XMLDBField Type
+    /// xmldb_field Type
         $typeoptions = array (XMLDB_TYPE_INTEGER => $field->getXMLDBTypeName(XMLDB_TYPE_INTEGER),
                               XMLDB_TYPE_NUMBER  => $field->getXMLDBTypeName(XMLDB_TYPE_NUMBER),
                               XMLDB_TYPE_FLOAT   => $field->getXMLDBTypeName(XMLDB_TYPE_FLOAT),
@@ -137,37 +147,31 @@ class edit_field extends XMLDBAction {
         if ($field->getType() != XMLDB_TYPE_DATETIME) {
             unset ($typeoptions[XMLDB_TYPE_DATETIME]);
         }
+        $select = html_writer::select($typeoptions, 'type', $field->getType(), false);
         $o.= '      <tr valign="top"><td><label for="menutype" accesskey="t">Type:</label></td>';
-        $o.= '        <td colspan="2">' . choose_from_menu($typeoptions, 'type', $field->getType(), '', '', '', true) . '</td></tr>';
-    /// XMLDBField Length
+        $o.= '        <td colspan="2">' . $select . '</td></tr>';
+    /// xmldb_field Length
         $o.= '      <tr valign="top"><td><label for="length" accesskey="l">Length:</label></td>';
         $o.= '        <td colspan="2"><input name="length" type="text" size="6" maxlength="6" id="length" value="' . s($field->getLength()) . '" /><span id="lengthtip"></span></td></tr>';
-    /// XMLDBField Decimals
+    /// xmldb_field Decimals
         $o.= '      <tr valign="top"><td><label for="decimals" accesskey="d">Decimals:</label></td>';
         $o.= '        <td colspan="2"><input name="decimals" type="text" size="6" maxlength="6" id="decimals" value="' . s($field->getDecimals()) . '" /><span id="decimalstip"></span></td></tr>';
-    /// XMLDBField Unsigned
+    /// xmldb_field Unsigned
         $unsignedoptions = array (0 => 'signed', 1 => 'unsigned');
+        $select = html_writer::select($unsignedoptions, 'unsigned', $field->getUnsigned(), false);
         $o.= '      <tr valign="top"><td><label for="menuunsigned" accesskey="u">Unsigned:</label></td>';
-        $o.= '        <td colspan="2">' . choose_from_menu($unsignedoptions, 'unsigned', $field->getUnsigned(), '', '', '', true) . '</td></tr>';
-    /// XMLDBField NotNull
+        $o.= '        <td colspan="2">' . $select . '</td></tr>';
+    /// xmldb_field NotNull
         $notnulloptions = array (0 => 'null', 'not null');
+        $select = html_writer::select($notnulloptions, 'notnull', $field->getNotNull(), false);
         $o.= '      <tr valign="top"><td><label for="menunotnull" accesskey="n">Not Null:</label></td>';
-        $o.= '        <td colspan="2">' . choose_from_menu($notnulloptions, 'notnull', $field->getNotNull(), '', '', '', true) . '</td></tr>';
-    /// XMLDBField Sequence
+        $o.= '        <td colspan="2">' . $select . '</td></tr>';
+    /// xmldb_field Sequence
         $sequenceoptions = array (0 => $this->str['no'], 1 => 'auto-numbered');
+        $select = html_writer::select($sequenceoptions, 'sequence', $field->getSequence(), false);
         $o.= '      <tr valign="top"><td><label for="menusequence" accesskey="s">Sequence:</label></td>';
-        $o.= '        <td colspan="2">' . choose_from_menu($sequenceoptions, 'sequence', $field->getSequence(), '', '', '', true) . '</td></tr>';
-    /// XMLDBField Enum and enumvalues
-        $enumoptions = array (0 => $this->str['no'], 1 => $this->str['yes']);
-        $o.= '      <tr valign="top"><td><label for="menuenum" accesskey="s">Enum:</label></td>';
-        $o.= '        <td>' . choose_from_menu($enumoptions, 'enum', $field->getEnum(), '', '', '', true) . '</td>';
-        if (is_array($field->getEnumValues())) {
-            $enumvalues = implode(', ', $field->getEnumValues());
-        } else {
-            $enumvalues = '';
-        }
-        $o.= '            <td><textarea name="enumvalues" rows="3" cols="70" id="enumvalues">' . s($enumvalues) . '</textarea></td></tr>';
-    /// XMLDBField Default
+        $o.= '        <td colspan="2">' . $select . '</td></tr>';
+    /// xmldb_field Default
         $o.= '      <tr valign="top"><td><label for="default" accesskey="d">Default:</label></td>';
         $o.= '        <td colspan="2"><input type="text" name="default" size="30" maxlength="80" id="default" value="' . s($field->getDefault()) . '" /></td></tr>';
     /// Change button
@@ -204,4 +208,4 @@ class edit_field extends XMLDBAction {
         return $result;
     }
 }
-?>
+
