diff --git a/admin/report/backups/index.php b/admin/report/backups/index.php
index 60ff814..a4006cd 100644
--- a/admin/report/backups/index.php
+++ b/admin/report/backups/index.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
       // index.php - scheduled backup logs
 
     require_once('../../../config.php');
@@ -7,18 +7,13 @@
 
     $courseid = optional_param('courseid', 0, PARAM_INT);
 
-    admin_externalpage_setup('reportbackups');
-    admin_externalpage_print_header();
+    admin_externalpage_setup('reportbackups', '', null, '', array('pagelayout'=>'report'));
+    echo $OUTPUT->header();
 
-/// Scheduled backups are disabled by the server admin
-    if (!empty($CFG->disablescheduledbackups)) {
-        print_error('scheduledbackupsdisabled', '', '', NULL, true);
-    }
-
-/// Scheduled backups aren't active by the site admin
+/// Automated backups aren't active by the site admin
     $backup_config = backup_get_config();
-    if (empty($backup_config->backup_sche_active)) {
-        notify(get_string('scheduledbackupsinactive'));
+    if (empty($backup_config->backup_auto_active)) {
+        echo $OUTPUT->notification(get_string('automatedbackupsinactive', 'backup'));
     }
 
 /// Get needed strings
@@ -43,13 +38,13 @@
 /// Decide when to show last execution logs or detailed logs
 /// Lastlog view
     if (!$courseid) {
-        print_heading($backuploglaststatus);
-        print_simple_box_start('center');
+        echo $OUTPUT->heading($backuploglaststatus);
+        echo $OUTPUT->box_start();
     /// Now, get every record from backup_courses
-        $courses = get_records("backup_courses");
+        $courses = $DB->get_records("backup_courses");
 
         if (!$courses) {
-            notify(get_string('nologsfound'));
+            echo $OUTPUT->notification(get_string('nologsfound'));
         } else {
             echo "<table border=\"0\" align=\"center\" cellpadding=\"3\" cellspacing=\"3\">";
             //Print table header
@@ -60,7 +55,7 @@
             echo "<td nowrap=\"nowrap\" align=\"center\"><font size=\"3\">$strnext</font></td></tr>";
             foreach ($courses as $course) {
             /// Get the course shortname
-                $coursename = get_field ("course","fullname","id",$course->courseid);
+                $coursename = $DB->get_field ("course", "fullname", array("id"=>$course->courseid));
                 if ($coursename) {
                     echo "<tr>";
                     echo "<td nowrap=\"nowrap\"><font size=\"2\"><a href=\"index.php?courseid=$course->courseid\">".$coursename."</a></font></td>";
@@ -82,38 +77,37 @@
             }
             echo "</table>";
         }
-        print_simple_box_end();
+        echo $OUTPUT->box_end();
 /// Detailed View !!
     } else {
-        print_heading($backuplogdetailed);
+        echo $OUTPUT->heading($backuplogdetailed);
 
-        $coursename = get_field("course","fullname","id","$courseid");
-        print_heading("$strcourse: $coursename");
+        $coursename = $DB->get_field("course", "fullname", array("id"=>"$courseid"));
+        echo $OUTPUT->heading("$strcourse: $coursename");
 
-        print_simple_box_start('center');
+        echo $OUTPUT->box_start();
 
     /// First, me get all the distinct backups for that course in backup_log
-        $executions = get_records_sql("SELECT DISTINCT laststarttime
-                                       FROM {$CFG->prefix}backup_log
-                                       WHERE courseid = '$courseid'
-                                       ORDER BY laststarttime DESC");
+        $executions = $DB->get_records_sql("SELECT DISTINCT laststarttime
+                                              FROM {backup_log}
+                                             WHERE courseid = ? AND backuptype = ?
+                                          ORDER BY laststarttime DESC", array($courseid,'scheduledbackup'));
 
     /// Iterate over backup executions
         if (!$executions) {
-            notify(get_string('nologsfound'));
+            echo $OUTPUT->notification(get_string('nologsfound'));
         } else {
             echo "<table border=\"0\" align=\"center\" cellpadding=\"3\" cellspacing=\"3\">";
             foreach($executions as $execution) {
                 echo "<tr>";
                 echo "<td nowrap=\"nowrap\" align=\"center\" colspan=\"3\">";
-                print_simple_box("<center>".userdate($execution->laststarttime)."</center>", "center");
+                echo $OUTPUT->box(userdate($execution->laststarttime));
                 echo "</td>";
                 echo "</tr>";
-                $logs = get_records_sql("SELECT *
-                                         FROM {$CFG->prefix}backup_log
-                                         WHERE courseid = '$courseid'  AND
-                                               laststarttime = '$execution->laststarttime'
-                                         ORDER BY id");
+                $logs = $DB->get_records_sql("SELECT *
+                                                FROM {backup_log}
+                                               WHERE courseid = ? AND laststarttime = ? AND backuptype = ?
+                                            ORDER BY id", array($courseid, $execution->laststarttime,'scheduledbackup'));
                 if ($logs) {
                     foreach ($logs as $log) {
                         echo "<tr>";
@@ -127,9 +121,8 @@
             }
             echo "</table>";
         }
-        print_simple_box_end();
+        echo $OUTPUT->box_end();
     }
 
-    print_footer();
+    echo $OUTPUT->footer();
 
-?>
