diff --git a/blocks/moodleblock.class.php b/blocks/moodleblock.class.php
index 09cffbd..de0b674 100644
--- a/blocks/moodleblock.class.php
+++ b/blocks/moodleblock.class.php
@@ -424,6 +424,8 @@ class block_base {
     }
 
     function get_required_javascript() {
+        // RL EDIT: Block navigation menus don't expand w/o following line
+        $this->page->requires->js('/blocks/navigation/yui/navigation/navigation.js');
         if ($this->instance_can_be_docked() && !$this->hide_header()) {
             $this->page->requires->js_init_call('M.core_dock.init_genericblock', array($this->instance->id));
             user_preference_allow_ajax_update('docked_block_instance_'.$this->instance->id, PARAM_INT);
