diff --git a/userpix/index.php b/userpix/index.php
index 6564de2..652df70 100644
--- a/userpix/index.php
+++ b/userpix/index.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
       // This simple script displays all the users with pictures on one page.
       // By default it is not linked anywhere on the site.  If you want to
       // make it available you should link it in yourself from somewhere.
@@ -6,32 +6,31 @@
       // to administrators only (see below)
 
 
-    include('../config.php');
+require('../config.php');
 
-    require_login();
+$PAGE->set_url('/userpix/index.php');
 
-/// Remove the following three lines if you want everyone to access it
-    require_capability('moodle/site:config', get_context_instance(CONTEXT_SYSTEM));
-
-    if (!$users = get_records("user", "picture", "1", "lastaccess DESC", "id,firstname,lastname")) {
-        error("no users!");
-    }
-
-    $title = get_string("users");
+require_login();
 
-    print_header($title, $title, build_navigation(array(array('name' => $title, 'link' => null, 'type' => 'misc'))));
-
-    foreach ($users as $user) {
+/// Remove the following three lines if you want everyone to access it
+$syscontext = get_context_instance(CONTEXT_SYSTEM);
+require_capability('moodle/site:config', $syscontext);
+
+$title = get_string("users");
+$PAGE->set_context($syscontext);
+$PAGE->navbar->add($title);
+$PAGE->set_title($title);
+$PAGE->set_heading($title);
+echo $OUTPUT->header();
+
+$rs = $DB->get_recordset_select("user", "deleted = 0 AND picture > 0", array(), "lastaccess DESC", user_picture::fields());
+foreach ($rs as $user) {
         $fullname = fullname($user);
         echo "<a href=\"$CFG->wwwroot/user/view.php?id=$user->id&amp;course=1\" ".
              "title=\"$fullname\">";
-        require_once($CFG->libdir.'/filelib.php');
-        $userpictureurl = get_file_url($user->id.'/f1.jpg', null, 'user');
-        echo '<img src="'. $userpictureurl .'"'.
-            ' style="border:0px; width:100px; height:100px" alt="'.$fullname.'" />';
+    echo $OUTPUT->user_picture($user);
         echo "</a> \n";
-    }
-
-    print_footer();
+}
+$rs->close();
 
-?>
+echo $OUTPUT->footer();
