diff --git a/question/format/missingword/format.php b/question/format/missingword/format.php
index 0f1e952..15e1d71 100644
--- a/question/format/missingword/format.php
+++ b/question/format/missingword/format.php
@@ -1,33 +1,54 @@
-<?php  // $Id$ 
-/// Modified by Tom Robb 12 June 2003 to include percentage and comment insertion
-/// facility.
-
-////////////////////////////////////////////////////////////////////////////
-/// MISSING WORD FORMAT
-///
-/// This Moodle class provides all functions necessary to import and export 
-/// one-correct-answer multiple choice questions in this format:
-///
-///    As soon as we begin to explore our body parts as infants
-///    we become students of {=anatomy and physiology ~reflexology 
-///    ~science ~experiment}, and in a sense we remain students for life.
-/// 
-/// Each answer is separated with a tilde ~, and the correct answer is 
-/// prefixed with an equals sign =
-///
-/// Percentage weights can be included by following the tilde with the
-/// desired percent.  Comments can be included for each choice by following
-/// the comment with a hash mark ("#") and the comment.  Example:
-///
-///    This is {=the best answer#comment on the best answer ~75%a good
-///    answer#comment on the good answer ~a wrong one#comment on the bad answer}
-///
-////////////////////////////////////////////////////////////////////////////
-
-// Based on format.php, included by ../../import.php
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Missing word question importer.
+ *
+ * @package    qformat
+ * @subpackage missingword
+ * @copyright  1999 onwards Martin Dougiamas {@link http://moodle.com}
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+
+defined('MOODLE_INTERNAL') || die();
+
+
 /**
- * @package questionbank
- * @subpackage importexport
+ * Missing word question importer.
+ *
+ * This Moodle class provides all functions necessary to import and export
+ * one-correct-answer multiple choice questions in this format:
+ *
+ *    As soon as we begin to explore our body parts as infants
+ *    we become students of {=anatomy and physiology ~reflexology
+ *    ~science ~experiment}, and in a sense we remain students for life.
+ *
+ * Each answer is separated with a tilde ~, and the correct answer is
+ * prefixed with an equals sign =
+ *
+ * Percentage weights can be included by following the tilde with the
+ * desired percent.  Comments can be included for each choice by following
+ * the comment with a hash mark ("#") and the comment.  Example:
+ *
+ *    This is {=the best answer#comment on the best answer ~75%a good
+ *    answer#comment on the good answer ~a wrong one#comment on the bad answer}
+ *
+ * @copyright  1999 onwards Martin Dougiamas {@link http://moodle.com}
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
 class qformat_missingword extends qformat_default {
 
@@ -67,7 +88,7 @@ class qformat_missingword extends qformat_default {
         $answertext = substr($text, $answerstart + 1, $answerlength - 1);
 
         /// Save the new question text
-        $question->questiontext = addslashes(substr_replace($text, "_____", $answerstart, $answerlength+1));
+        $question->questiontext = substr_replace($text, "_____", $answerstart, $answerlength+1);
         $question->name = $question->questiontext;
 
 
@@ -97,7 +118,7 @@ class qformat_missingword extends qformat_default {
                 if ($answer[0] == "=") {
                     $answer = substr($answer, 1);
                 }
-                $question->answer[]   = addslashes($answer);
+                $question->answer[]   = $answer;
                 $question->fraction[] = 1;
                 $question->feedback[] = "";
     
@@ -130,7 +151,7 @@ class qformat_missingword extends qformat_default {
 
                     if (strpos($answer,"#") > 0){
                         $hashpos = strpos($answer,"#");
-                        $comment = addslashes(substr(($answer),$hashpos+1));
+                        $comment = substr(($answer),$hashpos+1);
                         $answer  = substr($answer,0,$hashpos);
                     } else {
                         $comment = " ";
@@ -145,7 +166,7 @@ class qformat_missingword extends qformat_default {
 #                       $question->fraction[$key] = 0;
                         $question->fraction[$key] = $answeight;
                     }
-                    $question->answer[$key]   = addslashes($answer);
+                    $question->answer[$key]   = $answer;
                     $question->feedback[$key] = $comment;
                 }
     
@@ -154,4 +175,4 @@ class qformat_missingword extends qformat_default {
     }
 }
 
-?>
+
