diff --git a/auth/fc/fcFPP.php b/auth/fc/fcFPP.php
index d7db72e..c082202 100644
--- a/auth/fc/fcFPP.php
+++ b/auth/fc/fcFPP.php
@@ -51,7 +51,7 @@ class fcFPP
     $conn = fsockopen($host, $port, $errno, $errstr, 5);
     if (!$conn)
     {
-        print_error('auth_fcconnfail','auth', '', array($errno, $errstr));
+        print_error('auth_fcconnfail','auth_fc', '', array('no'=>$errno, 'str'=>$errstr));
         return false;
     }
 
