diff --git a/admin/xmldb/actions/edit_xml_file_save/edit_xml_file_save.class.php b/admin/xmldb/actions/edit_xml_file_save/edit_xml_file_save.class.php
index 2147a79..9319bbc 100644
--- a/admin/xmldb/actions/edit_xml_file_save/edit_xml_file_save.class.php
+++ b/admin/xmldb/actions/edit_xml_file_save/edit_xml_file_save.class.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 
 ///////////////////////////////////////////////////////////////////////////
 //                                                                       //
@@ -61,16 +61,16 @@ class edit_xml_file_save extends XMLDBAction {
 
     /// Do the job, setting result as needed
 
-        if (!data_submitted('nomatch')) { ///Basic prevention
-            error('Wrong action call');
+        if (!data_submitted()) { ///Basic prevention
+            print_error('wrongcall', 'error');
         }
 
     /// Get parameters
         $dirpath = required_param('dir', PARAM_PATH);
-        $dirpath = $CFG->dirroot . stripslashes_safe($dirpath);
+        $dirpath = $CFG->dirroot . $dirpath;
 
         $comment = required_param('comment', PARAM_CLEAN);
-        $comment = stripslashes_safe($comment);
+        $comment = $comment;
 
     /// Set comment and recalculate hash
         $editeddir =& $XMLDB->editeddirs[$dirpath];
@@ -97,4 +97,4 @@ class edit_xml_file_save extends XMLDBAction {
         return $result;
     }
 }
-?>
+
