diff --git a/search/Zend/Search/Lucene/Storage/Directory/Filesystem.php b/search/Zend/Search/Lucene/Storage/Directory/Filesystem.php
index 48195cf..9220ad2 100644
--- a/search/Zend/Search/Lucene/Storage/Directory/Filesystem.php
+++ b/search/Zend/Search/Lucene/Storage/Directory/Filesystem.php
@@ -183,8 +183,7 @@ class Zend_Search_Lucene_Storage_Directory_Filesystem extends Zend_Search_Lucene
         $this->_fileHandlers[$filename] = new Zend_Search_Lucene_Storage_File_Filesystem($this->_dirPath . '/' . $filename, 'w+b');
 
         global $php_errormsg;
-        $trackErrors = ini_get('track_errors'); 
-        ini_set('track_errors', '1');
+        $trackErrors = ini_get('track_errors'); ini_set('track_errors', '1');
         if (!@chmod($this->_dirPath . '/' . $filename, self::$_defaultFilePermissions)) {
             ini_set('track_errors', $trackErrors);
             throw new Zend_Search_Lucene_Exception($php_errormsg);
@@ -213,9 +212,9 @@ class Zend_Search_Lucene_Storage_Directory_Filesystem extends Zend_Search_Lucene
         $trackErrors = ini_get('track_errors'); ini_set('track_errors', '1');
         if (!@unlink($this->_dirPath . '/' . $filename)) {
             ini_set('track_errors', $trackErrors);
+            throw $e;
             
-            // keep loose on failed deletion.
-            // throw new Zend_Search_Lucene_Exception('Can\'t delete file: ' . $php_errormsg);
+            throw new Zend_Search_Lucene_Exception('Can\'t delete file: ' . $php_errormsg);
         }
         ini_set('track_errors', $trackErrors);
     }
