diff --git a/mod/quiz/accessrule/safebrowser/rule.php b/mod/quiz/accessrule/safebrowser/rule.php
index e73c5b8..763acdc 100644
--- a/mod/quiz/accessrule/safebrowser/rule.php
+++ b/mod/quiz/accessrule/safebrowser/rule.php
@@ -61,6 +61,9 @@ class quizaccess_safebrowser extends quiz_access_rule_base {
     public function setup_attempt_page($page) {
         $page->set_title($this->quizobj->get_course()->shortname . ': ' . $page->title);
         $page->set_cacheable(false);
+        $page->set_popup_notification_allowed(false); // Prevent message notifications.
+        $page->set_heading($page->title);
+        $page->set_pagelayout('secure');
     }
 
     /**
