diff --git a/question/type/random/edit_random_form.php b/question/type/random/edit_random_form.php
index 9e0a74d..de58ca4 100644
--- a/question/type/random/edit_random_form.php
+++ b/question/type/random/edit_random_form.php
@@ -1,18 +1,39 @@
-<?php  // $Id$
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
 /**
  * Defines the editing form for the random question type.
  *
- * @copyright &copy; 2007 Jamie Pratt
- * @author Jamie Pratt me@jamiep.org
- * @license http://www.gnu.org/copyleft/gpl.html GNU Public License
- * @package questionbank
- * @subpackage questiontypes
+ * @package    qtype
+ * @subpackage random
+ * @copyright  2007 Jamie Pratt me@jamiep.org
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
 
+
+defined('MOODLE_INTERNAL') || die();
+
+
 /**
  * random editing form definition.
+ *
+ * @copyright  2007 Jamie Pratt me@jamiep.org
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
-class question_edit_random_form extends question_edit_form {
+class qtype_random_edit_form extends question_edit_form {
     /**
      * Build the form definition.
      *
@@ -20,26 +41,31 @@ class question_edit_random_form extends question_edit_form {
      * If your question type does not support all these fields, then you can
      * override this method and remove the ones you don't want with $mform->removeElement().
      */
-    function definition() {
+    protected function definition() {
         global $COURSE, $CFG;
 
         $qtype = $this->qtype();
         $langfile = "qtype_$qtype";
 
-        $mform =& $this->_form;
+        $mform = $this->_form;
 
         // Standard fields at the start of the form.
         $mform->addElement('header', 'generalheader', get_string("general", 'form'));
 
-        $mform->addElement('questioncategory', 'category', get_string('category', 'quiz'), 
+        $mform->addElement('questioncategory', 'category', get_string('category', 'question'),
                 array('contexts' => $this->contexts->having_cap('moodle/question:useall')));
 
-        $mform->addElement('advcheckbox', 'questiontext', get_string("recurse", "quiz"), null, null, array(0, 1));
+        $mform->addElement('advcheckbox', 'questiontext[text]',
+                get_string('includingsubcategories', 'qtype_random'), null, null, array(0, 1));
 
         $mform->addElement('hidden', 'name');
         $mform->setType('name', PARAM_ALPHA);
         $mform->setDefault('name', '');
 
+        $mform->addElement('hidden', 'tags[]');
+        $mform->setType('tags[]', PARAM_ALPHA);
+        $mform->setDefault('tags[]', '');
+
         // Standard fields at the end of the form.
         $mform->addElement('hidden', 'questiontextformat', 0);
         $mform->setType('questiontextformat', PARAM_INT);
@@ -69,19 +95,25 @@ class question_edit_random_form extends question_edit_form {
         $mform->setDefault('returnurl', 0);
 
         $buttonarray = array();
-        $buttonarray[] = &$mform->createElement('submit', 'submitbutton', get_string('savechanges'));
-
-        $buttonarray[] = &$mform->createElement('cancel');
+        $buttonarray[] = $mform->createElement('submit', 'submitbutton', get_string('savechanges'));
+        $buttonarray[] = $mform->createElement('cancel');
         $mform->addGroup($buttonarray, 'buttonar', '', array(' '), false);
         $mform->closeHeaderBefore('buttonar');
     }
-    function validation($fromform, $files) {
+
+    public function set_data($question) {
+        $question->questiontext = array('text' => $question->questiontext);
+        // We don't want the complex stuff in the base class to run.
+        moodleform::set_data($question);
+    }
+
+    public function validation($fromform, $files) {
         //validation of category
         //is not relevant for this question type
         return array();
     }
-    function qtype() {
+
+    public function qtype() {
         return 'random';
     }
 }
-?>
\ No newline at end of file
