diff --git a/mod/scorm/grade.php b/mod/scorm/grade.php
index 5b0c07d..3e239f6 100644
--- a/mod/scorm/grade.php
+++ b/mod/scorm/grade.php
@@ -5,15 +5,15 @@
     $id   = required_param('id', PARAM_INT);          // Course module ID
 
     if (! $cm = get_coursemodule_from_id('scorm', $id)) {
-        error('Course Module ID was incorrect');
+        print_error('invalidcoursemodule');
     }
 
-    if (! $scorm = get_record('scorm', 'id',$cm->instance)) {
-        error('scorm ID was incorrect');
+    if (! $scorm = $DB->get_record('scorm', array('id'=>$cm->instance))) {
+        print_error('invalidcoursemodule');
     }
 
-    if (! $course = get_record('course','id', $scorm->course)) {
-        error('Course is misconfigured');
+    if (! $course = $DB->get_record('course', array('id'=> $scorm->course))) {
+        print_error('coursemisconf');
     }
 
     require_login($course->id, false, $cm);
