diff --git a/mod/scorm/player.php b/mod/scorm/player.php
old mode 100755
new mode 100644
index ce0dd83..a0f0a71
--- a/mod/scorm/player.php
+++ b/mod/scorm/player.php
@@ -1,19 +1,18 @@
-<?PHP  // $Id$
+<?PHP
 
 /// This page prints a particular instance of aicc/scorm package
 
     require_once('../../config.php');
     require_once($CFG->dirroot.'/mod/scorm/locallib.php');
+    require_once($CFG->libdir . '/completionlib.php');
 
-    //
-    // Checkin' script parameters
-    //
-    $id = optional_param('id', '', PARAM_INT);       // Course Module ID, or
+    $id = optional_param('cm', '', PARAM_INT);       // Course Module ID, or
     $a = optional_param('a', '', PARAM_INT);         // scorm ID
     $scoid = required_param('scoid', PARAM_INT);  // sco ID
     $mode = optional_param('mode', 'normal', PARAM_ALPHA); // navigation mode
     $currentorg = optional_param('currentorg', '', PARAM_RAW); // selected organization
     $newattempt = optional_param('newattempt', 'off', PARAM_ALPHA); // the user request to start a new attempt
+    $displaymode = optional_param('display','',PARAM_ALPHA);
 
     //IE 6 Bug workaround
     if (strpos($_SERVER['HTTP_USER_AGENT'], 'MSIE 6') !== false) {
@@ -21,28 +20,52 @@
         @apache_setenv('no-gzip', 1);
     }
     
+    //IE 9 workaround for Flash bug: MDL-29213
+    if (strpos($_SERVER['HTTP_USER_AGENT'], 'MSIE 9') !== false) {
+        if (!isset($CFG->additionalhtmlhead)) { //check to make sure set first - that way we can use .=
+            $CFG->additionalhtmlhead = '';
+        }
+        $CFG->additionalhtmlhead .= '<meta http-equiv="X-UA-Compatible" content="IE=8" />';
+    }
+
     if (!empty($id)) {
         if (! $cm = get_coursemodule_from_id('scorm', $id)) {
-            error("Course Module ID was incorrect");
+            print_error('invalidcoursemodule');
         }
-        if (! $course = get_record("course", "id", $cm->course)) {
-            error("Course is misconfigured");
+        if (! $course = $DB->get_record("course", array("id"=>$cm->course))) {
+            print_error('coursemisconf');
         }
-        if (! $scorm = get_record("scorm", "id", $cm->instance)) {
-            error("Course module is incorrect");
+        if (! $scorm = $DB->get_record("scorm", array("id"=>$cm->instance))) {
+            print_error('invalidcoursemodule');
         }
     } else if (!empty($a)) {
-        if (! $scorm = get_record("scorm", "id", $a)) {
-            error("Course module is incorrect");
+        if (! $scorm = $DB->get_record("scorm", array("id"=>$a))) {
+            print_error('invalidcoursemodule');
         }
-        if (! $course = get_record("course", "id", $scorm->course)) {
-            error("Course is misconfigured");
+        if (! $course = $DB->get_record("course", array("id"=>$scorm->course))) {
+            print_error('coursemisconf');
         }
         if (! $cm = get_coursemodule_from_instance("scorm", $scorm->id, $course->id)) {
-            error("Course Module ID was incorrect");
+            print_error('invalidcoursemodule');
         }
     } else {
-        error('A required parameter is missing');
+        print_error('missingparameter');
+    }
+
+    $url = new moodle_url('/mod/scorm/player.php', array('scoid'=>$scoid, 'cm'=>$cm->id));
+    if ($mode !== 'normal') {
+        $url->param('mode', $mode);
+    }
+    if ($currentorg !== '') {
+        $url->param('currentorg', $currentorg);
+    }
+    if ($newattempt !== 'off') {
+        $url->param('newattempt', $newattempt);
+    }
+    $PAGE->set_url($url);
+    $forcejs = get_config('scorm','forcejavascript');
+    if (!empty($forcejs)) {
+        $PAGE->add_body_class('forcejavascript');
     }
 
     require_login($course->id, false, $cm);
@@ -52,28 +75,40 @@
     $strpopup = get_string('popup','scorm');
     $strexit = get_string('exitactivity','scorm');
 
-    $navlinks = array();
+    $coursecontext = get_context_instance(CONTEXT_COURSE, $course->id);
     
-    if ($course->id != SITEID) {
-        if ($scorms = get_all_instances_in_course('scorm', $course)) {
-            // The module SCORM/AICC activity with the first id is the course  
-            $firstscorm = current($scorms);
-            if (!(($course->format == 'scorm') && ($firstscorm->id == $scorm->id))) {
-                $navlinks[] = array('name' => $strscorms, 'link' => "index.php?id=$course->id", 'type' => 'activity');
-            }
-        }
+    if ($displaymode == 'popup') {
+        $PAGE->set_pagelayout('popup');
+    } else {
+        $shortname = format_string($course->shortname, true, array('context' => $coursecontext));
+        $pagetitle = strip_tags("$shortname: ".format_string($scorm->name));
+        $PAGE->set_title($pagetitle);
+        $PAGE->set_heading($course->fullname);
     }
 
-    $pagetitle = strip_tags("$course->shortname: ".format_string($scorm->name));
-
-    if (!$cm->visible and !has_capability('moodle/course:viewhiddenactivities', get_context_instance(CONTEXT_COURSE,$course->id))) {
-        $navlinks[] = array('name' => format_string($scorm->name,true), 'link' => "view.php?id=$cm->id", 'type' => 'activityinstance');
-        $navigation = build_navigation($navlinks);
-        
-        print_header($pagetitle, $course->fullname, $navigation,
-                 '', '', true, update_module_button($cm->id, $course->id, $strscorm), '', false);
+    if (!$cm->visible and !has_capability('moodle/course:viewhiddenactivities', $coursecontext)) {
+        echo $OUTPUT->header();
         notice(get_string("activityiscurrentlyhidden"));
+        echo $OUTPUT->footer();
+        die;
+    }
+
+    //check if scorm closed
+    $timenow = time();
+    if ($scorm->timeclose !=0) {
+        if ($scorm->timeopen > $timenow) {
+            echo $OUTPUT->header();
+            echo $OUTPUT->box(get_string("notopenyet", "scorm", userdate($scorm->timeopen)), "generalbox boxaligncenter");
+            echo $OUTPUT->footer();
+            die;
+        } elseif ($timenow > $scorm->timeclose) {
+            echo $OUTPUT->header();
+            echo $OUTPUT->box(get_string("expired", "scorm", userdate($scorm->timeclose)), "generalbox boxaligncenter");
+            echo $OUTPUT->footer();
+            die;
+        }
     }
+
     //
     // TOC processing
     //
@@ -89,7 +124,7 @@
     }
     $attemptstr = '&amp;attempt=' . $attempt;
 
-    $result = scorm_get_toc($USER,$scorm,'structurelist',$currentorg,$scoid,$mode,$attempt,true);
+    $result = scorm_get_toc($USER, $scorm, $cm->id, TOCJSLINK, $currentorg, $scoid, $mode, $attempt, true, true);
     $sco = $result->sco;
 
     if (($mode == 'browse') && ($scorm->hidebrowse == 1)) {
@@ -107,7 +142,8 @@
         }
     }
 
-    add_to_log($course->id, 'scorm', 'view', "player.php?id=$cm->id&scoid=$sco->id", "$scorm->id", $cm->id);
+    add_to_log($course->id, 'scorm', 'view', "player.php?cm=$cm->id&scoid=$sco->id", "$scorm->id", $cm->id);
+
 
     $scoidstr = '&amp;scoid='.$sco->id;
     $scoidpop = '&scoid='.$sco->id;
@@ -124,217 +160,111 @@
     $SESSION->scorm_mode = $mode;
     $SESSION->scorm_attempt = $attempt;
 
+    // Mark module viewed
+    $completion = new completion_info($course);
+    $completion->set_module_viewed($cm);
+
     //
     // Print the page header
     //
-    $bodyscript = '';
-    if ($scorm->popup == 1) {
-        $bodyscript = 'onunload="main.close();"';
+    if (empty($scorm->popup) || $displaymode=='popup') {
+        $exitlink = '<a href="'.$CFG->wwwroot.'/course/view.php?id='.$scorm->course.'" title="'.$strexit.'">'.$strexit.'</a> ';
+        $PAGE->set_button($exitlink);
     }
 
-    $navlinks[] = array('name' => format_string($scorm->name,true), 'link' => "view.php?id=$cm->id", 'type' => 'activityinstance');
-    $navigation = build_navigation($navlinks);
-    $exitlink = '<a href="'.$CFG->wwwroot.'/course/view.php?id='.$scorm->course.'" title="'.$strexit.'">'.$strexit.'</a> ';
+    $PAGE->requires->yui2_lib('connection');
+    $PAGE->requires->data_for_js('scormplayerdata', Array('cwidth'=>$scorm->width,'cheight'=>$scorm->height,
+                                                          'popupoptions' => $scorm->options), true);
+    $PAGE->requires->js('/mod/scorm/request.js', true);
+    $PAGE->requires->js('/lib/cookies.js', true);
+    $PAGE->requires->css('/mod/scorm/styles.css');
+    echo $OUTPUT->header();
+
+    // NEW IMS TOC
+    $PAGE->requires->string_for_js('navigation', 'scorm');
+    $PAGE->requires->string_for_js('toc', 'scorm');
+    $PAGE->requires->string_for_js('hide', 'moodle');
+    $PAGE->requires->string_for_js('show', 'moodle');
+    $PAGE->requires->string_for_js('popupsblocked', 'scorm');
+
+    $name = false;
     
-    print_header($pagetitle, $course->fullname,
-                 $navigation,
-                 '', '', true, $exitlink.update_module_button($cm->id, $course->id, $strscorm), '', false, $bodyscript);
-?>
-    <script type="text/javascript" src="request.js"></script>
-    <script type="text/javascript" src="api.php?id=<?php echo $cm->id.$scoidstr.$modestr.$attemptstr ?>"></script>
-    <script type="text/javascript" src="<?php echo $CFG->wwwroot; ?>/mod/scorm/rd.js"></script>
-    <script type="text/javascript">
-    <!--    
-        window.onresize = function() {
-            scorm_resize(<?php echo $scorm->width.", ".$scorm->height; ?>);
-        };
-    -->  
-    </script>
-<?php
-    //}
-    if (($sco->previd != 0) && ((!isset($sco->previous)) || ($sco->previous == 0))) {
-        $scostr = '&scoid='.$sco->previd;
-        echo '    <script type="text/javascript">'."\n//<![CDATA[\n".'var prev="'.$CFG->wwwroot.'/mod/scorm/player.php?id='.$cm->id.$orgstr.$modepop.$scostr."\";\n//]]>\n</script>\n";
-    } else {
-        echo '    <script type="text/javascript">'."\n//<![CDATA[\n".'var prev="'.$CFG->wwwroot.'/mod/scorm/view.php?id='.$cm->id."\";\n//]]>\n</script>\n";
-    }
-    if (($sco->nextid != 0) && ((!isset($sco->next)) || ($sco->next == 0))) {
-        $scostr = '&scoid='.$sco->nextid;
-        echo '    <script type="text/javascript">'."\n//<![CDATA[\n".'var next="'.$CFG->wwwroot.'/mod/scorm/player.php?id='.$cm->id.$orgstr.$modepop.$scostr."\";\n//]]>\n</script>\n";
-    } else {
-        echo '    <script type="text/javascript">'."\n//<![CDATA[\n".'var next="'.$CFG->wwwroot.'/mod/scorm/view.php?id='.$cm->id."\";\n//]]>\n</script>\n";
-    }
 ?>
     <div id="scormpage">
-<?php  
-    if ($scorm->hidetoc == 0) {
-?>
+
         <div id="tocbox">
+        <div id='scormapi-parent'>
+            <script id="external-scormapi" type="text/JavaScript"></script>
+        </div>
 <?php
-        if ($scorm->hidenav ==0){
-?>
-            <!-- Bottons nav at left-->
-            <div id="tochead">
-                <form name="tochead" method="post" action="player.php?id=<?php echo $cm->id ?>" target="_top">
-<?php
-            $orgstr = '&amp;currentorg='.$currentorg;
-            if (($scorm->hidenav == 0) && ($sco->previd != 0) && (!isset($sco->previous) || $sco->previous == 0)) {
-                // Print the prev LO button
-                $scostr = '&amp;scoid='.$sco->previd;
-                $url = $CFG->wwwroot.'/mod/scorm/player.php?id='.$cm->id.$orgstr.$modestr.$scostr;
-?>
-                    <input name="prev" type="button" value="<?php print_string('prev','scorm') ?>" onClick="document.location.href=' <?php echo $url; ?> '"/>
-<?php
-            }
-            if (($scorm->hidenav == 0) && ($sco->nextid != 0) && (!isset($sco->next) || $sco->next == 0)) {
-                // Print the next LO button
-                $scostr = '&amp;scoid='.$sco->nextid;
-                $url = $CFG->wwwroot.'/mod/scorm/player.php?id='.$cm->id.$orgstr.$modestr.$scostr;
-?>
-                    <input name="next" type="button" value="<?php print_string('next','scorm') ?>" onClick="document.location.href=' <?php echo $url; ?> '"/>
-<?php
-            }
-?>
-                </form>
-            </div> <!-- tochead -->
-<?php
-        }
+if ($scorm->hidetoc == SCORM_TOC_POPUP or $mode=='browse' or $mode=='review') {
+    echo '<div id="scormtop">';
+    echo $mode == 'browse' ? '<div id="scormmode" class="scorm-left">'.get_string('browsemode', 'scorm')."</div>\n" : '';
+    echo $mode == 'review' ? '<div id="scormmode" class="scorm-left">'.get_string('reviewmode', 'scorm')."</div>\n" : '';
+    if ($scorm->hidetoc == SCORM_TOC_POPUP) {
+        echo '<div id="scormnav" class="scorm-right">'.$result->tocmenu.'</div>';
+    }
+    echo '</div>';
+}
 ?>
-            <div id="toctree" class="generalbox">
-            <?php echo $result->toc; ?>
+            <div id="toctree">
+                <?php
+                if (empty($scorm->popup) || $displaymode == 'popup') {
+                    echo $result->toc;
+                } else {
+                    //Added incase javascript popups are blocked we don't provide a direct link to the pop-up as JS communication can fail - the user must disable their pop-up blocker.
+                    $linkcourse = '<a href="'.$CFG->wwwroot.'/course/view.php?id='.$scorm->course.'">' . get_string('finishscormlinkname', 'scorm') . '</a>';
+                    echo $OUTPUT->box(get_string('finishscorm', 'scorm', $linkcourse), 'generalbox', 'altfinishlink');
+                }?>
             </div> <!-- toctree -->
         </div> <!--  tocbox -->
-<?php
-        $class = ' class="toc"';
-    } else {
-        $class = ' class="no-toc"';
-    }
-?>
-        <div id="scormbox"<?php echo $class; if(($scorm->hidetoc == 2) || ($scorm->hidetoc == 1)){echo 'style="width:100%"';}?>>
-<?php
-    // This very big test check if is necessary the "scormtop" div
-    if (
-           ($mode != 'normal') ||  // We are not in normal mode so review or browse text will displayed
-           (
-               ($scorm->hidenav == 0) &&  // Teacher want to display navigation links
-               ($scorm->hidetoc != 0) &&  // The buttons has not been displayed
-               (
-                   (
-                       ($sco->previd != 0) &&  // This is not the first learning object of the package
-                       ((!isset($sco->previous)) || ($sco->previous == 0))   // Moodle must manage the previous link
-                   ) || 
-                   (
-                       ($sco->nextid != 0) &&  // This is not the last learning object of the package
-                       ((!isset($sco->next)) || ($sco->next == 0))       // Moodle must manage the next link
-                   ) 
-               )
-           ) || ($scorm->hidetoc == 2)      // Teacher want to display toc in a small dropdown menu 
-       ) {
-?>
-            <div id="scormtop">
-        <?php echo $mode == 'browse' ? '<div id="scormmode" class="scorm-left">'.get_string('browsemode','scorm')."</div>\n" : ''; ?>
-        <?php echo $mode == 'review' ? '<div id="scormmode" class="scorm-left">'.get_string('reviewmode','scorm')."</div>\n" : ''; ?>
-<?php
-       if (($scorm->hidenav == 0) || ($scorm->hidetoc == 2) || ($scorm->hidetoc == 1)) {
-?>
-                <div id="scormnav" class="scorm-right">
-        <?php
-            $orgstr = '&amp;currentorg='.$currentorg;
-            if (($scorm->hidenav == 0) && ($sco->previd != 0) && (!isset($sco->previous) || $sco->previous == 0) && (($scorm->hidetoc == 2) || ($scorm->hidetoc == 1)) ) {
- 
-                // Print the prev LO button
-                $scostr = '&amp;scoid='.$sco->previd;
-                $url = $CFG->wwwroot.'/mod/scorm/player.php?id='.$cm->id.$orgstr.$modestr.$scostr;
-?>
-                    <form name="scormnavprev" method="post" action="player.php?id=<?php echo $cm->id ?>" target="_top" style= "display:inline">
-                        <input name="prev" type="button" value="<?php print_string('prev','scorm') ?>" onClick="document.location.href=' <?php echo $url; ?> '"/>
-                    </form>
-<?php
-            }
-            if ($scorm->hidetoc == 2) {
-                echo $result->tocmenu;
-            }
-            if (($scorm->hidenav == 0) && ($sco->nextid != 0) && (!isset($sco->next) || $sco->next == 0) && (($scorm->hidetoc == 2) || ($scorm->hidetoc == 1))) {
-                // Print the next LO button
-                $scostr = '&amp;scoid='.$sco->nextid;
-                $url = $CFG->wwwroot.'/mod/scorm/player.php?id='.$cm->id.$orgstr.$modestr.$scostr;
-?>
-                    <form name="scormnavnext" method="post" action="player.php?id=<?php echo $cm->id ?>" target="_top" style= "display:inline">
-                        <input name="next" type="button" value="<?php print_string('next','scorm') ?>" onClick="document.location.href=' <?php echo $url; ?> '"/>
-                    </form>
-<?php
-            }
-        ?>
-                </div>
-<?php
-        } 
-?>
-            </div> <!-- Scormtop -->
-<?php
-    } // The end of the very big test
-?>
-            <div id="scormobject" class="scorm-right">
                 <noscript>
                     <div id="noscript">
                         <?php print_string('noscriptnoscorm','scorm'); // No Martin(i), No Party ;-) ?>
+
                     </div>
                 </noscript>
 <?php
     if ($result->prerequisites) {
-        if ($scorm->popup == 0) {
-            echo "                <script type=\"text/javascript\">scorm_resize(".$scorm->width.", ".$scorm->height.");</script>\n";
-            $fullurl="loadSCO.php?id=".$cm->id.$scoidstr.$modestr;
-            echo "                <iframe id=\"scoframe1\" class=\"scoframe\" name=\"scoframe1\" src=\"{$fullurl}\"></iframe>\n";   
-        } else {
+        if ($scorm->popup != 0 && $displaymode !=='popup') {
             // Clean the name for the window as IE is fussy
-            $name = ereg_replace("[^A-Za-z0-9]", "", $scorm->name);
+            $name = preg_replace("/[^A-Za-z0-9]/", "", $scorm->name);
             if (!$name) {
                 $name = 'DefaultPlayerWindow';
             }
             $name = 'scorm_'.$name;
-            ?>
-                    <script type="text/javascript">
-                    //<![CDATA[
-                        scorm_resize(<?php echo $scorm->width.", ". $scorm->height; ?>);
-                        function openpopup(url,name,options,width,height) {
-                            fullurl = "<?php echo $CFG->wwwroot.'/mod/scorm/' ?>" + url;
-                            windowobj = window.open(fullurl,name,options);
-                            if ((width==100) && (height==100)) {
-                                // Fullscreen
-                                windowobj.moveTo(0,0);
-                            } 
-                            if (width<=100) {
-                                width = Math.round(screen.availWidth * width / 100);
-                            }
-                            if (height<=100) {
-                                height = Math.round(screen.availHeight * height / 100);
-                            }
-                            windowobj.resizeTo(width,height);
-                            windowobj.focus();
-                            return windowobj;
-                        }
 
-                        url = "loadSCO.php?id=<?php echo $cm->id.$scoidpop ?>";
-                        width = <?php p($scorm->width) ?>;
-                        height = <?php p($scorm->height) ?>;
-                        var main = openpopup(url, "<?php p($name) ?>", "<?php p($scorm->options) ?>", width, height);
-                    //]]>
-                    </script>
+            echo html_writer::script('', $CFG->wwwroot.'/mod/scorm/player.js');
+            $url = new moodle_url($PAGE->url, array('scoid' => $sco->id, 'display' => 'popup'));
+            echo html_writer::script(
+            js_writer::function_call('scorm_openpopup', Array($url->out(false),
+                                                       $name, $scorm->options,
+                                                       $scorm->width, $scorm->height)));
+            ?>
                     <noscript>
-                    <iframe id="main" class="scoframe" src="loadSCO.php?id=<?php echo $cm->id.$scoidstr.$modestr ?>">
-                    </iframe>
+            <!--[if IE]>
+                <iframe id="main" class="scoframe" name="main" src="loadSCO.php?id=<?php echo $cm->id.$scoidstr.$modestr; ?>"></iframe>
+            <![endif]-->
+            <!--[if !IE]>
+                <object id="main" class="scoframe" type="text/html" data="loadSCO.php?id=<?php echo $cm->id.$scoidstr.$modestr; ?>"></object>
+            <![endif]-->
                     </noscript>
 <?php            
-            //Added incase javascript popups are blocked
-            print_simple_box(get_string('popupblockmessage','scorm'),'center','','',5,'generalbox','altpopuplink');
-            $linkcourse = '<a href="'.$CFG->wwwroot.'/course/view.php?id='.$scorm->course.'">' . get_string('finishscormlinkname','scorm') . '</a>';
-            print_simple_box(get_string('finishscorm','scorm',$linkcourse),'center','','',5,'generalbox','altfinishlink');
         }
     } else {
-        print_simple_box(get_string('noprerequisites','scorm'),'center');
+        echo $OUTPUT->box(get_string('noprerequisites','scorm'));
     }
 ?>
-            </div> <!-- SCORM object -->
-        </div> <!-- SCORM box  -->
     </div> <!-- SCORM page -->
-<?php print_footer('none'); ?>
\ No newline at end of file
+<?php
+// NEW IMS TOC
+if (empty($scorm->popup) || $displaymode == 'popup') {
+    if (!isset($result->toctitle)) {
+        $result->toctitle = get_string('toc', 'scorm');
+    }
+    $PAGE->requires->js_init_call('M.mod_scorm.init', array($scorm->hidenav, $scorm->hidetoc, $result->toctitle, $name, $sco->id));
+}
+if (!empty($forcejs)) {
+    echo $OUTPUT->box(get_string("forcejavascriptmessage", "scorm"), "generalbox boxaligncenter forcejavascriptmessage");
+}
+echo $OUTPUT->footer();
