diff --git a/mod/scorm/rd.js b/mod/scorm/rd.js
index 98570e4..f24a121 100644
--- a/mod/scorm/rd.js
+++ b/mod/scorm/rd.js
@@ -1,4 +1,17 @@
 <!--
+function attach_resize_event() {
+    YAHOO.util.Event.addListener(window, 'resize', scorm_resize);
+}
+
+var prev='';
+var next='';
+function scorm_set_prev(url) {
+    prev = url;
+}
+function scorm_set_next(url) {
+    next = url;
+}
+
 function scorm_get_element_style(obj, prop, cssProp) {
     var ret = '';
     
@@ -13,8 +26,9 @@ function scorm_get_element_style(obj, prop, cssProp) {
     return ret;
 }
 
-function scorm_resize (cwidth, cheight) {
-
+function scorm_resize () {
+    var cwidth = scormplayerdata.cwidth;
+    var cheight = scormplayerdata.cheight;
     var winwidth = 0, winheight = 0;
     if( typeof( window.innerWidth ) == 'number' ) {
         //Non-IE
@@ -31,7 +45,7 @@ function scorm_resize (cwidth, cheight) {
     }
                               
     var header = document.getElementById('header');   
-    var content = document.getElementById('content');
+    var content = document.getElementById('region-content');
     var headerheight = 0;
     if (content) {
         headerheight = content.offsetTop;
@@ -82,8 +96,8 @@ function scorm_resize (cwidth, cheight) {
         var scoframe2 = document.getElementById('scoframe1');
         document.getElementById('scormobject').style.height = finalheight + 'px';
     }else{
-        var scoframe2 = document.getElementById('scoframe1');
         document.getElementById('scormobject').style.height = finalheight + 'px';
+        document.getElementById('scormpage').style.height = finalheight + 'px';
     }
 
 	// resize the content container too to move the footer below the SCORM content
@@ -91,14 +105,14 @@ function scorm_resize (cwidth, cheight) {
 	if (contenti3) {
     	contenti3.style.height = (winheight - totalheight + 30) + 'px';
 	} else {
-	   document.getElementById('content').style.height = (finalheight + 30) + 'px';
+       document.getElementById('region-main-box').style.height = (finalheight + 30) + 'px';
     }
      // resize the content container too to move the footer below the SCORM content
     var contenti3 = document.getElementById('content-i3');
     if (contenti3) {
         contenti3.style.height = (finalheight + 30) + 'px';
     } else {
-        document.getElementById('content').style.height = (finalheight + 30) + 'px';
+        document.getElementById('region-main-box').style.height = (finalheight + 30) + 'px';
     }
 }
 -->
\ No newline at end of file
