diff --git a/mod/scorm/version.php b/mod/scorm/version.php
old mode 100755
new mode 100644
index f6cd9ea..e6c1847
--- a/mod/scorm/version.php
+++ b/mod/scorm/version.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 
 /////////////////////////////////////////////////////////////////////////////////
 ///  Code fragment to define the version of scorm
@@ -6,12 +6,7 @@
 /////////////////////////////////////////////////////////////////////////////////
 
 
-// NOTE  The version below was accidentally set a month into the future!  We need to 
-//       catch up now, so until 27th October please only increment in very tiny steps 
-//       in HEAD, until we get past that date..
-
-$module->version  = 2007110503;   // The (date) version of this module
-$module->requires = 2007101509;   // The version of Moodle that is required
+$module->version  = 2011021403;   // The (date) version of this module
+$module->requires = 2010080300;   // The version of Moodle that is required
 $module->cron     = 300;            // How often should cron check this module (seconds)?
 
-?>
\ No newline at end of file
