diff --git a/mod/scorm/view.php b/mod/scorm/view.php
old mode 100755
new mode 100644
index 65ef272..fb7d1a8
--- a/mod/scorm/view.php
+++ b/mod/scorm/view.php
@@ -1,4 +1,4 @@
-<?php  // $Id$
+<?php
 
     require_once("../../config.php");
     require_once($CFG->dirroot.'/mod/scorm/locallib.php');
@@ -9,26 +9,36 @@
 
     if (!empty($id)) {
         if (! $cm = get_coursemodule_from_id('scorm', $id)) {
-            error("Course Module ID was incorrect");
+            print_error('invalidcoursemodule');
         }
-        if (! $course = get_record("course", "id", $cm->course)) {
-            error("Course is misconfigured");
+        if (! $course = $DB->get_record("course", array("id"=>$cm->course))) {
+            print_error('coursemisconf');
         }
-        if (! $scorm = get_record("scorm", "id", $cm->instance)) {
-            error("Course module is incorrect");
+        if (! $scorm = $DB->get_record("scorm", array("id"=>$cm->instance))) {
+            print_error('invalidcoursemodule');
         }
     } else if (!empty($a)) {
-        if (! $scorm = get_record("scorm", "id", $a)) {
-            error("Course module is incorrect");
+        if (! $scorm = $DB->get_record("scorm", array("id"=>$a))) {
+            print_error('invalidcoursemodule');
         }
-        if (! $course = get_record("course", "id", $scorm->course)) {
-            error("Course is misconfigured");
+        if (! $course = $DB->get_record("course", array("id"=>$scorm->course))) {
+            print_error('coursemisconf');
         }
         if (! $cm = get_coursemodule_from_instance("scorm", $scorm->id, $course->id)) {
-            error("Course Module ID was incorrect");
+            print_error('invalidcoursemodule');
         }
     } else {
-        error('A required parameter is missing');
+        print_error('missingparameter');
+    }
+
+    $url = new moodle_url('/mod/scorm/view.php', array('id'=>$cm->id));
+    if ($organization !== '') {
+        $url->param('organization', $organization);
+    }
+    $PAGE->set_url($url);
+    $forcejs = get_config('scorm','forcejavascript');
+    if (!empty($forcejs)) {
+        $PAGE->add_body_class('forcejavascript');
     }
 
     require_login($course->id, false, $cm);
@@ -47,32 +57,43 @@
 
     add_to_log($course->id, 'scorm', 'pre-view', 'view.php?id='.$cm->id, "$scorm->id", $cm->id);
 
-    if ((has_capability('mod/scorm:skipview', $contextmodule)) && scorm_simple_play($scorm,$USER, $contextmodule)) {
+    if ((has_capability('mod/scorm:skipview', $contextmodule)) && scorm_simple_play($scorm,$USER, $contextmodule,$cm->id)) {
         exit;
     }
 
     //
     // Print the page header
     //
-    $navlinks = array();
-    $navigation = build_navigation($navlinks,$cm);
+    $PAGE->set_title($pagetitle);
+    $PAGE->set_heading($course->fullname);
+    echo $OUTPUT->header();
     
-    print_header($pagetitle, $course->fullname, $navigation,
-                 '', '', true, update_module_button($cm->id, $course->id, $strscorm), navmenu($course, $cm));
+    $currenttab = 'info';
+    require($CFG->dirroot . '/mod/scorm/tabs.php');
 
-    if (has_capability('mod/scorm:viewreport', $context)) {
+    // Print the main part of the page
+    echo $OUTPUT->heading(format_string($scorm->name));
+    $attemptstatus = '';
+    if ($scorm->displayattemptstatus == 1) {
+        $attemptstatus = scorm_get_attempt_status($USER,$scorm);
+    }
+    echo $OUTPUT->box(format_module_intro('scorm', $scorm, $cm->id).$attemptstatus, 'generalbox boxaligncenter boxwidthwide', 'intro');
         
-        $trackedusers = scorm_get_count_users($scorm->id, $cm->groupingid);
-        if ($trackedusers > 0) {
-            echo "<div class=\"reportlink\"><a $CFG->frametarget href=\"report.php?id=$cm->id\"> ".get_string('viewalluserreports','scorm',$trackedusers).'</a></div>';
-        } else {
-            echo '<div class="reportlink">'.get_string('noreports','scorm').'</div>';
+    $scormopen = true;
+    $timenow = time();
+    if (!empty($scorm->timeopen) && $scorm->timeopen > $timenow) {
+        echo $OUTPUT->box(get_string("notopenyet", "scorm", userdate($scorm->timeopen)), "generalbox boxaligncenter");
+        $scormopen = false;
         }
+    if (!empty($scorm->timeclose) && $timenow > $scorm->timeclose) {
+        echo $OUTPUT->box(get_string("expired", "scorm", userdate($scorm->timeclose)), "generalbox boxaligncenter");
+        $scormopen = false;
     }
-
-    // Print the main part of the page
-    print_heading(format_string($scorm->name));
-    print_box(format_text($scorm->summary), 'generalbox', 'intro');
+    if ($scormopen) {
     scorm_view_display($USER, $scorm, 'view.php?id='.$cm->id, $cm);
-    print_footer($course);
-?>
+    }
+    if (!empty($forcejs)) {
+        echo $OUTPUT->box(get_string("forcejavascriptmessage", "scorm"), "generalbox boxaligncenter forcejavascriptmessage");
+    }
+    echo $OUTPUT->footer();
+
