diff --git a/admin/report/unittest/settings.php b/admin/report/unittest/settings.php
index e0f81a7..8826c31 100644
--- a/admin/report/unittest/settings.php
+++ b/admin/report/unittest/settings.php
@@ -1,3 +1,8 @@
-<?php  //$Id$
-$ADMIN->add('reports', new admin_externalpage('reportunittest', get_string('simpletest', 'admin'), "$CFG->wwwroot/$CFG->admin/report/unittest/index.php", 'report/unittest:view'));
-?>
\ No newline at end of file
+<?php
+
+defined('MOODLE_INTERNAL') || die;
+
+if ($hassiteconfig) {
+    $ADMIN->add('development', new admin_externalpage('reportsimpletest', get_string('simpletest', 'admin'), "$CFG->wwwroot/$CFG->admin/report/unittest/index.php",'report/unittest:view'));
+    $ADMIN->add('development', new admin_externalpage('reportdbtest', get_string('dbtest', 'admin'), "$CFG->wwwroot/$CFG->admin/report/unittest/dbtest.php",'report/unittest:view'));
+}
