diff --git a/admin/report/unittest/version.php b/admin/report/unittest/version.php
index 5892742..f734c56 100644
--- a/admin/report/unittest/version.php
+++ b/admin/report/unittest/version.php
@@ -1,4 +1,4 @@
-<?PHP // $Id$
+<?PHP
 
 ///////////////////////////////////////////////////////////////////////////
 //                                                                       //
@@ -23,7 +23,6 @@
 //                                                                       //
 ///////////////////////////////////////////////////////////////////////////
 
-$plugin->version  = 2007101501;
-$plugin->requires = 2007101532;
+$plugin->version  = 2010090501;
+$plugin->requires = 2010090501;
 
-?>
