diff --git a/tag/locallib.php b/tag/locallib.php
index 615d189..74c788e 100644
--- a/tag/locallib.php
+++ b/tag/locallib.php
@@ -1,9 +1,11 @@
-<?php // $Id$
+<?php
+
+require_once($CFG->dirroot.'/tag/lib.php');
+require_once($CFG->libdir.'/filelib.php');
 
 /**
  * locallib.php - moodle tag local library - output functions
  *
- * @version: $Id$
  * @licence http://www.gnu.org/copyleft/gpl.html GNU Public License
  * @package moodlecore
  *
@@ -16,16 +18,15 @@
  * @param $return if true return html string
  */
 function tag_print_cloud($nr_of_tags=150, $return=false) {
-
-    global $CFG;
+    global $CFG, $DB;
 
     $can_manage_tags = has_capability('moodle/tag:manage', get_context_instance(CONTEXT_SYSTEM));
 
-    if ( !$tagsincloud = get_records_sql('SELECT tg.rawname, tg.id, tg.name, tg.tagtype, COUNT(ti.id) AS count, tg.flag '.
-        'FROM '. $CFG->prefix .'tag_instance ti INNER JOIN '. $CFG->prefix .'tag tg ON tg.id = ti.tagid '.
-        'WHERE ti.itemtype != \'tag\' '.
-        'GROUP BY tg.id, tg.rawname, tg.name, tg.flag, tg.tagtype '.
-        'ORDER BY count DESC, tg.name ASC', 0, $nr_of_tags) ) {
+    if ( !$tagsincloud = $DB->get_records_sql('SELECT tg.rawname, tg.id, tg.name, tg.tagtype, COUNT(ti.id) AS count, tg.flag
+                                                 FROM {tag_instance} ti JOIN {tag} tg ON tg.id = ti.tagid
+                                                WHERE ti.itemtype <> \'tag\'
+                                             GROUP BY tg.id, tg.rawname, tg.name, tg.flag, tg.tagtype
+                                             ORDER BY count DESC, tg.name ASC', null, 0, $nr_of_tags) ) {
         $tagsincloud = array();
     }
 
@@ -93,12 +94,12 @@ function tag_cloud_sort($a, $b) {
 /**
  * Prints a box with the description of a tag and its related tags
  *
- * @param unknown_type $tag_object
+ * @param stdClass $tag_object
  * @param $return if true return html string
  */
 function tag_print_description_box($tag_object, $return=false) {
 
-    global $USER, $CFG;
+    global $USER, $CFG, $OUTPUT;
 
     $max_tags_displayed = 10; // todo: turn this into a system setting
 
@@ -109,12 +110,14 @@ function tag_print_description_box($tag_object, $return=false) {
     $output = '';
 
     if ($content) {
-        $output .= print_box_start('generalbox', 'tag-description', true);
+        $output .= $OUTPUT->box_start('generalbox', 'tag-description');
     }
 
     if (!empty($tag_object->description)) {
-        $options = new object();
+        $options = new stdClass();
         $options->para = false;
+        $options->overflowdiv = true;
+        $tag_object->description = file_rewrite_pluginfile_urls($tag_object->description, 'pluginfile.php', get_context_instance(CONTEXT_SYSTEM)->id, 'tag', 'description', $tag_object->id);
         $output .= format_text($tag_object->description, $tag_object->descriptionformat, $options);
     }
 
@@ -131,7 +134,7 @@ function tag_print_description_box($tag_object, $return=false) {
     }
 
     if ($content) {
-        $output .= print_box_end(true);
+        $output .= $OUTPUT->box_end();
     }
 
     if ($return) {
@@ -149,13 +152,13 @@ function tag_print_description_box($tag_object, $return=false) {
  */
 function tag_print_management_box($tag_object, $return=false) {
 
-    global $USER, $CFG;
+    global $USER, $CFG, $OUTPUT;
 
     $tagname  = tag_display_name($tag_object);
     $output = '';
 
     if (!isguestuser()) {
-        $output .= print_box_start('box','tag-management-box', true);
+        $output .= $OUTPUT->box_start('box','tag-management-box');
         $systemcontext   = get_context_instance(CONTEXT_SYSTEM);
         $links = array();
         
@@ -176,7 +179,7 @@ function tag_print_management_box($tag_object, $return=false) {
         }
 
         $output .= implode(' | ', $links);
-        $output .= print_box_end(true);
+        $output .= $OUTPUT->box_end();
     }
 
     if ($return) {
@@ -192,16 +195,16 @@ function tag_print_management_box($tag_object, $return=false) {
  * @param bool $return if true return html string
  */
 function tag_print_search_box($return=false) {
-    global $CFG;
+    global $CFG, $OUTPUT;
 
-    $output = print_box_start('','tag-search-box', true);
+    $output = $OUTPUT->box_start('','tag-search-box');
     $output .= '<form action="'.$CFG->wwwroot.'/tag/search.php" style="display:inline">';
     $output .= '<div>';
     $output .= '<input id="searchform_search" name="query" type="text" size="40" />';
     $output .= '<button id="searchform_button" type="submit">'. get_string('search', 'tag') .'</button><br />';
     $output .= '</div>';
     $output .= '</form>';
-    $output .= print_box_end(true);
+    $output .= $OUTPUT->box_end();
 
     if ($return) {
         return $output;
@@ -221,7 +224,7 @@ function tag_print_search_box($return=false) {
  */
 function tag_print_search_results($query,  $page, $perpage, $return=false) {
 
-    global $CFG, $USER;
+    global $CFG, $USER, $OUTPUT;
 
     $query = array_shift(tag_normalize($query, TAG_CASE_ORIGINAL));
 
@@ -243,11 +246,11 @@ function tag_print_search_results($query,  $page, $perpage, $return=false) {
     }
 
     if ( !empty($tags) ) { // there are results to display!!
-        $output .= print_heading(get_string('searchresultsfor', 'tag', htmlspecialchars($query)) ." : {$count}", '', 3, 'main', true);
+        $output .= $OUTPUT->heading(get_string('searchresultsfor', 'tag', htmlspecialchars($query)) ." : {$count}", 3, 'main');
 
         //print a link "Add $query to my interests"
         if (!empty($addtaglink)) {
-            $output .= print_box($addtaglink, 'box', 'tag-management-box', true);
+            $output .= $OUTPUT->box($addtaglink, 'box', 'tag-management-box');
         }
 
         $nr_of_lis_per_ul = 6;
@@ -265,14 +268,14 @@ function tag_print_search_results($query,  $page, $perpage, $return=false) {
         $output .= '</ul>';
         $output .= '<div>&nbsp;</div>'; // <-- small layout hack in order to look good in Firefox
 
-        $output .= print_paging_bar($count, $page, $perpage, $baseurl .'&amp;', 'page', false, true);
+        $output .= $OUTPUT->paging_bar($count, $page, $perpage, $baseurl);
     }
     else { //no results were found!!
-        $output .= print_heading(get_string('noresultsfor', 'tag', htmlspecialchars($query)), '', 3, 'main' , true);
+        $output .= $OUTPUT->heading(get_string('noresultsfor', 'tag', htmlspecialchars($query)), 3, 'main');
 
         //print a link "Add $query to my interests"
         if (!empty($addtaglink)) {
-            $output .= print_box($addtaglink, 'box', 'tag-management-box', true);
+            $output .= $OUTPUT->box($addtaglink, 'box', 'tag-management-box');
         }
     }
 
@@ -315,17 +318,17 @@ function tag_print_tagged_users_table($tag_object, $limitfrom='' , $limitnum='',
  * @param $return if true return html string
  */
 function tag_print_user_box($user, $return=false) {
-    global $CFG;
+    global $CFG, $OUTPUT;
 
     $textlib = textlib_get_instance();
     $usercontext = get_context_instance(CONTEXT_USER, $user->id);
     $profilelink = '';
 
-    if ( has_capability('moodle/user:viewdetails', $usercontext) || isteacherinanycourse($user->id) ) {
+    if ( has_capability('moodle/user:viewdetails', $usercontext) || has_coursecontact_role($user->id) ) {
         $profilelink = $CFG->wwwroot .'/user/view.php?id='. $user->id;
     }
 
-    $output = print_box_start('user-box', 'user'. $user->id, true);
+    $output = $OUTPUT->box_start('user-box', 'user'. $user->id);
     $fullname = fullname($user);
     $alt = '';
 
@@ -334,13 +337,7 @@ function tag_print_user_box($user, $return=false) {
         $alt = $fullname;
     }
 
-    //print user image - if image is only content of link it needs ALT text!
-    if ($user->picture) {
-        $output .= '<img alt="'. $alt .'" class="user-image" src="'. $CFG->wwwroot .'/user/pix.php/'. $user->id .'/f1.jpg" />';
-    } else {
-        $output .= '<img alt="'. $alt .'" class="user-image" src="'. $CFG->wwwroot .'/pix/u/f1.png" />';
-    }
-    
+    $output .= $OUTPUT->user_picture($user, array('size'=>100));
     $output .= '<br />';
 
     if (!empty($profilelink)) {
@@ -353,7 +350,7 @@ function tag_print_user_box($user, $return=false) {
     }
 
     $output .= '<strong>'. $fullname .'</strong>';
-    $output .= print_box_end(true);
+    $output .= $OUTPUT->box_end();
 
     if ($return) {
         return $output;
@@ -384,6 +381,3 @@ function tag_print_user_list($userlist, $return=false) {
         echo $output;
     }
 }
-
-
-?>
