diff --git a/tag/user.php b/tag/user.php
index b683b88..20d21b2 100644
--- a/tag/user.php
+++ b/tag/user.php
@@ -1,4 +1,4 @@
-<?php  // $Id$
+<?php
 
 require_once('../config.php');
 require_once('lib.php');
@@ -10,7 +10,7 @@ $tag = optional_param('tag', '', PARAM_TAG);
 require_login();
 
 if (empty($CFG->usetags)) {
-    error('Tags are disabled!');
+    print_error('tagdisabled');
 }
 
 if (isguestuser()) {
@@ -51,8 +51,6 @@ switch ($action) {
         break;
 
     default:
-        error('No action was specified');
+        print_error('unknowaction');
         break;
 }
-
-?>
