diff --git a/filter/tex/displaytex.php b/filter/tex/displaytex.php
index d352f85..ddc6ee6 100644
--- a/filter/tex/displaytex.php
+++ b/filter/tex/displaytex.php
@@ -1,20 +1,55 @@
-<?php // $Id$
-      // This script displays tex source code.
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
 
-    require_once('../../config.php');
+/**
+ * This script displays tex source code, it is used also from the algebra filter.
+ *
+ * @package    filter
+ * @subpackage tex
+ * @copyright  2004 Zbigniew Fiedorowicz fiedorow@math.ohio-state.edu
+ *             Originally based on code provided by Bruno Vernier bruno@vsbeducation.ca
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+
+define('NO_MOODLE_COOKIES', true); // Because it interferes with caching
+
+require('../../config.php');
+
+if (!filter_is_enabled('filter/tex') and !filter_is_enabled('filter/algebra')) {
+    print_error('filternotenabled');
+}
+
+$texexp = optional_param('texexp', '', PARAM_RAW);
+
+$title = get_string('source', 'filter_tex')
 
-    $texexp = urldecode($_SERVER['QUERY_STRING']);
-    // entities are usually encoded twice, first in HTML editor then in tex/filter.php
-    $texexp = html_entity_decode(html_entity_decode($texexp));
-    // encode all entities (saves non-ISO)
-    $texexp = htmlentities($texexp,ENT_COMPAT,'utf-8');
 ?>
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
 <html>
   <head>
-    <title>TeX Source</title>
-    <meta http-equiv="content-type" content="text/html; charset=utf-8" />
+    <title><?php echo $title; ?></title>
+    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
   </head>
-  <body bgcolor="#FFFFFF">
-    <?php echo $texexp; ?>
+  <body>
+    <div>
+      <dl>
+      <dt><?php echo $title; ?>:</dt>
+        <dd><?php p($texexp); ?></dd>
+      </dl>
+    </div>
   </body>
 </html>
\ No newline at end of file
