diff --git a/filter/tex/latex.php b/filter/tex/latex.php
index 0402469..47c7c3b 100644
--- a/filter/tex/latex.php
+++ b/filter/tex/latex.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
     // latex.php
     // render TeX stuff using latex - this will not work on all platforms
     // or configurations. Only works on Linux and Mac with appropriate 
@@ -20,13 +20,7 @@
             
             // construct directory structure
             $this->temp_dir = $CFG->dataroot . "/temp/latex";
-            if (!file_exists("$CFG->dataroot/temp")) {
-                mkdir( "$CFG->dataroot/temp", $CFG->directorypermissions );
-            }
-            if (!file_exists( $this->temp_dir )) {
-                mkdir( $this->temp_dir, $CFG->directorypermissions );
-            }
-
+            make_upload_directory('temp/latex');
         }
 
         /**
@@ -46,7 +40,7 @@
         function construct_latex_document( $formula, $fontsize=12 ) {
             global $CFG;
 
-            $formula = tex_sanitize_formula($formula);
+            $formula = filter_tex_sanitize_formula($formula);
 
             // $fontsize don't affects to formula's size. $density can change size
             $doc =  "\\documentclass[{$fontsize}pt]{article}\n"; 
@@ -160,4 +154,4 @@
     }
 
 
-?>
+
