diff --git a/filter/tex/texdebug.php b/filter/tex/texdebug.php
index bea689a..35f1afb 100644
--- a/filter/tex/texdebug.php
+++ b/filter/tex/texdebug.php
@@ -1,17 +1,35 @@
-<?php // $Id$
-      // This function fetches math. images from the data directory
-      // If not, it obtains the corresponding TeX expression from the cache_tex db table
-      // and uses mimeTeX to create the image file
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * This function fetches math. images from the data directory
+ * If not, it obtains the corresponding TeX expression from the cache_tex db table
+ * and uses mimeTeX to create the image file
+ *
+ * @package    filter
+ * @subpackage tex
+ * @copyright  2004 Zbigniew Fiedorowicz fiedorow@math.ohio-state.edu
+ *             Originally based on code provided by Bruno Vernier bruno@vsbeducation.ca
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
 
     require_once("../../config.php");
 
-    if (empty($CFG->textfilters)) {
-        error ('Filter not enabled!');
-    } else {
-        $filters = explode(',', $CFG->textfilters);
-        if (array_search('filter/tex', $filters) === FALSE) {
-            error ('Filter not enabled!');
-        }
+    if (!filter_is_enabled('filter/tex')) {
+        print_error('filternotenabled');
     }
 
     require_once($CFG->libdir.'/filelib.php');
@@ -24,14 +42,12 @@
     require_login();
     require_capability('moodle/site:config', get_context_instance(CONTEXT_SYSTEM), $USER->id); /// Required cap to run this. MDL-18552
 
-    $query = urldecode($_SERVER['QUERY_STRING']);
-    error_reporting(DEBUG_ALL);
     $output = '';
 
     // look up in cache if required
     if ($action=='ShowDB' or $action=='DeleteDB') {
         $md5 = md5($texexp);
-        $texcache = get_record("cache_filters","filter","tex", "md5key", $md5);
+        $texcache = $DB->get_record("cache_filters", array("filter"=>"tex", "md5key"=>$md5));
     }
 
     // Action: Show DB Entry
@@ -53,7 +69,7 @@
     if ($action=='DeleteDB') {
         if ($texcache) {
             $output = "Deleting DB cache_filters entry for $texexp\n";
-            $result =  delete_records("cache_filters","id",$texcache->id);
+            $result =  $DB->delete_records("cache_filters", array("id"=>$texcache->id));
             if ($result) {
                 $result = 1;
             } else {
@@ -102,13 +118,10 @@
 
 
     function outputText($texexp) {
-        header("Content-type: text/html");
+        header("Content-type: text/html; charset=utf-8");
         echo "<html><body><pre>\n";
         if ($texexp) {
-            $texexp = str_replace('<', '&lt;', $texexp);
-            $texexp = str_replace('>', '&gt;', $texexp);
-            $texexp = str_replace('"', '&quot;', $texexp);
-            echo "$texexp\n\n";
+            echo s($texexp)."\n\n";
         } else {
             echo "No text output available\n\n";
         }
@@ -123,7 +136,6 @@
             return;
         }
 
-        $texexp = stripslashes($texexp);
         $image  = md5($texexp) . ".gif";
         $filetype = 'image/gif';
         if (!file_exists("$CFG->dataroot/filter/tex")) {
@@ -135,8 +147,8 @@
         }
 
         $texexp = '\Large '.$texexp;
-        $commandpath = tex_filter_get_executable(true);
-        $cmd = tex_filter_get_cmd($pathname, $texexp);
+        $commandpath = filter_tex_get_executable(true);
+        $cmd = filter_tex_get_cmd($pathname, $texexp);
         system($cmd, $status);
 
         if ($return) {
@@ -221,7 +233,6 @@
         $img = "$latex->temp_dir/$md5.{$CFG->filter_tex_convertformat}";
 
         // put the expression as a file into the temp area
-        $expression = stripslashes($expression);
         $expression = html_entity_decode($expression);
         $output .= "<p>Processing TeX expression:</p><pre>$expression</pre>\n";
         $doc = $latex->construct_latex_document($expression);
@@ -245,9 +256,11 @@
         $output .= execute($cmd);
 
         if (!$graphic) {
-            echo($output);
-        } else {
+            echo $output;
+        } else if (file_exists($img)){
             send_file($img, "$md5.{$CFG->filter_tex_convertformat}");
+        } else {
+            echo "Error creating image, see command execution output for more details.";
          }
     }
 
