diff --git a/grade/import/grade_import_form.php b/grade/import/grade_import_form.php
old mode 100755
new mode 100644
index 3d74664..84629f8
--- a/grade/import/grade_import_form.php
+++ b/grade/import/grade_import_form.php
@@ -35,12 +35,11 @@ class grade_import_form extends moodleform {
         }
 
         // course id needs to be passed for auth purposes
-        $mform->addElement('hidden', 'id', optional_param('id'));
+        $mform->addElement('hidden', 'id', optional_param('id', 0, PARAM_INT));
         $mform->setType('id', PARAM_INT);
         $mform->addElement('header', 'general', get_string('importfile', 'grades'));
         // file upload
-        $mform->addElement('file', 'userfile', get_string('file'));
-        $mform->setType('userfile', PARAM_FILE);
+        $mform->addElement('filepicker', 'userfile', get_string('file'));
         $mform->addRule('userfile', null, 'required');
         $textlib = textlib_get_instance();
         $encodings = $textlib->get_encodings();
@@ -87,18 +86,16 @@ class grade_import_mapping_form extends moodleform {
             }
         }
         $mform->addElement('select', 'mapfrom', get_string('mapfrom', 'grades'), $mapfromoptions);
-        //choose_from_menu($mapfromoptions, 'mapfrom');
 
         $maptooptions = array('userid'=>'userid', 'username'=>'username', 'useridnumber'=>'useridnumber', 'useremail'=>'useremail', '0'=>'ignore');
-        //choose_from_menu($maptooptions, 'mapto');
         $mform->addElement('select', 'mapto', get_string('mapto', 'grades'), $maptooptions);
 
         $mform->addElement('header', 'general', get_string('mappings', 'grades'));
 
         // add a comment option
 
-        if ($gradeitems = $this->_customdata['gradeitems']) {
             $comments = array();
+        if ($gradeitems = $this->_customdata['gradeitems']) {
             foreach ($gradeitems as $itemid => $itemname) {
                 $comments['feedback_'.$itemid] = 'comments for '.$itemname;
             }
@@ -107,11 +104,9 @@ class grade_import_mapping_form extends moodleform {
         if ($header) {
             $i = 0; // index
             foreach ($header as $h) {
-
                 $h = trim($h);
                 // this is what each header maps to
-                $mform->addElement('selectgroups',
-                                   'mapping_'.$i, s($h),
+                $mform->addElement('selectgroups', 'mapping_'.$i, s($h),
                                    array('others'=>array('0'=>'ignore', 'new'=>'new gradeitem'),
                                          'gradeitems'=>$gradeitems,
                                          'comments'=>$comments));
@@ -136,4 +131,3 @@ class grade_import_mapping_form extends moodleform {
 
     }
 }
-?>
