diff --git a/mod/survey/mod_form.php b/mod/survey/mod_form.php
index 48242cc..d12ad22 100644
--- a/mod/survey/mod_form.php
+++ b/mod/survey/mod_form.php
@@ -1,11 +1,15 @@
 <?php
+if (!defined('MOODLE_INTERNAL')) {
+    die('Direct access to this script is forbidden.');    ///  It must be included from a Moodle page
+}
+
 require_once ($CFG->dirroot.'/course/moodleform_mod.php');
 
 class mod_survey_mod_form extends moodleform_mod {
 
     function definition() {
+        global $CFG, $DB;
 
-        global $CFG;
         $mform =& $this->_form;
 
         $strrequired = get_string('required');
@@ -17,12 +21,12 @@ class mod_survey_mod_form extends moodleform_mod {
         if (!empty($CFG->formatstringstriptags)) {
             $mform->setType('name', PARAM_TEXT);
         } else {
-            $mform->setType('name', PARAM_CLEAN);
+            $mform->setType('name', PARAM_CLEANHTML);
         }
         $mform->addRule('name', null, 'required', null, 'client');
 
-        if (!$options = get_records_menu("survey", "template", 0, "name", "id, name")) {
-            error('No survey templates found!');
+        if (!$options = $DB->get_records_menu("survey", array("template"=>0), "name", "id, name")) {
+            print_error('cannotfindsurveytmpt', 'survey');
         }
 
         foreach ($options as $id => $name) {
@@ -31,17 +35,11 @@ class mod_survey_mod_form extends moodleform_mod {
         $options = array(''=>get_string('choose').'...') + $options;
         $mform->addElement('select', 'template', get_string("surveytype", "survey"), $options);
         $mform->addRule('template', $strrequired, 'required', null, 'client');
-        $mform->setHelpButton('template', array('surveys', get_string('helpsurveys', 'survey')));
+        $mform->addHelpButton('template', 'surveytype', 'survey');
 
+        $this->add_intro_editor(false, get_string('customintro', 'survey'));
 
-        $mform->addElement('textarea', 'intro', get_string('customintro', 'survey'), 'wrap="virtual" rows="20" cols="75"');
-        $mform->setType('intro', PARAM_RAW);
-
-        $features = new stdClass;
-        $features->groups = true;
-        $features->groupings = true;
-        $features->groupmembersonly = true;
-        $this->standard_coursemodule_elements($features);
+        $this->standard_coursemodule_elements();
 
 //-------------------------------------------------------------------------------
         // buttons
@@ -50,4 +48,4 @@ class mod_survey_mod_form extends moodleform_mod {
 
 
 }
-?>
\ No newline at end of file
+
