diff --git a/auth/shibboleth/index.php b/auth/shibboleth/index.php
index 982084f..9f1e9f2 100644
--- a/auth/shibboleth/index.php
+++ b/auth/shibboleth/index.php
@@ -1,9 +1,12 @@
-<?php // $Id$
+<?php
+
       // Designed to be redirected from moodle/login/index.php
 
     require('../../config.php');
 
-    if (isloggedin() && $USER->username != 'guest') {      // Nothing to do
+    $PAGE->set_url('/auth/shibboleth/index.php');
+
+    if (isloggedin() && !isguestuser()) {      // Nothing to do
         if (isset($SESSION->wantsurl) and (strpos($SESSION->wantsurl, $CFG->wwwroot) === 0)) {
             $urltogo = $SESSION->wantsurl;    /// Because it's an address in this site
             unset($SESSION->wantsurl);
@@ -43,8 +46,8 @@
             
             update_user_login_times();
             
-            // Don't show username on login page
-            set_moodle_cookie('nobody');
+            // Don't show previous shibboleth username on login page
+            set_moodle_cookie('');
 
             set_login_session_preferences();
             
@@ -66,14 +69,14 @@
                 unset($SESSION->wantsurl);         /// Just in case
             }
 
-            /// Go to my-moodle page instead of homepage if mymoodleredirect enabled
-            if (!has_capability('moodle/site:config',get_context_instance(CONTEXT_SYSTEM)) and !empty($CFG->mymoodleredirect) and !isguest()) {
+            /// Go to my-moodle page instead of homepage if defaulthomepage enabled
+            if (!has_capability('moodle/site:config',get_context_instance(CONTEXT_SYSTEM)) and !empty($CFG->defaulthomepage) && $CFG->defaulthomepage == HOMEPAGE_MY and !isguestuser()) {
                 if ($urltogo == $CFG->wwwroot or $urltogo == $CFG->wwwroot.'/' or $urltogo == $CFG->wwwroot.'/index.php') {
                     $urltogo = $CFG->wwwroot.'/my/';
                 }
             }
 
-            check_enrolment_plugins($USER);
+            enrol_check_plugins($USER);
             load_all_capabilities();     /// This is what lets the user do anything on the site  :-)
 
             redirect($urltogo);
@@ -94,4 +97,4 @@
         print_error('shib_not_set_up_error', 'auth');
     }
 
-?>
+
