diff --git a/search/Zend/Search/Lucene/Storage/File/Filesystem.php b/search/Zend/Search/Lucene/Storage/File/Filesystem.php
index 496551d..d508ce9 100644
--- a/search/Zend/Search/Lucene/Storage/File/Filesystem.php
+++ b/search/Zend/Search/Lucene/Storage/File/Filesystem.php
@@ -220,4 +220,3 @@ class Zend_Search_Lucene_Storage_File_Filesystem extends Zend_Search_Lucene_Stor
     }
 }
 
-?>
\ No newline at end of file
