diff --git a/blocks/mentees/block_mentees.php b/blocks/mentees/block_mentees.php
old mode 100755
new mode 100644
index 6af6fc8..45d9463
--- a/blocks/mentees/block_mentees.php
+++ b/blocks/mentees/block_mentees.php
@@ -3,8 +3,7 @@
 class block_mentees extends block_base {
 
     function init() {
-        $this->title = get_string('blockname', 'block_mentees');
-        $this->version = 2007101509;
+        $this->title = get_string('pluginname', 'block_mentees');
     }
 
     function applicable_formats() {
@@ -20,21 +19,19 @@ class block_mentees extends block_base {
     }
 
     function get_content() {
+        global $CFG, $USER, $DB;
         
-        global $CFG, $USER;
         if ($this->content !== NULL) {
             return $this->content;
         }
 
         // get all the mentees, i.e. users you have a direct assignment to
-        if ($usercontexts = get_records_sql("SELECT c.instanceid, c.instanceid, u.firstname, u.lastname
-                                         FROM {$CFG->prefix}role_assignments ra,
-                                              {$CFG->prefix}context c,
-                                              {$CFG->prefix}user u
-                                         WHERE ra.userid = $USER->id
+        if ($usercontexts = $DB->get_records_sql("SELECT c.instanceid, c.instanceid, u.firstname, u.lastname
+                                                    FROM {role_assignments} ra, {context} c, {user} u
+                                                   WHERE ra.userid = ?
                                          AND   ra.contextid = c.id
                                          AND   c.instanceid = u.id
-                                         AND   c.contextlevel = ".CONTEXT_USER)) {
+                                                         AND c.contextlevel = ".CONTEXT_USER, array($USER->id))) {
                                            
             $this->content->text = '<ul>';
             foreach ($usercontexts as $usercontext) {
@@ -43,10 +40,9 @@ class block_mentees extends block_base {
             $this->content->text .= '</ul>';                    
         }
                                 
-        
         $this->content->footer = '';
 
         return $this->content;
     }
 }
-?>
+
