diff --git a/mod/assignment/type/online/file.php b/mod/assignment/type/online/file.php
index 2a61336..c7164f2 100644
--- a/mod/assignment/type/online/file.php
+++ b/mod/assignment/type/online/file.php
@@ -1,44 +1,48 @@
-<?php  // $Id$
+<?php
 
-    require("../../../../config.php");
-    require("../../lib.php");
-    require("assignment.class.php");
+require("../../../../config.php");
+require("../../lib.php");
+require("assignment.class.php");
  
-    $id     = required_param('id', PARAM_INT);      // Course Module ID
-    $userid = required_param('userid', PARAM_INT);  // User ID
+$id     = required_param('id', PARAM_INT);      // Course Module ID
+$userid = required_param('userid', PARAM_INT);  // User ID
 
-    if (! $cm = get_coursemodule_from_id('assignment', $id)) {
-        error("Course Module ID was incorrect");
-    }
+$PAGE->set_url('/mod/assignment/type/online/file.php', array('id'=>$id, 'userid'=>$userid));
 
-    if (! $assignment = get_record("assignment", "id", $cm->instance)) {
-        error("Assignment ID was incorrect");
-    }
+if (! $cm = get_coursemodule_from_id('assignment', $id)) {
+    print_error('invalidcoursemodule');
+}
 
-    if (! $course = get_record("course", "id", $assignment->course)) {
-        error("Course is misconfigured");
-    }
+if (! $assignment = $DB->get_record("assignment", array("id"=>$cm->instance))) {
+    print_error('invalidid', 'assignment');
+}
 
-    if (! $user = get_record("user", "id", $userid)) {
-        error("User is misconfigured");
-    }
+if (! $course = $DB->get_record("course", array("id"=>$assignment->course))) {
+    print_error('coursemisconf', 'assignment');
+}
 
-    require_login($course->id, false, $cm);
+if (! $user = $DB->get_record("user", array("id"=>$userid))) {
+    print_error('usermisconf', 'assignment');
+}
 
-    if (($USER->id != $user->id) && !has_capability('mod/assignment:grade', get_context_instance(CONTEXT_MODULE, $cm->id))) {
-        error("You can not view this assignment");
-    }
+require_login($course->id, false, $cm);
 
-    if ($assignment->assignmenttype != 'online') {
-        error("Incorrect assignment type");
-    }
+$context = get_context_instance(CONTEXT_MODULE, $cm->id);
+if (($USER->id != $user->id) && !has_capability('mod/assignment:grade', $context)) {
+    print_error('cannotviewassignment', 'assignment');
+}
 
-    $assignmentinstance = new assignment_online($cm->id, $assignment, $cm, $course);
+if ($assignment->assignmenttype != 'online') {
+    print_error('invalidtype', 'assignment');
+}
 
-    if ($submission = $assignmentinstance->get_submission($user->id)) {
-        print_header(fullname($user,true).': '.$assignment->name);
+$assignmentinstance = new assignment_online($cm->id, $assignment, $cm, $course);
 
-        print_simple_box_start('center', '', '', '', 'generalbox', 'dates');
+if ($submission = $assignmentinstance->get_submission($user->id)) {
+    $PAGE->set_pagelayout('popup');
+    $PAGE->set_title(fullname($user,true).': '.$assignment->name);
+    echo $OUTPUT->header();
+    echo $OUTPUT->box_start('generalbox boxaligcenter', 'dates');
         echo '<table>';
         if ($assignment->timedue) {
             echo '<tr><td class="c0">'.get_string('duedate','assignment').':</td>';
@@ -53,13 +57,12 @@
                 echo ' ('.get_string('numletters', '', count_letters(format_text($submission->data1, $submission->data2))).')</td></tr>';
             }
         echo '</table>';
-        print_simple_box_end();
+    echo $OUTPUT->box_end();
 
-        print_simple_box(format_text($submission->data1, $submission->data2), 'center', '100%');
-        close_window_button();
-        print_footer('none');
-    } else {
+    $text = file_rewrite_pluginfile_urls($submission->data1, 'pluginfile.php', $context->id, 'mod_assignment', $assignmentinstance->filearea, $submission->id);
+    echo $OUTPUT->box(format_text($text, $submission->data2, array('overflowdiv'=>true)), 'generalbox boxaligncenter boxwidthwide');
+    echo $OUTPUT->close_window_button();
+    echo $OUTPUT->footer();
+} else {
         print_string('emptysubmission', 'assignment');
-    }
-
-?>
+}
\ No newline at end of file
