diff --git a/mod/assignment/type/online/version.php b/mod/assignment/type/online/version.php
index ceba567..e43ffab 100644
--- a/mod/assignment/type/online/version.php
+++ b/mod/assignment/type/online/version.php
@@ -1,11 +1,11 @@
-<?PHP // $Id$
+<?php
 
 ////////////////////////////////////////////////////////////////////////////////
 //  Code fragment to define the module version etc.
 //  This fragment is called by /admin/index.php
 ////////////////////////////////////////////////////////////////////////////////
 
-$plugin->version   = 2005042900;
-$plugin->requires  = 2007101000;
+$plugin->version   = 2010090501;
+$plugin->requires  = 2010090501;
+
 
-?>
