diff --git a/question/format/examview/format.php b/question/format/examview/format.php
old mode 100755
new mode 100644
index 32bf0e5..53bb360
--- a/question/format/examview/format.php
+++ b/question/format/examview/format.php
@@ -1,16 +1,43 @@
-<?php // $Id$
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
 
-// Based on default.php, included by ../import.php
 /**
- * @package questionbank
- * @subpackage importexport
+ * Examview question importer.
+ *
+ * @package    qformat
+ * @subpackage examview
+ * @copyright  2005 Howard Miller
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
 
-require_once("$CFG->libdir/xmlize.php");
 
+defined('MOODLE_INTERNAL') || die();
+
+require_once ($CFG->libdir . '/xmlize.php');
+
+
+/**
+ * Examview question importer.
+ *
+ * @copyright  2005 Howard Miller
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
 class qformat_examview extends qformat_default {
     
-    var $qtypes = array(
+    public $qtypes = array(
         'tf' => TRUEFALSE,
         'mc' => MULTICHOICE,
         'yn' => TRUEFALSE,
@@ -25,7 +52,7 @@ class qformat_examview extends qformat_default {
         'sa' => ESSAY
         );
     
-    var $matching_questions = array();
+    public $matching_questions = array();
 
     function provide_import() {
         return true;
@@ -40,7 +67,7 @@ class qformat_examview extends qformat_default {
     function unxmlise( $xml ) {
         // if it's not an array then it's probably just data
         if (!is_array($xml)) {
-            $text = s(addslashes($xml));
+            $text = s($xml);
         }
         else {
             // otherwise parse the array
@@ -101,17 +128,17 @@ class qformat_examview extends qformat_default {
         }
         foreach($this->matching_questions as $match_group) {
             $question = $this->defaultquestion();
-            $htmltext = s(addslashes($match_group->questiontext));
+            $htmltext = s($match_group->questiontext);
             $question->questiontext = $htmltext;
             $question->name = $question->questiontext;
             $question->qtype = MATCH;
             $question->subquestions = array();
             $question->subanswers = array();
             foreach($match_group->subquestions as $key => $value) {
-                $htmltext = s(addslashes($value));
+                $htmltext = s($value);
                 $question->subquestions[] = $htmltext;
 
-                $htmltext = s(addslashes($match_group->subanswers[$key]));
+                $htmltext = s($match_group->subanswers[$key]);
                 $question->subanswers[] = $htmltext;
             }
             $questions[] = $question;
@@ -245,7 +272,7 @@ class qformat_examview extends qformat_default {
         foreach($answers as $key => $value) {
             $value = trim($value);
             if (strlen($value) > 0) {
-                $question->answer[$key] = addslashes($value);
+                $question->answer[$key] = $value;
                 $question->fraction[$key] = 1;
                 $question->feedback[$key] = "Correct";
             }
@@ -283,4 +310,4 @@ class qformat_examview extends qformat_default {
 }
 // end class
 
-?>
+
