diff --git a/lib/pear/Console/Getopt.php b/lib/pear/Console/Getopt.php
index 8da9e7e..b256d0e 100644
--- a/lib/pear/Console/Getopt.php
+++ b/lib/pear/Console/Getopt.php
@@ -16,7 +16,7 @@
 // | Author: Andrei Zmievski <andrei@php.net>                             |
 // +----------------------------------------------------------------------+
 //
-// $Id: Getopt.php,v 1.1.2.2 2008/02/27 02:58:24 martinlanghoff Exp $
+// $Id$
 
 require_once 'PEAR.php';
 
