diff --git a/enrol/paypal/version.php b/enrol/paypal/version.php
index 6be8fd7..9a5ce19 100644
--- a/enrol/paypal/version.php
+++ b/enrol/paypal/version.php
@@ -1,12 +1,30 @@
-<?PHP // $Id$
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
 
-/////////////////////////////////////////////////////////////////////////////////
-///  Code fragment to define the version of this enrolment module
-///  This fragment is called by admin/index.php
-/////////////////////////////////////////////////////////////////////////////////
+/**
+ * Paypal enrolment plugin version specification.
+ *
+ * @package    enrol
+ * @subpackage paypal
+ * @copyright  2010 Eugene Venter
+ * @author     Eugene Venter
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
 
-$plugin->version  = 2006092200;   // This module's version
+defined('MOODLE_INTERNAL') || die();
 
-$plugin->requires = 2007101000;   // Requires this Moodle version
-
-?>
+$plugin->version  = 2010073100;
+$plugin->requires = 2010090501;   // Requires this Moodle version
