diff --git a/backup/moodle2/backup_plan_builder.class.php b/backup/moodle2/backup_plan_builder.class.php
index a9240fc..fa70a86 100644
--- a/backup/moodle2/backup_plan_builder.class.php
+++ b/backup/moodle2/backup_plan_builder.class.php
@@ -166,7 +166,7 @@ abstract class backup_plan_builder {
     /**
      * Build one 1-course backup
      */
-    static protected function build_course_plan($controller, $id) {
+    static public function build_course_plan($controller, $id) {
 
         $plan = $controller->get_plan();
 
