diff --git a/blocks/course_list/settings.php b/blocks/course_list/settings.php
index 899af4a..1af7620 100644
--- a/blocks/course_list/settings.php
+++ b/blocks/course_list/settings.php
@@ -1,13 +1,15 @@
-<?php  //$Id$
+<?php
 
+defined('MOODLE_INTERNAL') || die;
 
-$options = array('all'=>get_string('allcourses', 'block_course_list'), 'own'=>get_string('owncourses', 'block_course_list'));
+if ($ADMIN->fulltree) {
+    $options = array('all'=>get_string('allcourses', 'block_course_list'), 'own'=>get_string('owncourses', 'block_course_list'));
 
-$settings->add(new admin_setting_configselect('block_course_list_adminview', get_string('adminview', 'block_course_list'),
+    $settings->add(new admin_setting_configselect('block_course_list_adminview', get_string('adminview', 'block_course_list'),
                    get_string('configadminview', 'block_course_list'), 'all', $options));
 
-$settings->add(new admin_setting_configcheckbox('block_course_list_hideallcourseslink', get_string('hideallcourseslink', 'block_course_list'),
+    $settings->add(new admin_setting_configcheckbox('block_course_list_hideallcourseslink', get_string('hideallcourseslink', 'block_course_list'),
                    get_string('confighideallcourseslink', 'block_course_list'), 0));
+}
 
 
-?>
