diff --git a/admin/xmldb/actions/edit_field_save/edit_field_save.class.php b/admin/xmldb/actions/edit_field_save/edit_field_save.class.php
index 41c2a8d..b762063 100644
--- a/admin/xmldb/actions/edit_field_save/edit_field_save.class.php
+++ b/admin/xmldb/actions/edit_field_save/edit_field_save.class.php
@@ -1,31 +1,34 @@
-<?php // $Id$
+<?php
 
-///////////////////////////////////////////////////////////////////////////
-//                                                                       //
-// NOTICE OF COPYRIGHT                                                   //
-//                                                                       //
-// Moodle - Modular Object-Oriented Dynamic Learning Environment         //
-//          http://moodle.com                                            //
-//                                                                       //
-// Copyright (C) 1999 onwards Martin Dougiamas        http://dougiamas.com  //
-//           (C) 2001-3001 Eloy Lafuente (stronk7) http://contiento.com  //
-//                                                                       //
-// This program is free software; you can redistribute it and/or modify  //
-// it under the terms of the GNU General Public License as published by  //
-// the Free Software Foundation; either version 2 of the License, or     //
-// (at your option) any later version.                                   //
-//                                                                       //
-// This program is distributed in the hope that it will be useful,       //
-// but WITHOUT ANY WARRANTY; without even the implied warranty of        //
-// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         //
-// GNU General Public License for more details:                          //
-//                                                                       //
-//          http://www.gnu.org/copyleft/gpl.html                         //
-//                                                                       //
-///////////////////////////////////////////////////////////////////////////
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
 
-/// This class will save the changes performed to one field
+/**
+ * @package   xmldb-editor
+ * @copyright 2003 onwards Eloy Lafuente (stronk7) {@link http://stronk7.com}
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
 
+/**
+ * This class verifies all the data introduced when editing a field for correctness,
+ * performing changes / displaying errors depending of the results.
+ *
+ * @package   xmldb-editor
+ * @copyright 2003 onwards Eloy Lafuente (stronk7) {@link http://stronk7.com}
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
 class edit_field_save extends XMLDBAction {
 
     /**
@@ -49,9 +52,8 @@ class edit_field_save extends XMLDBAction {
             'binaryincorrectlength' => 'xmldb',
             'numberincorrectdecimals' => 'xmldb',
             'floatincorrectdecimals' => 'xmldb',
-            'enumvaluesincorrect' => 'xmldb',
-            'wronglengthforenum' => 'xmldb',
             'defaultincorrect' => 'xmldb',
+            'back' => 'xmldb',
             'administration' => ''
         ));
     }
@@ -67,28 +69,28 @@ class edit_field_save extends XMLDBAction {
         $result = true;
 
     /// Set own core attributes
-        $this->does_generate = ACTION_NONE;
-        //$this->does_generate = ACTION_GENERATE_HTML;
+        //$this->does_generate = ACTION_NONE;
+        $this->does_generate = ACTION_GENERATE_HTML;
 
     /// These are always here
         global $CFG, $XMLDB;
 
     /// Do the job, setting result as needed
 
-        if (!data_submitted('nomatch')) { ///Basic prevention
-            error('Wrong action call');
+        if (!data_submitted()) { ///Basic prevention
+            print_error('wrongcall', 'error');
         }
 
     /// Get parameters
         $dirpath = required_param('dir', PARAM_PATH);
-        $dirpath = $CFG->dirroot . stripslashes_safe($dirpath);
+        $dirpath = $CFG->dirroot . $dirpath;
 
         $tableparam = strtolower(required_param('table', PARAM_PATH));
         $fieldparam = strtolower(required_param('field', PARAM_PATH));
         $name = substr(trim(strtolower(optional_param('name', $fieldparam, PARAM_PATH))),0,30);
 
         $comment = required_param('comment', PARAM_CLEAN);
-        $comment = trim(stripslashes_safe($comment));
+        $comment = trim($comment);
 
         $type       = required_param('type', PARAM_INT);
         $length     = strtolower(optional_param('length', NULL, PARAM_ALPHANUM));
@@ -96,11 +98,8 @@ class edit_field_save extends XMLDBAction {
         $unsigned   = optional_param('unsigned', false, PARAM_BOOL);
         $notnull    = optional_param('notnull', false, PARAM_BOOL);
         $sequence   = optional_param('sequence', false, PARAM_BOOL);
-        $enum       = optional_param('enum', false, PARAM_BOOL);
-        $enumvalues = optional_param('enumvalues', 0, PARAM_CLEAN);
-        $enumvalues = trim(stripslashes_safe($enumvalues));
         $default    = optional_param('default', NULL, PARAM_PATH);
-        $default    = trim(stripslashes_safe($default));
+        $default    = trim($default);
 
         $editeddir =& $XMLDB->editeddirs[$dirpath];
         $structure =& $editeddir->xml_file->getStructure();
@@ -110,25 +109,18 @@ class edit_field_save extends XMLDBAction {
 
         $errors = array();    /// To store all the errors found
 
-    /// Perform some automatic asumptions
+    /// Perform some automatic assumptions
         if ($sequence) {
             $unsigned = true;
             $notnull  = true;
-            $enum     = false;
             $default  = NULL;
         }
         if ($type != XMLDB_TYPE_NUMBER && $type != XMLDB_TYPE_FLOAT) {
             $decimals = NULL;
         }
-        if ($type != XMLDB_TYPE_CHAR && $type != XMLDB_TYPE_TEXT) {
-            $enum = false;
-        }
         if ($type == XMLDB_TYPE_BINARY) {
             $default = NULL;
         }
-        if (!$enum) {
-            $enumvalues = NULL;
-        }
         if ($default === '') {
             $default = NULL;
         }
@@ -232,67 +224,22 @@ class edit_field_save extends XMLDBAction {
                 $errors[] = $this->str['binaryincorrectlength'];
             }
         }
-    /// Enum checks
-        if ($enum) {
-            $enumerr = false;
-            $enumarr = explode(',',$enumvalues);
-            $maxlength = 0;
-            if ($enumarr) {
-                foreach ($enumarr as $key => $enumelement) {
-                /// Clear some spaces
-                    $enumarr[$key] = trim($enumelement);
-                    $enumelement = trim($enumelement);
-                /// Calculate needed length
-                    $le = strlen(str_replace("'", '', $enumelement));
-                    if ($le > $maxlength) {
-                        $maxlength = $le;
-                    }
-                /// Skip if under error
-                    if ($enumerr) {
-                        continue;
-                    }
-                /// Look for quoted strings
-                    if (substr($enumelement, 0, 1) != "'" ||
-                        substr($enumelement, -1, 1) != "'") {
-                        $enumerr = true;
-                    }
-                }
-            } else {
-                $enumerr = true;
-            }
-            if ($enumerr) {
-                $errors[] = $this->str['enumvaluesincorrect'];
-            } else {
-                $enumvalues = $enumarr;
-            }
-            if ($length < $maxlength) {
-                $errors[] = $this->str['wronglengthforenum'];
-            }
-        }
 
         if (!empty($errors)) {
-            $tempfield = new XMLDBField($name);
+            $tempfield = new xmldb_field($name);
             $tempfield->setType($type);
             $tempfield->setLength($length);
             $tempfield->setDecimals($decimals);
             $tempfield->setUnsigned($unsigned);
             $tempfield->setNotNull($notnull);
             $tempfield->setSequence($sequence);
-            $tempfield->setEnum($enum);
-            $tempfield->setEnumValues($enumvalues);
             $tempfield->setDefault($default);
         /// Prepare the output
-            $site = get_site();
-            $navlinks = array();
-            $navlinks[] = array('name' => $this->str['administration'], 'link' => '../index.php', 'type' => 'misc');
-            $navlinks[] = array('name' => 'XMLDB', 'link' => 'index.php', 'type' => 'misc');
-            $navigation = build_navigation($navlinks);
-            print_header("$site->shortname: XMLDB", "$site->fullname", $navigation);
-            notice ('<p>' .implode(', ', $errors) . '</p>
-                     <p>' . $tempfield->readableInfo(),
-                    'index.php?action=edit_field&amp;field=' .$field->getName() . '&amp;table=' . $table->getName()
-                    . '&amp;dir=' . urlencode(str_replace($CFG->dirroot, '', $dirpath)));
-            die; /// re-die :-P
+            $o = '<p>' .implode(', ', $errors) . '</p>
+                  <p>' . $name . ': ' . $tempfield->readableInfo() . '</p>';
+            $o.= '<a href="index.php?action=edit_field&amp;field=' . $field->getName() . '&amp;table=' . $table->getName() .
+                 '&amp;dir=' . urlencode(str_replace($CFG->dirroot, '', $dirpath)) . '">[' . $this->str['back'] . ']</a>';
+            $this->output = $o;
         }
 
     /// Continue if we aren't under errors
@@ -323,8 +270,6 @@ class edit_field_save extends XMLDBAction {
             $field->setUnsigned($unsigned);
             $field->setNotNull($notnull);
             $field->setSequence($sequence);
-            $field->setEnum($enum);
-            $field->setEnumValues($enumvalues);
             $field->setDefault($default);
 
         /// If the hash has changed from the old one, change the version
@@ -350,4 +295,4 @@ class edit_field_save extends XMLDBAction {
         return $result;
     }
 }
-?>
+
