diff --git a/lang/en/plugin.php b/lang/en/plugin.php
index 235dcf4..d3a1571 100644
--- a/lang/en/plugin.php
+++ b/lang/en/plugin.php
@@ -63,6 +63,7 @@ $string['plugindisable'] = 'Disable';
 $string['plugindisabled'] = 'Disabled';
 $string['pluginenable'] = 'Enable';
 $string['pluginenabled'] = 'Enabled';
+$string['release'] = 'Release';
 $string['requiredby'] = 'Required by: {$a}';
 $string['requires'] = 'Requires';
 $string['rootdir'] = 'Directory';
@@ -94,6 +95,8 @@ $string['type_coursereport'] = 'Course report';
 $string['type_coursereport_plural'] = 'Course reports';
 $string['type_editor'] = 'Editor';
 $string['type_editor_plural'] = 'Editors';
+$string['type_elis'] = 'ELIS';
+$string['type_elis_plural'] = 'ELIS';
 $string['type_enrol'] = 'Enrolment method';
 $string['type_enrol_plural'] = 'Enrolment methods';
 $string['type_filter'] = 'Text filter';
