diff --git a/lang/en/search.php b/lang/en/search.php
index 4e2f25a..4510cbe 100644
--- a/lang/en/search.php
+++ b/lang/en/search.php
@@ -76,5 +76,5 @@ $string['totalsize'] = 'Total size';
 $string['type'] = 'Type';
 $string['uncompleteindexingerror'] = 'Indexing was not successfully completed, please restart it.';
 $string['versiontoolow'] = 'Sorry, global search requires PHP 5.0.0 or later';
-$string['whichmodulestosearch?'] = 'Which modules to search?';
+$string['whichmodulestosearch'] = 'Which modules to search?';
 $string['wordsintitle'] = 'Words in title';
