diff --git a/course/report/outline/index.php b/course/report/outline/index.php
index dbe273e..dc1dd10 100644
--- a/course/report/outline/index.php
+++ b/course/report/outline/index.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 
 // Display user activity reports for a course (totals)
 
@@ -7,10 +7,13 @@
 
     $id = required_param('id',PARAM_INT);       // course id
 
-    if (!$course = get_record('course', 'id', $id)) {
-        error('Course id is incorrect.');
+    if (!$course = $DB->get_record('course', array('id'=>$id))) {
+        print_error('invalidcourseid');
     }
 
+    $PAGE->set_url('/course/report/outline/index.php', array('id'=>$id));
+    $PAGE->set_pagelayout('report');
+
     require_login($course);
     $context = get_context_instance(CONTEXT_COURSE, $course->id);
     require_capability('coursereport/outline:view', $context);
@@ -29,89 +32,126 @@
     $strlast           = get_string('lastaccess');
     $strreports        = get_string('reports');
     $strviews          = get_string('views');
+    $strrelatedblogentries = get_string('relatedblogentries', 'blog');
 
-    $navlinks = array();
-    $navlinks[] = array('name' => $strreports, 'link' => "../../report.php?id=$course->id", 'type' => 'misc');
-    $navlinks[] = array('name' => $stractivityreport, 'link' => null, 'type' => 'misc');
-    $navigation = build_navigation($navlinks);
+    $PAGE->set_title($course->shortname .': '. $stractivityreport);
+    $PAGE->set_heading($course->fullname);
+    echo $OUTPUT->header();
+    echo $OUTPUT->heading(format_string($course->fullname));
 
-    print_header("$course->shortname: $stractivityreport", $course->fullname, $navigation);
+    if (!$logstart = $DB->get_field_sql("SELECT MIN(time) FROM {log}")) {
+        print_error('logfilenotavailable');
+    }
 
-    print_heading(format_string($course->fullname));
+    echo $OUTPUT->container(get_string('computedfromlogs', 'admin', userdate($logstart)), 'loginfo');
 
-    if (!$logstart = get_field_sql("SELECT MIN(time) FROM {$CFG->prefix}log")) {
-        error('Logs not available');
-    }
+    $outlinetable = new html_table();
+    $outlinetable->attributes['class'] = 'generaltable boxaligncenter';
+    $outlinetable->cellpadding = 5;
+    $outlinetable->id = 'outlinetable';
+    $outlinetable->head = array($stractivity, $strviews);
 
-    echo '<div class="loginfo">'.get_string('computedfromlogs', 'admin', userdate($logstart)).'</div>';
+    if ($CFG->useblogassociations) {
+        $outlinetable->head[] = $strrelatedblogentries;
+    }
 
-    echo '<table id="outlinetable" class="generaltable boxaligncenter" cellpadding="5"><tr>';
-    echo '<th class="header c0" scope="col">'.$stractivity.'</th>';
-    echo '<th class="header c1" scope="col">'.$strviews.'</th>';
     if ($showlastaccess) {
-        echo '<th class="header c2" scope="col">'.$strlast.'</th>';
+        $outlinetable->head[] = $strlast;
     }
-    echo '</tr>';
 
     $modinfo = get_fast_modinfo($course);
+    $sections = get_all_sections($course->id);
 
     $sql = "SELECT cm.id, COUNT('x') AS numviews, MAX(time) AS lasttime
-              FROM {$CFG->prefix}course_modules cm
-                   JOIN {$CFG->prefix}modules m ON m.id = cm.module
-                   JOIN {$CFG->prefix}log l     ON l.cmid = cm.id
-             WHERE cm.course = $course->id AND l.action LIKE 'view%' AND m.visible = 1
+              FROM {course_modules} cm
+                   JOIN {modules} m ON m.id = cm.module
+                   JOIN {log} l     ON l.cmid = cm.id
+             WHERE cm.course = ? AND l.action LIKE 'view%' AND m.visible = 1
           GROUP BY cm.id";
-    $views = get_records_sql($sql);
+    $views = $DB->get_records_sql($sql, array($course->id));
 
-    $ri = 0;
     $prevsecctionnum = 0;
     foreach ($modinfo->sections as $sectionnum=>$section) {
         foreach ($section as $cmid) {
             $cm = $modinfo->cms[$cmid];
-            if ($cm->modname == 'label') {
+            if (!$cm->has_view()) {
                 continue;
             }
             if (!$cm->uservisible) {
                 continue;
             }
             if ($prevsecctionnum != $sectionnum) {
-                echo '<tr class="r'.$ri++.' section"><td colspan="3"><h3>';
-                switch ($course->format) {
-                    case 'weeks': print_string('week'); break;
-                    case 'topics': print_string('topic'); break;
-                    default: print_string('section'); break;
-                }
-                echo ' '.$sectionnum.'</h3></td></tr>';
+                $sectionrow = new html_table_row();
+                $sectionrow->attributes['class'] = 'section';
+                $sectioncell = new html_table_cell();
+                $sectioncell->colspan = count($outlinetable->head);
+
+                $sectiontitle = get_section_name($course, $sections[$sectionnum]);
+
+                $sectioncell->text = $OUTPUT->heading($sectiontitle, 3);
+                $sectionrow->cells[] = $sectioncell;
+                $outlinetable->data[] = $sectionrow;
 
                 $prevsecctionnum = $sectionnum;
             }
 
             $dimmed = $cm->visible ? '' : 'class="dimmed"';
             $modulename = get_string('modulename', $cm->modname);
-            echo '<tr class="r'.$ri++.'">';
-            echo "<td class=\"cell c0 actvity\"><img src=\"$CFG->modpixpath/$cm->modname/icon.gif\" class=\"icon\" alt=\"$modulename\" />";
-            echo "<a $dimmed title=\"$modulename\" href=\"$CFG->wwwroot/mod/$cm->modname/view.php?id=$cm->id\">".format_string($cm->name)."</a></td>";
-            echo "<td class=\"cell c1 numviews\">";
+
+            $reportrow = new html_table_row();
+            $activitycell = new html_table_cell();
+            $activitycell->attributes['class'] = 'activity';
+
+            $activityicon = $OUTPUT->pix_icon('icon', $modulename, $cm->modname, array('class'=>'icon'));
+
+            $attributes = array();
+            if (!$cm->visible) {
+                $attributes['class'] = 'dimmed';
+            }
+
+            $activitycell->text = $activityicon . html_writer::link("$CFG->wwwroot/mod/$cm->modname/view.php?id=$cm->id", format_string($cm->name), $attributes);;
+
+            $reportrow->cells[] = $activitycell;
+
+            $numviewscell = new html_table_cell();
+            $numviewscell->attributes['class'] = 'numviews';
+
             if (!empty($views[$cm->id]->numviews)) {
-                echo $views[$cm->id]->numviews;
+                $numviewscell->text = $views[$cm->id]->numviews;
             } else {
-                echo '-';
+                $numviewscell->text = '-';
             }
-            echo "</td>";
+
+            $reportrow->cells[] = $numviewscell;
+
+            if ($CFG->useblogassociations) {
+                require_once($CFG->dirroot.'/blog/lib.php');
+                $blogcell = new html_table_cell();
+                $blogcell->attributes['class'] = 'blog';
+                if ($blogcount = blog_get_associated_count($course->id, $cm->id)) {
+                    $blogcell->text = html_writer::link('/blog/index.php?modid='.$cm->id, $blogcount);
+                } else {
+                    $blogcell->text = '-';
+                }
+                $reportrow->cells[] = $blogcell;
+            }
+
             if ($showlastaccess) {
-                echo "<td class=\"cell c2 lastaccess\">";
+                $lastaccesscell = new html_table_cell();
+                $lastaccesscell->attributes['class'] = 'lastaccess';
+
                 if (isset($views[$cm->id]->lasttime)) {
                     $timeago = format_time(time() - $views[$cm->id]->lasttime);
-                    echo userdate($views[$cm->id]->lasttime)." ($timeago)";
+                    $lastaccesscell->text = userdate($views[$cm->id]->lasttime)." ($timeago)";
                 }
-                echo "</td>";
+                $reportrow->cells[] = $lastaccesscell;
             }
-            echo '</tr>';
+            $outlinetable->data[] = $reportrow;
         }
     }
-    echo '</table>';
+    echo html_writer::table($outlinetable);
+
+    echo $OUTPUT->footer();
 
-    print_footer($course);
 
 
-?>
