diff --git a/course/report/outline/version.php b/course/report/outline/version.php
index 5892742..819c2f1 100644
--- a/course/report/outline/version.php
+++ b/course/report/outline/version.php
@@ -1,4 +1,4 @@
-<?PHP // $Id$
+<?php
 
 ///////////////////////////////////////////////////////////////////////////
 //                                                                       //
@@ -23,7 +23,7 @@
 //                                                                       //
 ///////////////////////////////////////////////////////////////////////////
 
-$plugin->version  = 2007101501;
-$plugin->requires = 2007101532;
+$plugin->version  = 2010090501;
+$plugin->requires = 2010090501;
+
 
-?>
