diff --git a/grade/edit/tree/category.php b/grade/edit/tree/category.php
index e308aeb..0c37f8c 100644
--- a/grade/edit/tree/category.php
+++ b/grade/edit/tree/category.php
@@ -23,7 +23,14 @@ require_once 'category_form.php';
 $courseid = required_param('courseid', PARAM_INT);
 $id       = optional_param('id', 0, PARAM_INT); // grade_category->id
 
-if (!$course = get_record('course', 'id', $courseid)) {
+$url = new moodle_url('/grade/edit/tree/category.php', array('courseid'=>$courseid));
+if ($id !== 0) {
+    $url->param('id', $id);
+}
+$PAGE->set_url($url);
+$PAGE->set_pagelayout('admin');
+
+if (!$course = $DB->get_record('course', array('id' => $courseid))) {
     print_error('nocourseid');
 }
 
@@ -40,7 +47,7 @@ $heading = get_string('categoryedit', 'grades');
 
 if ($id) {
     if (!$grade_category = grade_category::fetch(array('id'=>$id, 'courseid'=>$course->id))) {
-        error('Incorrect category id!');
+        print_error('invalidcategory');
     }
     $grade_category->apply_forced_settings();
     $category = $grade_category->get_record_data();
@@ -148,7 +155,7 @@ if ($mform->is_cancelled()) {
 
     $convert = array('grademax', 'grademin', 'gradepass', 'multfactor', 'plusfactor', 'aggregationcoef');
     foreach ($convert as $param) {
-        if (array_key_exists($param, $itemdata)) {
+        if (property_exists($itemdata, $param)) {
             $itemdata->$param = unformat_float($itemdata->$param);
         }
     }
@@ -177,7 +184,7 @@ if ($mform->is_cancelled()) {
     }
 
     // Handle null decimals value - must be done before update!
-    if (!array_key_exists('decimals', $itemdata) or $itemdata->decimals < 0) {
+    if (!property_exists($itemdata, 'decimals') or $itemdata->decimals < 0) {
         $grade_item->decimals = null;
     }
 
@@ -203,10 +210,12 @@ if ($mform->is_cancelled()) {
     redirect($returnurl);
 }
 
-
-print_grade_page_head($courseid, 'edittree', null, $heading);
+$return = false;
+$buttons = false;
+$shownavigation = false;
+print_grade_page_head($courseid, 'edittree', null, $heading, $return, $buttons, $shownavigation);
 
 $mform->display();
 
-print_footer($course);
+echo $OUTPUT->footer();
 die;
