diff --git a/grade/edit/tree/outcomeitem_form.php b/grade/edit/tree/outcomeitem_form.php
index e36f0da..4cd5a71 100644
--- a/grade/edit/tree/outcomeitem_form.php
+++ b/grade/edit/tree/outcomeitem_form.php
@@ -34,10 +34,10 @@ class edit_outcomeitem_form extends moodleform {
         $mform->addRule('itemname', get_string('required'), 'required', null, 'client');
 
         $mform->addElement('text', 'iteminfo', get_string('iteminfo', 'grades'));
-        $mform->setHelpButton('iteminfo', array('iteminfo', get_string('iteminfo', 'grades'), 'grade'), true);
+        $mform->addHelpButton('iteminfo', 'iteminfo', 'grades');
 
         $mform->addElement('text', 'idnumber', get_string('idnumbermod'));
-        $mform->setHelpButton('idnumber', array('idnumber', get_string('idnumber', 'grades'), 'grade'), true);
+        $mform->addHelpButton('idnumber', 'idnumbermod');
 
         // allow setting of outcomes on module items too
         $options = array();
@@ -48,7 +48,7 @@ class edit_outcomeitem_form extends moodleform {
         }
         $mform->addElement('selectwithlink', 'outcomeid', get_string('outcome', 'grades'), $options, null,
             array('link' => $CFG->wwwroot.'/grade/edit/outcome/course.php?id='.$COURSE->id, 'label' => get_string('outcomeassigntocourse', 'grades')));
-        $mform->setHelpButton('outcomeid', array('outcomeid', get_string('outcomeid', 'grades'), 'grade'), true);
+        $mform->addHelpButton('outcomeid', 'outcome', 'grades');
         $mform->addRule('outcomeid', get_string('required'), 'required');
 
         $options = array(0=>get_string('none'));
@@ -60,26 +60,20 @@ class edit_outcomeitem_form extends moodleform {
             }
         }
         $mform->addElement('select', 'cmid', get_string('linkedactivity', 'grades'), $options);
-        $mform->setHelpButton('cmid', array('linkedactivity', get_string('linkedactivity', 'grades'), 'grade'), true);
+        $mform->addHelpButton('cmid', 'linkedactivity', 'grades');
         $mform->setDefault('cmid', 0);
 
-        /*$mform->addElement('text', 'gradepass', get_string('gradepass', 'grades'));
-        $mform->setHelpButton('gradepass', array(false, get_string('gradepass', 'grades'),
-                false, true, false, get_string('gradepasshelp', 'grades')));*/
-
         /// hiding
         /// advcheckbox is not compatible with disabledIf !!
         $mform->addElement('checkbox', 'hidden', get_string('hidden', 'grades'));
-        $mform->setHelpButton('hidden', array('hidden', get_string('hidden', 'grades'), 'grade'));
+        $mform->addHelpButton('hidden', 'hidden', 'grades');
         $mform->addElement('date_time_selector', 'hiddenuntil', get_string('hiddenuntil', 'grades'), array('optional'=>true));
-        $mform->setHelpButton('hiddenuntil', array('hiddenuntil', get_string('hiddenuntil', 'grades'), 'grade'));
         $mform->disabledIf('hidden', 'hiddenuntil[off]', 'notchecked');
 
         //locking
         $mform->addElement('advcheckbox', 'locked', get_string('locked', 'grades'));
-        $mform->setHelpButton('locked', array('locked', get_string('locked', 'grades'), 'grade'));
+        $mform->addHelpButton('locked', 'locked', 'grades');
         $mform->addElement('date_time_selector', 'locktime', get_string('locktime', 'grades'), array('optional'=>true));
-        $mform->setHelpButton('locktime', array('lockedafter', get_string('locktime', 'grades'), 'grade'));
 
 /// parent category related settings
         $mform->addElement('header', 'headerparent', get_string('parentcategory', 'grades'));
@@ -102,7 +96,7 @@ class edit_outcomeitem_form extends moodleform {
                     $coefstring = ($coefstring=='' or $coefstring=='aggregationcoefextrasum') ? 'aggregationcoefextrasum' : 'aggregationcoef';
 
                 } else if ($cat->aggregation == GRADE_AGGREGATE_EXTRACREDIT_MEAN) {
-                    $coefstring = ($coefstring=='' or $coefstring=='aggregationcoefextra') ? 'aggregationcoefextra' : 'aggregationcoef';
+                    $coefstring = ($coefstring=='' or $coefstring=='aggregationcoefextraweight') ? 'aggregationcoefextraweight' : 'aggregationcoef';
 
                 } else if ($cat->aggregation == GRADE_AGGREGATE_SUM) {
                     $coefstring = ($coefstring=='' or $coefstring=='aggregationcoefextrasum') ? 'aggregationcoefextrasum' : 'aggregationcoef';
@@ -127,7 +121,7 @@ class edit_outcomeitem_form extends moodleform {
             } else {
                 $mform->addElement('text', 'aggregationcoef', get_string($coefstring, 'grades'));
             }
-            $mform->setHelpButton('aggregationcoef', array($coefstring, get_string($coefstring, 'grades'), 'grade'), true);
+            $mform->addHelpButton('aggregationcoef', $coefstring, 'grades');
         }
 
 /// hidden params
@@ -204,7 +198,7 @@ class edit_outcomeitem_form extends moodleform {
                         $aggcoef = 'aggregationcoefextrasum';
 
                     } else if ($parent_category->aggregation == GRADE_AGGREGATE_EXTRACREDIT_MEAN) {
-                        $aggcoef = 'aggregationcoefextra';
+                        $aggcoef = 'aggregationcoefextraweight';
 
                     } else if ($parent_category->aggregation == GRADE_AGGREGATE_SUM) {
                         $aggcoef = 'aggregationcoefextrasum';
@@ -212,7 +206,7 @@ class edit_outcomeitem_form extends moodleform {
 
                     if ($aggcoef !== '') {
                         $agg_el->setLabel(get_string($aggcoef, 'grades'));
-                        $mform->setHelpButton('aggregationcoef', array($aggcoef, get_string($aggcoef, 'grades'), 'grade'));
+                        $mform->addHelpButton('aggregationcoef', $aggcoef, 'grades');
                     }
                 }
             }
@@ -247,4 +241,4 @@ class edit_outcomeitem_form extends moodleform {
     }
 
 }
-?>
+
