diff --git a/blocks/online_users/db/access.php b/blocks/online_users/db/access.php
index d22923f..d042fac 100644
--- a/blocks/online_users/db/access.php
+++ b/blocks/online_users/db/access.php
@@ -1,4 +1,4 @@
-<?php  // $Id$
+<?php
 
 ///////////////////////////////////////////////////////////////////////////
 //                                                                       //
@@ -23,21 +23,21 @@
 //                                                                       //
 ///////////////////////////////////////////////////////////////////////////
 
-$block_online_users_capabilities = array(
+$capabilities = array(
 
     'block/online_users:viewlist' => array(
 
         'captype' => 'read',
         'contextlevel' => CONTEXT_BLOCK,
-        'legacy' => array(
+        'archetypes' => array(
             'user' => CAP_ALLOW,
             'guest' => CAP_ALLOW,
             'student' => CAP_ALLOW,
             'teacher' => CAP_ALLOW,
             'editingteacher' => CAP_ALLOW,
-            'admin' => CAP_ALLOW
+            'manager' => CAP_ALLOW
         )
     )
 );
 
-?>
+
