diff --git a/auth/pop3/auth.php b/auth/pop3/auth.php
index 6f52fee..8fe4f94 100644
--- a/auth/pop3/auth.php
+++ b/auth/pop3/auth.php
@@ -35,18 +35,18 @@ class auth_plugin_pop3 extends auth_plugin_base {
      * Returns true if the username and password work and false if they are
      * wrong or don't exist.
      *
-     * @param string $username The username (with system magic quotes)
-     * @param string $password The password (with system magic quotes)
+     * @param string $username The username
+     * @param string $password The password
      * @return bool Authentication success or failure.
      */
     function user_login($username, $password) {
         if (! function_exists('imap_open')) {
-            print_error('auth_pop3notinstalled','auth');
+            print_error('auth_pop3notinstalled','auth_pop3');
             exit;
         }
 
         global $CFG;
-        $hosts = split(';', $this->config->host);   // Could be multiple hosts
+        $hosts = explode(';', $this->config->host);   // Could be multiple hosts
         foreach ($hosts as $host) {                 // Try each host in turn
             $host = trim($host);
 
@@ -70,7 +70,7 @@ class auth_plugin_pop3 extends auth_plugin_base {
             }
 
             error_reporting(0);
-            $connection = imap_open($host, stripslashes($username), stripslashes($password));
+            $connection = imap_open($host, $username, $password);
             error_reporting($CFG->debug);
 
             if ($connection) {
@@ -108,10 +108,10 @@ class auth_plugin_pop3 extends auth_plugin_base {
      * Returns the URL for changing the user's pw, or false if the default can
      * be used.
      *
-     * @return bool
+     * @return moodle_url
      */
     function change_password_url() {
-        return $this->config->changepasswordurl;
+        return new moodle_url($this->config->changepasswordurl);
     }
 
     /**
@@ -123,6 +123,8 @@ class auth_plugin_pop3 extends auth_plugin_base {
      * @param array $page An object containing all the data for this page.
      */
     function config_form($config, $err, $user_fields) {
+        global $OUTPUT;
+
         include "config.html";
     }
 
@@ -159,4 +161,4 @@ class auth_plugin_pop3 extends auth_plugin_base {
 
 }
 
-?>
+
