diff --git a/install/lang/ko/install.php b/install/lang/ko/install.php
index 974921b..a3862a2 100644
--- a/install/lang/ko/install.php
+++ b/install/lang/ko/install.php
@@ -60,9 +60,10 @@ $string['memorylimithelp'] = '<p>현재 서버의 PHP 메모리 사용량은 {$a
 
 <li>만약 php.ini 파일에 접근 가능하다면 당신은 <b>memory_limit 40M</b> 처럼 값을 바꿀 수 있을것입니다. 만약 여러분이 직접 접근 할 수 없다면 서버 관리자에게 요청하여 처리하실 수 있습니다.</li>
 
-<li>또 도저히 php.ini 안에 있는 값을 바꿀 수가 없다면 무들 디렉토리에 아래와 같은 내용을 포함하는 .htaccess 를 넣어두면 됩니다.
-<P><blockquote>php_value memory_limit 40M<blockquote></p>
-<p>그러나 어떤 서버에서는 이러한 기능이 모든 PHP페이지에 적용되어 버릴 수도 있게 되는 데 (당신이 페이지를 살펴보았을때 문제를 찾을 것이다) 이 때에는 .htaccess 를 제거해야 하고 다른 방안을 찾아봐야 할 것입니다.</p></li></ol>';
+<li>어떤 PHP 서버에서는  .htaccess  파일을 무들 디렉토리에 만들고 다음 내용을 포함하는 줄을 입력하면 됩니다.
+
+<P><blockquote><div>php_value memory_limit 40M</div><blockquote></p>
+<p>그러나 어떤 서버에서는 이러한 기능이 모든 PHP페이지에 적용되지 않을  수 있는데 (페이지에 오류가 나타날 것입니다) .htaccess 를 제거해야만 할 수도 있습니다.</p></li></ol>';
 $string['paths'] = '경로';
 $string['pathserrcreatedataroot'] = '데이터 디렉토리 ({$a->dataroot})가 설치스크립트에 의해 생성될 수 없습니다.';
 $string['pathshead'] = '경로 확인';
