diff --git a/grade/import/xml/import.php b/grade/import/xml/import.php
index cb21310..3c941f0 100644
--- a/grade/import/xml/import.php
+++ b/grade/import/xml/import.php
@@ -23,7 +23,13 @@ $url       = required_param('url', PARAM_URL); // only real urls here
 $id        = required_param('id', PARAM_INT); // course id
 $feedback  = optional_param('feedback', 0, PARAM_BOOL);
 
-if (!$course = get_record('course', 'id', $id)) {
+$url = new moodle_url('/grade/import/xml/import.php', array('id'=>$id,'url'=>$url));
+if ($feedback !== 0) {
+    $url->param('feedback', $feedback);
+}
+$PAGE->set_url($url);
+
+if (!$course = $DB->get_record('course', array('id'=>$id))) {
     print_error('nocourseid');
 }
 
@@ -38,28 +44,25 @@ require_capability('gradeimport/xml:view', $context);
 // that we'll take longer, and that the process should be recycled soon
 // to free up memory.
 @set_time_limit(0);
-@raise_memory_limit("256M");
-if (function_exists('apache_child_terminate')) {
-    @apache_child_terminate();
-}
+raise_memory_limit(MEMORY_EXTRA);
 
 $text = download_file_content($url);
 if ($text === false) {
-    error('Can not read file');
+    print_error('cannotreadfile');
 }
 
 $error = '';
 $importcode = import_xml_grades($text, $course, $error);
 
 if ($importcode !== false) {
-    /// comit the code if we are up this far
+    /// commit the code if we are up this far
 
     if (defined('USER_KEY_LOGIN')) {
         if (grade_import_commit($id, $importcode, $feedback, false)) {
             echo 'ok';
             die;
         } else {
-            error('Grade import error'); //TODO: localize
+            print_error('cannotimportgrade'); //TODO: localize
         }
 
     } else {
@@ -67,12 +70,12 @@ if ($importcode !== false) {
 
         grade_import_commit($id, $importcode, $feedback, true);
 
-        print_footer();
+        echo $OUTPUT->footer();
         die;
     }
 
 } else {
-    error($error);
+    print_error('error', 'gradeimport_xml');
 }
 
-?>
+
