diff --git a/lib/adodb/drivers/adodb-mssql_n.inc.php b/lib/adodb/drivers/adodb-mssql_n.inc.php
index 3808298..78d2c6c 100644
--- a/lib/adodb/drivers/adodb-mssql_n.inc.php
+++ b/lib/adodb/drivers/adodb-mssql_n.inc.php
@@ -9,7 +9,7 @@
 // ADOdb  - Database Abstraction Library for PHP                         //
 //          http://adodb.sourceforge.net/                                //
 //                                                                       //
-// Copyright (C) 2000-2008 John Lim (jlim\@natsoft.com.my)               //
+// Copyright (c) 2000-2010 John Lim (jlim\@natsoft.com.my)               //
 //          All rights reserved.                                         //
 //          Released under both BSD license and LGPL library license.    //
 //          Whenever there is any discrepancy between the two licenses,  //
@@ -58,7 +58,7 @@ class ADODB_mssql_n extends ADODB_mssql {
 		ADODB_mssql::ADODB_mssql();
 	}
 
-	function _query($sql,$inputarr)
+	function _query($sql,$inputarr=false)
 	{
         $sql = $this->_appendN($sql);
 		return ADODB_mssql::_query($sql,$inputarr);
