diff --git a/lib/adodb/drivers/adodb-netezza.inc.php b/lib/adodb/drivers/adodb-netezza.inc.php
index 7db95b4..0f5a1ee 100644
--- a/lib/adodb/drivers/adodb-netezza.inc.php
+++ b/lib/adodb/drivers/adodb-netezza.inc.php
@@ -1,6 +1,6 @@
 <?php
 /*
-  V4.98 13 Feb 2008  (c) 2000-2008 John Lim (jlim#natsoft.com.my). All rights reserved.
+  V5.11 5 May 2010   (c) 2000-2010 John Lim (jlim#natsoft.com). All rights reserved.
  
   First cut at the Netezza Driver by Josh Eldridge joshuae74#hotmail.com
  Based on the previous postgres drivers.
@@ -53,7 +53,7 @@ class ADODB_netezza extends ADODB_postgres64 {
 	
 	}
 	
-	function &MetaColumns($table,$upper=true) 
+	function MetaColumns($table,$upper=true) 
 	{
 	
 	// Changed this function to support Netezza which has no concept of keys
@@ -67,7 +67,7 @@ class ADODB_netezza extends ADODB_postgres64 {
 		$ADODB_FETCH_MODE = ADODB_FETCH_NUM;
 		if ($this->fetchMode !== false) $savem = $this->SetFetchMode(false);
 		
-		$rs =& $this->Execute(sprintf($this->metaColumnsSQL,$table,$table));
+		$rs = $this->Execute(sprintf($this->metaColumnsSQL,$table,$table));
 		if (isset($savem)) $this->SetFetchMode($savem);
 		$ADODB_FETCH_MODE = $save;
 		
