diff --git a/lib/adodb/drivers/adodb-pdo_mssql.inc.php b/lib/adodb/drivers/adodb-pdo_mssql.inc.php
index 2a9958c..5a7dd51 100644
--- a/lib/adodb/drivers/adodb-pdo_mssql.inc.php
+++ b/lib/adodb/drivers/adodb-pdo_mssql.inc.php
@@ -2,7 +2,7 @@
 
 
 /*
-V4.98 13 Feb 2008  (c) 2000-2008 John Lim (jlim#natsoft.com.my). All rights reserved.
+V5.11 5 May 2010   (c) 2000-2010 John Lim (jlim#natsoft.com). All rights reserved.
   Released under both BSD license and Lesser GPL library license. 
   Whenever there is any discrepancy between the two licenses, 
   the BSD license will take precedence.
@@ -47,12 +47,12 @@ class ADODB_pdo_mssql extends ADODB_pdo {
 		$this->Execute("SET TRANSACTION ".$transaction_mode);
 	}
 	
-	function MetaTables()
+	function MetaTables($ttype=false,$showSchema=false,$mask=false) 
 	{
 		return false;
 	}
 	
-	function MetaColumns()
+	function MetaColumns($table,$normalize=true)
 	{
 		return false;
 	}
