diff --git a/lib/adodb/drivers/adodb-sqlanywhere.inc.php b/lib/adodb/drivers/adodb-sqlanywhere.inc.php
index c842a76..3933d85 100644
--- a/lib/adodb/drivers/adodb-sqlanywhere.inc.php
+++ b/lib/adodb/drivers/adodb-sqlanywhere.inc.php
@@ -1,6 +1,6 @@
 <?php
 /* 
-version V4.98 13 Feb 2008 (c) 2000-2008  John Lim (jlim#natsoft.com.my).  All rights
+version V5.06 16 Oct 2008  (c) 2000-2010  John Lim (jlim#natsoft.com).  All rights
 reserved.
   Released under both BSD license and Lesser GPL library license. 
   Whenever there is any discrepancy between the two licenses, 
