diff --git a/mod/assignment/index.php b/mod/assignment/index.php
index f68146b..71eb023 100644
--- a/mod/assignment/index.php
+++ b/mod/assignment/index.php
@@ -1,65 +1,69 @@
-<?php // $Id$
+<?php
 
-    require_once("../../config.php");
-    require_once("lib.php");
-    require_once($CFG->libdir.'/gradelib.php');
+require_once("../../config.php");
+require_once("lib.php");
+require_once($CFG->libdir.'/gradelib.php');
 
-    $id = required_param('id', PARAM_INT);   // course
+$id = required_param('id', PARAM_INT);   // course
 
-    if (! $course = get_record("course", "id", $id)) {
-        error("Course ID is incorrect");
-    }
+if (!$course = $DB->get_record('course', array('id'=>$id))) {
+    print_error('invalidcourseid');
+}
+
+require_course_login($course);
+$PAGE->set_pagelayout('incourse');
 
-    require_course_login($course);
-    add_to_log($course->id, "assignment", "view all", "index.php?id=$course->id", "");
+add_to_log($course->id, "assignment", "view all", "index.php?id=$course->id", "");
 
-    $strassignments = get_string("modulenameplural", "assignment");
-    $strassignment = get_string("modulename", "assignment");
-    $strassignmenttype = get_string("assignmenttype", "assignment");
-    $strweek = get_string("week");
-    $strtopic = get_string("topic");
-    $strname = get_string("name");
-    $strduedate = get_string("duedate", "assignment");
-    $strsubmitted = get_string("submitted", "assignment");
-    $strgrade = get_string("grade");
+$strassignments = get_string("modulenameplural", "assignment");
+$strassignment = get_string("modulename", "assignment");
+$strassignmenttype = get_string("assignmenttype", "assignment");
+$strsectionname  = get_string('sectionname', 'format_'.$course->format);
+$strname = get_string("name");
+$strduedate = get_string("duedate", "assignment");
+$strsubmitted = get_string("submitted", "assignment");
+$strgrade = get_string("grade");
 
-    $navlinks = array();
-    $navlinks[] = array('name' => $strassignments, 'link' => '', 'type' => 'activity');
-    $navigation = build_navigation($navlinks);
 
-    print_header_simple($strassignments, "", $navigation, "", "", true, "", navmenu($course));
+$PAGE->set_url('/mod/assignment/index.php', array('id'=>$course->id));
+$PAGE->navbar->add($strassignments);
+$PAGE->set_title($strassignments);
+$PAGE->set_heading($course->fullname);
+echo $OUTPUT->header();
 
-    if (!$cms = get_coursemodules_in_course('assignment', $course->id, 'm.assignmenttype, m.timedue')) {
+if (!$cms = get_coursemodules_in_course('assignment', $course->id, 'cm.idnumber, m.assignmenttype, m.timedue')) {
         notice(get_string('noassignments', 'assignment'), "../../course/view.php?id=$course->id");
         die;
-    }
+}
 
-    $timenow = time();
+$usesections = course_format_uses_sections($course->format);
+if ($usesections) {
+    $sections = get_all_sections($course->id);
+}
 
-    if ($course->format == "weeks") {
-        $table->head  = array ($strweek, $strname, $strassignmenttype, $strduedate, $strsubmitted, $strgrade);
-        $table->align = array ("center", "left", "left", "left", "right");
-    } else if ($course->format == "topics") {
-        $table->head  = array ($strtopic, $strname, $strassignmenttype, $strduedate, $strsubmitted, $strgrade);
-        $table->align = array ("center", "left", "left", "left", "right");
-    } else {
+$timenow = time();
+
+$table = new html_table();
+
+if ($usesections) {
+    $table->head  = array ($strsectionname, $strname, $strassignmenttype, $strduedate, $strsubmitted, $strgrade);
+} else {
         $table->head  = array ($strname, $strassignmenttype, $strduedate, $strsubmitted, $strgrade);
-        $table->align = array ("left", "left", "left", "right");
-    }
+}
 
-    $currentsection = "";
+$currentsection = "";
 
-    $types = assignment_types();
+$types = assignment_types();
 
-    $modinfo = get_fast_modinfo($course);
-    foreach ($modinfo->instances['assignment'] as $cm) {
+$modinfo = get_fast_modinfo($course);
+foreach ($modinfo->instances['assignment'] as $cm) {
         if (!$cm->uservisible) {
             continue;
         }
 
         $cm->timedue        = $cms[$cm->id]->timedue;
         $cm->assignmenttype = $cms[$cm->id]->assignmenttype;
-        $cm->idnumber       = get_field('course_modules', 'idnumber', 'id', $cm->id); //hack
+    $cm->idnumber       = $cms[$cm->id]->idnumber;
 
         //Show dimmed if the mod is hidden
         $class = $cm->visible ? '' : 'class="dimmed"';
@@ -67,15 +71,17 @@
         $link = "<a $class href=\"view.php?id=$cm->id\">".format_string($cm->name)."</a>";
 
         $printsection = "";
+    if ($usesections) {
         if ($cm->sectionnum !== $currentsection) {
             if ($cm->sectionnum) {
-                $printsection = $cm->sectionnum;
+                $printsection = get_section_name($course, $sections[$cm->sectionnum]);
             }
             if ($currentsection !== "") {
                 $table->data[] = 'hr';
             }
             $currentsection = $cm->sectionnum;
         }
+    }
 
         if (!file_exists($CFG->dirroot.'/mod/assignment/type/'.$cm->assignmenttype.'/assignment.class.php')) {
             continue;
@@ -97,18 +103,24 @@
 
         $type = $types[$cm->assignmenttype];
 
+    // if type has an 'all.php' defined, make this a link
+    $pathtoall = "{$CFG->dirroot}/mod/assignment/type/{$cm->assignmenttype}/all.php";
+    if (file_exists($pathtoall)) {
+        $type = "<a href=\"{$CFG->wwwroot}/mod/assignment/type/{$cm->assignmenttype}/".
+            "all.php?id={$course->id}\">$type</a>";
+    }
+
         $due = $cm->timedue ? userdate($cm->timedue) : '-';
 
-        if ($course->format == "weeks" or $course->format == "topics") {
+    if ($usesections) {
             $table->data[] = array ($printsection, $link, $type, $due, $submitted, $grade);
         } else {
             $table->data[] = array ($link, $type, $due, $submitted, $grade);
         }
-    }
+}
 
-    echo "<br />";
+echo "<br />";
 
-    print_table($table);
+echo html_writer::table($table);
 
-    print_footer($course);
-?>
+echo $OUTPUT->footer();
