diff --git a/mod/assignment/version.php b/mod/assignment/version.php
index dcaf4a1..8bcda08 100644
--- a/mod/assignment/version.php
+++ b/mod/assignment/version.php
@@ -1,12 +1,12 @@
-<?PHP // $Id$
+<?php
 
 ////////////////////////////////////////////////////////////////////////////////
 //  Code fragment to define the module version etc.
 //  This fragment is called by /admin/index.php
 ////////////////////////////////////////////////////////////////////////////////
 
-$module->version  = 2007101511;
-$module->requires = 2007101509;  // Requires this Moodle version
+$module->version  = 2010102600;
+$module->requires = 2010102600;  // Requires this Moodle version
 $module->cron     = 60;
 
-?>
+
