diff --git a/grade/export/txt/dump.php b/grade/export/txt/dump.php
index 079aec8..305b480 100644
--- a/grade/export/txt/dump.php
+++ b/grade/export/txt/dump.php
@@ -15,18 +15,18 @@
 // You should have received a copy of the GNU General Public License
 // along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
 
-$nomoodlecookie = true; // session not used here
+define('NO_MOODLE_COOKIES', true); // session not used here
 require_once '../../../config.php';
 
 $id = required_param('id', PARAM_INT); // course id
-if (!$course = get_record('course', 'id', $id)) {
+if (!$course = $DB->get_record('course', array('id'=>$id))) {
     print_error('nocourseid');
 }
 
 require_user_key_login('grade/export', $id); // we want different keys for each course
 
 if (empty($CFG->gradepublishing)) {
-    error('Grade publishing disabled');
+    print_error('gradepubdisable');
 }
 
 $context = get_context_instance(CONTEXT_COURSE, $id);
@@ -35,4 +35,4 @@ require_capability('gradeexport/txt:publish', $context);
 // use the same page parameters as export.php and append &key=sdhakjsahdksahdkjsahksadjksahdkjsadhksa
 require 'export.php';
 
-?>
+
