diff --git a/grade/export/txt/grade_export_txt.php b/grade/export/txt/grade_export_txt.php
old mode 100755
new mode 100644
index c69f955..77f804b
--- a/grade/export/txt/grade_export_txt.php
+++ b/grade/export/txt/grade_export_txt.php
@@ -19,29 +19,27 @@ require_once($CFG->dirroot.'/grade/export/lib.php');
 
 class grade_export_txt extends grade_export {
 
-    var $plugin = 'txt';
+    public $plugin = 'txt';
 
-    var $separator; // default separator
+    public $separator; // default separator
 
-    function grade_export_txt($course, $groupid=0, $itemlist='', $export_feedback=false, $updatedgradesonly = false, $displaytype = GRADE_DISPLAY_TYPE_REAL, $decimalpoints = 2, $separator='comma') {
+    public function grade_export_txt($course, $groupid=0, $itemlist='', $export_feedback=false, $updatedgradesonly = false, $displaytype = GRADE_DISPLAY_TYPE_REAL, $decimalpoints = 2, $separator='comma') {
         $this->grade_export($course, $groupid, $itemlist, $export_feedback, $updatedgradesonly, $displaytype, $decimalpoints);
         $this->separator = $separator;
     }
 
-    function process_form($formdata) {
-        parent::process_form($formdata);
-        if (isset($formdata->separator)) {
-            $this->separator = $formdata->separator;
-        }
+    public function __construct($course, $groupid=0, $itemlist='', $export_feedback=false, $updatedgradesonly = false, $displaytype = GRADE_DISPLAY_TYPE_REAL, $decimalpoints = 2, $separator='comma') {
+        parent::__construct($course, $groupid, $itemlist, $export_feedback, $updatedgradesonly, $displaytype, $decimalpoints);
+        $this->separator = $separator;
     }
 
-    function get_export_params() {
+    public function get_export_params() {
         $params = parent::get_export_params();
         $params['separator'] = $this->separator;
         return $params;
     }
 
-    function print_grades() {
+    public function print_grades() {
         global $CFG;
 
         $export_tracking = $this->track_exports();
@@ -68,7 +66,8 @@ class grade_export_txt extends grade_export {
             @header('Pragma: no-cache');
         }
         header("Content-Type: application/download\n");
-        $downloadfilename = clean_filename("{$this->course->shortname} $strgrades");
+        $shortname = format_string($this->course->shortname, true, array('context' => get_context_instance(CONTEXT_COURSE, $this->course->id)));
+        $downloadfilename = clean_filename("$shortname $strgrades");
         header("Content-Disposition: attachment; filename=\"$downloadfilename.txt\"");
 
 /// Print names of all the fields
@@ -119,4 +118,4 @@ class grade_export_txt extends grade_export {
     }
 }
 
-?>
+
