diff --git a/grade/export/txt/index.php b/grade/export/txt/index.php
old mode 100755
new mode 100644
index 81b808e..47de3e7
--- a/grade/export/txt/index.php
+++ b/grade/export/txt/index.php
@@ -21,7 +21,9 @@ require_once 'grade_export_txt.php';
 
 $id = required_param('id', PARAM_INT); // course id
 
-if (!$course = get_record('course', 'id', $id)) {
+$PAGE->set_url('/grade/export/txt/index.php', array('id'=>$id));
+
+if (!$course = $DB->get_record('course', array('id'=>$id))) {
     print_error('nocourseid');
 }
 
@@ -31,7 +33,7 @@ $context = get_context_instance(CONTEXT_COURSE, $id);
 require_capability('moodle/grade:export', $context);
 require_capability('gradeexport/txt:view', $context);
 
-print_grade_page_head($COURSE->id, 'export', 'txt', get_string('exportto', 'grades') . ' ' . get_string('modulename', 'gradeexport_txt'));
+print_grade_page_head($COURSE->id, 'export', 'txt', get_string('exportto', 'grades') . ' ' . get_string('pluginname', 'gradeexport_txt'));
 
 if (!empty($CFG->gradepublishing)) {
     $CFG->gradepublishing = has_capability('gradeexport/txt:publish', $context);
@@ -42,21 +44,21 @@ $mform = new grade_export_form(null, array('includeseparator'=>true, 'publishing
 $groupmode    = groups_get_course_groupmode($course);   // Groups are being used
 $currentgroup = groups_get_course_group($course, true);
 if ($groupmode == SEPARATEGROUPS and !$currentgroup and !has_capability('moodle/site:accessallgroups', $context)) {
-    print_heading(get_string("notingroup"));
-    print_footer($course);
+    echo $OUTPUT->heading(get_string("notingroup"));
+    echo $OUTPUT->footer();
     die;    
 }
 
 // process post information
 if ($data = $mform->get_data()) {
-    $export = new grade_export_txt($course, $currentgroup, '', false, false, $data->display, $data->decimals);
+    $export = new grade_export_txt($course, $currentgroup, '', false, false, $data->display, $data->decimals, $data->separator);
 
     // print the grades on screen for feedback
 
     $export->process_form($data);
     $export->print_continue();
     $export->display_preview();
-    print_footer($course);
+    echo $OUTPUT->footer();
     exit;
 }
 
@@ -65,5 +67,5 @@ echo '<div class="clearer"></div>';
 
 $mform->display();
 
-print_footer();
-?>
+echo $OUTPUT->footer();
+
