diff --git a/blocks/login/block_login.php b/blocks/login/block_login.php
index 44b174b..0955972 100644
--- a/blocks/login/block_login.php
+++ b/blocks/login/block_login.php
@@ -1,9 +1,8 @@
-<?PHP //$Id$
+<?php
 
 class block_login extends block_base {
     function init() {
-        $this->title = get_string('login');
-        $this->version = 2007101509; 
+        $this->title = get_string('pluginname', 'block_login');
     }
 
     function applicable_formats() {
@@ -11,7 +10,7 @@ class block_login extends block_base {
     }
 
     function get_content () {
-        global $USER, $CFG;
+        global $USER, $CFG, $SESSION;
         $wwwroot = '';
         $signup = '';
 
@@ -36,14 +35,14 @@ class block_login extends block_base {
         // TODO: now that we have multiauth it is hard to find out if there is a way to change password
         $forgot = $wwwroot . '/login/forgot_password.php';
 
-        $username = get_moodle_cookie() === 'nobody' ? '' : get_moodle_cookie();
+        $username = get_moodle_cookie();
 
         $this->content->footer = '';
         $this->content->text = '';
 
         if (!isloggedin() or isguestuser()) {   // Show the block
 
-            $this->content->text .= "\n".'<form class="loginform" id="login" method="post" action="'.$wwwroot.'/login/index.php">';
+            $this->content->text .= "\n".'<form class="loginform" id="login" method="post" action="'.get_login_url().'">';
 
             $this->content->text .= '<div class="c1 fld username"><label for="login_username">'.get_string('username').'</label>';
             $this->content->text .= '<input type="text" name="username" id="login_username" value="'.s($username).'" /></div>';
@@ -72,4 +71,4 @@ class block_login extends block_base {
     }
 }
 
-?>
+
