diff --git a/course/view.php b/course/view.php
index 6f40c36..a1c8fd3 100644
--- a/course/view.php
+++ b/course/view.php
@@ -122,7 +122,13 @@
 
     $PAGE->set_pagelayout('course');
     $PAGE->set_pagetype('course-view-' . $course->format);
+    $PAGE->set_other_editing_capability('moodle/course:update');
     $PAGE->set_other_editing_capability('moodle/course:manageactivities');
+    $PAGE->set_other_editing_capability('moodle/course:activityvisibility');
+    if (course_format_uses_sections($course->format)) {
+        $PAGE->set_other_editing_capability('moodle/course:sectionvisibility');
+        $PAGE->set_other_editing_capability('moodle/course:movesections');
+    }
 
     // Preload course format renderer before output starts.
     // This is a little hacky but necessary since
@@ -189,9 +195,8 @@
             }
         }
 
-        if (has_capability('moodle/course:update', $context)) {
-            if (!empty($section)) {
-                if (!empty($move) and has_capability('moodle/course:movesections', $context) and confirm_sesskey()) {
+        if (!empty($section) && !empty($move) &&
+                has_capability('moodle/course:movesections', $context) && confirm_sesskey()) {
             $destsection = $section + $move;
             if (move_section_to($course, $section, $destsection)) {
                 if ($course->id == SITEID) {
@@ -203,8 +208,6 @@
                 echo $OUTPUT->notification('An error occurred while moving a section');
             }
         }
-            }
-        }
     } else {
         $USER->editing = 0;
     }
