diff --git a/course/view.php b/course/view.php
index 18f0313..8a74b55 100644
--- a/course/view.php
+++ b/course/view.php
@@ -124,7 +124,12 @@
 
     $PAGE->set_pagelayout('course');
     $PAGE->set_pagetype('course-view-' . $course->format);
+    $PAGE->set_other_editing_capability('moodle/course:update');
     $PAGE->set_other_editing_capability('moodle/course:manageactivities');
+    $PAGE->set_other_editing_capability('moodle/course:activityvisibility');
+    if (course_format_uses_sections($course->format)) {
+        $PAGE->set_other_editing_capability('moodle/course:sectionvisibility');
+    }
 
     if ($reset_user_allowed_editing) {
         // ugly hack
