diff --git a/lib/htmlpurifier/HTMLPurifier/Definition.php b/lib/htmlpurifier/HTMLPurifier/Definition.php
index a7408c9..c7f82eb 100644
--- a/lib/htmlpurifier/HTMLPurifier/Definition.php
+++ b/lib/htmlpurifier/HTMLPurifier/Definition.php
@@ -13,6 +13,17 @@ abstract class HTMLPurifier_Definition
     public $setup = false;
 
     /**
+     * If true, write out the final definition object to the cache after
+     * setup.  This will be true only if all invocations to get a raw
+     * definition object are also optimized.  This does not cause file
+     * system thrashing because on subsequent calls the cached object
+     * is used and any writes to the raw definition object are short
+     * circuited.  See enduser-customize.html for the high-level
+     * picture.
+     */
+    public $optimized = null;
+
+    /**
      * What type of definition is it?
      */
     public $type;
