diff --git a/search/documents/physical_ppt.php b/search/documents/physical_ppt.php
index 875184f..b91628a 100644
--- a/search/documents/physical_ppt.php
+++ b/search/documents/physical_ppt.php
@@ -6,9 +6,9 @@
 * @category core
 * @subpackage document_wrappers
 * @author Valery Fremaux [valery.fremaux@club-internet.fr] > 1.8
-* @contributor Tatsuva Shirai 20090530
 * @date 2008/03/31
 * @license http://www.gnu.org/copyleft/gpl.html GNU Public License
+* @version revised for Moodle 2.0
 *
 * this is a format handler for getting text out of a proprietary binary format 
 * so it can be indexed by Lucene search engine
@@ -30,6 +30,7 @@
 
 /**
 * @param object $resource
+* @param string $directfile if the resource is given as a direct file path, use it as reference to the file
 * @uses $CFG
 */
 function get_text_for_indexing_ppt(&$resource, $directfile = ''){
@@ -38,7 +39,7 @@ function get_text_for_indexing_ppt(&$resource, $directfile = ''){
     $indextext = null;
     
     // SECURITY : do not allow non admin execute anything on system !!
-    if (!has_capability('moodle/site:doanything', get_context_instance(CONTEXT_SYSTEM))) return;
+    if (!has_capability('moodle/site:config', get_context_instance(CONTEXT_SYSTEM))) return;
 
     if ($directfile == ''){
         $text = implode('', file("{$CFG->dataroot}/{$resource->course}/{$resource->reference}"));
@@ -89,7 +90,7 @@ function get_text_for_indexing_ppt(&$resource, $directfile = ''){
         $indextext = shorten_text($text, $CFG->block_search_limit_index_body);
     }
 
-    $indextext = mb_convert_encoding($indextext, 'UTF-8', 'auto'); // Shirai 20090530 - MDL19342
+    $indextext = mb_convert_encoding($indextext, 'UTF-8', 'auto');
     return $indextext;
 }
 ?>
\ No newline at end of file
