diff --git a/search/documents/physical_odt.php b/search/documents/physical_odt.php
index d34e60c..289ed25 100644
--- a/search/documents/physical_odt.php
+++ b/search/documents/physical_odt.php
@@ -6,9 +6,9 @@
 * @category core
 * @subpackage document_wrappers
 * @author Valery Fremaux [valery.fremaux@club-internet.fr] > 1.8
-* @contributor Tatsuva Shirai 20090530
 * @date 2008/03/31
 * @license http://www.gnu.org/copyleft/gpl.html GNU Public License
+* @version revised for Moodle 2.0
 *
 * this is a format handler for getting text out of the opensource ODT binary format 
 * so it can be indexed by Lucene search engine
@@ -17,22 +17,26 @@
 /**
 * OpenOffice Odt extractor
 * @param object $resource 
+* @param string $directfile if the resource is given as a direct file path, use it as reference to the file
 * @uses $CFG
 */
 function get_text_for_indexing_odt(&$resource, $directfile = ''){
     global $CFG;
     
     // SECURITY : do not allow non admin execute anything on system !!
-    if (!has_capability('moodle/site:doanything', get_context_instance(CONTEXT_SYSTEM))) return;
+    if (!has_capability('moodle/site:config', get_context_instance(CONTEXT_SYSTEM))) return;
 
-    $moodleroot = (@$CFG->block_search_usemoodleroot) ? "{$CFG->dirroot}/" : '' ;
+    // adds moodle root switch if none was defined
+    if (!isset($CFG->block_search_usemoodleroot)){
+        set_config('block_search_usemoodleroot', 1);
+    }
+    
+    $moodleroot = ($CFG->block_search_usemoodleroot) ? "{$CFG->dirroot}/" : '' ;
 
     // just call pdftotext over stdout and capture the output
     if (!empty($CFG->block_search_odt_to_text_cmd)){
-        // we need to remove any line command options...
-        preg_match("/^\S+/", $CFG->block_search_odt_to_text_cmd, $matches);
-        if (!file_exists("{$moodleroot}{$matches[0]}")){
-            mtrace('Error with OpenOffice ODT to text converter command : executable not found at '.$moodleroot.$CFG->block_search_odt_to_text_cmd);
+        if (!file_exists("{$moodleroot}{$CFG->block_search_odt_to_text_cmd}")){
+            mtrace('Error with OpenOffice ODT to text converter command : exectuable not found at '.$moodleroot.$CFG->block_search_odt_to_text_cmd);
         } else {
             if ($directfile == ''){
                 $file = escapeshellarg("{$CFG->dataroot}/{$resource->course}/{$resource->reference}");
