diff --git a/lib/base32.php b/lib/base32.php
index 7bb720f..831c712 100644
--- a/lib/base32.php
+++ b/lib/base32.php
@@ -23,6 +23,12 @@
 //
 
 /**
+ * @package moodlecore
+ * @copyright (c) 2001 The PHP Group
+ * @license GNU GPL v2.0 http://www.php.net/license/2_02.txt
+ */
+
+/**
  * Base32 encode a binary string
  *
  * @param    $inString   Binary string to base32 encode
@@ -83,7 +89,7 @@ function base32_encode ($inString)
     }
     
     /* Create an array by chunking it every 5 chars */
-    $fiveBitsArray = split("\n",rtrim(chunk_split($compBits, 5, "\n"))); 
+    $fiveBitsArray = explode("\n",rtrim(chunk_split($compBits, 5, "\n"))); 
     
     /* Look-up each chunk and add it to $outstring */
     foreach($fiveBitsArray as $fiveBitsString) { 
@@ -188,5 +194,3 @@ function Base32_decode($inString) {
     
     return $outString;
 }
-
-?>
\ No newline at end of file
