diff --git a/lib/cookies.js b/lib/cookies.js
index 966b7c2..c9c2004 100644
--- a/lib/cookies.js
+++ b/lib/cookies.js
@@ -39,7 +39,7 @@ function cookie(name, value, expires, path, domain, secure)
 			if (i == 0) break;
 		}
 		return null;
-	}
+	};
 
 	// Set cookie
 	this.set=function()
@@ -63,7 +63,7 @@ function cookie(name, value, expires, path, domain, secure)
 
 		// Re-initialize
 		this.expires=expStore;
-	}
+	};
 
 	// Kill cookie
 	this.kill=function()
@@ -72,13 +72,13 @@ function cookie(name, value, expires, path, domain, secure)
 				((this.path) ? "; path=" + this.path : "") +
 				((this.domain) ? "; domain=" + this.domain : "") +
 				"; expires=Thu, 01-Jan-70 00:00:01 GMT";
-	}
+	};
 
 	// Change cookie settings.
-	this.changeName=function(chName) { this.kill(); this.name=chName; this.set(); }
-	this.changeVal=function(chVal) { this.kill(); this.value=chVal; this.set(); }
-	this.changeExp=function(chExp) { this.kill(); this.expires=chExp; this.set(); }
-	this.changePath=function(chPath) { this.kill(); this.path=chPath; this.set(); }
-	this.changeDomain=function(chDom) { this.kill(); this.domain=chDom; this.set(); }
+	this.changeName=function(chName) { this.kill(); this.name=chName; this.set(); };
+	this.changeVal=function(chVal) { this.kill(); this.value=chVal; this.set(); };
+	this.changeExp=function(chExp) { this.kill(); this.expires=chExp; this.set(); };
+	this.changePath=function(chPath) { this.kill(); this.path=chPath; this.set(); };
+	this.changeDomain=function(chDom) { this.kill(); this.domain=chDom; this.set(); };
 	this.changeSecurity=function(chSec) { this.kill(); this.secure=chSec; this.set(); }
 }
