diff --git a/lib/cronlib.php b/lib/cronlib.php
index bc5e29a..8ed8a10 100644
--- a/lib/cronlib.php
+++ b/lib/cronlib.php
@@ -424,6 +424,11 @@ function cron_run() {
     cron_execute_plugin_type('tool', 'admin tools');
 
 
+    // run the ELIS cron jobs
+    mtrace("Running ELIS cron tasks...");
+    require_once($CFG->dirroot.'/elis/core/cron.php');
+    elis_cron();
+
     // and finally run any local cronjobs, if any
     if ($locals = get_plugin_list('local')) {
         mtrace('Processing customized cron scripts ...', '');
