diff --git a/lib/formslib.php b/lib/formslib.php
index 5a7a7fe..3301c0b 100644
--- a/lib/formslib.php
+++ b/lib/formslib.php
@@ -1306,10 +1306,14 @@ class MoodleQuickForm extends HTML_QuickForm_DHTMLRulesTableless {
      * @access   public
      */
     function MoodleQuickForm($formName, $method, $action, $target='', $attributes=null){
-        global $CFG, $OUTPUT;
+        global $CFG, $OUTPUT, $PAGE;
 
         static $formcounter = 1;
-
+        if ($PAGE->pagelayout == 'embedded' && $formcounter == 1) {
+            // ELIS-3679, ELIS-3643: don't reuse PAGE form ids for embedded content
+            //error_log("MoodleQuickForm: re-initializing formcounter to 999!");
+            $formcounter = 999;
+        }
         HTML_Common::HTML_Common($attributes);
         $target = empty($target) ? array() : array('target' => $target);
         $this->_formName = $formName;
@@ -1690,6 +1694,9 @@ class MoodleQuickForm extends HTML_QuickForm_DHTMLRulesTableless {
                 if (PEAR::isError($value)) {
                     return $value;
                 }
+                if (!is_array($value)) { // ELIS-4509: setselect::setupForm()
+                    $value = array($elementName => $value);
+                }
                 //oh, stock QuickFOrm was returning array of arrays!
                 $unfiltered = HTML_QuickForm::arrayMerge($unfiltered, $value);
             }
