diff --git a/lib/htaccess b/lib/htaccess
index 223396c..d0597fc 100644
--- a/lib/htaccess
+++ b/lib/htaccess
@@ -27,7 +27,7 @@ DirectoryIndex index.php index.html index.htm
 
 ### Thirdly, set up some PHP variables that Moodle needs
 
-php_flag magic_quotes_gpc        1
+php_flag magic_quotes_gpc        0
 php_flag magic_quotes_runtime    0
 php_flag register_globals        0
 php_flag file_uploads            1
