diff --git a/lib/html2text.php b/lib/html2text.php
index 720893d..e360b2c 100644
--- a/lib/html2text.php
+++ b/lib/html2text.php
@@ -1,36 +1,22 @@
 <?php
 
-/*************************************************************************
- *                                                                       *
- * class.html2text.inc                                                   *
- *                                                                       *
- *************************************************************************
- *                                                                       *
- * Converts HTML to formatted plain text                                 *
- *                                                                       *
- * Copyright (c) 2005-2007 Jon Abernathy <jon@chuggnutt.com>             *
- * All rights reserved.                                                  *
- *                                                                       *
- * This script is free software; you can redistribute it and/or modify   *
- * it under the terms of the GNU General Public License as published by  *
- * the Free Software Foundation; either version 2 of the License, or     *
- * (at your option) any later version.                                   *
- *                                                                       *
- * The GNU General Public License can be found at                        *
- * http://www.gnu.org/copyleft/gpl.html.                                 *
- *                                                                       *
- * This script is distributed in the hope that it will be useful,        *
- * but WITHOUT ANY WARRANTY; without even the implied warranty of        *
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the          *
- * GNU General Public License for more details.                          *
- *                                                                       *
- * Author(s): Jon Abernathy <jon@chuggnutt.com>                          *
- *                                                                       *
- * Last modified: 08/08/07                                               *
- *                                                                       *
- *************************************************************************/
-
-require_once "$CFG->libdir/textlib.class.php";
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+//
+// Author(s): Jon Abernathy <jon@chuggnutt.com>
+// Copyright (c) 2005-2007 Jon Abernathy <jon@chuggnutt.com>
 
 /**
  *  Takes HTML and converts it to formatted, plain text.
@@ -101,9 +87,15 @@ require_once "$CFG->libdir/textlib.class.php";
  *
  *  *** End of the housecleaning updates. Updated 08/08/07.
  *
- *  @author Jon Abernathy <jon@chuggnutt.com>
- *  @version 1.0.0
- *  @since PHP 4.0.2
+ * @package   moodlecore
+ * @copyright Jon Abernathy <jon@chuggnutt.com>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+/**
+ * @package   moodlecore
+ * @copyright Jon Abernathy <jon@chuggnutt.com>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
  */
 class html2text
 {
@@ -177,6 +169,7 @@ class html2text
         '/&(bull|#149|#8226);/i',                // Bullet
         '/&(pound|#163);/i',                     // Pound sign
         '/&(euro|#8364);/i',                     // Euro sign
+        '/[ ]+([\n\t])/',                        // Trailing spaces before newline or tab
         '/[ ]{2,}/'                              // Runs of spaces, post-handling
     );
 
@@ -220,6 +213,7 @@ class html2text
         '*',
         '£',
         'EUR',                                  // Euro sign. € ?
+        '\\1',                                  // Trailing spaces before newline or tab
         ' '                                     // Runs of spaces, post-handling
     );
 
@@ -463,18 +457,17 @@ class html2text
         $this->_link_count = 0;
         $this->_link_list = '';
 
-        $text = trim(stripslashes($this->html));
+        $text = trim($this->html);
 
         // Convert <PRE>
         $this->_convert_pre($text);
 
         // Run our defined search-and-replace
         $text = preg_replace($this->search, $this->replace, $text);
-        $text = preg_replace_callback($this->callback_search, array(&$this, '_preg_callback'), $text);
+        $text = preg_replace_callback($this->callback_search, array('html2text', '_preg_callback'), $text);
 
         // Replace known html entities
-        $tl=textlib_get_instance();
-        $text = $tl->entities_to_utf8($text, true);
+        $text = html_entity_decode($text, ENT_COMPAT, 'UTF-8');
 
         // Remove unknown/unhandled entities (this cannot be done in search-and-replace block)
         $text = preg_replace('/&[^&;]+;/i', '', $text); 
@@ -545,21 +538,27 @@ class html2text
     /**
      *  Helper function for PRE body conversion.
      *
-     *  @param string HTML content
+     *  @param string $text HTML content
      *  @access private
      */
     function _convert_pre(&$text)
     {
-        while(preg_match('/<pre[^>]*>(.*)<\/pre>/ismU', $text, $matches)) {
-            $result = preg_replace($this->pre_search, $this->pre_replace, $matches[1]);
-            $text = preg_replace('/<pre[^>]*>.*<\/pre>/ismU', '<div><br>' . $result . '<br></div>', $text, 1);
+         while (preg_match('/<pre[^>]*>(.*)<\/pre>/ismU', $text, $matches)) {
+            // convert the content
+            $this->pre_content = sprintf('<div><br>%s<br></div>',
+                preg_replace($this->pre_search, $this->pre_replace, $matches[1]));
+            // replace the content (use callback because content can contain $0 variable)
+            $text = preg_replace_callback('/<pre[^>]*>.*<\/pre>/ismU',
+                array('html2text', '_preg_pre_callback'), $text, 1);
+            // free memory
+            $this->pre_content = '';
         }
     }
 
     /**
      *  Callback function for preg_replace_callback use.
      *
-     *  @param  array PREG matches
+     *  @param  array $matches PREG matches
      *  @return string
      *  @access private
      */
@@ -581,9 +580,20 @@ class html2text
     }
     
     /**
+     *  Callback function for preg_replace_callback use in PRE content handler.
+     *
+     *  @param  array PREG matches
+     *  @return string
+     */
+    private function _preg_pre_callback($matches)
+    {
+        return $this->pre_content;
+    }
+
+    /**
      *  Strtoupper multibyte wrapper function
      *
-     *  @param  string
+     *  @param  string $str
      *  @return string
      *  @access private
      */
@@ -593,5 +603,3 @@ class html2text
         return $tl->strtoupper($str);
     }
 }
-
-?>
