diff --git a/lib/javascript-static.js b/lib/javascript-static.js
index 2d5bc24..cff1d44 100644
--- a/lib/javascript-static.js
+++ b/lib/javascript-static.js
@@ -527,7 +527,7 @@ function fix_column_widths() {
 function fix_column_width(colName) {
     if(column = document.getElementById(colName)) {
         if(!column.offsetWidth) {
-            setTimeout("fix_column_width('" + colName + "')", 20);
+            //sanity check - this should be set
             return;
         }
 
