diff --git a/lib/javascript.php b/lib/javascript.php
index fe7f14f..edc9451 100644
--- a/lib/javascript.php
+++ b/lib/javascript.php
@@ -66,6 +66,14 @@ if (!$jsfiles) {
     die();
 }
 
+if (!empty($CFG->nojsminify)) {
+    header('Content-Type: application/x-javascript; charset=utf-8');
+    foreach ($jsfiles as $jsfile) {
+        readfile($jsfile);
+    }
+    die;
+}
+
 minify($jsfiles);
 
 function minify($files) {
