diff --git a/lib/moodlelib.php b/lib/moodlelib.php
index 0b0ee4a..9ee15b9 100644
--- a/lib/moodlelib.php
+++ b/lib/moodlelib.php
@@ -4226,6 +4226,9 @@ function update_internal_user_password($user, $password) {
     if ($user->password !== $hashedpassword) {
         $DB->set_field('user', 'password',  $hashedpassword, array('id'=>$user->id));
         $user->password = $hashedpassword;
+
+        // Trigger user updated event
+        events_trigger('user_updated', $user);
     }
 
     return true;
@@ -5064,6 +5067,7 @@ function get_mailer($action='get') {
                 $mailer->SMTPDebug = true;
             }
             $mailer->Host          = $CFG->smtphosts;        // specify main and backup servers
+            $mailer->SMTPSecure    = $CFG->smtpsecure;       // specify secure connection protocol
             $mailer->SMTPKeepAlive = $prevkeepalive;         // use previous keepalive
 
             if ($CFG->smtpuser) {                            // Use SMTP authentication
@@ -5412,7 +5416,14 @@ function setnew_password_and_mail($user) {
 
     $newpassword = generate_password();
 
-    $DB->set_field('user', 'password', hash_internal_user_password($newpassword), array('id'=>$user->id));
+    $hashedpassword = hash_internal_user_password($newpassword);
+
+    $DB->set_field('user', 'password', $hashedpassword, array('id'=>$user->id));
+
+    $user->password = $hashedpassword;
+
+    // Trigger user updated event
+    events_trigger('user_updated', $user);
 
     $a = new stdClass();
     $a->firstname   = fullname($user, true);
@@ -6370,6 +6381,8 @@ class core_string_manager implements string_manager {
             if ($plugintype === 'mod') {
                 // bloody mod hack
                 $file = $pluginname;
+            } elseif (strncmp($plugintype, 'elis:', 5) === 0) {
+                $file = $pluginname;
             } else {
                 $file = $plugintype . '_' . $pluginname;
             }
@@ -7736,6 +7749,7 @@ function get_plugin_types($fullpaths=true) {
                       'qbehaviour'    => 'question/behaviour',
                       'qformat'       => 'question/format',
                       'plagiarism'    => 'plagiarism',
+                      'elis'          => 'elis',
                       'tool'          => $CFG->admin.'/tool',
                       'theme'         => 'theme',  // this is a bit hacky, themes may be in $CFG->themedir too
         );
@@ -7751,6 +7765,28 @@ function get_plugin_types($fullpaths=true) {
             }
         }
 
+        $blocks = get_plugin_list('block');
+        foreach ($blocks as $block => $blockdir) {
+            if (file_exists("$blockdir/db/subplugins.php")) {
+                $subplugins = array();
+                include("$blockdir/db/subplugins.php");
+                foreach ($subplugins as $subtype=>$dir) {
+                    $info[$subtype] = $dir;
+                }
+            }
+        }
+
+        $plugins = get_plugin_list('elis');
+        foreach ($plugins as $plugin => $plugindir) {
+            if (file_exists("$plugindir/db/subplugins.php")) {
+                $subplugins = array();
+                include("$plugindir/db/subplugins.php");
+                foreach ($subplugins as $subtype=>$dir) {
+                    $info[$subtype] = $dir;
+                }
+            }
+        }
+
         // local is always last!
         $info['local'] = 'local';
 
@@ -7790,6 +7826,14 @@ function get_plugin_list($plugintype) {
         // mod is an exception because we have to call this function from get_plugin_types()
         $fulldirs[] = $CFG->dirroot.'/mod';
 
+    } else if ($plugintype === 'block') {
+        // block is similarly an exception because we have to call this function from get_plugin_types()
+        $fulldirs[] = $CFG->dirroot.'/blocks';
+
+    } else if ($plugintype === 'elis') {
+        // elis is an exception because we have to call this function from get_plugin_types()
+        $fulldirs[] = $CFG->dirroot.'/elis';
+
     } else if ($plugintype === 'theme') {
         $fulldirs[] = $CFG->dirroot.'/theme';
         // themes are special because they may be stored also in separate directory
