diff --git a/lib/moodlelib.php b/lib/moodlelib.php
index cac0785..f3d109b 100644
--- a/lib/moodlelib.php
+++ b/lib/moodlelib.php
@@ -1141,7 +1141,6 @@ function make_timestamp($year, $month=1, $day=1, $hour=0, $minute=0, $second=0,
     }
 
     return $time;
-
 }
 
 /**
@@ -1475,7 +1474,7 @@ function get_user_timezone($tz = 99) {
 
     $tz = 99;
 
-    while(($tz == '' || $tz == 99 || $tz == NULL) && $next = each($timezones)) {
+    while(($tz === '' || $tz == 99 || $tz === NULL) && $next = each($timezones)) {
         $tz = $next['value'];
     }
 
@@ -5297,18 +5296,24 @@ function clean_getstring_data( $a ) {
 function places_to_search_for_lang_strings() {
     global $CFG;
 
-    return array(
+    static $places = null;
+
+    if ($places === null) {
+        $places = array(
         '__exceptions' => array('moodle', 'langconfig'),
         'assignment_' => array('mod/assignment/type'),
         'auth_' => array('auth'),
         'block_' => array('blocks'),
+        'crlm_' => array('curriculum/plugins'),
         'datafield_' => array('mod/data/field'),
         'datapreset_' => array('mod/data/preset'),
+        'elis_' => array('elis'),
         'enrol_' => array('enrol'),
         'filter_' => array('filter'),
         'format_' => array('course/format'),
         'qtype_' => array('question/type'),
         'report_' => array($CFG->admin.'/report', 'course/report', 'mod/quiz/report'),
+        'rlreport_' => array('blocks/php_report/instances'),
         'resource_' => array('mod/resource/type'),
         'gradereport_' => array('grade/report'),
         'gradeimport_' => array('grade/import'),
@@ -5317,6 +5322,13 @@ function places_to_search_for_lang_strings() {
         'profilefield_' => array('user/profile/field'),
         '' => array('mod')
     );
+        $elis_plugins = get_list_of_plugins('elis');
+        foreach ($elis_plugins as $plugin) {
+            $places["elis:{$plugin}_"] = array('elis/{$plugin}/plugins');
+        }
+    }
+
+    return $places;
 }
 
 /**
