diff --git a/lib/moodlelib.php b/lib/moodlelib.php
index 04c0eb1..93225cf 100644
--- a/lib/moodlelib.php
+++ b/lib/moodlelib.php
@@ -4538,6 +4538,9 @@ function update_internal_user_password($user, $password) {
     if ($passwordchanged || $algorithmchanged) {
         $DB->set_field('user', 'password',  $hashedpassword, array('id'=>$user->id));
         $user->password = $hashedpassword;
+
+        // Trigger user updated event
+        events_trigger('user_updated', $user);
     }
 
     return true;
@@ -5760,6 +5763,11 @@ function setnew_password_and_mail($user, $fasthash = false) {
     $hashedpassword = hash_internal_user_password($newpassword, $fasthash);
     $DB->set_field('user', 'password', $hashedpassword, array('id'=>$user->id));
 
+    $user->password = $hashedpassword;
+
+    // Trigger user updated event
+    events_trigger('user_updated', $user);
+
     $a = new stdClass();
     $a->firstname   = fullname($user, true);
     $a->sitename    = format_string($site->fullname);
@@ -6777,6 +6785,10 @@ class core_string_manager implements string_manager {
             if ($plugintype === 'mod') {
                 // bloody mod hack
                 $file = $pluginname;
+            // RL EDIT: BJB130213
+            } else if (strncmp($plugintype, 'elis:', 5) === 0) {
+                $file = $pluginname;
+            // End RL EDIT
             } else {
                 $file = $plugintype . '_' . $pluginname;
             }
@@ -8268,6 +8280,9 @@ function get_plugin_types($fullpaths=true) {
                       'cachestore'    => 'cache/stores',
                       'cachelock'     => 'cache/locks',
                       'theme'         => 'theme',  // this is a bit hacky, themes may be in $CFG->themedir too
+                      // RL EDIT: BJB130213
+                      'elis'          => 'elis',  // ELIS-5770 -- include ELIS plugins
+                      // End RL EDIT
         );
 
         $subpluginowners = array_merge(array_values(get_plugin_list('mod')),
@@ -8282,6 +8297,32 @@ function get_plugin_types($fullpaths=true) {
             }
         }
 
+        // RL EDIT: BJB130213
+        // ELIS-2455 -- Needed for Integration Point sub-plugins
+        $blocks = get_plugin_list('block');
+        foreach ($blocks as $block => $blockdir) {
+            if (file_exists("$blockdir/db/subplugins.php")) {
+                $subplugins = array();
+                include("$blockdir/db/subplugins.php");
+                foreach ($subplugins as $subtype=>$dir) {
+                    $info[$subtype] = $dir;
+                }
+            }
+        }
+
+        // ELIS-5770 -- Needed for ELIS sub-plugins
+        $elisplugs = get_plugin_list('elis');
+        foreach ($elisplugs as $elisplug => $plugdir) {
+            if (file_exists($plugdir.'/db/subplugins.php')) {
+                $subplugins = array();
+                include($plugdir.'/db/subplugins.php');
+                foreach ($subplugins as $subtype => $dir) {
+                    $info[$subtype] = $dir;
+                }
+            }
+        }
+        // End RL EDIT
+
         // local is always last!
         $info['local'] = 'local';
 
@@ -8343,7 +8384,12 @@ function get_plugin_list($plugintype) {
     if ($plugintype === 'mod') {
         // mod is an exception because we have to call this function from get_plugin_types()
         $fulldirs[] = $CFG->dirroot.'/mod';
-
+    // RL EDIT: BJB130213
+    } else if ($plugintype === 'block') {
+        // ELIS-2455 -  Needed for Integration Point sub-plugins
+        // block is similarly an exception because we have to call this function from get_plugin_types()
+        $fulldirs[] = $CFG->dirroot.'/blocks';
+    // End RL EDIT
     } else if ($plugintype === 'editor') {
         // Exception also needed for editor for same reason.
         $fulldirs[] = $CFG->dirroot . '/lib/editor';
@@ -8354,7 +8400,12 @@ function get_plugin_list($plugintype) {
         if (!empty($CFG->themedir) and file_exists($CFG->themedir) and is_dir($CFG->themedir) ) {
             $fulldirs[] = $CFG->themedir;
         }
-
+    // RL EDIT: BJB130213
+    } else if ($plugintype === 'elis') {
+        // ELIS-5770 -  Ensure that we setup for ELIS plugins to be loaded correctly
+        // elis is an exception because we have to call this function from get_plugin_types()
+        $fulldirs[] = $CFG->dirroot.'/elis';
+    // End RL EDIT
     } else {
         $types = get_plugin_types(true);
         if (!array_key_exists($plugintype, $types)) {
