diff --git a/lib/moodlelib.php b/lib/moodlelib.php
index 047430b..cce31fd 100644
--- a/lib/moodlelib.php
+++ b/lib/moodlelib.php
@@ -1944,7 +1944,7 @@ function get_user_timezone($tz = 99) {
 
     $tz = 99;
 
-    while(($tz == '' || $tz == 99 || $tz == NULL) && $next = each($timezones)) {
+    while(($tz === '' || $tz == 99 || $tz === NULL) && $next = each($timezones)) {
         $tz = $next['value'];
     }
 
@@ -5786,6 +5786,8 @@ class core_string_manager implements string_manager {
             if ($plugintype === 'mod') {
                 // bloody mod hack
                 $file = $pluginname;
+            } elseif (strncmp($plugintype, 'elis:', 5) === 0) {
+                $file = $pluginname;
             } else {
                 $file = $plugintype . '_' . $pluginname;
             }
@@ -7126,6 +7128,7 @@ function get_plugin_types($fullpaths=true) {
                       'qbehaviour'    => 'question/behaviour',
                       'qformat'       => 'question/format',
                       'plagiarism'    => 'plagiarism',
+                      'elis'          => 'elis',
                       'theme'         => 'theme'); // this is a bit hacky, themes may be in $CFG->themedir too
 
         $mods = get_plugin_list('mod');
@@ -7139,6 +7142,28 @@ function get_plugin_types($fullpaths=true) {
             }
         }
 
+        $blocks = get_plugin_list('block');
+        foreach ($blocks as $block => $blockdir) {
+            if (file_exists("$blockdir/db/subplugins.php")) {
+                $subplugins = array();
+                include("$blockdir/db/subplugins.php");
+                foreach ($subplugins as $subtype=>$dir) {
+                    $info[$subtype] = $dir;
+                }
+            }
+        }
+
+        $plugins = get_plugin_list('elis');
+        foreach ($plugins as $plugin => $plugindir) {
+            if (file_exists("$plugindir/db/subplugins.php")) {
+                $subplugins = array();
+                include("$plugindir/db/subplugins.php");
+                foreach ($subplugins as $subtype=>$dir) {
+                    $info[$subtype] = $dir;
+                }
+            }
+        }
+
         // local is always last!
         $info['local'] = 'local';
 
@@ -7178,6 +7203,14 @@ function get_plugin_list($plugintype) {
         // mod is an exception because we have to call this function from get_plugin_types()
         $fulldirs[] = $CFG->dirroot.'/mod';
 
+    } else if ($plugintype === 'block') {
+        // block is similarly an exception because we have to call this function from get_plugin_types()
+        $fulldirs[] = $CFG->dirroot.'/blocks';
+
+    } else if ($plugintype === 'elis') {
+        // elis is an exception because we have to call this function from get_plugin_types()
+        $fulldirs[] = $CFG->dirroot.'/elis';
+
     } else if ($plugintype === 'theme') {
         $fulldirs[] = $CFG->dirroot.'/theme';
         // themes are special because they may be stored also in separate directory
