diff --git a/lib/navigationlib.php b/lib/navigationlib.php
index c574db7..aadec41 100644
--- a/lib/navigationlib.php
+++ b/lib/navigationlib.php
@@ -3604,7 +3604,9 @@ class settings_navigation extends navigation_node {
                 $editstring = get_string('turneditingon');
             }
             $coursenode->add($editstring, $editurl, self::TYPE_SETTING, null, null, new pix_icon('i/edit', ''));
+        }
 
+        if (has_capability('moodle/course:manageactivities', $coursecontext)) {
             // Add the module chooser toggle
             if ($this->page->user_is_editing() && has_capability('moodle/course:manageactivities', $coursecontext)
                     && course_ajax_enabled($course)) {
