diff --git a/lib/outputlib.php b/lib/outputlib.php
index 285085e..3b26828 100644
--- a/lib/outputlib.php
+++ b/lib/outputlib.php
@@ -894,6 +894,11 @@ class theme_config {
             $css = $csspostprocess($css, $this);
         }
 
+        // ELIS-3985: remove image urls that point to site index page
+        $match = $CFG->wwwroot;
+        $css = preg_replace("|.*background:[ ]*url\({$match}\).*;|", '', $css);
+        $css = preg_replace("|.*background:[ ]*([#,0-9,A-Z,a-z]*)[ ]*url\({$match}\).*;|", ' background-color: $1;', $css);
+
         return $css;
     }
 
