diff --git a/lib/pluginlib.php b/lib/pluginlib.php
index 1bc425a..f2a7ae4 100644
--- a/lib/pluginlib.php
+++ b/lib/pluginlib.php
@@ -157,7 +157,9 @@ class plugin_manager {
 
         if ($disablecache or is_null($this->subpluginsinfo)) {
             $this->subpluginsinfo = array();
-            $mods = get_plugin_list('mod');
+            $parentplugins = array('mod', 'block', 'elis');
+            foreach ($parentplugins as $parent) {
+                $mods = get_plugin_list($parent);
             foreach ($mods as $mod => $moddir) {
                 $modsubplugins = array();
                 if (file_exists($moddir . '/db/subplugins.php')) {
@@ -168,7 +170,8 @@ class plugin_manager {
                         $subplugin->typerootdir = $subplugintyperootdir;
                         $modsubplugins[$subplugintype] = $subplugin;
                     }
-                $this->subpluginsinfo['mod_' . $mod] = $modsubplugins;
+                        $this->subpluginsinfo[$parent . '_' . $mod] = $modsubplugins;
+                    }
                 }
             }
         }
