diff --git a/lib/pluginlib.php b/lib/pluginlib.php
index a8eca2b..f3ef7e7 100644
--- a/lib/pluginlib.php
+++ b/lib/pluginlib.php
@@ -131,7 +131,9 @@ class plugin_manager {
 
         if ($disablecache or is_null($this->subpluginsinfo)) {
             $this->subpluginsinfo = array();
-            $mods = get_plugin_list('mod');
+            $parentplugins = array('mod', 'block', 'elis');
+            foreach ($parentplugins as $parent) {
+                $mods = get_plugin_list($parent);
                 foreach ($mods as $mod => $moddir) {
                     $modsubplugins = array();
                     if (file_exists($moddir . '/db/subplugins.php')) {
@@ -142,7 +144,8 @@ class plugin_manager {
                             $subplugin->typerootdir = $subplugintyperootdir;
                             $modsubplugins[$subplugintype] = $subplugin;
                         }
-                $this->subpluginsinfo['mod_' . $mod] = $modsubplugins;
+                        $this->subpluginsinfo[$parent . '_' . $mod] = $modsubplugins;
+                    }
                 }
             }
         }
