diff --git a/lib/setup.php b/lib/setup.php
index 3037136..ecb2050 100644
--- a/lib/setup.php
+++ b/lib/setup.php
@@ -227,6 +227,15 @@ if (!defined('AJAX_SCRIPT')) {
     define('AJAX_SCRIPT', false);
 }
 
+// Begin Remote-Learner edit
+// detect NanoGong plugin
+if (!defined('NANOGONG')) {
+    if (file_exists($CFG->libdir.'/editor/tinymce/tiny_mce/3.5.7b/plugins/nanogong/nanogong.php')) {
+        define('NANOGONG', '3.5.7b');
+    }
+}
+// End Remote-Learner edit
+
 // File permissions on created directories in the $CFG->dataroot
 if (empty($CFG->directorypermissions)) {
     $CFG->directorypermissions = 02777;      // Must be octal (that's why it's here)
