diff --git a/lib/simpletestlib/authentication.php b/lib/simpletestlib/authentication.php
index 98fe411..382f128 100644
--- a/lib/simpletestlib/authentication.php
+++ b/lib/simpletestlib/authentication.php
@@ -1,21 +1,21 @@
 <?php
-    /**
+/**
      *	Base include file for SimpleTest
      *	@package	SimpleTest
      *	@subpackage	WebTester
      *	@version	$Id$
      */
-    /**
+/**
      *	include http class
      */
-    require_once(dirname(__FILE__) . '/http.php');
+require_once(dirname(__FILE__) . '/http.php');
     
-    /**
+/**
      *    Represents a single security realm's identity.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleRealm {
+class SimpleRealm {
         var $_type;
         var $_root;
         var $_username;
@@ -121,14 +121,14 @@
         function _isIn($part, $whole) {
             return strpos($whole, $part) === 0;
         }
-    }
+}
     
-    /**
+/**
      *    Manages security realms.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleAuthenticator {
+class SimpleAuthenticator {
         var $_realms;
         
         /**
@@ -234,5 +234,5 @@
                         'Authorization: Basic ' . base64_encode("$username:$password"));
             }
         }
-    }
+}
 ?>
\ No newline at end of file
