diff --git a/lib/simpletestlib/collector.php b/lib/simpletestlib/collector.php
index 439f6b5..3e91154 100644
--- a/lib/simpletestlib/collector.php
+++ b/lib/simpletestlib/collector.php
@@ -1,5 +1,5 @@
 <?php
-    /**
+/**
      * This file contains the following classes: {@link SimpleCollector},
      * {@link SimplePatternCollector}.
      *
@@ -9,14 +9,14 @@
      * @version $Id$
      */
     
-    /**
+/**
      * The basic collector for {@link GroupTest}
      *
      * @see collect(), GroupTest::collect()
      * @package SimpleTest
      * @subpackage UnitTester
      */
-    class SimpleCollector {
+class SimpleCollector {
     
         /**
          * Strips off any kind of slash at the end so as to normalise the path.
@@ -43,6 +43,9 @@
             $path = $this->_removeTrailingSlash($path);
             if ($handle = opendir($path)) {
                 while (($entry = readdir($handle)) !== false) {
+                if ($this->_isHidden($entry)) {
+                    continue;
+                }
                     $this->_handle($test, $path . DIRECTORY_SEPARATOR . $entry);
                 }
                 closedir($handle);
@@ -63,13 +66,25 @@
          * @access protected
          */
         function _handle(&$test, $file) {
-            if (! is_dir($file)) {
-                $test->addTestFile($file);
-            }
+        if (is_dir($file)) {
+            return;
         }
+        $test->addTestFile($file);
     }
     
     /**
+     *  Tests for hidden files so as to skip them. Currently
+     *  only tests for Unix hidden files.
+     *  @param string $filename        Plain filename.
+     *  @return boolean                True if hidden file.
+     *  @access private
+     */
+    function _isHidden($filename) {
+        return strncmp($filename, '.', 1) == 0;
+    }
+}
+
+/**
      * An extension to {@link SimpleCollector} that only adds files matching a
      * given pattern.
      *
@@ -77,7 +92,7 @@
      * @subpackage UnitTester
      * @see SimpleCollector
      */
-    class SimplePatternCollector extends SimpleCollector {
+class SimplePatternCollector extends SimpleCollector {
         var $_pattern;
     
         /**
@@ -103,5 +118,5 @@
                 parent::_handle($test, $filename);
             }
         }
-    }
+}
 ?>
\ No newline at end of file
