diff --git a/lib/simpletestlib/compatibility.php b/lib/simpletestlib/compatibility.php
index eefb602..bd65a96 100644
--- a/lib/simpletestlib/compatibility.php
+++ b/lib/simpletestlib/compatibility.php
@@ -1,16 +1,16 @@
 <?php
-    /**
+/**
      *	base include file for SimpleTest
      *	@package	SimpleTest
      *	@version	$Id$
      */
     
-    /**
+/**
      *  Static methods for compatibility between different
      *  PHP versions.
      *  @package	SimpleTest
      */
-    class SimpleTestCompatibility {
+class SimpleTestCompatibility {
     	
     	/**
     	 *	  Creates a copy whether in PHP5 or PHP4.
@@ -38,12 +38,12 @@
          *    @static
          */
         function isIdentical($first, $second) {
-            if ($first != $second) {
-                return false;
-            }
             if (version_compare(phpversion(), '5') >= 0) {
                 return SimpleTestCompatibility::_isIdenticalType($first, $second);
             }
+        if ($first != $second) {
+            return false;
+        }
             return ($first === $second);
         }
         
@@ -169,5 +169,5 @@
                 set_socket_timeout($handle, $timeout, 0);
             }
         }
-    }
+}
 ?>
\ No newline at end of file
