diff --git a/lib/simpletestlib/cookies.php b/lib/simpletestlib/cookies.php
index b346a79..abbc2ff 100644
--- a/lib/simpletestlib/cookies.php
+++ b/lib/simpletestlib/cookies.php
@@ -1,18 +1,18 @@
 <?php
-    /**
+/**
      *	Base include file for SimpleTest
      *	@package	SimpleTest
      *	@subpackage	WebTester
      *	@version	$Id$
      */
 
-    /**#@+
+/**#@+
      *	include other SimpleTest class files
      */
-    require_once(dirname(__FILE__) . '/url.php');
-    /**#@-*/
+require_once(dirname(__FILE__) . '/url.php');
+/**#@-*/
     
-    /**
+/**
      *    Cookie data holder. Cookie rules are full of pretty
      *    arbitary stuff. I have used...
      *    http://wp.netscape.com/newsref/std/cookie_spec.html
@@ -20,7 +20,7 @@
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleCookie {
+class SimpleCookie {
         var $_host;
         var $_name;
         var $_value;
@@ -217,15 +217,15 @@
             }
             return $path;
         }
-    }
+}
     
-    /**
+/**
      *    Repository for cookies. This stuff is a
      *    tiny bit browser dependent.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleCookieJar {
+class SimpleCookieJar {
         var $_cookies;
         
         /**
@@ -376,5 +376,5 @@
             }
             return $pairs;
         }
-    }
+}
 ?>
\ No newline at end of file
