diff --git a/lib/simpletestlib/dumper.php b/lib/simpletestlib/dumper.php
index ea21e64..d228103 100644
--- a/lib/simpletestlib/dumper.php
+++ b/lib/simpletestlib/dumper.php
@@ -1,23 +1,23 @@
 <?php
-    /**
+/**
      *	base include file for SimpleTest
      *	@package	SimpleTest
      *	@subpackage	UnitTester
      *	@version	$Id$
      */
-    /**
+/**
      * does type matter
      */
-    if (! defined('TYPE_MATTERS')) {
+if (! defined('TYPE_MATTERS')) {
         define('TYPE_MATTERS', true);
-    }
+}
     
-    /**
+/**
      *    Displays variables as text and does diffs.
 	 *	  @package	SimpleTest
 	 *	  @subpackage	UnitTester
      */
-    class SimpleDumper {
+class SimpleDumper {
         
         /**
          *    Renders a variable in a shorter form than print_r().
@@ -356,5 +356,5 @@
             ob_end_clean();
             return $formatted;
         }
-    }
+}
 ?>
\ No newline at end of file
