diff --git a/lib/simpletestlib/exceptions.php b/lib/simpletestlib/exceptions.php
index 407b1c2..70a243a 100644
--- a/lib/simpletestlib/exceptions.php
+++ b/lib/simpletestlib/exceptions.php
@@ -1,25 +1,25 @@
 <?php
-    /**
+/**
      *	base include file for SimpleTest
      *	@package	SimpleTest
      *	@subpackage	UnitTester
      *	@version	$Id$
      */
 
-    /**#@+
-     * Includes SimpleTest files and defined the root constant
-     * for dependent libraries.
+/**#@+
+ * Include required SimpleTest files 
      */
-    require_once(dirname(__FILE__) . '/invoker.php');
-    require_once(dirname(__FILE__) . '/expectation.php');
+require_once dirname(__FILE__) . '/invoker.php';
+require_once dirname(__FILE__) . '/expectation.php';
+/**#@-*/
 
-    /**
+/**
      *    Extension that traps exceptions and turns them into
      *    an error message. PHP5 only.
 	 *	  @package SimpleTest
 	 *	  @subpackage UnitTester
      */
-    class SimpleExceptionTrappingInvoker extends SimpleInvokerDecorator {
+class SimpleExceptionTrappingInvoker extends SimpleInvokerDecorator {
 
         /**
          *    Stores the invoker to be wrapped.
@@ -39,8 +39,10 @@
             $trap = SimpleTest::getContext()->get('SimpleExceptionTrap');
             $trap->clear();
             try {
+            $has_thrown = false;
                 parent::invoke($method);
             } catch (Exception $exception) {
+            $has_thrown = true;
 	            if (! $trap->isExpected($this->getTestCase(), $exception)) {
 	                $this->getTestCase()->exception($exception);
 	            }
@@ -49,10 +51,15 @@
 			if ($message = $trap->getOutstanding()) {
 				$this->getTestCase()->fail($message);
 			}
+        if ($has_thrown) {
+            try {
+                parent::getTestCase()->tearDown();
+            } catch (Exception $e) { }
         }
     }
+}
 
-    /**
+/**
      *    Tests exceptions either by type or the exact
      *    exception. This could be improved to accept
      *    a pattern expectation to test the error
@@ -60,7 +67,7 @@
 	 *	  @package SimpleTest
 	 *	  @subpackage UnitTester
      */
-    class ExceptionExpectation extends SimpleExpectation {
+class ExceptionExpectation extends SimpleExpectation {
         private $expected;
 
         /**
@@ -115,19 +122,19 @@
          *    @param Exception $compare     Exception to describe.
          *    @return string                Text description.
          */
-        function describeException($exception) {
+    protected function describeException($exception) {
             return get_class($exception) . ": " . $exception->getMessage();
         }
-    }
+}
 
-    /**
+/**
      *    Stores expected exceptions for when they
      *    get thrown. Saves the irritating try...catch
      *    block.
 	 *	  @package	SimpleTest
 	 *	  @subpackage	UnitTester
      */
-    class SimpleExceptionTrap {
+class SimpleExceptionTrap {
         private $expected;
         private $message;
 
@@ -187,5 +194,5 @@
             $this->expected = false;
             $this->message = false;
         }
-    }
+}
 ?>
\ No newline at end of file
