diff --git a/lib/simpletestlib/form.php b/lib/simpletestlib/form.php
index f05a7f7..937f262 100644
--- a/lib/simpletestlib/form.php
+++ b/lib/simpletestlib/form.php
@@ -1,25 +1,25 @@
 <?php
-    /**
+/**
      *	Base include file for SimpleTest.
      *	@package	SimpleTest
      *	@subpackage	WebTester
      *	@version	$Id$
      */
      
-    /**#@+
+/**#@+
      * include SimpleTest files
      */
-    require_once(dirname(__FILE__) . '/tag.php');
-    require_once(dirname(__FILE__) . '/encoding.php');
-    require_once(dirname(__FILE__) . '/selector.php');
-    /**#@-*/
+require_once(dirname(__FILE__) . '/tag.php');
+require_once(dirname(__FILE__) . '/encoding.php');
+require_once(dirname(__FILE__) . '/selector.php');
+/**#@-*/
     
-    /**
+/**
      *    Form tag class to hold widget values.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleForm {
+class SimpleForm {
         var $_method;
         var $_action;
         var $_encoding;
@@ -34,11 +34,11 @@
         /**
          *    Starts with no held controls/widgets.
          *    @param SimpleTag $tag        Form tag to read.
-         *    @param SimpleUrl $url        Location of holding page.
+     *    @param SimplePage $page      Holding page.
          */
-        function SimpleForm($tag, $url) {
+    function SimpleForm($tag, &$page) {
             $this->_method = $tag->getAttribute('method');
-            $this->_action = $this->_createAction($tag->getAttribute('action'), $url);
+        $this->_action = $this->_createAction($tag->getAttribute('action'), $page);
             $this->_encoding = $this->_setEncodingClass($tag);
             $this->_default_target = false;
             $this->_id = $tag->getAttribute('id');
@@ -90,12 +90,11 @@
          *    @param SimpleUrl $base   Page location.
          *    @return SimpleUrl        Absolute form target.
          */
-        function _createAction($action, $base) {
+    function _createAction($action, &$page) {
             if (($action === '') || ($action === false)) {
-                return $base;
+            return $page->expandUrl($page->getUrl());
             }
-            $url = new SimpleUrl($action);
-            return $url->makeAbsolute($base);
+        return $page->expandUrl(new SimpleUrl($action));;
         }
         
         /**
@@ -173,7 +172,7 @@
          */
         function _addRadioButton(&$tag) {
             if (! isset($this->_radios[$tag->getName()])) {
-                $this->_widgets[] = &new SimpleRadioGroup();
+            $this->_widgets[] = new SimpleRadioGroup();
                 $this->_radios[$tag->getName()] = count($this->_widgets) - 1;
             }
             $this->_widgets[$this->_radios[$tag->getName()]]->addWidget($tag);
@@ -192,7 +191,7 @@
                 $index = $this->_checkboxes[$tag->getName()];
                 if (! SimpleTestCompatibility::isA($this->_widgets[$index], 'SimpleCheckboxGroup')) {
                     $previous = &$this->_widgets[$index];
-                    $this->_widgets[$index] = &new SimpleCheckboxGroup();
+                $this->_widgets[$index] = new SimpleCheckboxGroup();
                     $this->_widgets[$index]->addWidget($previous);
                 }
                 $this->_widgets[$index]->addWidget($tag);
@@ -229,15 +228,19 @@
          *                                      present, nothing will be set.
          *    @access public
          */
-        function setField($selector, $value) {
+    function setField($selector, $value, $position=false) {
             $success = false;
+        $_position = 0;
             for ($i = 0, $count = count($this->_widgets); $i < $count; $i++) {
                 if ($selector->isMatch($this->_widgets[$i])) {
+                $_position++;
+                if ($position === false or $_position === (int)$position) {
                     if ($this->_widgets[$i]->setValue($value)) {
                         $success = true;
                     }
                 }
             }
+        }
             return $success;
         }
         
@@ -348,5 +351,5 @@
         function submit() {
             return $this->_encode();
         }
-    }
+}
 ?>
