diff --git a/lib/simpletestlib/invoker.php b/lib/simpletestlib/invoker.php
index 22867fa..b9f81ff 100644
--- a/lib/simpletestlib/invoker.php
+++ b/lib/simpletestlib/invoker.php
@@ -1,33 +1,33 @@
 <?php
-    /**
+/**
      *	Base include file for SimpleTest
      *	@package	SimpleTest
      *	@subpackage	UnitTester
      *	@version	$Id$
      */
 
-    /**#@+
+/**#@+
      * Includes SimpleTest files and defined the root constant
      * for dependent libraries.
      */
-    require_once(dirname(__FILE__) . '/errors.php');
-    require_once(dirname(__FILE__) . '/compatibility.php');
-    require_once(dirname(__FILE__) . '/scorer.php');
-    require_once(dirname(__FILE__) . '/expectation.php');
-    require_once(dirname(__FILE__) . '/dumper.php');
-    if (! defined('SIMPLE_TEST')) {
+require_once(dirname(__FILE__) . '/errors.php');
+require_once(dirname(__FILE__) . '/compatibility.php');
+require_once(dirname(__FILE__) . '/scorer.php');
+require_once(dirname(__FILE__) . '/expectation.php');
+require_once(dirname(__FILE__) . '/dumper.php');
+if (! defined('SIMPLE_TEST')) {
         define('SIMPLE_TEST', dirname(__FILE__) . '/');
-    }
-    /**#@-*/
+}
+/**#@-*/
 
-    /**
+/**
      *    This is called by the class runner to run a
      *    single test method. Will also run the setUp()
      *    and tearDown() methods.
 	 *	  @package SimpleTest
 	 *	  @subpackage UnitTester
      */
-    class SimpleInvoker {
+class SimpleInvoker {
         var $_test_case;
 
         /**
@@ -65,8 +65,19 @@
          */
         function invoke($method) {
             $this->_test_case->setUp();
+        // moodle hack start
+        // note: this breaks PHP4 compatibility!
+        $rethrow = null;
+        try {
             $this->_test_case->$method();
+        } catch (Exception $e) {
+            $rethrow = $e;
+        }
             $this->_test_case->tearDown();
+        if ($rethrow) {
+            throw $rethrow;
+        }
+        // moodle hack end
         }
 
         /**
@@ -78,15 +89,15 @@
         function after($method) {
             $this->_test_case->after($method);
         }
-    }
+}
 
-    /**
+/**
      *    Do nothing decorator. Just passes the invocation
      *    straight through.
 	 *	  @package SimpleTest
 	 *	  @subpackage UnitTester
      */
-    class SimpleInvokerDecorator {
+class SimpleInvokerDecorator {
         var $_invoker;
 
         /**
@@ -135,5 +146,5 @@
         function after($method) {
             $this->_invoker->after($method);
         }
-    }
+}
 ?>
