diff --git a/lib/simpletestlib/reflection_php4.php b/lib/simpletestlib/reflection_php4.php
index a624106..c30eea8 100644
--- a/lib/simpletestlib/reflection_php4.php
+++ b/lib/simpletestlib/reflection_php4.php
@@ -1,17 +1,18 @@
 <?php
-    /**
+/**
      *	base include file for SimpleTest
      *	@package	SimpleTest
      *	@subpackage	UnitTester
      *	@version	$Id$
      */
 
-    /**
+/**
      *    Version specific reflection API.
 	 *	  @package SimpleTest
 	 *	  @subpackage UnitTester
+ *  @ignore duplicate with reflection_php5.php
      */
-    class SimpleReflection {
+class SimpleReflection {
         var $_interface;
 
         /**
@@ -103,6 +104,26 @@
         }
 
         /**
+     *    Determines if the the entity is an interface, which for PHP 4
+     *    will never be the case.
+     *    @returns boolean      True if interface.
+     *    @access public
+     */
+    function isInterface() {
+        return false;
+    }
+        
+    /**
+     *    Scans for final methods, but as it's PHP 4 there
+     *    aren't any.
+     *    @returns boolean   True if the class has a final method.
+     *    @access public
+     */
+    function hasFinal() {
+        return false;
+    }
+
+    /**
          *	  Gets the source code matching the declaration
          *	  of a method.
          * 	  @param string $method		  Method name.
@@ -111,5 +132,5 @@
         function getSignature($method) {
         	return "function &$method()";
         }
-    }
+}
 ?>
\ No newline at end of file
