diff --git a/lib/simpletestlib/selector.php b/lib/simpletestlib/selector.php
index ec8f354..7c5962e 100644
--- a/lib/simpletestlib/selector.php
+++ b/lib/simpletestlib/selector.php
@@ -1,25 +1,25 @@
 <?php
-    /**
+/**
      *	Base include file for SimpleTest.
      *	@package	SimpleTest
      *	@subpackage	WebTester
      *	@version	$Id$
      */
 
-    /**#@+
+/**#@+
      * include SimpleTest files
      */
-    require_once(dirname(__FILE__) . '/tag.php');
-    require_once(dirname(__FILE__) . '/encoding.php');
-    /**#@-*/
+require_once(dirname(__FILE__) . '/tag.php');
+require_once(dirname(__FILE__) . '/encoding.php');
+/**#@-*/
 
-    /**
+/**
      *    Used to extract form elements for testing against.
      *    Searches by name attribute.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleByName {
+class SimpleByName {
         var $_name;
 
         /**
@@ -30,6 +30,10 @@
             $this->_name = $name;
         }
 
+    function getName() {
+        return $this->_name;
+    }
+
         /**
          *    Compares with name attribute of widget.
          *    @param SimpleWidget $widget    Control to compare.
@@ -38,15 +42,15 @@
         function isMatch($widget) {
             return ($widget->getName() == $this->_name);
         }
-    }
+}
 
-    /**
+/**
      *    Used to extract form elements for testing against.
      *    Searches by visible label or alt text.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleByLabel {
+class SimpleByLabel {
         var $_label;
 
         /**
@@ -69,15 +73,15 @@
             }
             return $widget->isLabel($this->_label);
         }
-    }
+}
 
-    /**
+/**
      *    Used to extract form elements for testing against.
      *    Searches dy id attribute.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleById {
+class SimpleById {
         var $_id;
 
         /**
@@ -96,15 +100,15 @@
         function isMatch($widget) {
             return $widget->isId($this->_id);
         }
-    }
+}
 
-    /**
+/**
      *    Used to extract form elements for testing against.
      *    Searches by visible label, name or alt text.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleByLabelOrName {
+class SimpleByLabelOrName {
         var $_label;
 
         /**
@@ -129,5 +133,5 @@
             }
             return ($widget->getName() == $this->_label);
         }
-    }
+}
 ?>
\ No newline at end of file
