diff --git a/lib/simpletestlib/tag.php b/lib/simpletestlib/tag.php
index c713d0a..b0d037a 100644
--- a/lib/simpletestlib/tag.php
+++ b/lib/simpletestlib/tag.php
@@ -1,24 +1,24 @@
 <?php
-    /**
+/**
      *	Base include file for SimpleTest.
      *	@package	SimpleTest
      *	@subpackage	WebTester
      *	@version	$Id$
      */
      
-    /**#@+
+/**#@+
      * include SimpleTest files
      */
-    require_once(dirname(__FILE__) . '/parser.php');
-    require_once(dirname(__FILE__) . '/encoding.php');
-    /**#@-*/
+require_once(dirname(__FILE__) . '/parser.php');
+require_once(dirname(__FILE__) . '/encoding.php');
+/**#@-*/
    
-    /**
+/**
      *    HTML or XML tag.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleTag {
+class SimpleTag {
         var $_name;
         var $_attributes;
         var $_content;
@@ -148,14 +148,40 @@
         function isId($id) {
             return ($this->getAttribute('id') == $id);
         }
+}
+
+/**
+ *    Base url.
+ *    @package SimpleTest
+ *    @subpackage WebTester
+ */
+class SimpleBaseTag extends SimpleTag {
+    
+    /**
+     *    Starts with a named tag with attributes only.
+     *    @param hash $attributes    Attribute names and
+     *                               string values.
+     */
+    function SimpleBaseTag($attributes) {
+        $this->SimpleTag('base', $attributes);
     }
     
     /**
+     *    Base tag is not a block tag.
+     *    @return boolean       false
+     *    @access public
+     */
+    function expectEndTag() {
+        return false;
+    }
+}
+
+/**
      *    Page title.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleTitleTag extends SimpleTag {
+class SimpleTitleTag extends SimpleTag {
         
         /**
          *    Starts with a named tag with attributes only.
@@ -165,14 +191,14 @@
         function SimpleTitleTag($attributes) {
             $this->SimpleTag('title', $attributes);
         }
-    }
+}
     
-    /**
+/**
      *    Link.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleAnchorTag extends SimpleTag {
+class SimpleAnchorTag extends SimpleTag {
         
         /**
          *    Starts with a named tag with attributes only.
@@ -195,14 +221,14 @@
             }
             return $url;
         }
-    }
+}
     
-    /**
+/**
      *    Form element.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleWidget extends SimpleTag {
+class SimpleWidget extends SimpleTag {
         var $_value;
         var $_label;
         var $_is_set;
@@ -304,14 +330,14 @@
                 $encoding->add($this->getName(), $this->getValue());
             }
         }
-    }
+}
     
-    /**
+/**
      *    Text, password and hidden field.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleTextTag extends SimpleWidget {
+class SimpleTextTag extends SimpleWidget {
         
         /**
          *    Starts with a named tag with attributes only.
@@ -347,14 +373,14 @@
             }
             return parent::setValue($value);
         }
-    }
+}
     
-    /**
+/**
      *    Submit button as input tag.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleSubmitTag extends SimpleWidget {
+class SimpleSubmitTag extends SimpleWidget {
         
         /**
          *    Starts with a named tag with attributes only.
@@ -405,14 +431,14 @@
         function isLabel($label) {
             return trim($label) == trim($this->getLabel());
         }
-    }
+}
       
-    /**
+/**
      *    Image button as input tag.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleImageSubmitTag extends SimpleWidget {
+class SimpleImageSubmitTag extends SimpleWidget {
         
         /**
          *    Starts with a named tag with attributes only.
@@ -480,14 +506,14 @@
                 $encoding->add('y', $y);
             }
         }
-    }
+}
       
-    /**
+/**
      *    Submit button as button tag.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleButtonTag extends SimpleWidget {
+class SimpleButtonTag extends SimpleWidget {
         
         /**
          *    Starts with a named tag with attributes only.
@@ -537,14 +563,14 @@
         function isLabel($label) {
             return trim($label) == trim($this->getLabel());
         }
-    }
+}
   
-    /**
+/**
      *    Content tag for text area.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleTextAreaTag extends SimpleWidget {
+class SimpleTextAreaTag extends SimpleWidget {
         
         /**
          *    Starts with a named tag with attributes only.
@@ -622,14 +648,14 @@
         function isPrivateContent() {
             return true;
         }
-    }
+}
     
-    /**
+/**
      *    File upload widget.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleUploadTag extends SimpleWidget {
+class SimpleUploadTag extends SimpleWidget {
         
         /**
          *    Starts with attributes only.
@@ -663,14 +689,14 @@
                     implode('', file($this->getValue())),
                     basename($this->getValue()));
         }
-    }
+}
     
-    /**
+/**
      *    Drop down widget.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleSelectionTag extends SimpleWidget {
+class SimpleSelectionTag extends SimpleWidget {
         var $_options;
         var $_choice;
         
@@ -750,14 +776,14 @@
             }
             return $this->_options[$this->_choice]->getValue();
         }
-    }
+}
     
-    /**
+/**
      *    Drop down widget.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class MultipleSelectionTag extends SimpleWidget {
+class MultipleSelectionTag extends SimpleWidget {
         var $_options;
         var $_values;
         
@@ -845,14 +871,14 @@
             }
             return $this->_values;
         }
-    }
+}
     
-    /**
+/**
      *    Option for selection field.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleOptionTag extends SimpleWidget {
+class SimpleOptionTag extends SimpleWidget {
         
         /**
          *    Stashes the attributes.
@@ -906,14 +932,14 @@
         function isPrivateContent() {
             return true;
         }
-    }
+}
     
-    /**
+/**
      *    Radio button.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleRadioButtonTag extends SimpleWidget {
+class SimpleRadioButtonTag extends SimpleWidget {
         
         /**
          *    Stashes the attributes.
@@ -946,7 +972,7 @@
             if ($value === false) {
                 return parent::setValue($value);
             }
-            if ($value !== $this->getAttribute('value')) {
+        if ($value != $this->getAttribute('value')) {
                 return false;
             }
             return parent::setValue($value);
@@ -963,14 +989,14 @@
             }
             return false;
         }
-    }
+}
     
-    /**
+/**
      *    Checkbox widget.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleCheckboxTag extends SimpleWidget {
+class SimpleCheckboxTag extends SimpleWidget {
         
         /**
          *    Starts with attributes only.
@@ -1027,14 +1053,14 @@
             }
             return false;
         }
-    }
+}
     
-    /**
+/**
      *    A group of multiple widgets with some shared behaviour.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleTagGroup {
+class SimpleTagGroup {
         var $_widgets = array();
 
         /**
@@ -1117,14 +1143,14 @@
         function write(&$encoding) {
             $encoding->add($this->getName(), $this->getValue());
         }
-    }
+}
 
-    /**
+/**
      *    A group of tags with the same name within a form.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleCheckboxGroup extends SimpleTagGroup {
+class SimpleCheckboxGroup extends SimpleTagGroup {
         
         /**
          *    Accessor for current selected widget or false
@@ -1239,15 +1265,15 @@
             }
             return $value;
         }
-    }
+}
 
-    /**
+/**
      *    A group of tags with the same name within a form.
      *    Used for radio buttons.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleRadioGroup extends SimpleTagGroup {
+class SimpleRadioGroup extends SimpleTagGroup {
         
         /**
          *    Each tag is tried in turn until one is
@@ -1319,14 +1345,14 @@
             }
             return false;
         }
-    }
+}
     
-    /**
+/**
      *    Tag to keep track of labels.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleLabelTag extends SimpleTag {
+class SimpleLabelTag extends SimpleTag {
         
         /**
          *    Starts with a named tag with attributes only.
@@ -1345,14 +1371,14 @@
         function getFor() {
             return $this->getAttribute('for');
         }
-    }
+}
     
-    /**
+/**
      *    Tag to aid parsing the form.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleFormTag extends SimpleTag {
+class SimpleFormTag extends SimpleTag {
         
         /**
          *    Starts with a named tag with attributes only.
@@ -1362,14 +1388,14 @@
         function SimpleFormTag($attributes) {
             $this->SimpleTag('form', $attributes);
         }
-    }
+}
     
-    /**
+/**
      *    Tag to aid parsing the frames in a page.
 	 *    @package SimpleTest
 	 *    @subpackage WebTester
      */
-    class SimpleFrameTag extends SimpleTag {
+class SimpleFrameTag extends SimpleTag {
         
         /**
          *    Starts with a named tag with attributes only.
@@ -1388,5 +1414,5 @@
         function expectEndTag() {
             return false;
         }
-    }
+}
 ?>
\ No newline at end of file
