diff --git a/lib/excel/Worksheet.php b/lib/excel/Worksheet.php
index f512fb8..d7c5cfc 100644
--- a/lib/excel/Worksheet.php
+++ b/lib/excel/Worksheet.php
@@ -909,7 +909,7 @@ class Worksheet extends BIFFwriter
         $row     = $match[2];
     
         // Convert base26 column string to number
-        $chars = split('', $col);
+        $chars = str_split($col);
         $expn  = 0;
         $col   = 0;
     
@@ -1264,10 +1264,10 @@ class Worksheet extends BIFFwriter
         }
     
         // Strip the '=' or '@' sign at the beginning of the formula string
-        if (ereg("^=",$formula)) {
+        if (preg_match("/^=/",$formula)) {
             $formula = preg_replace("/(^=)/","",$formula);
         }
-        elseif(ereg("^@",$formula)) {
+        elseif(preg_match("/^@/",$formula)) {
             $formula = preg_replace("/(^@)/","",$formula);
         }
         else {
@@ -1530,13 +1530,13 @@ class Worksheet extends BIFFwriter
         // Determine if the link contains a sheet reference and change some of the
         // parameters accordingly.
         // Split the dir name and sheet name (if it exists)
-        list($dir_long , $sheet) = split('/\#/', $url);
+        list($dir_long , $sheet) = explode('/#/', $url);
         $link_type               = 0x01 | $absolute;
     
         if (isset($sheet)) {
             $link_type |= 0x08;
             $sheet_len  = pack("V", strlen($sheet) + 0x01);
-            $sheet      = join("\0", split('', $sheet));
+            $sheet      = join("\0", str_split($sheet));
             $sheet     .= "\0\0\0";
         }
         else {
@@ -1555,7 +1555,7 @@ class Worksheet extends BIFFwriter
         $dir_short   = preg_replace('/\.\.\\/', '', $dir_long) . "\0";
     
         // Store the long dir name as a wchar string (non-null terminated)
-        $dir_long       = join("\0", split('', $dir_long));
+        $dir_long       = join("\0", str_split($dir_long));
         $dir_long       = $dir_long . "\0";
     
         // Pack the lengths of the dir strings
@@ -1644,7 +1644,7 @@ class Worksheet extends BIFFwriter
         if (defined $sheet) {
             $link_type |= 0x08;
             $sheet_len  = pack("V", length($sheet) + 0x01);
-            $sheet      = join("\0", split('', $sheet));
+            $sheet      = join("\0", str_split($sheet));
             $sheet     .= "\0\0\0";
     }
         else {
@@ -1665,7 +1665,7 @@ class Worksheet extends BIFFwriter
     
     
         # Store the long dir name as a wchar string (non-null terminated)
-        $dir_long       = join("\0", split('', $dir_long));
+        $dir_long       = join("\0", str_split($dir_long));
         $dir_long       = $dir_long . "\0";
     
     
