diff --git a/lib/pear/Spreadsheet/Excel/Writer/Worksheet.php b/lib/pear/Spreadsheet/Excel/Writer/Worksheet.php
index 1b53972..07ed2ef 100644
--- a/lib/pear/Spreadsheet/Excel/Writer/Worksheet.php
+++ b/lib/pear/Spreadsheet/Excel/Writer/Worksheet.php
@@ -1319,7 +1319,7 @@ class Spreadsheet_Excel_Writer_Worksheet extends Spreadsheet_Excel_Writer_BIFFwr
         $row     = $match[2];
 
         // Convert base26 column string to number
-        $chars = split('', $col);
+        $chars = str_split($col);
         $expn  = 0;
         $col   = 0;
 
@@ -2038,7 +2038,7 @@ class Spreadsheet_Excel_Writer_Worksheet extends Spreadsheet_Excel_Writer_BIFFwr
         // parameters accordingly.
         // Split the dir name and sheet name (if it exists)
         /*if (preg_match("/\#/", $url)) {
-            list($dir_long, $sheet) = split("\#", $url);
+            list($dir_long, $sheet) = explode("#", $url);
         } else {
             $dir_long = $url;
         }
@@ -2046,7 +2046,7 @@ class Spreadsheet_Excel_Writer_Worksheet extends Spreadsheet_Excel_Writer_BIFFwr
         if (isset($sheet)) {
             $link_type |= 0x08;
             $sheet_len  = pack("V", strlen($sheet) + 0x01);
-            $sheet      = join("\0", split('', $sheet));
+            $sheet      = join("\0", str_split($sheet));
             $sheet     .= "\0\0\0";
         } else {
             $sheet_len   = '';
@@ -2070,7 +2070,7 @@ class Spreadsheet_Excel_Writer_Worksheet extends Spreadsheet_Excel_Writer_BIFFwr
         $dir_short   = preg_replace("/\.\.\\\/", '', $dir_long) . "\0";
     
         // Store the long dir name as a wchar string (non-null terminated)
-        //$dir_long       = join("\0", split('', $dir_long));
+        //$dir_long       = join("\0", str_split($dir_long));
         $dir_long       = $dir_long . "\0";
     
         // Pack the lengths of the dir strings
