diff --git a/course/format/topics/format.php b/course/format/topics/format.php
index 6045f65..8ad112e 100644
--- a/course/format/topics/format.php
+++ b/course/format/topics/format.php
@@ -1,179 +1,158 @@
-<?php // $Id$
-      // Display the whole course as "topics" made of of modules
-      // In fact, this is very similar to the "weeks" format, in that
-      // each "topic" is actually a week.  The main difference is that
-      // the dates aren't printed - it's just an aesthetic thing for
-      // courses that aren't so rigidly defined by time.
-      // Included from "view.php"
-      
-
-    require_once($CFG->libdir.'/ajax/ajaxlib.php');
-  
-    $topic = optional_param('topic', -1, PARAM_INT);
-
-    // Bounds for block widths
-    // more flexible for theme designers taken from theme config.php
-    $lmin = (empty($THEME->block_l_min_width)) ? 100 : $THEME->block_l_min_width;
-    $lmax = (empty($THEME->block_l_max_width)) ? 210 : $THEME->block_l_max_width;
-    $rmin = (empty($THEME->block_r_min_width)) ? 100 : $THEME->block_r_min_width;
-    $rmax = (empty($THEME->block_r_max_width)) ? 210 : $THEME->block_r_max_width;
-
-    define('BLOCK_L_MIN_WIDTH', $lmin);
-    define('BLOCK_L_MAX_WIDTH', $lmax);
-    define('BLOCK_R_MIN_WIDTH', $rmin);
-    define('BLOCK_R_MAX_WIDTH', $rmax);
-
-    $preferred_width_left  = bounded_number(BLOCK_L_MIN_WIDTH, blocks_preferred_width($pageblocks[BLOCK_POS_LEFT]),  
-                                            BLOCK_L_MAX_WIDTH);
-    $preferred_width_right = bounded_number(BLOCK_R_MIN_WIDTH, blocks_preferred_width($pageblocks[BLOCK_POS_RIGHT]), 
-                                            BLOCK_R_MAX_WIDTH);
-
-    if ($topic != -1) {
+<?php
+
+// Display the whole course as "topics" made of of modules
+// Included from "view.php"
+/**
+ * Evaluation topics format for course display - NO layout tables, for accessibility, etc.
+ *
+ * A duplicate course format to enable the Moodle development team to evaluate
+ * CSS for the multi-column layout in place of layout tables.
+ * Less risk for the Moodle 1.6 beta release.
+ *   1. Straight copy of topics/format.php
+ *   2. Replace <table> and <td> with DIVs; inline styles.
+ *   3. Reorder columns so that in linear view content is first then blocks;
+ * styles to maintain original graphical (side by side) view.
+ *
+ * Target: 3-column graphical view using relative widths for pixel screen sizes
+ * 800x600, 1024x768... on IE6, Firefox. Below 800 columns will shift downwards.
+ *
+ * http://www.maxdesign.com.au/presentation/em/ Ideal length for content.
+ * http://www.svendtofte.com/code/max_width_in_ie/ Max width in IE.
+ *
+ * @copyright &copy; 2006 The Open University
+ * @author N.D.Freear@open.ac.uk, and others.
+ * @license http://www.gnu.org/copyleft/gpl.html GNU Public License
+ * @package
+ */
+
+defined('MOODLE_INTERNAL') || die();
+
+require_once($CFG->libdir.'/filelib.php');
+require_once($CFG->libdir.'/completionlib.php');
+
+$topic = optional_param('topic', -1, PARAM_INT);
+
+if ($topic != -1) {
         $displaysection = course_set_display($course->id, $topic);
-    } else {
-        if (isset($USER->display[$course->id])) {       // for admins, mostly
-            $displaysection = $USER->display[$course->id];
-        } else {
-            $displaysection = course_set_display($course->id, 0);
-        }
-    }
+} else {
+    $displaysection = course_get_display($course->id);
+}
 
-    $context = get_context_instance(CONTEXT_COURSE, $course->id);
+$context = get_context_instance(CONTEXT_COURSE, $course->id);
 
-    if (($marker >=0) && has_capability('moodle/course:setcurrentsection', $context) && confirm_sesskey()) {
+if (($marker >=0) && has_capability('moodle/course:setcurrentsection', $context) && confirm_sesskey()) {
         $course->marker = $marker;
-        if (! set_field("course", "marker", $marker, "id", $course->id)) {
-            error("Could not mark that topic for this course");
-        }
-    }
-
-    $streditsummary   = get_string('editsummary');
-    $stradd           = get_string('add');
-    $stractivities    = get_string('activities');
-    $strshowalltopics = get_string('showalltopics');
-    $strtopic         = get_string('topic');
-    $strgroups        = get_string('groups');
-    $strgroupmy       = get_string('groupmy');
-    $editing          = $PAGE->user_is_editing();
-
-    if ($editing) {
-        $strstudents = moodle_strtolower($course->students);
-        $strtopichide = get_string('topichide', '', $strstudents);
-        $strtopicshow = get_string('topicshow', '', $strstudents);
+    $DB->set_field("course", "marker", $marker, array("id"=>$course->id));
+}
+
+$streditsummary  = get_string('editsummary');
+$stradd          = get_string('add');
+$stractivities   = get_string('activities');
+$strshowalltopics = get_string('showalltopics');
+$strtopic         = get_string('topic');
+$strgroups       = get_string('groups');
+$strgroupmy      = get_string('groupmy');
+$editing         = $PAGE->user_is_editing();
+
+if ($editing) {
+    $strtopichide = get_string('hidetopicfromothers');
+    $strtopicshow = get_string('showtopicfromothers');
         $strmarkthistopic = get_string('markthistopic');
         $strmarkedthistopic = get_string('markedthistopic');
         $strmoveup = get_string('moveup');
         $strmovedown = get_string('movedown');
-    }
+}
 
+// Print the Your progress icon if the track completion is enabled
+$completioninfo = new completion_info($course);
+echo $completioninfo->display_help_icon();
 
-/// Layout the whole page as three big columns.
-    echo '<table id="layout-table" cellspacing="0" summary="'.get_string('layouttable').'"><tr>';
-
-/// The left column ...
-    $lt = (empty($THEME->layouttable)) ? array('left', 'middle', 'right') : $THEME->layouttable;
-    foreach ($lt as $column) {
-        switch ($column) {
-            case 'left':
-
-    if (blocks_have_content($pageblocks, BLOCK_POS_LEFT) || $editing) {
-        echo '<td style="width:'.$preferred_width_left.'px" id="left-column">';
-        print_container_start();
-        blocks_print_group($PAGE, $pageblocks, BLOCK_POS_LEFT);
-        print_container_end();
-        echo '</td>';
-    }
-
-            break;
-            case 'middle':
-/// Start main column
-    echo '<td id="middle-column">';
-    print_container_start();
-    echo skip_main_destination();
+echo $OUTPUT->heading(get_string('topicoutline'), 2, 'headingblock header outline');
 
-    print_heading_block(get_string('topicoutline'), 'outline');
-
-    echo '<table class="topics" width="100%" summary="'.get_string('layouttable').'">';
+// Note, an ordered list would confuse - "1" could be the clipboard or summary.
+echo "<ul class='topics'>\n";
 
 /// If currently moving a file then show the current clipboard
-    if (ismoving($course->id)) {
-        $stractivityclipboard = strip_tags(get_string('activityclipboard', '', addslashes($USER->activitycopyname)));
+if (ismoving($course->id)) {
+    $stractivityclipboard = strip_tags(get_string('activityclipboard', '', $USER->activitycopyname));
         $strcancel= get_string('cancel');
-        echo '<tr class="clipboard">';
-        echo '<td colspan="3">';
-        echo $stractivityclipboard.'&nbsp;&nbsp;(<a href="mod.php?cancelcopy=true&amp;sesskey='.$USER->sesskey.'">'.$strcancel.'</a>)';
-        echo '</td>';
-        echo '</tr>';
+    echo '<li class="clipboard">';
+    echo $stractivityclipboard.'&nbsp;&nbsp;(<a href="mod.php?cancelcopy=true&amp;sesskey='.sesskey().'">'.$strcancel.'</a>)';
+    echo "</li>\n";
+}
+
+/// Print Section 0 with general activities
+
+$section = 0;
+$thissection = $sections[$section];
+unset($sections[0]);
+
+if ($thissection->summary or $thissection->sequence or $PAGE->user_is_editing()) {
+
+    // Note, no need for a 'left side' cell or DIV.
+    // Note, 'right side' is BEFORE content.
+    echo '<li id="section-0" class="section main clearfix" >';
+    echo '<div class="left side">&nbsp;</div>';
+    echo '<div class="right side" >&nbsp;</div>';
+    echo '<div class="content">';
+    if (!is_null($thissection->name)) {
+        echo $OUTPUT->heading(format_string($thissection->name, true, array('context' => $context)), 3, 'sectionname');
     }
-
-/// Print Section 0
-
-    $section = 0;
-    $thissection = $sections[$section];
-
-    if ($thissection->summary or $thissection->sequence or isediting($course->id)) {
-        echo '<tr id="section-0" class="section main">';
-        echo '<td id="sectionblock-0" class="left side">&nbsp;</td>';
-        echo '<td class="content">';
-        
         echo '<div class="summary">';
+
+    $coursecontext = get_context_instance(CONTEXT_COURSE, $course->id);
+    $summarytext = file_rewrite_pluginfile_urls($thissection->summary, 'pluginfile.php', $coursecontext->id, 'course', 'section', $thissection->id);
+    $summaryformatoptions = new stdClass();
         $summaryformatoptions->noclean = true;
-        echo format_text($thissection->summary, FORMAT_HTML, $summaryformatoptions);
+    $summaryformatoptions->overflowdiv = true;
+    echo format_text($summarytext, $thissection->summaryformat, $summaryformatoptions);
 
-        if (isediting($course->id) && has_capability('moodle/course:update', get_context_instance(CONTEXT_COURSE, $course->id))) {
+    if ($PAGE->user_is_editing() && has_capability('moodle/course:update', $coursecontext)) {
             echo '<a title="'.$streditsummary.'" '.
-                 ' href="editsection.php?id='.$thissection->id.'"><img src="'.$CFG->pixpath.'/t/edit.gif" '.
-                 ' alt="'.$streditsummary.'" /></a><br /><br />';
+             ' href="editsection.php?id='.$thissection->id.'"><img src="'.$OUTPUT->pix_url('t/edit') . '" '.
+             ' class="icon edit" alt="'.$streditsummary.'" /></a>';
         }
         echo '</div>';
 
         print_section($course, $thissection, $mods, $modnamesused);
 
-        if (isediting($course->id)) {
+    if ($PAGE->user_is_editing()) {
             print_section_add_menus($course, $section, $modnames);
         }
 
-        echo '</td>';
-        echo '<td class="right side">&nbsp;</td>';
-        echo '</tr>';
-        echo '<tr class="section separator"><td colspan="3" class="spacer"></td></tr>';
-    }
+    echo '</div>';
+    echo "</li>\n";
+}
 
 
 /// Now all the normal modules by topic
 /// Everything below uses "section" terminology - each "section" is a topic.
 
-    $timenow = time();
-    $section = 1;
-    $sectionmenu = array();
+$timenow = time();
+$section = 1;
+$sectionmenu = array();
 
-    while ($section <= $course->numsections) {
+while ($section <= $course->numsections) {
 
         if (!empty($sections[$section])) {
             $thissection = $sections[$section];
 
         } else {
-            unset($thissection);
+        $thissection = new stdClass;
             $thissection->course = $course->id;   // Create a new section structure
             $thissection->section = $section;
+        $thissection->name    = null;
             $thissection->summary = '';
+        $thissection->summaryformat = FORMAT_HTML;
             $thissection->visible = 1;
-            if (!$thissection->id = insert_record('course_sections', $thissection)) {
-                notify('Error inserting new topic!');
-            }
+        $thissection->id = $DB->insert_record('course_sections', $thissection);
         }
 
         $showsection = (has_capability('moodle/course:viewhiddensections', $context) or $thissection->visible or !$course->hiddensections);
 
-        if (!empty($displaysection) and $displaysection != $section) {
+    if (!empty($displaysection) and $displaysection != $section) {  // Check this topic is visible
             if ($showsection) {
-                $strsummary = strip_tags(format_string($thissection->summary,true));
-                if (strlen($strsummary) < 57) {
-                    $strsummary = ' - '.$strsummary;
-                } else {
-                    $strsummary = ' - '.substr($strsummary, 0, 60).'...';
-                }
-                $sectionmenu['topic='.$section] = s($section.$strsummary);
+            $sectionmenu[$section] = get_section_name($course, $thissection);
             }
             $section++;
             continue;
@@ -193,102 +172,118 @@
                 $sectionstyle = '';
             }
 
-            echo '<tr id="section-'.$section.'" class="section main'.$sectionstyle.'">';
-            echo '<td id="sectionblock-'.$section.'" class="left side">'.$currenttext.$section.'</td>';
-
-            echo '<td class="content">';
-            if (!has_capability('moodle/course:viewhiddensections', $context) and !$thissection->visible) {   // Hidden for students
-                echo get_string('notavailable');
-            } else {
-                echo '<div class="summary">';
-                $summaryformatoptions->noclean = true;
-                echo format_text($thissection->summary, FORMAT_HTML, $summaryformatoptions);
-
-                if (isediting($course->id) && has_capability('moodle/course:update', get_context_instance(CONTEXT_COURSE, $course->id))) {
-                    echo ' <a title="'.$streditsummary.'" href="editsection.php?id='.$thissection->id.'">'.
-                         '<img src="'.$CFG->pixpath.'/t/edit.gif" alt="'.$streditsummary.'" /></a><br /><br />';
-                }
-                echo '</div>';
-
-                print_section($course, $thissection, $mods, $modnamesused);
+        echo '<li id="section-'.$section.'" class="section main clearfix'.$sectionstyle.'" >'; //'<div class="left side">&nbsp;</div>';
 
-                if (isediting($course->id)) {
-                    print_section_add_menus($course, $section, $modnames);
-                }
-            }
-            echo '</td>';
+            echo '<div class="left side">'.$currenttext.$section.'</div>';
+        // Note, 'right side' is BEFORE content.
+        echo '<div class="right side">';
 
-            echo '<td class="right side">';
             if ($displaysection == $section) {      // Show the zoom boxes
                 echo '<a href="view.php?id='.$course->id.'&amp;topic=0#section-'.$section.'" title="'.$strshowalltopics.'">'.
-                     '<img src="'.$CFG->pixpath.'/i/all.gif" class="icon topicall" alt="'.$strshowalltopics.'" /></a><br />';
+                 '<img src="'.$OUTPUT->pix_url('i/all') . '" class="icon" alt="'.$strshowalltopics.'" /></a><br />';
             } else {
-                $strshowonlytopic = get_string('showonlytopic', '', $section);
+            $strshowonlytopic = get_string("showonlytopic", "", $section);
                 echo '<a href="view.php?id='.$course->id.'&amp;topic='.$section.'" title="'.$strshowonlytopic.'">'.
-                     '<img src="'.$CFG->pixpath.'/i/one.gif" class="icon topicone" alt="'.$strshowonlytopic.'" /></a><br />';
+                 '<img src="'.$OUTPUT->pix_url('i/one') . '" class="icon" alt="'.$strshowonlytopic.'" /></a><br />';
             }
 
-            if (isediting($course->id) && has_capability('moodle/course:update', get_context_instance(CONTEXT_COURSE, $course->id))) {
+        if ($PAGE->user_is_editing() && has_capability('moodle/course:update', get_context_instance(CONTEXT_COURSE, $course->id))) {
+
                 if ($course->marker == $section) {  // Show the "light globe" on/off
-                    echo '<a href="view.php?id='.$course->id.'&amp;marker=0&amp;sesskey='.$USER->sesskey.'#section-'.$section.'" title="'.$strmarkedthistopic.'">'.
-                         '<img src="'.$CFG->pixpath.'/i/marked.gif" alt="'.$strmarkedthistopic.'" /></a><br />';
+                echo '<a href="view.php?id='.$course->id.'&amp;marker=0&amp;sesskey='.sesskey().'#section-'.$section.'" title="'.$strmarkedthistopic.'">'.'<img src="'.$OUTPUT->pix_url('i/marked') . '" alt="'.$strmarkedthistopic.'" /></a><br />';
                 } else {
-                    echo '<a href="view.php?id='.$course->id.'&amp;marker='.$section.'&amp;sesskey='.$USER->sesskey.'#section-'.$section.'" title="'.$strmarkthistopic.'">'.
-                         '<img src="'.$CFG->pixpath.'/i/marker.gif" alt="'.$strmarkthistopic.'" /></a><br />';
+                echo '<a href="view.php?id='.$course->id.'&amp;marker='.$section.'&amp;sesskey='.sesskey().'#section-'.$section.'" title="'.$strmarkthistopic.'">'.'<img src="'.$OUTPUT->pix_url('i/marker') . '" alt="'.$strmarkthistopic.'" /></a><br />';
                 }
 
                 if ($thissection->visible) {        // Show the hide/show eye
-                    echo '<a href="view.php?id='.$course->id.'&amp;hide='.$section.'&amp;sesskey='.$USER->sesskey.'#section-'.$section.'" title="'.$strtopichide.'">'.
-                         '<img src="'.$CFG->pixpath.'/i/hide.gif" alt="'.$strtopichide.'" /></a><br />';
+                echo '<a href="view.php?id='.$course->id.'&amp;hide='.$section.'&amp;sesskey='.sesskey().'#section-'.$section.'" title="'.$strtopichide.'">'.
+                     '<img src="'.$OUTPUT->pix_url('i/hide') . '" class="icon hide" alt="'.$strtopichide.'" /></a><br />';
                 } else {
-                    echo '<a href="view.php?id='.$course->id.'&amp;show='.$section.'&amp;sesskey='.$USER->sesskey.'#section-'.$section.'" title="'.$strtopicshow.'">'.
-                         '<img src="'.$CFG->pixpath.'/i/show.gif" alt="'.$strtopicshow.'" /></a><br />';
+                echo '<a href="view.php?id='.$course->id.'&amp;show='.$section.'&amp;sesskey='.sesskey().'#section-'.$section.'" title="'.$strtopicshow.'">'.
+                     '<img src="'.$OUTPUT->pix_url('i/show') . '" class="icon hide" alt="'.$strtopicshow.'" /></a><br />';
                 }
-
                 if ($section > 1) {                       // Add a arrow to move section up
-                    echo '<a href="view.php?id='.$course->id.'&amp;random='.rand(1,10000).'&amp;section='.$section.'&amp;move=-1&amp;sesskey='.$USER->sesskey.'#section-'.($section-1).'" title="'.$strmoveup.'">'.
-                         '<img src="'.$CFG->pixpath.'/t/up.gif" alt="'.$strmoveup.'" /></a><br />';
+                echo '<a href="view.php?id='.$course->id.'&amp;random='.rand(1,10000).'&amp;section='.$section.'&amp;move=-1&amp;sesskey='.sesskey().'#section-'.($section-1).'" title="'.$strmoveup.'">'.
+                     '<img src="'.$OUTPUT->pix_url('t/up') . '" class="icon up" alt="'.$strmoveup.'" /></a><br />';
                 }
 
                 if ($section < $course->numsections) {    // Add a arrow to move section down
-                    echo '<a href="view.php?id='.$course->id.'&amp;random='.rand(1,10000).'&amp;section='.$section.'&amp;move=1&amp;sesskey='.$USER->sesskey.'#section-'.($section+1).'" title="'.$strmovedown.'">'.
-                         '<img src="'.$CFG->pixpath.'/t/down.gif" alt="'.$strmovedown.'" /></a><br />';
+                echo '<a href="view.php?id='.$course->id.'&amp;random='.rand(1,10000).'&amp;section='.$section.'&amp;move=1&amp;sesskey='.sesskey().'#section-'.($section+1).'" title="'.$strmovedown.'">'.
+                     '<img src="'.$OUTPUT->pix_url('t/down') . '" class="icon down" alt="'.$strmovedown.'" /></a><br />';
+            }
                 }
+        echo '</div>';
 
+        echo '<div class="content">';
+        if (!has_capability('moodle/course:viewhiddensections', $context) and !$thissection->visible) {   // Hidden for students
+            echo get_string('notavailable');
+        } else {
+            if (!is_null($thissection->name)) {
+                echo $OUTPUT->heading(format_string($thissection->name, true, array('context' => $context)), 3, 'sectionname');
+            }
+            echo '<div class="summary">';
+            if ($thissection->summary) {
+                $coursecontext = get_context_instance(CONTEXT_COURSE, $course->id);
+                $summarytext = file_rewrite_pluginfile_urls($thissection->summary, 'pluginfile.php', $coursecontext->id, 'course', 'section', $thissection->id);
+                $summaryformatoptions = new stdClass();
+                $summaryformatoptions->noclean = true;
+                $summaryformatoptions->overflowdiv = true;
+                echo format_text($summarytext, $thissection->summaryformat, $summaryformatoptions);
+            } else {
+               echo '&nbsp;';
             }
 
-            echo '</td></tr>';
-            echo '<tr class="section separator"><td colspan="3" class="spacer"></td></tr>';
+            if ($PAGE->user_is_editing() && has_capability('moodle/course:update', get_context_instance(CONTEXT_COURSE, $course->id))) {
+                echo ' <a title="'.$streditsummary.'" href="editsection.php?id='.$thissection->id.'">'.
+                     '<img src="'.$OUTPUT->pix_url('t/edit') . '" class="icon edit" alt="'.$streditsummary.'" /></a><br /><br />';
         }
+            echo '</div>';
 
-        $section++;
+            print_section($course, $thissection, $mods, $modnamesused);
+            echo '<br />';
+            if ($PAGE->user_is_editing()) {
+                print_section_add_menus($course, $section, $modnames);
+            }
     }
-    echo '</table>';
 
-    if (!empty($sectionmenu)) {
-        echo '<div class="jumpmenu">';
-        echo popup_form($CFG->wwwroot.'/course/view.php?id='.$course->id.'&amp;', $sectionmenu,
-                   'sectionmenu', '', get_string('jumpto'), '', '', true);
         echo '</div>';
+        echo "</li>\n";
     }
 
-    print_container_end();
-    echo '</td>';
-
-            break;
-            case 'right':
-    // The right column
-    if (blocks_have_content($pageblocks, BLOCK_POS_RIGHT) || $editing) {
-        echo '<td style="width:'.$preferred_width_right.'px" id="right-column">';
-        print_container_start();
-        blocks_print_group($PAGE, $pageblocks, BLOCK_POS_RIGHT);
-        print_container_end();
-        echo '</td>';
-    }
+    unset($sections[$section]);
+    $section++;
+}
 
-            break;
+if (!$displaysection and $PAGE->user_is_editing() and has_capability('moodle/course:update', get_context_instance(CONTEXT_COURSE, $course->id))) {
+    // print stealth sections if present
+    $modinfo = get_fast_modinfo($course);
+    foreach ($sections as $section=>$thissection) {
+        if (empty($modinfo->sections[$section])) {
+            continue;
         }
+
+        echo '<li id="section-'.$section.'" class="section main clearfix orphaned hidden">'; //'<div class="left side">&nbsp;</div>';
+
+        echo '<div class="left side">';
+        echo '</div>';
+        // Note, 'right side' is BEFORE content.
+        echo '<div class="right side">';
+        echo '</div>';
+        echo '<div class="content">';
+        echo $OUTPUT->heading(get_string('orphanedactivities'), 3, 'sectionname');
+        print_section($course, $thissection, $mods, $modnamesused);
+        echo '</div>';
+        echo "</li>\n";
     }
-    echo '</tr></table>';
+}
+
+
+echo "</ul>\n";
     
-?>
+if (!empty($sectionmenu)) {
+    $select = new single_select(new moodle_url('/course/view.php', array('id'=>$course->id)), 'topic', $sectionmenu);
+    $select->label = get_string('jumpto');
+    $select->class = 'jumpmenu';
+    $select->formid = 'sectionmenu';
+    echo $OUTPUT->render($select);
+}
