diff --git a/admin/auth.php b/admin/auth.php
index 41faab5..8345aad 100644
--- a/admin/auth.php
+++ b/admin/auth.php
@@ -14,14 +14,13 @@ require_once($CFG->libdir.'/tablelib.php');
 require_login();
 require_capability('moodle/site:config', get_context_instance(CONTEXT_SYSTEM));
 
-$returnurl = "$CFG->wwwroot/$CFG->admin/settings.php?section=manageauths";
+$returnurl = new moodle_url('/admin/settings.php', array('section'=>'manageauths'));
+
+$PAGE->set_url($returnurl);
 
 $action = optional_param('action', '', PARAM_ACTION);
 $auth   = optional_param('auth', '', PARAM_SAFEDIR);
 
-// get currently installed and enabled auth plugins
-$authsavailable = get_list_of_plugins('auth');
-
 get_enabled_auth_plugins(true); // fix the list of enabled auths
 if (empty($CFG->auth)) {
     $authsenabled = array();
@@ -30,7 +29,7 @@ if (empty($CFG->auth)) {
 }
 
 if (!empty($auth) and !exists_auth_plugin($auth)) {
-    print_error('pluginnotinstalled', 'auth', $url, $auth);
+    print_error('pluginnotinstalled', 'auth', $returnurl, $auth);
 }
 
 ////////////////////////////////////////////////////////////////////////////////
@@ -52,6 +51,7 @@ switch ($action) {
         if ($auth == $CFG->registerauth) {
             set_config('registerauth', '');
         }
+        session_gc(); // remove stale sessions
         break;
 
     case 'enable':
@@ -61,13 +61,14 @@ switch ($action) {
             $authsenabled = array_unique($authsenabled);
             set_config('auth', implode(',', $authsenabled));
         }
+        session_gc(); // remove stale sessions
         break;
 
     case 'down':
         $key = array_search($auth, $authsenabled);
         // check auth plugin is valid
         if ($key === false) {
-            print_error('pluginnotenabled', 'auth', $url, $auth);
+            print_error('pluginnotenabled', 'auth', $returnurl, $auth);
         }
         // move down the list
         if ($key < (count($authsenabled) - 1)) {
@@ -82,7 +83,7 @@ switch ($action) {
         $key = array_search($auth, $authsenabled);
         // check auth is valid
         if ($key === false) {
-            print_error('pluginnotenabled', 'auth', $url, $auth);
+            print_error('pluginnotenabled', 'auth', $returnurl, $auth);
         }
         // move up the list
         if ($key >= 1) {
@@ -97,6 +98,6 @@ switch ($action) {
         break;
 }
 
-redirect ($returnurl);
+redirect($returnurl);
+
 
-?>
