diff --git a/admin/auth_config.php b/admin/auth_config.php
index 0281b12..71dafdf 100644
--- a/admin/auth_config.php
+++ b/admin/auth_config.php
@@ -7,8 +7,7 @@ require_once '../config.php';
 require_once $CFG->libdir.'/adminlib.php';
 
 $auth = required_param('auth', PARAM_SAFEDIR);
-
-$CFG->pagepath = 'auth/' . $auth;
+$PAGE->set_pagetype('admin-auth-' . $auth);
 
 admin_externalpage_setup('authsetting'.$auth);
 
@@ -19,7 +18,6 @@ $returnurl = "$CFG->wwwroot/$CFG->admin/settings.php?section=manageauths";
 
 // save configuration changes
 if ($frm = data_submitted() and confirm_sesskey()) {
-    $frm = stripslashes_recursive($frm);
 
     $authplugin->validate_form($frm, $err);
 
@@ -33,9 +31,7 @@ if ($frm = data_submitted() and confirm_sesskey()) {
                 if (preg_match('/^lockconfig_(.+?)$/', $name, $matches)) {
                     $plugin = "auth/$auth";
                     $name   = $matches[1];
-                    if (!set_config($name, $value, $plugin)) {
-                        error("Problem saving config $name as $value for plugin $plugin");
-                    }
+                    set_config($name, $value, $plugin);
                 }
             }
             redirect($returnurl);
@@ -47,7 +43,9 @@ if ($frm = data_submitted() and confirm_sesskey()) {
         }
     }
 } else {
-    $frm = get_config("auth/$auth");
+    $frmlegacystyle = get_config('auth/'.$auth);
+    $frmnewstyle    = get_config('auth_'.$auth);
+    $frm = (object)array_merge((array)$frmlegacystyle, (array)$frmnewstyle);
 }
 
 $user_fields = $authplugin->userfields;
@@ -59,28 +57,28 @@ $user_fields = $authplugin->userfields;
     $authdescription = $authplugin->get_description();
 
 // output configuration form
-admin_externalpage_print_header();
+echo $OUTPUT->header();
 
 // choose an authentication method
-echo "<form $CFG->frametarget id=\"authmenu\" method=\"post\" action=\"auth_config.php\">\n";
+echo "<form id=\"authmenu\" method=\"post\" action=\"auth_config.php\">\n";
 echo "<div>\n";
-echo "<input type=\"hidden\" name=\"sesskey\" value=\"".$USER->sesskey."\" />\n";
+echo "<input type=\"hidden\" name=\"sesskey\" value=\"".sesskey()."\" />\n";
 echo "<input type=\"hidden\" name=\"auth\" value=\"".$auth."\" />\n";
 
 // auth plugin description
-print_simple_box_start('center', '80%');
-print_heading($authtitle);
-print_simple_box_start('center', '80%', '', 5, 'informationbox');
+echo $OUTPUT->box_start();
+echo $OUTPUT->heading($authtitle);
+echo $OUTPUT->box_start('informationbox');
 echo $authdescription;
-print_simple_box_end();
+echo $OUTPUT->box_end();
 echo "<hr />\n";
 $authplugin->config_form($frm, $err, $user_fields);
-print_simple_box_end();
+echo $OUTPUT->box_end();
 echo '<p style="text-align: center"><input type="submit" value="' . get_string("savechanges") . "\" /></p>\n";
 echo "</div>\n";
 echo "</form>\n";
 
-admin_externalpage_print_footer();
+echo $OUTPUT->footer();
 exit;
 
 /// Functions /////////////////////////////////////////////////////////////////
@@ -90,12 +88,12 @@ exit;
 // other options
 // Note: lockconfig_ fields have special handling.
 function print_auth_lock_options ($auth, $user_fields, $helptext, $retrieveopts, $updateopts) {
-
+    global $OUTPUT;
     echo '<tr><td colspan="3">';
     if ($retrieveopts) {
-        print_heading(get_string('auth_data_mapping', 'auth'));
+        echo $OUTPUT->heading(get_string('auth_data_mapping', 'auth'));
     } else {
-        print_heading(get_string('auth_fieldlocks', 'auth'));
+        echo $OUTPUT->heading(get_string('auth_fieldlocks', 'auth'));
     }
     echo '</td></tr>';
 
@@ -151,23 +149,23 @@ function print_auth_lock_options ($auth, $user_fields, $helptext, $retrieveopts,
             echo "<input id=\"lockconfig_{$varname}\" name=\"lockconfig_{$varname}\" type=\"text\" size=\"30\" value=\"{$pluginconfig->$varname}\" />";
             echo '<div style="text-align: right">';
             echo '<label for="menulockconfig_field_updatelocal_'.$field.'">'.get_string('auth_updatelocal', 'auth') . '</label>&nbsp;';
-            choose_from_menu($updatelocaloptions, "lockconfig_field_updatelocal_{$field}", $pluginconfig->{"field_updatelocal_$field"}, "");
+            echo html_writer::select($updatelocaloptions, "lockconfig_field_updatelocal_{$field}", $pluginconfig->{"field_updatelocal_$field"}, false);
             echo '<br />';
             if ($updateopts) {
                 echo '<label for="menulockconfig_field_updateremote_'.$field.'">'.get_string('auth_updateremote', 'auth') . '</label>&nbsp;';
-                choose_from_menu($updateextoptions, "lockconfig_field_updateremote_{$field}", $pluginconfig->{"field_updateremote_$field"}, "");
+                echo html_writer::select($updateextoptions, "lockconfig_field_updateremote_{$field}", $pluginconfig->{"field_updateremote_$field"}, false);
                 echo '<br />';
 
 
             }
             echo '<label for="menulockconfig_field_lock_'.$field.'">'.get_string('auth_fieldlock', 'auth') . '</label>&nbsp;';
-            choose_from_menu($lockoptions, "lockconfig_field_lock_{$field}", $pluginconfig->{"field_lock_$field"}, "");
+            echo html_writer::select($lockoptions, "lockconfig_field_lock_{$field}", $pluginconfig->{"field_lock_$field"}, false);
             echo '</div>';
         } else {
             echo '<tr valign="top"><td align="right">';
             echo '<label for="menulockconfig_field_lock_'.$field.'">'.$fieldname.'</label>';
             echo '</td><td>';
-            choose_from_menu($lockoptions, "lockconfig_field_lock_{$field}", $pluginconfig->{"field_lock_$field"}, "");
+            echo html_writer::select($lockoptions, "lockconfig_field_lock_{$field}", $pluginconfig->{"field_lock_$field"}, false);
         }
         echo '</td>';
         if (!empty($helptext)) {
@@ -178,4 +176,4 @@ function print_auth_lock_options ($auth, $user_fields, $helptext, $retrieveopts,
     }
 }
 
-?>
+
