diff --git a/admin/block.php b/admin/block.php
index f516532..59938ec 100644
--- a/admin/block.php
+++ b/admin/block.php
@@ -1,29 +1,28 @@
-<?PHP  // $Id$
+<?php
 
 // block.php - allows admin to edit all local configuration variables for a block
 
     require_once('../config.php');
     require_once($CFG->libdir.'/adminlib.php');
-    require_once($CFG->libdir.'/blocklib.php');
 
     $blockid = required_param('block', PARAM_INT);
 
     if(!$blockrecord = blocks_get_record($blockid)) {
-        error('This block does not exist');
+        print_error('blockdoesnotexist', 'error');
     }
 
     admin_externalpage_setup('blocksetting'.$blockrecord->name);
 
     $block = block_instance($blockrecord->name);
     if($block === false) {
-        error('Problem in instantiating block object');
+        print_error('blockcannotinistantiate', 'error');
     }
 
     // Define the data we're going to silently include in the instance config form here,
     // so we can strip them from the submitted data BEFORE handling it.
     $hiddendata = array(
         'block' => $blockid,
-        'sesskey' => $USER->sesskey
+        'sesskey' => sesskey()
     );
 
     /// If data submitted, then process and store.
@@ -34,7 +33,7 @@
              print_error('confirmsesskeybad', 'error');
         }
         if(!$block->has_config()) {
-            error('This block does not support global configuration');
+            print_error('blockcannotconfig', 'error');
         }
         $remove = array_keys($hiddendata);
         foreach($remove as $item) {
@@ -50,11 +49,14 @@
     $strmanageblocks = get_string('manageblocks');
     $strblockname = $block->get_title();
 
-    admin_externalpage_print_header();
+    echo $OUTPUT->header();
 
-    print_heading($strblockname);
+    echo $OUTPUT->heading($strblockname);
 
-    print_simple_box(get_string('configwarning', 'admin'), 'center', '50%');
+    echo $OUTPUT->notification('This block still uses an old-style config_global.html file. ' .
+            'It must be updated by a developer to use a settings.php file.');
+
+    echo $OUTPUT->box(get_string('configwarning', 'admin'), 'generalbox boxwidthnormal boxaligncenter');
     echo '<br />';
 
     echo '<form method="post" action="block.php">';
@@ -63,8 +65,12 @@
         echo '<input type="hidden" name="'. $name .'" value="'. $val .'" />';
     }
     echo '</p>';
-    $block->config_print();
+
+    echo $OUTPUT->box_start();
+    include($CFG->dirroot.'/blocks/'. $block->name() .'/config_global.html');
+    echo $OUTPUT->box_end();
+
     echo '</form>';
-    print_footer();
+    echo $OUTPUT->footer();
+
 
-?>
