diff --git a/admin/cron.php b/admin/cron.php
index 4872d4a..e1a94b4 100644
--- a/admin/cron.php
+++ b/admin/cron.php
@@ -524,6 +524,11 @@
         }
     }
 
+    // run the ELIS cron jobs
+    mtrace("Running ELIS cron tasks...");
+    require_once($CFG->dirroot.'/elis/core/cron.php');
+    elis_cron();
+
     // run any customized cronjobs, if any
     // looking for functions in lib/local/cron.php
     if (file_exists($CFG->dirroot.'/local/cron.php')) {
