diff --git a/admin/dbperformance.php b/admin/dbperformance.php
index 648a2ec..d48b29a 100644
--- a/admin/dbperformance.php
+++ b/admin/dbperformance.php
@@ -1,10 +1,13 @@
-<?PHP  // $Id$
-       // dbperformance.php - shows latest ADOdb stats for the current server
+<?PHP
+// dbperformance.php - shows latest ADOdb stats for the current server
 
-    require_once('../config.php');
+// disable moodle specific debug messages and any errors in output
+define('NO_DEBUG_DISPLAY', true);
+
+require_once('../config.php');
+
+error('TODO: rewrite db perf code'); // TODO: rewrite
 
-    // disable moodle specific debug messages that would be breaking the frames
-    disable_debugging();
 
     $topframe    = optional_param('topframe', 0, PARAM_BOOL);
     $bottomframe = optional_param('bottomframe', 0, PARAM_BOOL);
@@ -18,26 +21,12 @@
     $stradministration = get_string("administration");
     $site = get_site();
 
-    $navigation = build_navigation(array(
-        array('name'=>$stradministration, 'link'=>'index.php', 'type'=>'misc'),
-        array('name'=>$strdatabaseperformance, 'link'=>null, 'type'=>'misc')));
     if (!empty($topframe)) {
-        print_header("$site->shortname: $strdatabaseperformance", "$site->fullname", $navigation);
+        $PAGE->set_url('/admin/dbperformance.php');
+        $PAGE->navbar->add($stradministration, new moodle_url('/admin/index.php'));
+        $PAGE->navbar->add($strdatabaseperformance);
+        $PAGE->set_title("$site->shortname: $strdatabaseperformance");
+        $PAGE->set_heading($site->fullname);
+        echo $OUTPUT->header();
         exit;
     }
-
-    if (!empty($bottomframe) or !empty($do)) {
-        $perf =&NewPerfMonitor($db);
-        $perf->UI($pollsecs=5);
-        exit;
-    }
-
-?>
-<head>
-<title><?php echo "$site->shortname: $strdatabaseperformance" ?></title>
-</head>
-
-<frameset rows="80,*">
-   <frame src="dbperformance.php?topframe=true">
-   <frame src="dbperformance.php?bottomframe=true">
-</frameset>
