diff --git a/admin/environment.php b/admin/environment.php
index 74eb8f0..66ccbea 100644
--- a/admin/environment.php
+++ b/admin/environment.php
@@ -1,4 +1,4 @@
-<?php  //$Id$
+<?php
 
 ///////////////////////////////////////////////////////////////////////////
 //                                                                       //
@@ -52,12 +52,12 @@
     $strmisc = get_string('miscellaneous');
 
 /// Print the header stuff
-    admin_externalpage_print_header();
+    echo $OUTPUT->header();
 
 /// Print the component download link
-    echo '<div class="reportlink"><a href="environment.php?action=updatecomponent&amp;sesskey='.$USER->sesskey.'">'.$strupdate.'</a></div>';
+    echo '<div class="reportlink"><a href="environment.php?action=updatecomponent&amp;sesskey='.sesskey().'">'.$strupdate.'</a></div>';
 
-    print_heading($strenvironment);
+    echo $OUTPUT->heading($strenvironment);
 
 /// Handle the 'updatecomponent' action
     if ($action == 'updatecomponent' && confirm_sesskey()) {
@@ -72,23 +72,23 @@
                         $a = new stdClass();
                         $a->url = 'http://download.moodle.org/environment/environment.zip';
                         $a->dest= $CFG->dataroot.'/';
-                        print_simple_box(get_string($cd->get_error(), 'error', $a), 'center', '', '', 5, 'errorbox');
+                        echo $OUTPUT->box(get_string($cd->get_error(), 'error', $a), 'errorbox');
                     } else {
-                        print_simple_box(get_string($cd->get_error(), 'error'), 'center', '', '', 5, 'errorbox');
+                        echo $OUTPUT->box(get_string($cd->get_error(), 'error'), 'errorbox');
                     }
                     break;
                 case COMPONENT_UPTODATE:
-                    print_simple_box(get_string($cd->get_error(), 'error'), 'center');
+                    echo $OUTPUT->box(get_string($cd->get_error(), 'error'));
                     break;
                 case COMPONENT_INSTALLED:
-                    print_simple_box(get_string('componentinstalled', 'admin'), 'center');
+                    echo $OUTPUT->box(get_string('componentinstalled', 'admin'));
                     break;
             }
         }
     }
 
 /// Start of main box
-    print_simple_box_start('center');
+    echo $OUTPUT->box_start();
 
     echo "<div style=\"text-align:center\">".$stradminhelpenvironment."</div><br />";
 
@@ -106,7 +106,7 @@
             if (empty($version)) {
                 $version =  $env_version;
             }
-        ///Iterate over each version, adding bigged than current
+        ///Iterate over each version, adding bigger than current
             foreach ($env_versions as $env_version) {
                 if (version_compare(normalize_version($current_version), $env_version, '<')) {
                     $versions[$env_version] = $env_version;
@@ -120,17 +120,18 @@
     }
 
 /// Print form and popup menu
-    echo '<div style="text-align:center">'.$strmoodleversion.' ';
-    popup_form("$CFG->wwwroot/$CFG->admin/environment.php?version=",
-        $versions, 'selectversion', $version, '');
+    echo '<div style="text-align:center"> ';
+    $select = new single_select(new moodle_url('/admin/environment.php'), 'version', $versions, $version, null);
+    $select->label = $strmoodleversion;
+    echo $OUTPUT->render($select);
     echo '</div>';
 
 /// End of main box
-    print_simple_box_end();
+    echo $OUTPUT->box_end();
 
 /// Gather and show results
     $status = check_moodle_environment($version, $environment_results);
 
 /// Print footer
-    admin_externalpage_print_footer();
-?>
+    echo $OUTPUT->footer();
+
