diff --git a/admin/handlevirus.php b/admin/handlevirus.php
index 7a09c1e..9e32cc0 100644
--- a/admin/handlevirus.php
+++ b/admin/handlevirus.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 /** This expects the output from a command like
  * clamscan -r --infected --no-summary <files> 2>&1 | php -d error_log=/path/to/log thisfile.php 
  * also it's important that the output of clamscan prints the FULL PATH to each infected file, so use absolute paths for area to scan
@@ -6,14 +6,15 @@
  * php -d error_log=/path/to/log thisfile.php will override the default error log for php cli, which is stderr, so if you want this script to just print stuff out, use php thisfile.php instead.
  */
 
+die('TODO: MDL-19380');
 
 $fd = fopen('php://stdin','r');
 if (!$fd) {
     exit();
 }
 
-$FULLME='cron';
 require_once(dirname(dirname(__FILE__)).'/config.php');
+require_once($CFG->libdir.'/eventslib.php');
 require_once($CFG->dirroot.'/lib/uploadlib.php'); // contains virus handling stuff.
 
 $site = get_site();
@@ -29,7 +30,7 @@ while(!feof($fd)) {
     $bits = explode('/',$file);
     $a->filename = $bits[count($bits)-1];
 
-    if (!$log = get_record("log","module","upload","info",$file,"action","upload")) {
+    if (!$log = $DB->get_record("log", array("module"=>"upload", "info"=>$file, "action"=>"upload"))) {
         $a->action = clam_handle_infected_file($file,0,false);
         clam_replace_infected_file($file);
         notify_admins_unknown($file,$a);
@@ -38,13 +39,17 @@ while(!feof($fd)) {
     $action = clam_handle_infected_file($file,$log->userid,true);
     clam_replace_infected_file($file);
     
-    $user = get_record("user","id",$log->userid);
-    $course = get_record("course","id",$log->course);
+    list($ctxselect, $ctxjoin) = context_instance_preload_sql('c.id', CONTEXT_COURSE, 'ctx');
+    $sql = "SELECT c.id, c.fullname $ctxselect FROM {course} c $ctxjoin WHERE c.id = :courseid";
+    $course = $DB->get_record_sql($sql, array('courseid' => $log->course));
+    context_instance_preload($course);
+
+    $user = $DB->get_record("user", array("id"=>$log->userid));
     $subject = get_string('virusfoundsubject','moodle',format_string($site->fullname));
     $a->date = userdate($log->time);
 
     $a->action = $action;
-    $a->course = $course->fullname;
+    $a->course = format_string($course->fullname, true, array('context' => get_context_instance(CONTEXT_COURSE, $course->id)));
     $a->user = fullname($user);
 
     notify_user($user,$subject,$a);
@@ -59,7 +64,17 @@ function notify_user($user,$subject,$a) {
         return false;
     }
     $body = get_string('virusfoundlater','moodle',$a);
-    email_to_user($user,get_admin(),$subject,$body);
+
+    $eventdata = new stdClass();
+    $eventdata->modulename        = 'moodle';
+    $eventdata->userfrom          = get_admin();
+    $eventdata->userto            = $user;
+    $eventdata->subject           = $subject;
+    $eventdata->fullmessage       = $body;
+    $eventdata->fullmessageformat = FORMAT_PLAIN;
+    $eventdata->fullmessagehtml   = '';
+    $eventdata->smallmessage      = '';
+    message_send($eventdata);
 }
 
 
@@ -69,7 +84,16 @@ function notify_admins($user,$subject,$a) {
 
     $body = get_string('virusfoundlateradmin','moodle',$a);
     foreach ($admins as $admin) {
-        email_to_user($admin,$admin,$subject,$body);
+        $eventdata = new stdClass();
+        $eventdata->modulename        = 'moodle';
+        $eventdata->userfrom          = $admin;
+        $eventdata->userto            = $admin;
+        $eventdata->subject           = $subject;
+        $eventdata->fullmessage       = $body;
+        $eventdata->fullmessageformat = FORMAT_PLAIN;
+        $eventdata->fullmessagehtml   = '';
+        $eventdata->smallmessage      = '';
+        message_send($eventdata);
     }
 }
 
@@ -81,7 +105,16 @@ function notify_admins_unknown($file,$a) {
     $subject = get_string('virusfoundsubject','moodle',format_string($site->fullname));
     $body = get_string('virusfoundlateradminnolog','moodle',$a);
     foreach ($admins as $admin) {
-        email_to_user($admin,$admin,$subject,$body);
+        $eventdata = new stdClass();
+        $eventdata->modulename        = 'moodle';
+        $eventdata->userfrom          = $admin;
+        $eventdata->userto            = $admin;
+        $eventdata->subject           = $subject;
+        $eventdata->fullmessage       = $body;
+        $eventdata->fullmessageformat = FORMAT_PLAIN;
+        $eventdata->fullmessagehtml   = '';
+        $eventdata->smallmessage      = '';
+        message_send($eventdata);
     }
 }
 
@@ -114,4 +147,4 @@ function validate_line($line) {
     return $file;
 }
 
-?>
+
