diff --git a/admin/mailout-debugger.php b/admin/mailout-debugger.php
index 855f3b4..d183fa4 100644
--- a/admin/mailout-debugger.php
+++ b/admin/mailout-debugger.php
@@ -63,7 +63,7 @@ fclose($stdin);
  *
  * Very useful for perl-like error-handling:
  *
- * do_somethting() or mdie("Something went wrong");
+ * do_something() or mdie("Something went wrong");
  *
  * @param string  $msg       Error message
  * @param integer $errorcode Error code to emit
@@ -74,4 +74,4 @@ function mdie($msg='', $errorcode=1) {
     exit($errorcode);
 }
 
-?>
+
