diff --git a/admin/process_email.php b/admin/process_email.php
old mode 100755
new mode 100644
index f07e8c9..20ad4c1
--- a/admin/process_email.php
+++ b/admin/process_email.php
@@ -1,6 +1,6 @@
 #!/usr/bin/php -f
-<?php // $Id$
-define('FULLME','cron'); // prevent warnings
+<?php
+
 //error_reporting(0);
 //ini_set('display_errors',0);
 require_once(dirname(dirname(__FILE__)).'/config.php');
@@ -9,6 +9,7 @@ $address = $tmp[0];
 
 // BOUNCE EMAILS TO NOREPLY
 if ($_ENV['RECIPIENT'] == $CFG->noreplyaddress) {
+    $user = new stdClass();
     $user->email = $_ENV['SENDER'];
 
     if (!validate_email($user->email)) {
@@ -53,7 +54,7 @@ if ($modid == '0') { // special
     $modname = 'moodle';
 }
 else {
-    $modname = get_field("modules","name","id",$modid);
+    $modname = $DB->get_field("modules", "name", array("id"=>$modid));
     include_once('mod/'.$modname.'/lib.php');
 }
 $function = $modname.'_process_email';
@@ -76,4 +77,4 @@ fclose($fd);
 
 
 
-?>
+
