diff --git a/admin/renderer.php b/admin/renderer.php
index 3080d1e..97192f4 100644
--- a/admin/renderer.php
+++ b/admin/renderer.php
@@ -1146,13 +1146,14 @@ class core_admin_renderer extends plugin_renderer_base {
             get_string('displayname', 'core_plugin'),
             get_string('source', 'core_plugin'),
             get_string('version', 'core_plugin'),
+            get_string('release', 'core_plugin'), // ELIS_2.0.0
             get_string('availability', 'core_plugin'),
             get_string('actions', 'core_plugin'),
             get_string('notes','core_plugin'),
         );
-        $table->headspan = array(1, 1, 1, 1, 2, 1);
+        $table->headspan = array(1, 1, 1, 1, 1, 2, 1);
         $table->colclasses = array(
-            'pluginname', 'source', 'version', 'availability', 'settings', 'uninstall', 'notes'
+            'pluginname', 'source', 'version', 'release', 'availability', 'settings', 'uninstall', 'notes' // ELIS 2.x
         );
 
         foreach ($plugininfo as $type => $plugins) {
@@ -1205,6 +1206,24 @@ class core_admin_renderer extends plugin_renderer_base {
 
                 $version = new html_table_cell($plugin->versiondb);
 
+                // ELIS_2.0.0 Release
+                $versionfile = $plugin->rootdir .'/version.php';
+                $module = new stdClass; // ELIS-7928: was getting previous value
+                if (file_exists($versionfile)) {
+                    include($versionfile);
+                }
+
+                // ELIS-7045 - Activity modules use a $module object in their version.php file so check that for a release property as well.
+                if (isset($plugin->release)) {
+                    $rel = $plugin->release;
+                } else if (isset($module->release)) {
+                    $rel = $module->release;
+                } else {
+                    $rel = '';
+                }
+
+                $release = new html_table_cell($rel);
+
                 $isenabled = $plugin->is_enabled();
                 if (is_null($isenabled)) {
                     $availability = new html_table_cell('');
@@ -1250,7 +1269,7 @@ class core_admin_renderer extends plugin_renderer_base {
                 $notes = new html_table_cell($requiredby.$updateinfo);
 
                 $row->cells = array(
-                    $pluginname, $source, $version, $availability, $settings, $uninstall, $notes
+                    $pluginname, $source, $version, $release, $availability, $settings, $uninstall, $notes
                 );
                 $table->data[] = $row;
             }
