diff --git a/admin/upgradesettings.php b/admin/upgradesettings.php
index 2f23407..5f56e33 100644
--- a/admin/upgradesettings.php
+++ b/admin/upgradesettings.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 
 // detects settings that were added during an upgrade, displays a screen for the admin to
 // modify them, and then processes modifications
@@ -11,13 +11,15 @@ $return = optional_param('return', '', PARAM_ALPHA);
 /// no guest autologin
 require_login(0, false);
 
-$adminroot =& admin_get_root(); // need all settings
 admin_externalpage_setup('upgradesettings'); // now hidden page
+$PAGE->set_pagelayout('maintenance'); // do not print any blocks or other rubbish, we want to force saving
+$PAGE->blocks->show_only_fake_blocks();
+$adminroot = admin_get_root(); // need all settings
 
 // now we'll deal with the case that the admin has submitted the form with new settings
 if ($data = data_submitted() and confirm_sesskey()) {
     $count = admin_write_settings($data);
-    $adminroot =& admin_get_root(true); //reload tree
+    $adminroot = admin_get_root(true); //reload tree
 }
 
 $newsettings = admin_output_new_settings_by_page($adminroot);
@@ -47,9 +49,11 @@ if (!empty($adminroot->errors)) {
 
 // and finally, if we get here, then there are new settings and we have to print a form
 // to modify them
-admin_externalpage_print_header($focus);
+echo $OUTPUT->header($focus);
 
-print_box(get_string('upgradesettingsintro','admin'), 'generalbox');
+if (!empty($SITE->fullname) and !empty($SITE->shortname)) {
+    echo $OUTPUT->box(get_string('upgradesettingsintro','admin'), 'generalbox');
+}
 
 echo '<form action="upgradesettings.php" method="post" id="adminsettings">';
 echo '<div>';
@@ -63,10 +67,6 @@ echo '<div class="form-buttons"><input class="form-submit" type="submit" value="
 echo '</div>';
 echo '</form>';
 
-if (!empty($CFG->adminusehtmleditor)) {
-    use_html_editor();
-}
+echo $OUTPUT->footer();
 
-print_footer();
 
-?>
