diff --git a/lib/bennu/bennu.class.php b/lib/bennu/bennu.class.php
index 40435c7..57d643f 100644
--- a/lib/bennu/bennu.class.php
+++ b/lib/bennu/bennu.class.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 
 /**
  *  BENNU - PHP iCalendar library
@@ -9,7 +9,6 @@
  *  See http://bennu.sourceforge.net/ for more information and downloads.
  *
  * @author Ioannis Papaioannou 
- * @version $Id$
  * @license http://www.gnu.org/copyleft/lesser.html GNU Lesser General Public License
  */
 
@@ -56,4 +55,3 @@ class Bennu {
     }
 }
 
-?>
