diff --git a/lib/bennu/iCalendar_properties.php b/lib/bennu/iCalendar_properties.php
index 3f3120e..bb621ce 100644
--- a/lib/bennu/iCalendar_properties.php
+++ b/lib/bennu/iCalendar_properties.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 
 /**
  *  BENNU - PHP iCalendar library
@@ -25,11 +25,7 @@ class iCalendar_property {
     var $val_multi        = false;
     var $val_default      = NULL;
 
-    function iCalendar_property() {
-        $this->construct();
-    }
-
-    function construct() {
+    function __construct() {
         $this->parameters = array();
     }
 
@@ -221,7 +217,7 @@ class iCalendar_property_calscale extends iCalendar_property {
     var $name        = 'CALSCALE';
     var $val_type    = RFC2445_TYPE_TEXT;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -238,7 +234,7 @@ class iCalendar_property_method extends iCalendar_property {
     var $name        = 'METHOD';
     var $val_type    = RFC2445_TYPE_TEXT;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -258,7 +254,7 @@ class iCalendar_property_prodid extends iCalendar_property {
     var $val_type    = RFC2445_TYPE_TEXT;
     var $val_default = NULL;
 
-    function construct() {
+    function __construct() {
         $this->val_default = '-//John Papaioannou/NONSGML Bennu '._BENNU_VERSION.'//EN';
 
         $this->valid_parameters = array(
@@ -273,7 +269,7 @@ class iCalendar_property_version extends iCalendar_property {
     var $val_type    = RFC2445_TYPE_TEXT;
     var $val_default = '2.0';
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -293,7 +289,7 @@ class iCalendar_property_attach extends iCalendar_property {
     var $name        = 'ATTACH';
     var $val_type    = RFC2445_TYPE_URI;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'FMTTYPE'     => RFC2445_OPTIONAL | RFC2445_ONCE,
             'ENCODING'    => RFC2445_OPTIONAL | RFC2445_ONCE,
@@ -339,7 +335,7 @@ class iCalendar_property_categories extends iCalendar_property {
     var $val_type    = RFC2445_TYPE_TEXT;
     var $val_multi   = true;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'LANGUAGE'    => RFC2445_OPTIONAL | RFC2445_ONCE,
             RFC2445_XNAME => RFC2445_OPTIONAL
@@ -353,7 +349,7 @@ class iCalendar_property_class extends iCalendar_property {
     var $val_type    = RFC2445_TYPE_TEXT;
     var $val_default = 'PUBLIC';
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -370,7 +366,7 @@ class iCalendar_property_comment extends iCalendar_property {
     var $name        = 'COMMENT';
     var $val_type    = RFC2445_TYPE_TEXT;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'ALTREP'      => RFC2445_OPTIONAL | RFC2445_ONCE,
             'LANGUAGE'    => RFC2445_OPTIONAL | RFC2445_ONCE,
@@ -384,7 +380,7 @@ class iCalendar_property_description extends iCalendar_property {
     var $name        = 'DESCRIPTION';
     var $val_type    = RFC2445_TYPE_TEXT;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'ALTREP'      => RFC2445_OPTIONAL | RFC2445_ONCE,
             'LANGUAGE'    => RFC2445_OPTIONAL | RFC2445_ONCE,
@@ -398,7 +394,7 @@ class iCalendar_property_geo extends iCalendar_property {
     var $name        = 'GEO';
     var $val_type    = RFC2445_TYPE_TEXT;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'ALTREP'      => RFC2445_OPTIONAL | RFC2445_ONCE,
             'LANGUAGE'    => RFC2445_OPTIONAL | RFC2445_ONCE,
@@ -438,7 +434,7 @@ class iCalendar_property_location extends iCalendar_property {
     var $name        = 'LOCATION';
     var $val_type    = RFC2445_TYPE_TEXT;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'ALTREP'      => RFC2445_OPTIONAL | RFC2445_ONCE,
             'LANGUAGE'    => RFC2445_OPTIONAL | RFC2445_ONCE,
@@ -452,7 +448,7 @@ class iCalendar_property_percent_complete extends iCalendar_property {
     var $name        = 'PERCENT-COMPLETE';
     var $val_type    = RFC2445_TYPE_INTEGER;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -474,7 +470,7 @@ class iCalendar_property_priority extends iCalendar_property {
     var $name        = 'PRIORITY';
     var $val_type    = RFC2445_TYPE_INTEGER;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -497,7 +493,7 @@ class iCalendar_property_resources extends iCalendar_property {
     var $val_type    = RFC2445_TYPE_TEXT;
     var $val_multi   = true;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'ALTREP'      => RFC2445_OPTIONAL | RFC2445_ONCE,
             'LANGUAGE'    => RFC2445_OPTIONAL | RFC2445_ONCE,
@@ -511,7 +507,7 @@ class iCalendar_property_status extends iCalendar_property {
     var $name        = 'STATUS';
     var $val_type    = RFC2445_TYPE_TEXT;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -541,7 +537,7 @@ class iCalendar_property_summary extends iCalendar_property {
     var $name        = 'SUMMARY';
     var $val_type    = RFC2445_TYPE_TEXT;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'ALTREP'      => RFC2445_OPTIONAL | RFC2445_ONCE,
             'LANGUAGE'    => RFC2445_OPTIONAL | RFC2445_ONCE,
@@ -558,7 +554,7 @@ class iCalendar_property_completed extends iCalendar_property {
     var $name        = 'COMPLETED';
     var $val_type    = RFC2445_TYPE_DATE_TIME;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -578,7 +574,7 @@ class iCalendar_property_dtend extends iCalendar_property {
     var $name        = 'DTEND';
     var $val_type    = RFC2445_TYPE_DATE_TIME;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'VALUE'       => RFC2445_OPTIONAL | RFC2445_ONCE,
             'TZID'        => RFC2445_OPTIONAL | RFC2445_ONCE,
@@ -620,7 +616,7 @@ class iCalendar_property_due extends iCalendar_property {
     var $name        = 'DUE';
     var $val_type    = RFC2445_TYPE_DATE_TIME;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'VALUE'       => RFC2445_OPTIONAL | RFC2445_ONCE,
             'TZID'        => RFC2445_OPTIONAL | RFC2445_ONCE,
@@ -662,7 +658,7 @@ class iCalendar_property_dtstart extends iCalendar_property {
     var $name        = 'DTSTART';
     var $val_type    = RFC2445_TYPE_DATE_TIME;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'VALUE'       => RFC2445_OPTIONAL | RFC2445_ONCE,
             'TZID'        => RFC2445_OPTIONAL | RFC2445_ONCE,
@@ -705,7 +701,7 @@ class iCalendar_property_duration extends iCalendar_property {
     var $name        = 'DURATION';
     var $val_type    = RFC2445_TYPE_DURATION;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -727,7 +723,7 @@ class iCalendar_property_freebusy extends iCalendar_property {
     var $val_type    = RFC2445_TYPE_PERIOD;
     var $val_multi   = true;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'FBTYPE'      => RFC2445_OPTIONAL | RFC2445_ONCE,
             RFC2445_XNAME => RFC2445_OPTIONAL
@@ -744,7 +740,7 @@ class iCalendar_property_freebusy extends iCalendar_property {
             // Start time MUST be in UTC
             return false;
         }
-        if($value{$pos + 1} != 'P' && $substr($value, -1) != 'Z') {
+        if($value{$pos + 1} != 'P' && substr($value, -1) != 'Z') {
             // If the second part is not a period, it MUST be in UTC
             return false;
         }
@@ -761,7 +757,7 @@ class iCalendar_property_transp extends iCalendar_property {
     var $val_type    = RFC2445_TYPE_TEXT;
     var $val_default = 'OPAQUE';
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -786,7 +782,7 @@ class iCalendar_property_attendee extends iCalendar_property {
     // TODO: MUST NOT be specified when the calendar object has METHOD=PUBLISH
     // TODO: standard has lots of detail here, make triple sure that we eventually conform
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'LANGUAGE'       => RFC2445_OPTIONAL | RFC2445_ONCE,
             'CN'             => RFC2445_OPTIONAL | RFC2445_ONCE,
@@ -826,7 +822,7 @@ class iCalendar_property_contact extends iCalendar_property {
     var $name        = 'CONTACT';
     var $val_type    = RFC2445_TYPE_TEXT;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'ALTREP'      => RFC2445_OPTIONAL | RFC2445_ONCE,
             'LANGUAGE'    => RFC2445_OPTIONAL | RFC2445_ONCE,
@@ -840,7 +836,7 @@ class iCalendar_property_organizer extends iCalendar_property {
     var $name        = 'ORGANIZER';
     var $val_type    = RFC2445_TYPE_CAL_ADDRESS;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'CN'          => RFC2445_OPTIONAL | RFC2445_ONCE,
             'DIR'         => RFC2445_OPTIONAL | RFC2445_ONCE,
@@ -880,7 +876,7 @@ class iCalendar_property_recurrence_id extends iCalendar_property {
     var $name        = 'RECURRENCE-ID';
     var $val_type    = RFC2445_TYPE_DATE_TIME;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'RANGE'       => RFC2445_OPTIONAL | RFC2445_ONCE,
             'TZID'        => RFC2445_OPTIONAL | RFC2445_ONCE,
@@ -912,7 +908,7 @@ class iCalendar_property_related_to extends iCalendar_property {
 
     // TODO: the value of this property must reference another component's UID
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'RELTYPE'     => RFC2445_OPTIONAL | RFC2445_ONCE,
             RFC2445_XNAME => RFC2445_OPTIONAL
@@ -925,7 +921,7 @@ class iCalendar_property_url extends iCalendar_property {
     var $name        = 'URL';
     var $val_type    = RFC2445_TYPE_URI;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -937,7 +933,7 @@ class iCalendar_property_uid extends iCalendar_property {
     var $name        = 'UID';
     var $val_type    = RFC2445_TYPE_TEXT;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -958,7 +954,7 @@ class iCalendar_property_exdate extends iCalendar_property {
     var $val_type    = RFC2445_TYPE_DATE_TIME;
     var $val_multi   = true;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'TZID'        => RFC2445_OPTIONAL | RFC2445_ONCE,
             'VALUE'       => RFC2445_OPTIONAL | RFC2445_ONCE,
@@ -987,7 +983,7 @@ class iCalendar_property_exrule extends iCalendar_property {
     var $name        = 'EXRULE';
     var $val_type    = RFC2445_TYPE_RECUR;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -1000,7 +996,7 @@ class iCalendar_property_rdate extends iCalendar_property {
     var $val_type    = RFC2445_TYPE_DATE_TIME;
     var $val_multi   = true;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'TZID'        => RFC2445_OPTIONAL | RFC2445_ONCE,
             'VALUE'       => RFC2445_OPTIONAL | RFC2445_ONCE,
@@ -1029,14 +1025,71 @@ class iCalendar_property_rrule extends iCalendar_property {
     var $name        = 'RRULE';
     var $val_type    = RFC2445_TYPE_RECUR;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
     }
 }
 
-// TODO: 4.8.6 Alarm Component Properties
+// 4.8.6 Alarm Component Properties
+// -------------------------------------------
+class iCalendar_property_action extends iCalendar_property {
+	var $name      = 'ACTION';
+    var $val_type   = RFC2445_TYPE_TEXT;
+    
+    function __construct() {
+        $this->valid_parameters = array(
+            RFC2445_XNAME => RFC2445_OPTIONAL
+        );
+    }
+    
+    function is_valid_value($value) {
+        if(!parent::is_valid_value($value)) {
+            return false;
+        }
+        
+        // Value must be one of the following, or an x-name.
+        $valid_values = array('ACTION', 'DISPLAY', 'EMAIL', 'PROCEDURE');
+        return(in_array($value, $valid_values) || rfc2445_is_xname($value));        
+        
+    }
+}
+
+class iCalendar_property_repeat extends iCalendar_property {
+    var $name      = 'REPEAT';
+    var $val_type   = RFC2445_TYPE_INTEGER;
+    
+    function __construct() {
+        $this->valid_parameters = array(
+            RFC2445_XNAME => RFC2445_OPTIONAL
+        );
+    }   
+}
+
+class iCalendar_property_trigger extends iCalendar_property {
+    var $name      = 'TRIGGER';
+    var $val_type   = RFC2445_TYPE_TEXT;
+    
+    function __construct() {
+        $this->valid_parameters = array(
+            'VALUE' => RFC2445_OPTIONAL | RFC2445_ONCE,
+            'RELATED' => RFC2445_OPTIONAL | RFC2445_ONCE,
+            RFC2445_XNAME => RFC2445_OPTIONAL
+        );
+    }
+    
+    function is_valid_value($value) {        
+    	if(!parent::is_valid_value($value)) {
+            return false;
+        }
+        // Must either be DURATION or DATE_TIME type
+        return(rfc2445_is_valid_value($value, RFC2445_TYPE_DURATION) 
+            || rfc2445_is_valid_value($value, RFC2445_TYPE_DATE_TIME));
+    }
+}
+
+
 
 // 4.8.7 Change Management Component Properties
 // --------------------------------------------
@@ -1046,7 +1099,7 @@ class iCalendar_property_created extends iCalendar_property {
     var $name        = 'CREATED';
     var $val_type    = RFC2445_TYPE_DATE_TIME;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -1066,7 +1119,7 @@ class iCalendar_property_dtstamp extends iCalendar_property {
     var $name        = 'DTSTAMP';
     var $val_type    = RFC2445_TYPE_DATE_TIME;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -1086,7 +1139,7 @@ class iCalendar_property_last_modified extends iCalendar_property {
     var $name        = 'LAST-MODIFIED';
     var $val_type    = RFC2445_TYPE_DATE_TIME;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -1107,7 +1160,7 @@ class iCalendar_property_sequence extends iCalendar_property {
     var $val_type    = RFC2445_TYPE_INTEGER;
     var $val_default = 0;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -1130,7 +1183,7 @@ class iCalendar_property_x extends iCalendar_property {
     var $name        = RFC2445_XNAME;
     var $val_type    = NULL;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'LANGUAGE'    => RFC2445_OPTIONAL | RFC2445_ONCE,
             RFC2445_XNAME => RFC2445_OPTIONAL
@@ -1160,7 +1213,7 @@ class iCalendar_property_request_status extends iCalendar_property {
     var $name        = 'REQUEST-STATUS';
     var $val_type    = RFC2445_TYPE_TEXT;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             'LANGUAGE'    => RFC2445_OPTIONAL | RFC2445_ONCE,
             RFC2445_XNAME => RFC2445_OPTIONAL
@@ -1280,6 +1333,88 @@ class iCalendar_property_request_status extends iCalendar_property {
 
 }
 
+class iCalendar_property_tzid extends iCalendar_property {
+
+    var $name        = 'TZID';
+    var $val_type    = RFC2445_TYPE_TEXT;
+
+    function __construct() {
+        $this->valid_parameters = array(
+            RFC2445_XNAME => RFC2445_OPTIONAL
+        );
+    }
+
+    function is_valid_value($value) {
+        if(!parent::is_valid_value($value)) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+}
+
+class iCalendar_property_tzname extends iCalendar_property {
+
+    var $name        = 'TZNAME';
+    var $val_type    = RFC2445_TYPE_TEXT;
+
+    function __construct() {
+        $this->valid_parameters = array(
+            'LANGUAGE' => RFC2445_OPTIONAL | RFC2445_ONCE,
+            RFC2445_XNAME => RFC2445_OPTIONAL
+        );
+    }
+
+    function is_valid_value($value) {
+        if(!parent::is_valid_value($value)) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+}
+
+class iCalendar_property_tzoffsetfrom extends iCalendar_property {
+
+    var $name        = 'TZOFFSETFROM';
+    var $val_type    = RFC2445_TYPE_UTC_OFFSET;
+
+    function __construct() {
+        $this->valid_parameters = array(
+            RFC2445_XNAME => RFC2445_OPTIONAL
+        );
+    }
+
+    function is_valid_value($value) {
+        if(!parent::is_valid_value($value)) {
+            return false;
+        } else {
+            return true;
+        }
+    }
+}
+
+class iCalendar_property_tzoffsetto extends iCalendar_property {
+
+    var $name        = 'TZOFFSETTO';
+    var $val_type    = RFC2445_TYPE_UTC_OFFSET;
+
+    function __construct() {
+        $this->valid_parameters = array(
+            RFC2445_XNAME => RFC2445_OPTIONAL
+        );
+    }
+
+    function is_valid_value($value) {
+        if(!parent::is_valid_value($value)) {
+            return false;
+        } else {
+        	return true;
+        }
+    }
+}
+
+
 
 #######################
 /*
@@ -1288,7 +1423,7 @@ class iCalendar_property_class extends iCalendar_property {
     var $name        = 'CLASS';
     var $val_type    = RFC2445_TYPE_TEXT;
 
-    function construct() {
+    function __construct() {
         $this->valid_parameters = array(
             RFC2445_XNAME => RFC2445_OPTIONAL
         );
@@ -1296,4 +1431,3 @@ class iCalendar_property_class extends iCalendar_property {
 }
 */
 
-?>
