diff --git a/calendar/index.php b/calendar/index.php
index 53490cb..ba009f8 100644
--- a/calendar/index.php
+++ b/calendar/index.php
@@ -1,7 +1,5 @@
-<?php // %Id%
+<?php
 
-    require('../config.php');
-
-    redirect($CFG->wwwroot.'/calendar/view.php');
-
-?>
+require('../config.php');
+$PAGE->set_url('/calendar/view.php');
+redirect($CFG->wwwroot.'/calendar/view.php');
\ No newline at end of file
