diff --git a/blocks/participants/block_participants.php b/blocks/participants/block_participants.php
index 02d55a6..5779210 100644
--- a/blocks/participants/block_participants.php
+++ b/blocks/participants/block_participants.php
@@ -1,37 +1,28 @@
-<?PHP //$Id$
+<?php
 
 class block_participants extends block_list {
     function init() {
-        $this->title = get_string('people');
-        $this->version = 2007101509;
+        $this->title = get_string('pluginname', 'block_participants');
     }
 
     function get_content() {
 
-        global $CFG, $COURSE;
+        global $CFG, $OUTPUT;
 
         if (empty($this->instance)) {
             $this->content = '';
             return $this->content;
         }
 
-        // the following 3 lines is need to pass _self_test();
-        if (empty($this->instance->pageid)) {
-            return '';
-        }
-        
-        $this->content = new object();
+        $this->content = new stdClass();
         $this->content->items = array();
         $this->content->icons = array();
         $this->content->footer = '';
         
         /// MDL-13252 Always get the course context or else the context may be incorrect in the user/index.php
-        if (!$currentcontext = get_context_instance(CONTEXT_COURSE, $COURSE->id)) {
-            $this->content = '';
-            return $this->content;
-        }
+        $currentcontext = $this->page->context;
         
-        if ($COURSE->id == SITEID) {
+        if ($this->page->course->id == SITEID) {
             if (!has_capability('moodle/site:viewparticipants', get_context_instance(CONTEXT_SYSTEM))) {
                 $this->content = '';
                 return $this->content;
@@ -42,10 +33,9 @@ class block_participants extends block_list {
                 return $this->content;
             }
         }
-
+        $icon = '<img src="'.$OUTPUT->pix_url('i/users') . '" class="icon" alt="" />&nbsp;';
         $this->content->items[] = '<a title="'.get_string('listofallpeople').'" href="'.
-                                  $CFG->wwwroot.'/user/index.php?contextid='.$currentcontext->id.'">'.get_string('participants').'</a>';
-        $this->content->icons[] = '<img src="'.$CFG->pixpath.'/i/users.gif" class="icon" alt="" />';
+                                  $CFG->wwwroot.'/user/index.php?contextid='.$currentcontext->id.'">'.$icon.get_string('participants').'</a>';
 
         return $this->content;
     }
@@ -57,4 +47,4 @@ class block_participants extends block_list {
 
 }
 
-?>
+
