diff --git a/theme/canvas/config.php b/theme/canvas/config.php
index 2a043b2..9848de1 100644
--- a/theme/canvas/config.php
+++ b/theme/canvas/config.php
@@ -166,7 +166,14 @@ $THEME->layouts = array(
         'file' => 'report.php',
         'regions' => array('side-pre'),
         'defaultregion' => 'side-pre',
-    )
+    ),
+    // The pagelayout used for safebrowser and securewindow.
+    'secure' => array(
+        'file' => 'general.php',
+        'regions' => array('side-pre', 'side-post'),
+        'defaultregion' => 'side-pre',
+        'options' => array('nofooter'=>true, 'nonavbar'=>true, 'nocustommenu'=>true, 'nologinlinks'=>true),
+    ),
 );
 
 /////////////////////////////////////////////////////////
