diff --git a/mod/scorm/datamodels/debug.js.php b/mod/scorm/datamodels/debug.js.php
index d55f64e..ce27636 100644
--- a/mod/scorm/datamodels/debug.js.php
+++ b/mod/scorm/datamodels/debug.js.php
@@ -1,20 +1,27 @@
-// hopefully fool ie IE proof way of getting DOM element
+<!--// hopefully fool ie IE proof way of getting DOM element
 function safeGetElement(doc, el) {
     return doc.ids ? doc.ids[el] : doc.getElementById ? doc.getElementById(el) : doc.all[el];
 }
+// Find elements by class name
+var aryClassElements = new Array();
+function getNextElementByClassName( strClassName, obj ) {
+    if ( obj.className == strClassName ) {
+        aryClassElements[aryClassElements.length] = obj;
+    }
+    for ( var i = 0; i < obj.childNodes.length; i++ )
+        getNextElementByClassName( strClassName, obj.childNodes[i] );
+}
 
-// Add in a JS controlled link for toggling the Debug logging
-var logButton = document.createElement('a');
-logButton.id = 'mod-scorm-log-toggle';
-logButton.name = 'logToggle';
-logButton.href = 'javascript:toggleLog();';
-if (getLoggingActive() == "A") {
-    logButton.innerHTML = '<?php print_string('scormloggingon','scorm') ?>';
-} else {
-    logButton.innerHTML = '<?php print_string('scormloggingoff','scorm') ?>';
+function getElementsByClassName( strClassName, obj ) {
+    aryClassElements = new Array();
+    getNextElementByClassName( strClassName, obj );
+    if (aryClassElements.length > 0) {
+        return aryClassElements[0];
+    }
+    else {
+        return null;
+    }
 }
-var content = safeGetElement(document, 'content');
-content.appendChild(logButton);
 
 // retrieve cookie data
 function getCookie (cookie_name){
@@ -45,7 +52,7 @@ function toggleLog () {
     if (getLoggingActive() == "A") {
         AppendToLog("Moodle Logging Deactivated", 0);
         setLoggingActive('N');
-        logButton.innerHTML = '<?php print_string('scormloggingoff','scorm') ?>';
+        logButton.innerHTML = '--><?php print_string('scormloggingoff','scorm') ?>';
     } else {
         setLoggingActive('A');
         AppendToLog("Moodle Logging Activated", 0);
@@ -714,3 +721,18 @@ function LogAPICall(func, nam, val, rc) {
     s += ' => ' + String(rc);
     AppendToLog(s, rc);
 }
+
+
+// Add in a JS controlled link for toggling the Debug logging
+var logButton = document.createElement('a');
+logButton.id = 'mod-scorm-log-toggle';
+logButton.name = 'logToggle';
+logButton.href = 'javascript:toggleLog();';
+if (getLoggingActive() == "A") {
+    logButton.innerHTML = '<?php print_string('scormloggingon','scorm') ?>';
+} else {
+    logButton.innerHTML = '<?php print_string('scormloggingoff','scorm') ?>';
+}
+var content = safeGetElement(document, 'scormpage');
+content.insertBefore(logButton, content.firstChild);
+
