diff --git a/mod/scorm/datamodels/scorm_13.js.php b/mod/scorm/datamodels/scorm_13.js.php
index 16ac757..54286a5 100644
--- a/mod/scorm/datamodels/scorm_13.js.php
+++ b/mod/scorm/datamodels/scorm_13.js.php
@@ -12,6 +12,9 @@
             }
         }
     }
+    if (!isset($currentorg)) {
+        $currentorg = '';
+    }
 ?>
 
 // Used need to debug cmi content (if you uncomment this, you must comment the definition inside SCORMapi1_3)
@@ -263,6 +266,7 @@ function SCORMapi1_3() {
     var Initialized = false;
     var Terminated = false;
     var diagnostic = "";
+    var errorCode = "0";
 
     function Initialize (param) {
         errorCode = "0";
@@ -271,8 +275,7 @@ function SCORMapi1_3() {
                 Initialized = true;
                 errorCode = "0";
                 <?php
-                    if (debugging('',DEBUG_DEVELOPER)) {
-//                        echo 'alert("Initialized SCORM 1.3");';
+                    if (scorm_debugging($scorm)) {
                         echo 'LogAPICall("Initialize", param, "", errorCode);';
                     }
                 ?>
@@ -288,34 +291,47 @@ function SCORMapi1_3() {
             errorCode = "201";
         }
         <?php
-            if (debugging('',DEBUG_DEVELOPER)) {
-//                echo 'alert("Initialize: "+GetErrorString(errorCode));';
+            if (scorm_debugging($scorm)) {
                 echo 'LogAPICall("Initialize", param, "", errorCode);';
             }
         ?>
         return "false";
     }
 
+
+<?php
+    // pull in the TOC callback
+    include_once($CFG->dirroot.'/mod/scorm/datamodels/callback.js.php');
+ ?>
+
+
     function Terminate (param) {
         errorCode = "0";
         if (param == "") {
             if ((Initialized) && (!Terminated)) {
+                var AJAXResult = StoreData(cmi,true);
+                <?php
+                    if (scorm_debugging($scorm)) {
+                        echo 'LogAPICall("Terminate", "AJAXResult", AJAXResult, 0);';
+                    }
+                ?>
+                result = ('true' == AJAXResult) ? 'true' : 'false';
+                errorCode = ('true' == result)? '0' : '101'; // General exception for any AJAX fault
                 <?php
-                    if (debugging('',DEBUG_DEVELOPER)) {
-//                        echo 'alert("Terminated SCORM 1.3");';
-                        echo 'LogAPICall("Terminate", param, "", 0);';
+                    if (scorm_debugging($scorm)) {
+                        echo 'LogAPICall("Terminate", "result", result, errorCode);';
                     }
                 ?>
+                if ('true' == result) {
                 Initialized = false;
                 Terminated = true;
-                var result = StoreData(cmi,true);
                 if (adl.nav.request != '_none_') {
                     switch (adl.nav.request) {
                         case 'continue':
-                            setTimeout('top.nextSCO();',500);
+                                setTimeout('scorm_get_next();',500);
                         break;
                         case 'previous':
-                            setTimeout('top.prevSCO();',500);
+                                setTimeout('scorm_get_prev();',500);
                         break;
                         case 'choice':
                         break;
@@ -330,10 +346,16 @@ function SCORMapi1_3() {
                     }
                 } else {
                     if (<?php echo $scorm->auto ?> == 1) {
-                        setTimeout('top.nextSCO();',500);
+                            setTimeout('scorm_get_next();',500);
                     }
                 }
-                return "true";
+                    // trigger TOC update
+                    var sURL = "<?php echo $CFG->wwwroot; ?>" + "/mod/scorm/prereqs.php?a=<?php echo $scorm->id ?>&scoid=<?php echo $scoid ?>&attempt=<?php echo $attempt ?>&mode=<?php echo $mode ?>&currentorg=<?php echo $currentorg ?>&sesskey=<?php echo sesskey(); ?>";
+                    YAHOO.util.Connect.asyncRequest('GET', sURL, this.connectPrereqCallback, null);
+                } else {
+                    diagnostic = "Failure calling the Terminate remote callback: the server replied with HTTP Status " + AJAXResult;
+                }
+                return result;
             } else {
                 if (Terminated) {
                     errorCode = "113";
@@ -345,8 +367,8 @@ function SCORMapi1_3() {
             errorCode = "201";
         }
         <?php
-            if (debugging('',DEBUG_DEVELOPER)) {
-                echo 'alert("Terminate: "+GetErrorString(errorCode));';
+            if (scorm_debugging($scorm)) {
+                echo 'LogAPICall("Terminate", param, "", errorCode);';
             }
         ?>
         return "false";
@@ -377,8 +399,7 @@ function SCORMapi1_3() {
                             if ((typeof eval(subelement) != "undefined") && (eval(subelement) != null)) {
                                 errorCode = "0";
                                 <?php
-                                    if (debugging('',DEBUG_DEVELOPER)) {
-//                                        echo 'alert("GetValue("+element+") -> "+eval(element));';
+                                    if (scorm_debugging($scorm)) {
                                         echo 'LogAPICall("GetValue", element, eval(element), 0);';
                                     }
                                 ?>
@@ -443,8 +464,7 @@ function SCORMapi1_3() {
             }
         }
         <?php
-            if (debugging('',DEBUG_DEVELOPER)) {
-//                echo 'alert("GetValue("+element+") -> "+GetErrorString(errorCode));';
+            if (scorm_debugging($scorm)) {
                 echo 'LogAPICall("GetValue", element, "", errorCode);';
             }
         ?>
@@ -740,8 +760,7 @@ function SCORMapi1_3() {
                                             eval(element+'=value;');
                                             errorCode = "0";
                                             <?php
-                                                if (debugging('',DEBUG_DEVELOPER)) {
-//                                                    echo 'alert("SetValue("+element+","+value+") -> OK");';
+                                                if (scorm_debugging($scorm)) {
                                                     echo 'LogAPICall("SetValue", element, value, errorCode);';
                                                 }
                                             ?>
@@ -756,8 +775,7 @@ function SCORMapi1_3() {
                                     eval(element+'=value;');
                                     errorCode = "0";
                                     <?php
-                                        if (debugging('',DEBUG_DEVELOPER)) {
-//                                           echo 'alert("SetValue("+element+","+value+") -> OK");';
+                                        if (scorm_debugging($scorm)) {
                                             echo 'LogAPICall("SetValue", element, value, errorCode);';
                                         }
                                     ?>
@@ -784,7 +802,7 @@ function SCORMapi1_3() {
             }
         }
         <?php
-            if (debugging('',DEBUG_DEVELOPER)) {
+            if (scorm_debugging($scorm)) {
                 echo 'LogAPICall("SetValue", element, value, errorCode);';
             }
         ?>
@@ -909,14 +927,23 @@ function SCORMapi1_3() {
         errorCode = "0";
         if (param == "") {
             if ((Initialized) && (!Terminated)) {
-                result = StoreData(cmi,false);
+                var AJAXResult = StoreData(cmi,false);
                 <?php
-                    if (debugging('',DEBUG_DEVELOPER)) {
-                        echo 'LogAPICall("Commit", param, "", 0);';
-                        //echo 'alert("Data Commited");';
+                    if (scorm_debugging($scorm)) {
+                        echo 'LogAPICall("Commit", "AJAXResult", AJAXResult, 0);';
                     }
                 ?>
-                return "true";
+                var result = ('true' == AJAXResult) ? 'true' : 'false';
+                errorCode = ('true' == result)? '0' : '101'; // General exception for any AJAX fault
+                <?php
+                    if (scorm_debugging($scorm)) {
+                        echo 'LogAPICall("Commit", "result", result, errorCode);';
+                    }
+                ?>
+                if ('false' == result) {
+                    diagnostic = "Failure calling the Commit remote callback: the server replied with HTTP Status " + AJAXResult;
+                }
+                return result;
             } else {
                 if (Terminated) {
                     errorCode = "143";
@@ -928,9 +955,8 @@ function SCORMapi1_3() {
             errorCode = "201";
         }
         <?php
-            if (debugging('',DEBUG_DEVELOPER)) {
-                echo 'LogAPICall("Commit", param, "", 0);';
-//                echo 'alert("Commit: "+GetErrorString(errorCode));';
+            if (scorm_debugging($scorm)) {
+                echo 'LogAPICall("Commit", param, "", errorCode);';
             }
         ?>
         return "false";
@@ -938,7 +964,7 @@ function SCORMapi1_3() {
 
     function GetLastError () {
      <?php
-        if (debugging('',DEBUG_DEVELOPER)) {
+        if (scorm_debugging($scorm)) {
             echo 'LogAPICall("GetLastError", "", "", errorCode);';
         }
     ?>
@@ -1029,14 +1055,14 @@ function SCORMapi1_3() {
                 break;
             }
             <?php
-            if (debugging('',DEBUG_DEVELOPER)) {
+            if (scorm_debugging($scorm)) {
                 echo 'LogAPICall("GetErrorString", param,  errorString, 0);';
             }
              ?>
             return errorString;
         } else {
            <?php
-            if (debugging('',DEBUG_DEVELOPER)) {
+            if (scorm_debugging($scorm)) {
                 echo 'LogAPICall("GetErrorString", param,  "No error string found!", 0);';
             }
              ?>
@@ -1047,14 +1073,14 @@ function SCORMapi1_3() {
     function GetDiagnostic (param) {
         if (diagnostic != "") {
             <?php
-                if (debugging('',DEBUG_DEVELOPER)) {
+                if (scorm_debugging($scorm)) {
                     echo 'LogAPICall("GetDiagnostic", param, diagnostic, 0);';
                 }
             ?>
             return diagnostic;
         }
         <?php
-            if (debugging('',DEBUG_DEVELOPER)) {
+            if (scorm_debugging($scorm)) {
                 echo 'LogAPICall("GetDiagnostic", param, param, 0);';
             }
         ?>
@@ -1098,7 +1124,7 @@ function SCORMapi1_3() {
 
     function AddTime (first, second) {
         <?php
-//            if (debugging('',DEBUG_DEVELOPER)) {
+//            if (scorm_debugging($scorm)) {
 //                echo 'alert("AddTime: "+first+" + "+second);';
 //            }
         ?>
@@ -1228,18 +1254,8 @@ function SCORMapi1_3() {
         datastring += navrequest;
         datastring += '&attempt=<?php echo $attempt ?>';
         datastring += '&scoid=<?php echo $scoid ?>';
-        <?php
-//            if (debugging('',DEBUG_DEVELOPER)) {
-//                echo 'popupwin(datastring);';
-//            }
-        ?>
         var myRequest = NewHttpReq();
-        var result = DoRequest(myRequest,"<?php p($CFG->wwwroot) ?>/mod/scorm/datamodel.php","id=<?php p($id) ?>&sesskey=<?php p($USER->sesskey) ?>"+datastring);
-        <?php
-//            if (debugging('',DEBUG_DEVELOPER)) {
-//                echo 'popupwin(result);';
-//            }
-        ?>
+        var result = DoRequest(myRequest,"<?php p($CFG->wwwroot) ?>/mod/scorm/datamodel.php","id=<?php p($id) ?>&a=<?php p($a) ?>&sesskey=<?php echo sesskey() ?>"+datastring);
         var results = String(result).split('\n');
         if ((results.length > 2) && (navrequest != '')) {
             eval(results[2]);
@@ -1263,7 +1279,7 @@ var API_1484_11 = new SCORMapi1_3();
 
 <?php
 // pull in the debugging utilities
-if (debugging('',DEBUG_DEVELOPER)) {
+if (scorm_debugging($scorm)) {
     include_once($CFG->dirroot.'/mod/scorm/datamodels/debug.js.php');
     echo 'AppendToLog("Moodle SCORM 1.3 API Loaded, Activity: '.$scorm->name.', SCO: '.$sco->identifier.'", 0);';
 }
