diff --git a/mod/data/css.php b/mod/data/css.php
old mode 100755
new mode 100644
index 00469ed..a6f9616
--- a/mod/data/css.php
+++ b/mod/data/css.php
@@ -1,40 +1,43 @@
-<?php  // $Id$
-///////////////////////////////////////////////////////////////////////////
-//                                                                       //
-// NOTICE OF COPYRIGHT                                                   //
-//                                                                       //
-// Moodle - Modular Object-Oriented Dynamic Learning Environment         //
-//          http://moodle.org                                            //
-//                                                                       //
-// Copyright (C) 2005 Martin Dougiamas  http://dougiamas.com             //
-//                                                                       //
-// This program is free software; you can redistribute it and/or modify  //
-// it under the terms of the GNU General Public License as published by  //
-// the Free Software Foundation; either version 2 of the License, or     //
-// (at your option) any later version.                                   //
-//                                                                       //
-// This program is distributed in the hope that it will be useful,       //
-// but WITHOUT ANY WARRANTY; without even the implied warranty of        //
-// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         //
-// GNU General Public License for more details:                          //
-//                                                                       //
-//          http://www.gnu.org/copyleft/gpl.html                         //
-//                                                                       //
-///////////////////////////////////////////////////////////////////////////
-
-    $lifetime  = 600;                                   // Seconds to cache this stylesheet
-    $nomoodlecookie = true;                             // Cookies prevent caching, so don't use them
-
-    require_once('../../config.php');
-
-    $d = optional_param('d', 0, PARAM_INT);   // database id
-
-    if ($data = get_record('data', 'id', $d)) {
+<?php
+
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * This file is part of the Database module for Moodle
+ *
+ * @copyright 2005 Martin Dougiamas  http://dougiamas.com
+ * @license http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @package mod-data
+ */
+
+define('NO_MOODLE_COOKIES', true); // session not used here
+
+require_once('../../config.php');
+
+$d = optional_param('d', 0, PARAM_INT);   // database id
+$lifetime  = 600;                                   // Seconds to cache this stylesheet
+
+$PAGE->set_url('/mod/data/css.php', array('d'=>$d));
+
+if ($data = $DB->get_record('data', array('id'=>$d))) {
         header('Last-Modified: ' . gmdate('D, d M Y H:i:s', time()) . ' GMT');
         header('Expires: ' . gmdate("D, d M Y H:i:s", time() + $lifetime) . ' GMT');
         header('Cache-control: max_age = '. $lifetime);
         header('Pragma: ');
-        header('Content-type: text/css');  // Correct MIME type
+    header('Content-type: text/css; charset=utf-8');  // Correct MIME type
 
         echo $data->csstemplate;
-    }
+}
\ No newline at end of file
