diff --git a/mnet/xmlrpc/xmlparser.php b/mnet/xmlrpc/xmlparser.php
index e62743d..95ccd2a 100644
--- a/mnet/xmlrpc/xmlparser.php
+++ b/mnet/xmlrpc/xmlparser.php
@@ -83,13 +83,10 @@ class mnet_encxml_parser {
      * |                                                |
      * |________________________________________________|
      *
-     * @uses $MNET
      * @param   string  $data   The XML that you want to parse
      * @return  bool            True on success - false on failure
      */
     function parse($data) {
-        global $MNET, $MNET_REMOTE_CLIENT;
-
         $p = xml_parse($this->parser, $data);
 
         if ($p == 0) {
@@ -308,4 +305,3 @@ class mnet_encxml_parser {
         return true;
     }
 }
-?>
