diff --git a/course/format/weeks/config.php b/course/format/weeks/config.php
old mode 100755
new mode 100644
index 2d3eaca..1a429b4
--- a/course/format/weeks/config.php
+++ b/course/format/weeks/config.php
@@ -6,8 +6,5 @@
 // format.
 //
 // The default blocks layout for this course format:
-    $format['defaultblocks'] = 'participants,activity_modules,search_forums,'.
-                               'admin,course_list:news_items,calendar_upcoming,'.
-                               'recent_activity';
+    $format['defaultblocks'] = ':search_forums,news_items,calendar_upcoming,recent_activity';
 //
-?>
\ No newline at end of file
