diff --git a/group/clientlib.js b/group/clientlib.js
index 5a3ddf7..9b0c41f 100644
--- a/group/clientlib.js
+++ b/group/clientlib.js
@@ -19,7 +19,6 @@ function UpdatableGroupsCombo(wwwRoot, courseId) {
         	if (o.responseText !== undefined) {
                 var groupsComboEl = document.getElementById("groups");
                 var membersComboEl = document.getElementById("members");
-
                 if (membersComboEl) {
                     // Clear the members list box.
                     while (membersComboEl.firstChild) {
@@ -184,7 +183,7 @@ var createLoaderImg = function (elClass, parentId, wwwRoot) {
     }
     var loadingImg = document.createElement("img");
 
-    loadingImg.setAttribute("src", wwwRoot+"/pix/i/ajaxloader.gif");
+    loadingImg.setAttribute("src", M.util.image_url('/i/ajaxloader', 'moodle'));
     loadingImg.setAttribute("class", elClass);
     loadingImg.setAttribute("alt", "Loading");
     loadingImg.setAttribute("id", "loaderImg");
@@ -201,3 +200,53 @@ var removeLoaderImgs = function (elClass, parentId) {
         parentEl.removeChild(loader);
     }
 };
+
+/**
+ * Updates the current groups information shown about a user when a user is selected.
+ *
+ * @global {Array} userSummaries
+ *      userSummaries is added to the page via /user/selector/lib.php - group_non_members_selector::print_user_summaries()
+ *      as a global that can be used by this function.
+ */
+function updateUserSummary() {
+    var selectEl = document.getElementById('addselect'),
+        summaryDiv = document.getElementById('group-usersummary'),
+        length = selectEl.length,
+        selectCnt = 0,
+        selectIdx = -1,
+        i;
+
+    for (i = 0; i < length; i++) {
+        if (selectEl.options[i].selected) {
+            selectCnt++;
+            selectIdx = i;
+        }
+    }
+
+    if (selectCnt == 1 && userSummaries[selectIdx]) {
+        summaryDiv.innerHTML = userSummaries[selectIdx];
+    } else {
+        summaryDiv.innerHTML = '';
+    }
+
+    return true;
+}
+
+function init_add_remove_members_page(Y) {
+    var add = Y.one('#add');
+    var addselect = M.core_user.get_user_selector('addselect');
+    add.set('disabled', addselect.is_selection_empty());
+    addselect.on('user_selector:selectionchanged', function(isempty) {
+        add.set('disabled', isempty);
+    });
+
+    var remove = Y.one('#remove');
+    var removeselect = M.core_user.get_user_selector('removeselect');
+    remove.set('disabled', removeselect.is_selection_empty());
+    removeselect.on('user_selector:selectionchanged', function(isempty) {
+        remove.set('disabled', isempty);
+    });
+
+    addselect = document.getElementById('addselect');
+    addselect.onchange = updateUserSummary;
+}
\ No newline at end of file
