diff --git a/group/delete.php b/group/delete.php
index 67f5d7c..3d5af15 100644
--- a/group/delete.php
+++ b/group/delete.php
@@ -16,9 +16,11 @@ $courseid = required_param('courseid', PARAM_INT);
 $groupids = required_param('groups', PARAM_SEQUENCE);
 $confirm = optional_param('confirm', 0, PARAM_BOOL);
 
+$PAGE->set_url('/group/delete.php', array('courseid'=>$courseid,'groups'=>$groupids));
+
 // Make sure course is OK and user has access to manage groups
-if (!$course = get_record('course', 'id', $courseid)) {
-    error('Course ID was incorrect');
+if (!$course = $DB->get_record('course', array('id' => $courseid))) {
+    print_error('invalidcourseid');
 }
 require_login($course);
 $context = get_context_instance(CONTEXT_COURSE, $course->id);
@@ -28,11 +30,11 @@ require_capability('moodle/course:managegroups', $context);
 $groupidarray = explode(',',$groupids);
 $groupnames = array();
 foreach($groupidarray as $groupid) {
-    if (!$group = get_record('groups', 'id', $groupid)) {
-        error('Group ID was incorrect');
+    if (!$group = $DB->get_record('groups', array('id' => $groupid))) {
+        print_error('invalidgroupid');
     }
     if ($courseid != $group->courseid) {
-        error('Group not on required course');
+        print_error('groupunknown', '', '', $group->courseid);
     }
     $groupnames[] = format_string($group->name);
 }
@@ -47,16 +49,16 @@ if ($confirm && data_submitted()) {
     if (!confirm_sesskey() ) {
         print_error('confirmsesskeybad','error',$returnurl);
     }
-    begin_sql();
+
     foreach($groupidarray as $groupid) {
-        if (!groups_delete_group($groupid)) {
-            print_error('erroreditgroup', 'group', $returnurl);
+        groups_delete_group($groupid);
         } 
-    }
-    commit_sql();
+
     redirect($returnurl);
 } else {
-    print_header(get_string('deleteselectedgroup', 'group'), get_string('deleteselectedgroup', 'group'));
+    $PAGE->set_title(get_string('deleteselectedgroup', 'group'));
+    $PAGE->set_heading($course->fullname . ': '. get_string('deleteselectedgroup', 'group'));
+    echo $OUTPUT->header();
     $optionsyes = array('courseid'=>$courseid, 'groups'=>$groupids, 'sesskey'=>sesskey(), 'confirm'=>1);
     $optionsno = array('id'=>$courseid);
     if(count($groupnames)==1) {
@@ -68,7 +70,8 @@ if ($confirm && data_submitted()) {
         }
         $message.='</ul>';
     }
-    notice_yesno($message, 'delete.php', 'index.php', $optionsyes, $optionsno, 'post', 'get');
-    print_footer();
+    $formcontinue = new single_button(new moodle_url('delete.php', $optionsyes), get_string('yes'), 'post');
+    $formcancel = new single_button(new moodle_url('index.php', $optionsno), get_string('no'), 'get');
+    echo $OUTPUT->confirm($message, $formcontinue, $formcancel);
+    echo $OUTPUT->footer();
 }
-?>
