diff --git a/admin/report/log/settings.php b/admin/report/log/settings.php
index be15280..ecaae3b 100644
--- a/admin/report/log/settings.php
+++ b/admin/report/log/settings.php
@@ -1,5 +1,7 @@
-<?php  // $Id$
+<?php
+
+defined('MOODLE_INTERNAL') || die;
+
 // just a link to course report
 $ADMIN->add('reports', new admin_externalpage('reportlog', get_string('log', 'admin'), "$CFG->wwwroot/course/report/log/index.php?id=".SITEID, 'coursereport/log:view'));
 $ADMIN->add('reports', new admin_externalpage('reportloglive', get_string('loglive', 'coursereport_log'), "$CFG->wwwroot/course/report/log/indexlive.php?id=".SITEID, 'coursereport/log:viewlive'));
-?>
\ No newline at end of file
