diff --git a/lib/simpletest/fixtures/user_agents.php b/lib/simpletest/fixtures/user_agents.php
index b1e7b0b..6a4b49c 100644
--- a/lib/simpletest/fixtures/user_agents.php
+++ b/lib/simpletest/fixtures/user_agents.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 
 ///////////////////////////////////////////////////////////////////////////
 //                                                                       //
@@ -54,4 +54,4 @@ $user_agents = array(
                         'Debian Linux' => 'Opera/9.01 (X11; Linux i686; U; en)')
     )
 );
-?>
+
