diff --git a/course/report/log/graph.php b/course/report/log/graph.php
index e3a5ea8..cf13b2c 100644
--- a/course/report/log/graph.php
+++ b/course/report/log/graph.php
@@ -1,37 +1,68 @@
-<?php // $Id$
-      // Produces a graph of log accesses
-
-    require_once("../../../config.php");
-    require_once("../../lib.php");
-    require_once("$CFG->libdir/graphlib.php");
-
-    $id   = required_param('id', PARAM_INT);    // Course ID
-    $type = required_param('type', PARAM_FILE);  // Graph Type
-    $user = optional_param('user', 0, PARAM_INT);  // Student ID
-    $date = optional_param('date', 0, PARAM_INT);  // A time of a day (in GMT)
-
-    if (! $course = get_record("course", "id", $id)) {
-        error("Course is misconfigured");
-    }
-
-    require_login($course);
-    $context = get_context_instance(CONTEXT_COURSE, $course->id);
-
-    if (!$course->showreports or $USER->id != $user) {
+<?php
+
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Produces a graph of log accesses
+ *
+ * @copyright 1999 Martin Dougiamas  http://dougiamas.com
+ * @license http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ * @package course
+ */
+
+require_once("../../../config.php");
+require_once("../../lib.php");
+require_once("$CFG->libdir/graphlib.php");
+
+$id   = required_param('id', PARAM_INT);    // Course ID
+$type = required_param('type', PARAM_FILE);  // Graph Type
+$user = optional_param('user', 0, PARAM_INT);  // Student ID
+$date = optional_param('date', 0, PARAM_INT);  // A time of a day (in GMT)
+
+$url = new moodle_url('/course/report/log/graph.php', array('id'=>$id,'type'=>$type));
+if ($user !== 0) {
+    $url->param('user', $user);
+}
+if ($date !== 0) {
+    $url->param('date', $date);
+}
+$PAGE->set_url($url);
+
+if (! $course = $DB->get_record("course", array("id"=>$id))) {
+    print_error('invalidcourseid');
+}
+
+require_login($course);
+$context = get_context_instance(CONTEXT_COURSE, $course->id);
+
+if (!$course->showreports or $USER->id != $user) {
         require_capability('coursereport/log:view', $context);
-    }
+}
 
-    if ($user) {
-        if (! $user = get_record("user", "id", $user)) {
-            error("Can not find that user");
-        }
+if ($user) {
+    if (! $user = $DB->get_record("user", array("id"=>$user))) {
+        print_error("nousers");
     }
+}
 
-    $logs = array();
+$logs = array();
 
-    $timenow = time();
+$timenow = time();
 
-    switch ($type) {
+switch ($type) {
      case "usercourse.png":
 
        $site = get_site();
@@ -63,6 +94,7 @@
            $reducedays = 0;
        }
 
+   $days = array();
        $i = 0;
        while ($timestart < $timenow) {
            $timefinish = $timestart + 86400;
@@ -88,7 +120,7 @@
 
        $graph = new graph(750, 400);
 
-       $a->coursename = $course->shortname;
+   $a->coursename = format_string($course->shortname, true, array('context' => $context));
        $a->username = fullname($user, true);
        $graph->parameter['title'] = get_string("hitsoncourse", "", $a);
 
@@ -136,6 +168,7 @@
        }
        $dayfinish = $daystart + 86400;
 
+   $hours = array();
        for ($i=0; $i<=23; $i++) {
            $logs[$i] = 0;
            $hour = $daystart + $i * 3600;
@@ -150,7 +183,7 @@
 
        $graph = new graph(750, 400);
 
-       $a->coursename = $course->shortname;
+   $a->coursename = format_string($course->shortname, true, array('context' => $context));
        $a->username = fullname($user, true);
        $graph->parameter['title'] = get_string("hitsoncoursetoday", "", $a);
 
@@ -179,6 +212,5 @@
 
      default:
        break;
-   }
+}
 
-?>
