diff --git a/course/report/log/indexlive.php b/course/report/log/indexlive.php
index adcb807..09978b9 100644
--- a/course/report/log/indexlive.php
+++ b/course/report/log/indexlive.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
       // Display link to live logs in separate window
 
     require_once('../../../config.php');
@@ -6,10 +6,10 @@
     require_once('lib.php');
     require_once($CFG->libdir.'/adminlib.php');
 
-    $id = required_param('id', PARAM_INT);// Course ID
+    $id = optional_param('id', 0, PARAM_INT);// Course ID
 
-    if (!$course = get_record('course', 'id', $id) ) {
-        error('That\'s an invalid course id'.$id);
+    if (!$course = $DB->get_record('course', array('id'=>$id)) ) {
+        print_error('invalidcourseid');
     }
 
     require_login($course);
@@ -21,23 +21,24 @@
     $strreports = get_string('reports');
 
     if ($course->id == SITEID) {
-        admin_externalpage_setup('reportloglive','',array('id' => $course->id));
-        admin_externalpage_print_header();
+        admin_externalpage_setup('reportloglive');
+        echo $OUTPUT->header();
 
     } else {
-        $navlinks = array();
-        $navlinks[] = array('name' => $strreports, 'link' => "$CFG->wwwroot/course/report.php?id=$course->id", 'type' => 'misc');
-        $navlinks[] = array('name' => $strlogs, 'link' => null, 'type' => 'misc');
-        $navigation = build_navigation($navlinks);
-        print_header($course->shortname .': '. $strlogs, $course->fullname, $navigation, '');
+        $PAGE->set_title($course->shortname .': '. $strlogs);
+        $PAGE->set_heading($course->fullname);
+        $PAGE->navbar->add($strreports, new moodle_url('/course/report.php', array('id'=>$course->id)), navigation_node::TYPE_CUSTOM);
+        $PAGE->navbar->add($strlogs);
+        echo $OUTPUT->header();
     }
 
-    print_heading(get_string('loglive', 'coursereport_log'));
+    echo $OUTPUT->heading(get_string('loglive', 'coursereport_log'));
 
-    echo '<div class="info">';
-    link_to_popup_window('/course/report/log/live.php?id='. $course->id,'livelog', get_string('livelogs'), 500, 800);
-    echo '<div>';
+    echo $OUTPUT->container_start('info');
+    $link = new moodle_url('/course/report/log/live.php?id='. $course->id);
+    echo $OUTPUT->action_link($link, get_string('livelogs'), new popup_action('click', $link, 'livelog', array('height' => 500, 'width' => 800)));
+    echo $OUTPUT->container_end();
+
+    echo $OUTPUT->footer();
 
-    print_footer($course);
 
-?>
