diff --git a/install/lang/sl/error.php b/install/lang/sl/error.php
index 9c82a4c..8ad559b 100644
--- a/install/lang/sl/error.php
+++ b/install/lang/sl/error.php
@@ -38,9 +38,9 @@ $string['cannotfindcomponent'] = 'Komponente ni možno najti.';
 $string['cannotsavemd5file'] = 'Datoteke MD5 ni možno shraniti.';
 $string['cannotsavezipfile'] = 'Datoteke ZIP ni možno shraniti.';
 $string['cannotunzipfile'] = 'Datoteke ni možno razširiti (unzip).';
-$string['componentisuptodate'] = 'Komponenta je posodobljena.';
+$string['componentisuptodate'] = 'Komponentaje posodobljena';
 $string['downloadedfilecheckfailed'] = 'Preverjanje prenesene datoteke je spodletelo.';
-$string['invalidmd5'] = 'Neveljaven MD5';
+$string['invalidmd5'] = 'Neveljaven MD5 -  poskusite ponovno';
 $string['missingrequiredfield'] = 'Eno izmed zahtevanih polj manjka';
 $string['remotedownloaderror'] = 'Prenos komponent na vaš strežnik ni uspel. Prosimo, da preverite nastavitve prehodnega strežnika. Priporočamo uporabo razširitve cURL za PHP.<br /><br />Datoteko <a href="{$a->url}">{$a->url}</a> morate prenesti ročno - skopirajte jo v imenik "{$a->dest}" in razširite.';
 $string['wrongdestpath'] = 'Napačna ciljna pot.';
