diff --git a/mod/assignment/type/offline/assignment.class.php b/mod/assignment/type/offline/assignment.class.php
index 464b5d2..04097dc 100644
--- a/mod/assignment/type/offline/assignment.class.php
+++ b/mod/assignment/type/offline/assignment.class.php
@@ -1,4 +1,4 @@
-<?php // $Id$
+<?php
 
 /**
  * Extend the base assignment class for offline assignments
@@ -19,7 +19,7 @@ class assignment_offline extends assignment_base {
     }
 
     function prepare_new_submission($userid) {
-        $submission = new Object;
+        $submission = new stdClass();
         $submission->assignment   = $this->assignment->id;
         $submission->userid       = $userid;
         $submission->timecreated  = time(); // needed for offline assignments
@@ -38,7 +38,7 @@ class assignment_offline extends assignment_base {
 
     // needed for the timemodified override
     function process_feedback() {
-        global $CFG, $USER;
+        global $CFG, $USER, $DB;
         require_once($CFG->libdir.'/gradelib.php');
 
         if (!$feedback = data_submitted() or !confirm_sesskey()) {      // No incoming data?
@@ -66,9 +66,8 @@ class assignment_offline extends assignment_base {
         if (!$grading_info->items[0]->grades[$feedback->userid]->locked and
             !$grading_info->items[0]->grades[$feedback->userid]->overridden) {
 
-            $submission->grade      = $feedback->grade;
-            $submission->submissioncomment    = $feedback->submissioncomment;
-            $submission->format     = $feedback->format;
+            $submission->grade      = $feedback->xgrade;
+            $submission->submissioncomment    = $feedback->submissioncomment_editor['text'];
             $submission->teacher    = $USER->id;
             $mailinfo = get_user_preferences('assignment_mailinfo', 0);
             if (!$mailinfo) {
@@ -85,11 +84,9 @@ class assignment_offline extends assignment_base {
                 $submission->timemodified = time();
             }
 
-            if (! update_record('assignment_submissions', $submission)) {
-                return false;
-            }
+            $DB->update_record('assignment_submissions', $submission);
 
-            // triger grade event
+            // trigger grade event
             $this->update_grade($submission);
 
             add_to_log($this->course->id, 'assignment', 'update grades',
@@ -102,4 +99,4 @@ class assignment_offline extends assignment_base {
 
 }
 
-?>
+
